<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="Toolkit for identification and statistical testing 
    of RNA editing signals from within R. Provides support for identifying
    sites from bulk-RNA-seq, single cell RNA-seq datasets, and general methods 
    for extraction of allelic read counts from alignment files. ">
<title>RNA editing tools in R • raer</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><link href="deps/_Source%20Sans%20Pro-0.4.1/font.css" rel="stylesheet">
<link href="deps/_Source%20Code%20Pro-0.4.1/font.css" rel="stylesheet">
<!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="RNA editing tools in R">
<meta property="og:description" content="Toolkit for identification and statistical testing 
    of RNA editing signals from within R. Provides support for identifying
    sites from bulk-RNA-seq, single cell RNA-seq datasets, and general methods 
    for extraction of allelic read counts from alignment files. ">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><script defer data-domain="rnabioco.github.io" src="https://plausible.io/js/plausible.js"></script>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-dark"><div class="container">
    
    <a class="navbar-brand me-2" href="index.html">raer</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="articles/raer.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="articles/novel-sites.html">Identifying RNA editing sites using RNA and DNA sequencing data</a>
    <a class="dropdown-item" href="articles/single-cell.html">Identifying RNA editing sites in droplet single cell datasets</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/rnabioco/raer/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header"><h1 id="raer">raer<a class="anchor" aria-label="anchor" href="#raer"></a>
</h1></div>
<!-- badges: start -->

<p>raer is an R package that facilitates rapid interactive analysis of RNA editing in R in the bioconductor ecosystem.</p>
<p><strong>raer is under active development and functionality may change</strong></p>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>You can install the development version of raer from <a href="https://github.com/" class="external-link">GitHub</a> with:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># install.packages("devtools")</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"rnabioco/raer"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="quick-start">Quick start<a class="anchor" aria-label="anchor" href="#quick-start"></a>
</h2>
<p>The raer package provides methods to compute per site read count summaries from bam files, either for known sites, or for all detected sites.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://rnabioco.github.io/raer">raer</a></span><span class="op">)</span></span>
<span><span class="va">bamfn</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/raer_example.html">raer_example</a></span><span class="op">(</span><span class="st">"SRR5564269_Aligned.sortedByCoord.out.md.bam"</span><span class="op">)</span></span>
<span><span class="va">bam2fn</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/raer_example.html">raer_example</a></span><span class="op">(</span><span class="st">"SRR5564277_Aligned.sortedByCoord.out.md.bam"</span><span class="op">)</span></span>
<span><span class="va">fafn</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/raer_example.html">raer_example</a></span><span class="op">(</span><span class="st">"human.fasta"</span><span class="op">)</span></span>
<span><span class="va">bedfn</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/raer_example.html">raer_example</a></span><span class="op">(</span><span class="st">"regions.bed"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/get_pileup.html">get_pileup</a></span><span class="op">(</span><span class="va">bamfn</span>, <span class="va">fafn</span>, bedfile <span class="op">=</span> <span class="va">bedfn</span><span class="op">)</span></span>
<span><span class="va">res</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="op">]</span></span>
<span><span class="co">#&gt; GRanges object with 5 ranges and 9 metadata columns:</span></span>
<span><span class="co">#&gt;       seqnames    ranges strand |         Ref         Var      nRef      nVar</span></span>
<span><span class="co">#&gt;          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;character&gt; &lt;integer&gt; &lt;integer&gt;</span></span>
<span><span class="co">#&gt;   [1]     SSR3       201      - |           A           -        14         0</span></span>
<span><span class="co">#&gt;   [2]     SSR3       202      - |           T           -        14         0</span></span>
<span><span class="co">#&gt;   [3]     SSR3       203      - |           T           -        14         0</span></span>
<span><span class="co">#&gt;   [4]     SSR3       204      - |           T           -        15         0</span></span>
<span><span class="co">#&gt;   [5]     SSR3       205      - |           T           -        16         0</span></span>
<span><span class="co">#&gt;              nA        nT        nC        nG        nN</span></span>
<span><span class="co">#&gt;       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;</span></span>
<span><span class="co">#&gt;   [1]        14         0         0         0         0</span></span>
<span><span class="co">#&gt;   [2]         0        14         0         0         0</span></span>
<span><span class="co">#&gt;   [3]         0        14         0         0         0</span></span>
<span><span class="co">#&gt;   [4]         0        15         0         0         0</span></span>
<span><span class="co">#&gt;   [5]         0        16         0         0         0</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   seqinfo: 3 sequences from an unspecified genome</span></span></code></pre></div>
<p>The <code><a href="reference/get_pileup.html">FilterParam()</a></code> class holds multiple options for customizing the output of <code><a href="reference/get_pileup.html">get_pileup()</a></code>.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fp</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/get_pileup.html">FilterParam</a></span><span class="op">(</span>only_keep_variants <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/get_pileup.html">get_pileup</a></span><span class="op">(</span><span class="va">bamfn</span>, <span class="va">fafn</span>, filterParam <span class="op">=</span> <span class="va">fp</span><span class="op">)</span></span>
<span><span class="va">res</span></span>
<span><span class="co">#&gt; GRanges object with 31 ranges and 9 metadata columns:</span></span>
<span><span class="co">#&gt;        seqnames    ranges strand |         Ref         Var      nRef      nVar</span></span>
<span><span class="co">#&gt;           &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;character&gt; &lt;integer&gt; &lt;integer&gt;</span></span>
<span><span class="co">#&gt;    [1]     SSR3       102      - |           T          TG        12         1</span></span>
<span><span class="co">#&gt;    [2]     SSR3       228      - |           A          AG        13         1</span></span>
<span><span class="co">#&gt;    [3]     SSR3       244      - |           A          AG        18         1</span></span>
<span><span class="co">#&gt;    [4]     SSR3       254      - |           A          AG        18         1</span></span>
<span><span class="co">#&gt;    [5]     SSR3       258      - |           G          GA         8        10</span></span>
<span><span class="co">#&gt;    ...      ...       ...    ... .         ...         ...       ...       ...</span></span>
<span><span class="co">#&gt;   [27]     DHFR       300      - |           A          AG        56         1</span></span>
<span><span class="co">#&gt;   [28]     DHFR       332      - |           A          AG        55         1</span></span>
<span><span class="co">#&gt;   [29]     DHFR       336      - |           G          GT        51         1</span></span>
<span><span class="co">#&gt;   [30]     DHFR       430      - |           A          AT        37         1</span></span>
<span><span class="co">#&gt;   [31]     DHFR       513      - |           T          TC        35         1</span></span>
<span><span class="co">#&gt;               nA        nT        nC        nG        nN</span></span>
<span><span class="co">#&gt;        &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;</span></span>
<span><span class="co">#&gt;    [1]         0        12         0         1         0</span></span>
<span><span class="co">#&gt;    [2]        13         0         0         1         0</span></span>
<span><span class="co">#&gt;    [3]        18         0         0         1         0</span></span>
<span><span class="co">#&gt;    [4]        18         0         0         1         0</span></span>
<span><span class="co">#&gt;    [5]        10         0         0         8         0</span></span>
<span><span class="co">#&gt;    ...       ...       ...       ...       ...       ...</span></span>
<span><span class="co">#&gt;   [27]        56         0         0         1         0</span></span>
<span><span class="co">#&gt;   [28]        55         0         0         1         0</span></span>
<span><span class="co">#&gt;   [29]         0         1         0        51         0</span></span>
<span><span class="co">#&gt;   [30]        37         1         0         0         0</span></span>
<span><span class="co">#&gt;   [31]         0        35         1         0         0</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   seqinfo: 3 sequences from an unspecified genome</span></span></code></pre></div>
<p>Multiple bam files can be processed, which enables rapid comparisons of RNA-Seq vs. WGS or WXS data, or RNA-Seq vs RNA-seq (ADAR WT VS ADAR KO).</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fp</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/get_pileup.html">FilterParam</a></span><span class="op">(</span>only_keep_variants <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                  library_type <span class="op">=</span> <span class="st">"fr-first-strand"</span>,</span>
<span>                  min_nucleotide_depth <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span></span>
<span><span class="va">plps</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/get_pileup.html">get_pileup</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">bam2fn</span>, <span class="va">bamfn</span><span class="op">)</span>, </span>
<span>                  <span class="va">fafn</span>, </span>
<span>                  filterParam <span class="op">=</span> <span class="va">fp</span><span class="op">)</span></span>
<span><span class="va">plps</span></span>
<span><span class="co">#&gt; [[1]]</span></span>
<span><span class="co">#&gt; GRanges object with 74 ranges and 9 metadata columns:</span></span>
<span><span class="co">#&gt;        seqnames    ranges strand |         Ref         Var      nRef      nVar</span></span>
<span><span class="co">#&gt;           &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;character&gt; &lt;integer&gt; &lt;integer&gt;</span></span>
<span><span class="co">#&gt;    [1]     SSR3       102      - |           T           -        15         0</span></span>
<span><span class="co">#&gt;    [2]     SSR3       125      - |           C          CG        21         1</span></span>
<span><span class="co">#&gt;    [3]     SSR3       156      - |           C          CA        25         1</span></span>
<span><span class="co">#&gt;    [4]     SSR3       176      - |           A          AG         8        16</span></span>
<span><span class="co">#&gt;    [5]     SSR3       198      - |           A          AG        24         1</span></span>
<span><span class="co">#&gt;    ...      ...       ...    ... .         ...         ...       ...       ...</span></span>
<span><span class="co">#&gt;   [70]     DHFR       397      - |           A          AT        31         1</span></span>
<span><span class="co">#&gt;   [71]     DHFR       399      - |           G          GA        28         1</span></span>
<span><span class="co">#&gt;   [72]     DHFR       423      - |           T          TC        31         1</span></span>
<span><span class="co">#&gt;   [73]     DHFR       430      - |           A           -        33         0</span></span>
<span><span class="co">#&gt;   [74]     DHFR       513      - |           T           -        21         0</span></span>
<span><span class="co">#&gt;               nA        nT        nC        nG        nN</span></span>
<span><span class="co">#&gt;        &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;</span></span>
<span><span class="co">#&gt;    [1]         0        15         0         0         0</span></span>
<span><span class="co">#&gt;    [2]         0         0        21         1         0</span></span>
<span><span class="co">#&gt;    [3]         1         0        25         0         0</span></span>
<span><span class="co">#&gt;    [4]         8         0         0        16         0</span></span>
<span><span class="co">#&gt;    [5]        24         0         0         1         0</span></span>
<span><span class="co">#&gt;    ...       ...       ...       ...       ...       ...</span></span>
<span><span class="co">#&gt;   [70]        31         1         0         0         0</span></span>
<span><span class="co">#&gt;   [71]         1         0         0        28         0</span></span>
<span><span class="co">#&gt;   [72]         0        31         1         0         0</span></span>
<span><span class="co">#&gt;   [73]        33         0         0         0         0</span></span>
<span><span class="co">#&gt;   [74]         0        21         0         0         0</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   seqinfo: 3 sequences from an unspecified genome</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[2]]</span></span>
<span><span class="co">#&gt; GRanges object with 74 ranges and 9 metadata columns:</span></span>
<span><span class="co">#&gt;        seqnames    ranges strand |         Ref         Var      nRef      nVar</span></span>
<span><span class="co">#&gt;           &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;character&gt; &lt;integer&gt; &lt;integer&gt;</span></span>
<span><span class="co">#&gt;    [1]     SSR3       102      - |           T          TG        12         1</span></span>
<span><span class="co">#&gt;    [2]     SSR3       125      - |           C           -        17         0</span></span>
<span><span class="co">#&gt;    [3]     SSR3       156      - |           C           -        16         0</span></span>
<span><span class="co">#&gt;    [4]     SSR3       176      - |           A           -        16         0</span></span>
<span><span class="co">#&gt;    [5]     SSR3       198      - |           A           -        15         0</span></span>
<span><span class="co">#&gt;    ...      ...       ...    ... .         ...         ...       ...       ...</span></span>
<span><span class="co">#&gt;   [70]     DHFR       397      - |           A           -        43         0</span></span>
<span><span class="co">#&gt;   [71]     DHFR       399      - |           G           -        43         0</span></span>
<span><span class="co">#&gt;   [72]     DHFR       423      - |           T           -        42         0</span></span>
<span><span class="co">#&gt;   [73]     DHFR       430      - |           A          AT        37         1</span></span>
<span><span class="co">#&gt;   [74]     DHFR       513      - |           T          TC        35         1</span></span>
<span><span class="co">#&gt;               nA        nT        nC        nG        nN</span></span>
<span><span class="co">#&gt;        &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;</span></span>
<span><span class="co">#&gt;    [1]         0        12         0         1         0</span></span>
<span><span class="co">#&gt;    [2]         0         0        17         0         0</span></span>
<span><span class="co">#&gt;    [3]         0         0        16         0         0</span></span>
<span><span class="co">#&gt;    [4]        16         0         0         0         0</span></span>
<span><span class="co">#&gt;    [5]        15         0         0         0         0</span></span>
<span><span class="co">#&gt;    ...       ...       ...       ...       ...       ...</span></span>
<span><span class="co">#&gt;   [70]        43         0         0         0         0</span></span>
<span><span class="co">#&gt;   [71]         0         0         0        43         0</span></span>
<span><span class="co">#&gt;   [72]         0        42         0         0         0</span></span>
<span><span class="co">#&gt;   [73]        37         1         0         0         0</span></span>
<span><span class="co">#&gt;   [74]         0        35         1         0         0</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   seqinfo: 3 sequences from an unspecified genome</span></span></code></pre></div>
<p>To facilitate comparisons across groups, the pileups can be stored in a <code>RangedSummarizedExperiment</code>.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/create_se.html">create_se</a></span><span class="op">(</span><span class="va">plps</span><span class="op">)</span></span>
<span><span class="co">#&gt; class: RangedSummarizedExperiment </span></span>
<span><span class="co">#&gt; dim: 74 2 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(7): Var nRef ... nC nG</span></span>
<span><span class="co">#&gt; rownames(74): SSR3_102_2 SSR3_125_2 ... DHFR_430_2 DHFR_513_2</span></span>
<span><span class="co">#&gt; rowData names(1): Ref</span></span>
<span><span class="co">#&gt; colnames(2): sample_1 sample_2</span></span>
<span><span class="co">#&gt; colData names(1): sample</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="related-work">Related work<a class="anchor" aria-label="anchor" href="#related-work"></a>
</h2>
<p>The functionality in <code>raer</code> builds off of previously published methods and software:</p>
<ul>
<li>Python package: <a href="https://github.com/BioinfoUNIBA/REDItools" class="external-link">REDItools</a> from <a href="https://doi.org/10.1093/bioinformatics/btt287" class="external-link">Picardi E, Pesole G</a><br>
</li>
<li>Java tool: <a href="https://github.com/dieterich-lab/JACUSA2" class="external-link">JACUSA2</a> from <a href="https://doi.org/10.1186/s12859-016-1432-8" class="external-link">Piechotta M et al</a><br>
</li>
<li>Python-based pipeline: <a href="https://github.com/a2iEditing/deNovo-Detect" class="external-link">deNovo-Detect</a> from <a href="https://doi.org/10.1038/s41467-022-28841-4" class="external-link">Gabey O et al</a><br>
</li>
<li>Java-based tool: <a href="https://github.com/a2iEditing/RNAEditingIndexer" class="external-link">RNAEditingIndexer</a> from <a href="https://doi.org/10.1038/s41592-019-0610-9" class="external-link">Roth SH et al</a>
</li>
</ul>
</div>
</div>
  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/rnabioco/raer/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/rnabioco/raer/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing raer</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>
<a href="http://github.com/kriemo" class="external-link">Kent Riemondy</a> <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0003-0750-1273" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Kristen Wells-Wrasman <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0002-7466-8164" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>
<a href="https://medschool.cuanschutz.edu/rbi" class="external-link"><img src="https://rnabioco.github.io/rbitemplate/rbi-pkgdown-logo.svg" alt="RBI" width="150"></a> <br><small class="roles"> Copyright holder, funder </small>  </li>
<li><a href="authors.html">More about authors...</a></li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/rnabioco/raer/actions/workflows/check-bioc.yml" class="external-link"><img src="https://github.com/rnabioco/raer/actions/workflows/check-bioc.yml/badge.svg" alt="R-CMD-check-bioc"></a></li>
</ul>
</div>

  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>An R package developed by the <a href="https://medschool.cuanschutz.edu/rbi" class="external-link">RNA Bioscience Initiative</a></p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.4.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
