---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# raer <a href="https://rnabioco.github.io/raer"><img src="man/figures/logo.png" align="right" height="138" /></a>

<!-- badges: start -->
[![R-CMD-check-bioc](https://github.com/rnabioco/raer/actions/workflows/check-bioc.yml/badge.svg)](https://github.com/rnabioco/raer/actions/workflows/check-bioc.yml)
[![Codecov test coverage](https://codecov.io/gh/rnabioco/raer/branch/main/graph/badge.svg)](https://app.codecov.io/gh/rnabioco/raer?branch=main)
<!-- badges: end -->

raer facilitates analysis of RNA adenosine editing in the
[Bioconductor](https://bioconductor.org/) ecosystem.

ðŸš§ **raer is under active development and functionality may change** ðŸš§

## Installation

You can install the development version of raer from [GitHub](https://github.com/rnabioco/raer) with:

``` r
# install.packages("BiocManager")
BiocManager::install("rnabioco/raer")
```

## Quick start

raer provides methods to compute per site read count summaries from BAM
alignment files, either for known sites, or for all detected sites.

```{r get_pileup}
library(raer)
bamfn <- raer_example("SRR5564269_Aligned.sortedByCoord.out.md.bam")
bam2fn <- raer_example("SRR5564277_Aligned.sortedByCoord.out.md.bam")
fafn <- raer_example("human.fasta")
bedfn <- raer_example("regions.bed")

res <- get_pileup(bamfn, fafn, bedfile = bedfn)
res[1:5, ]
```

Multiple BAM files can be processed, enabling rapid comparisons of RNA-Seq vs.
WGS or WXS data, or RNA-Seq vs RNA-seq (e.g., ADAR WT VS ADAR KO).

The `FilterParam()` class holds multiple options for customizing the output of 
`get_pileup()`.

To facilitate comparisons across groups, pileups can be combined in a 
`RangedSummarizedExperiment`. 

```{r create_se}
fp <- FilterParam(
  only_keep_variants = TRUE,
  library_type = "fr-first-strand",
  min_nucleotide_depth = 2
)

plps <- get_pileup(
  c(bam2fn, bamfn),
  fafn,
  filterParam = fp
)

create_se(plps)
```

## Related work

| **Name** | **Language** | **Bulk RNA?** | **Single-cell?** | **Publication** |
|:-------------:|:-------------:|:-------------:|:-------------:|:-------------:|
| [raer](https://github.com/rnabioco/raer) | R/Bioconductor | âœ…  |  âœ…   | Riemondy Unpublished |
| [REDItools](https://github.com/BioinfoUNIBA/REDItools) | Python | âœ…  |  âŒ   | [Picardi 2013](https://academic.oup.com/bioinformatics/article/29/14/1813/232807) |
| [JACUSA2](https://github.com/dieterich-lab/JACUSA2) | Java | âœ…  |  âŒ  |  [Piechotte 2017](https://doi.org/10.1186/s12859-016-1432-8)  |
| [RNAEditingIndexer](https://github.com/a2iEditing/RNAEditingIndexer) | Java | âœ…  |  âŒ  |  [Roth 2019](https://doi.org/10.1038/s41592-019-0610-9)  |
| [deNovo-Detect](https://github.com/a2iEditing/deNovo-Detect) | Python | âœ…  |  âŒ   |  [Gabay 2022](https://doi.org/10.1038/s41467-022-28841-4)  |
