<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Changelog • raer</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/Source_Sans_Pro-0.4.8/font.css" rel="stylesheet"><link href="../deps/Source_Code_Pro-0.4.8/font.css" rel="stylesheet"><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Changelog"><meta property="og:image" content="https://rnabioco.github.io/raer/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><script defer data-domain="rnabioco.github.io" src="https://plausible.io/js/plausible.js"></script></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-dark"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">raer</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.1.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/raer.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/rnabioco/raer/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-news">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Changelog</h1>
      <small>Source: <a href="https://github.com/rnabioco/raer/blob/HEAD/NEWS.md" class="external-link"><code>NEWS.md</code></a></small>
    </div>

    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.1.0" id="raer-110">raer 1.1.0<a class="anchor" aria-label="anchor" href="#raer-110"></a></h2>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.0.0" id="raer-100">raer 1.0.0<a class="anchor" aria-label="anchor" href="#raer-100"></a></h2>
<ul><li>raer added to the Bioconductor project</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.99.12" id="raer-09912">raer 0.99.12<a class="anchor" aria-label="anchor" href="#raer-09912"></a></h2>
<ul><li>Changes made to prepare for bioc submission</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.99.11" id="raer-09911">raer 0.99.11<a class="anchor" aria-label="anchor" href="#raer-09911"></a></h2>
<ul><li>Replaced base R fisher test with c-wrapper to call htslib fisher test, which speeds up execution many fold.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.99.10" id="raer-09910">raer 0.99.10<a class="anchor" aria-label="anchor" href="#raer-09910"></a></h2>
<ul><li>The options to write to tabix indexed output files have been removed from <code><a href="../reference/pileup_sites.html">pileup_sites()</a></code> as they have limited utility and introduce unwanted code complexity.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.99.9" id="raer-0999">raer 0.99.9<a class="anchor" aria-label="anchor" href="#raer-0999"></a></h2>
<ul><li>The <code>genomic-unstranded</code> option for the <code>library-type</code> argument in <code><a href="../reference/pileup_sites.html">FilterParam()</a></code> has been renamed to <code>unstranded</code>, and the <code>unstranded</code> option has been removed.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.99.8" id="raer-0998">raer 0.99.8<a class="anchor" aria-label="anchor" href="#raer-0998"></a></h2>
<ul><li>Function arguments involving a fasta file have been renamed to all be <code>fasta</code>
</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.99.7" id="raer-0997">raer 0.99.7<a class="anchor" aria-label="anchor" href="#raer-0997"></a></h2>
<ul><li>added a single cell specific AEI calculation (<code><a href="../reference/calc_scAEI.html">calc_scAEI()</a></code>)</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.99.6" id="raer-0996">raer 0.99.6<a class="anchor" aria-label="anchor" href="#raer-0996"></a></h2>
<ul><li>added method to count base consensus base when counting UMIs with <code><a href="../reference/pileup_cells.html">pileup_cells()</a></code> using the sum of base qualities to select consensus.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.99.5" id="raer-0995">raer 0.99.5<a class="anchor" aria-label="anchor" href="#raer-0995"></a></h2>
<ul><li><p><code><a href="../reference/pileup_cells.html">pileup_cells()</a></code> now allows for multiple alleles to be queried at a site.</p></li>
<li><p>Fixed an indexing bug in <code><a href="../reference/pileup_cells.html">pileup_cells()</a></code> that misassigned sites to counts.</p></li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.99.4" id="raer-0994">raer 0.99.4<a class="anchor" aria-label="anchor" href="#raer-0994"></a></h2>
<ul><li><p><code>annot_snps</code> will now compare the site allele to the SNP allele and return a column <code>snp_matches_site</code> indicating if the site matches the snp.</p></li>
<li><p>added new function, <code><a href="../reference/find_scde_sites.html">find_scde_sites()</a></code> to identify differentially editing sites in single cell data using fishers exact tests.</p></li>
<li><p><code>pileup_cells</code> now respects the <code>min_depth</code> and <code>min_variant_reads</code> FilterParameters.</p></li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.99.3" id="raer-0993">raer 0.99.3<a class="anchor" aria-label="anchor" href="#raer-0993"></a></h2>
<ul><li>support <code>BamFile</code> and <code>BamFileList</code> inputs to <code><a href="../reference/pileup_sites.html">pileup_sites()</a></code> and <code><a href="../reference/pileup_cells.html">pileup_cells()</a></code>, which provides an option to provide custom BAI index file names.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.99.2" id="raer-0992">raer 0.99.2<a class="anchor" aria-label="anchor" href="#raer-0992"></a></h2>
<ul><li>rename <code>prep_for_de()</code> and <code>perform_de()</code> to <code><a href="../reference/make_de_object.html">make_de_object()</a></code> and <code><a href="../reference/find_de_sites.html">find_de_sites()</a></code>.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.99.1" id="raer-0991">raer 0.99.1<a class="anchor" aria-label="anchor" href="#raer-0991"></a></h2>
<ul><li><p>default values for <code>edit_from</code> and <code>edit_to</code> for <code><a href="../reference/calc_edit_frequency.html">calc_edit_frequency()</a></code> have been changed to <code>A</code> and <code>G</code> respectively.</p></li>
<li><p>renamed <code>type</code> argument in <code>perform_de</code> to <code>test</code> and removed <code>type</code> argument in <code>prep_for_de</code></p></li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.99.0" id="raer-0990">raer 0.99.0<a class="anchor" aria-label="anchor" href="#raer-0990"></a></h2>
<ul><li><p>added support for processing multiple BAM files with <code><a href="../reference/calc_AEI.html">calc_AEI()</a></code>.</p></li>
<li><p>Dropped minimally used <code>bad_reads</code> and <code>reads</code> parameters from <code><a href="../reference/pileup_sites.html">pileup_sites()</a></code></p></li>
<li><p>Added utility to screen scRNA-seq bam files for regions with oligo-dT mispriming (<code><a href="../reference/find_mispriming_sites.html">find_mispriming_sites()</a></code>).</p></li>
<li><p>add option to query ref and alt SNP alleles</p></li>
<li><p>added tests for SummarizedExperiment filtering approaches</p></li>
<li><p>added a strand bias stat <code>sor</code> using approach from GATK (StrandOddsRatio), and a confidence score <code><a href="../reference/calc_confidence.html">calc_confidence()</a></code> from SAILOR pipeline.</p></li>
<li><p>‘N’ bases in read or reference are ignored</p></li>
<li>
<p>Removed outdated or unused functionality:</p>
<ul><li>bed indexing (<code>indexBed</code> and related C code)</li>
<li>bam tag indexing (<code>build_tag_index</code>, <code>show_tag_index</code>, <code>get_tag_bam</code>, )</li>
<li>bam tag index based single cell approach (<code>sc_editing</code>)</li>
<li>bam tag indexing C code from bri (<code>src/bri/*</code>)</li>
<li>sparse matrix merging for <code>merge_pileups()</code>.</li>
<li>unneeded utilities (<code>filter_by_coverage</code>)</li>
<li>Remaining (and mostly unused) Rcpp code</li>
<li>Removed fastmap, Rcpp, zlibbioc, RColorBrewer, and BiocGenerics dependencies</li>
<li>Removed system requirements for C libraries used by bri</li>
</ul></li>
<li><p>The bed indexing used in <code><a href="../reference/pileup_sites.html">pileup_sites()</a></code> has been replaced with the region indexing approach from <code><a href="../reference/pileup_cells.html">pileup_cells()</a></code>.</p></li>
<li><p><code><a href="../reference/pileup_sites.html">pileup_sites()</a></code> now requires a GRanges object rather than a bed file. The <code>bedfile</code>parameter has been removed and replaced with a <code>sites</code> parameter.</p></li>
<li><p>Renamed <code>Ref</code> and <code>Var</code> output columns to <code>REF</code> and <code>ALT</code> and <code>nVar</code> was renamed to <code>nAlt</code>. This provides consistency with VCF format and consistency across <code><a href="../reference/pileup_cells.html">pileup_cells()</a></code> and <code><a href="../reference/pileup_sites.html">pileup_sites()</a></code> function calls</p></li>
<li><p><code><a href="../reference/pileup_cells.html">pileup_cells()</a></code> gained functionality to process multiple smart-seq2 style bam files.</p></li>
<li><p>Changed <code>filterParam</code> argument in <code>pileup_sites</code> and <code>pileup_cells</code> to <code>param</code> for simplicity.</p></li>
<li><p>Added <code>FilterParam</code> to exclude multi-allelic sites <code>report_multiallelic</code>, or exclude reporting a variant in the Var assay based on allelic frequency (<code>min_allelic_freq</code>).</p></li>
<li><p>The <code>bam_flags</code> parameter used in <code>pileup_sites</code> and <code>pileup_cells</code> has been moved into the <code>FilterParam</code> class.</p></li>
<li><p>The <code>bedindex</code> parameter for <code>pileup_sites</code> has been removed. This option is not needed at the user level and is planned to be replaced by the regional indexing used in <code><a href="../reference/pileup_cells.html">pileup_cells()</a></code>.</p></li>
<li><p>Added <code>FilterParam</code> option to trim reads based on fractional distance from 5’ (<code>ftrim_5p</code>) or 3’ end (<code>ftrim_3p</code>).</p></li>
<li><p>Incorporated RBPZ and VDB statistics from bcftools, now returned as rowData columns when calling <code>pileup_sites</code>.</p></li>
<li><p>A <code>RangedSummarizedExperiment</code> object is now directly returned from <code>pileup_sites</code>. Using <code>merge_pileups</code> is no longer necessary and is not an exported function.</p></li>
<li><p>Renamed <code>get_pileup</code> to <code>pileup_sites</code> and <code>create_se</code> to <code>merge_pileups</code></p></li>
<li><p>Rename <code>remove_clustered_variants</code>, <code>remove_multiallelic</code>, and <code>remove_splice_variants</code> to <code>filter_*</code> for consistency.</p></li>
<li><p>Rewrote and renamed the single cell editing function <code>sc_editing</code> to <code><a href="../reference/pileup_cells.html">pileup_cells()</a></code>. <code><a href="../reference/pileup_cells.html">pileup_cells()</a></code> does not require sorting and index by cell barcode, uses a new format to specify sites to query and requires providing the reference and alternate alleles of interest, writes to disk in a sparse matrix compatible format to reduce memory usage, and should have more performance as there is no need to query a fasta index.</p></li>
<li><p>Implemented method to collapse reads with duplicate UMIs.</p></li>
<li><p>Added option to filter sites in pileup based on number of reads containing a variant (<a href="https://github.com/rnabioco/raer/issues/54" class="external-link">#54</a>)</p></li>
<li><p>Added a <code>NEWS.md</code> file to track changes to the package.</p></li>
</ul></div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>An R package developed by the <a href="https://medschool.cuanschutz.edu/rbi" class="external-link">RNA Bioscience Initiative</a></p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

