---
title: "RNA editing during viral infection"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{chikv-editing}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
editor_options: 
  markdown: 
    wrap: 72
  chunk_output_type: console
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

## Overview

This is an analysis of RNA editing during a Chikungunya (CHIKV) infection of
mouse lymph node tissue, described in:

Carpentier KS, Sheridan RM, Lucas CJ, Davenport BJ, Li FS, Lucas ED, McCarthy
MK, Reynoso GV, May NA, Tamburini BAJ, Hesselberth JR, Hickman HD, Morrison TE.
MARCO+ lymphatic endothelial cells sequester arthritogenic alphaviruses to limit
viremia and viral dissemination. EMBO J. 2021 Nov 15;40(22):e108966. doi:
10.15252/embj.2021108966. Epub 2021 Oct 7. [PMID: 34618370; PMCID:
PMC8591538](https://pubmed.ncbi.nlm.nih.gov/34618370/).

In this study, mice were infected with an attenuated strain of CHIKV and lymph
nodes were harvested after 24 hours and analyzed by 10x Genomics 3' single-cell
RNA sequencing, focusing on CD45- stromal cells.

Data from triplicate infection and mock samples were analyzed by the CellRanger
pipeline, and output BAM files serve as the input data below.

### Sample information

* Todo: Create table of samples, 10x summary

```{r libs, message=FALSE}
library(raer)
library(tidyverse)
library(ggtext)
library(cowplot)
library(fs)
library(glue)
library(qs)
library(janitor)
library(cli)
library(Seurat)

library(plyranges)
library(SummarizedExperiment)
library(SingleCellExperiment)
library(VariantAnnotation)
library(TxDb.Mmusculus.UCSC.mm10.ensGene)
library(BiocParallel)
```

```{r setup_common_data}
ext_path <- fs::path('~/projects/raer-chikv')
samples <- c('A1','A2','A3', 'M1', 'M2', 'M3')

fns = list(
  bams    = fs::path(ext_path, 'bam', str_c( samples, '.bam')),
  bais    = fs::path(ext_path, 'bam', str_c(samples, '.bam.bai')),
  fasta   = fs::path(ext_path, 'ref', 'mm10_CHIKV_AF15561.fa.gz'),
  rmsk_db = fs::path(ext_path, 'ref', 'rmsk.mm10.tsv.gz'),
  snp_db  = fs::path(ext_path, 'ref', 'snp142.mm10.bed.gz'),
  redi_db = fs::path(ext_path, "ref", 'rediportal.mm10.cleaned.tsv.gz')
)

names(fns$bams) <- samples
```

### External data sources

* `mm10_CHIKV_AF15561.fa.gz` is a combined FASTA of mouse mm10 (XXX: source?) and the
  CHIKV genome. It's the reference that was used during the upstream CellRanger step.

* `rmsk.mm10.tsv.gz` is the RepeatMasker table downloaded from UCSC.

* `snp142.mm10.bed.gz` is a SNP table downloaded from UCSC.

* `rediportal.small.mm10.tsv.gz` are sites from the [REDIportal database](http://srv00.recas.ba.infn.it/atlas/index.html). The full
  table is poorly formatted, having 25 column headers and 22 data columns, so
  these data are the first 22 columns, and anything after the first ~10 may
  not be useful.
  
## Pileup and filtering

This is the main raer pileup routine. On this data set, this step takes about 4
hours (parallelized) and generates a large `RangedSummarizedExperiment` (`rse`)
with \~48 million sites.

```{r run_pileup, eval=FALSE}
fp <- FilterParam(
  min_nucleotide_depth = 3,
  min_base_quality     = 30,
  min_mapq             = 255,
  library_type         = "fr-second-strand",
  trim_5p              = 5,
  trim_3p              = 5,
  indel_dist           = 4,
  homopolymer_len      = 6,
  max_mismatch_type    = c(3, 3),
  min_read_bqual       = c(0.25, 20),
  only_keep_variants   = TRUE
)

bpp <- MulticoreParam(
  workers = 24,
  progressbar = TRUE
)

all_rse <- pileup_sites(
  fns$bams,
  fafile = fns$fasta,
  filterParam = fp,
  BPPARAM = bpp
)

# remove multi-allelics now, they're not much use later.
# this step removes about 15% (~8 million) of the sites.
all_rse <- filter_multiallelic(all_rse)

chikv_rse <- subsetByOverlaps(
  all_rse,
  filter(rowRanges(all_rse), seqnames == 'CHIKV_AF15561')
)

mm_rse <- subsetByOverlaps(
  all_rse,
  filter(rowRanges(all_rse), seqnames != 'CHIKV_AF15561')
)

qs::qsave(all_rse, fs::path(ext_path, '/rds/all_rse.rds'), nthreads = 10)
qs::qsave(mm_rse, fs::path(ext_path, '/rds/mm_rse.rds'), nthreads = 10)
qs::qsave(chikv_rse, fs::path(ext_path, '/rds/chikv_rse.rds'), nthreads = 10)
```

```{r load_rds, eval = FALSE}
all_rse <- qs::qread(fs::path(ext_path, '/rds/all_rse.rds'), nthreads = 10)
mm_rse <- qs::qread(fs::path(ext_path, '/rds/mm_rse.rds'), nthreads = 10)
chikv_rse <- qs::qread(fs::path(ext_path, '/rds/chikv_rse.rds'), nthreads = 10)
```

### Filtering for high confidence sites

From the intial set of \~50 million sites from the pileup, we filter
based on several criteria:

1.  filter on `nVar`, requiring at least 3 reads for every sample among
    replicates

2.  remove clustered variants and variants near splice sites, as these
    tend to be lower confidence due to mapping artifacts

3.  after `calc_edit_frequency()`, filter on `depth` and `edit_freq`,
    requiring a minimum of 3 reads and an editing frequency of at least
    5%.

4.  annotate the remaining variants based on overlap with reference
    databases of repeats and SNPs, and remove variants based on ad hoc
    criteria

### Filtering sites by treshold

This is a helper function to filter samples within each replicate.

```{r filter_fxn}
filter_replicates <- function(rse, assay_name = NULL, thresh = NULL, n_rep = 3) {
 
  x <- assay(rse, assay_name)[, 1:6]
  
  mocks <- c('M1', 'M2', 'M3')
  virus <- c('A1', 'A2', 'A3')

  keep_mocks <- rowSums(x[, mocks] >= thresh) >= n_rep
  keep_virus <- rowSums(x[, virus] >= thresh) >= n_rep 

  # keep anything that passes the filter in either mock or virus samples
  keep_all <- keep_mocks | keep_virus
  
  rse[keep_all, ]
}
```

The order of filtering here is important: the first `nVar` filter
removes a large chunk of the data, so we only compute the next steps on
\~10% of the sites.

```{r filter_mm, eval = FALSE}
# filter for at least 3 reads at each site every sample in the replicates.
# this removes a large chunk of the data that is low coverage, ~90% or so
mm_rse_filt <- filter_replicates(mm_rse, "nVar", thresh = 3)

# variant position bias (vpb, from 0 -> 1, 1 is less bias) 
# filter on vpb, 0.015 or more
mm_rse_filt <- subsetByOverlaps(
  mm_rse_filt,
  filter(rowRanges(mm_rse_filt), vpb >= 0.015)
)

# Next, remove sites based on gene annotations
txdb <- TxDb.Mmusculus.UCSC.mm10.ensGene

mm_rse_filt <- filter_clustered_variants(
  mm_rse_filt, txdb, variant_dist = 100
  ) |>
  filter_splice_variants(txdb)

# calculate `edit_freq` and `depth`
mm_rse_filt <- calc_edit_frequency(mm_rse_filt)

# filter on site depth >= 3 for all samples within replicates (nVar + nRef)
mm_rse_filt <- filter_replicates(mm_rse_filt, "depth", thresh = 6)

# filter for edit_freq >= 0.05 for at least 3 samples
mm_rse_filt <- filter_replicates(mm_rse_filt, "edit_freq", thresh = 0.05)

mm_rse_filt
```

Now do the same for viral data --- these may be a bit harsh for the
virus coverage we have.

```{r filter_chikv, eval = FALSE}
chikv_rse_filt <- filter_replicates(chikv_rse, "nVar", thresh = 3)

chikv_rse_filt <- subsetByOverlaps(
  mm_rse_filt,
  filter(rowRanges(chikv_rse_filt), vpb >= 0.015)
)

# calculate edit frequencies and depth
chikv_rse_filt <- calc_edit_frequency(chikv_rse_filt)

# filter on site depth >= 3 for all samples within replicates 
chikv_rse_filt <- filter_replicates(chikv_rse_filt, "depth", thresh = 3)

# filter for edit_freq >= 0.05 for all replicates
chikv_rse_filt <- filter_replicates(chikv_rse_filt, "edit_freq", thresh = 0.05)

chikv_rse_filt
```

### Filtering by annotation

Filter the mouse data by annotation, comparing to reference SNP and
repeat databases.

```{r annotate_rse, eval = FALSE}
# load external annotations

rmsk_db <-  read_tsv(
  fns$rmsk_db,
  col_names = c('seqnames', 'start', 'end', 'strand', 'repName', 'repClass', 'repFamily')
) |>
  plyranges::as_granges()

snp_db  <- read_tsv(
  fns$snp_db,
  col_names = c('seqnames', 'start', 'end', 'snpId', 'score', 'strand')
) |>
  plyranges::as_granges()

# annotate the filtered sites
mm_rse_filt <- annot_from_gr(
  mm_rse_filt,
  rmsk_db,
  cols_to_map = c("repName", "repFamily")
)

mm_rse_filt <- annot_from_gr(
  mm_rse_filt,
  snp_db,
  cols_to_map = "snpId"
)

# filter out specific repeat classes
mm_rse_filt <- mm_rse_filt[
  !rowData(mm_rse_filt)$repFamily %in%
    c("Simple_repeat", "Low_complexity", "Low_complexity,Low_complexity")
]

# filter out common variants (i.e., already have a snpId)
mm_rse_filt <- mm_rse_filt[is.na(rowData(mm_rse_filt)$snpId)] 

mm_rse_filt
```

### Save filtered data

```{r save_filt_data, eval=FALSE}
qs::qsave(mm_rse_filt, fs::path(ext_path, '/rds/mm_rse_filt.rds'), nthreads = 10)
qs::qsave(chikv_rse_filt, fs::path(ext_path, '/rds/chikv_rse_filt.rds'), nthreads = 10)
# mm_rse_filt <- qs::qread(fs::path(ext_path, '/rds/mm_rse_filt.rds'), nthreads = 10)
```

## Analysis & annotation

### REDIportal annotation

```{r rediportal_annotation}
redi_tbl <- read_tsv(
  fns$redi_db,
  col_names = TRUE
) |>
  janitor::clean_names()

redi_gr <- 
  setNames(
    redi_tbl,
    names(redi_tbl) |>
      str_replace('_wg_encode_gencode_basic_vm16', '_gencode') |>
      str_replace('_ref_gene', '_refgene') |>
      str_replace('repeat', 'repeat_class')
  ) |>
  dplyr::rename(
    seqnames = region,
    start = position
  ) |>
  mutate(
    end = start + 1
  ) |>
  relocate(end, .after = 2) |>
  relocate(strand, .after = 3) |>
  # we'll only keep gencode annotations, through `exonic_func`
  dplyr::select(seqnames:exonic_func_gencode) |>
  rename_with(~ str_c('redi_', .x), .cols = ref:exonic_func_gencode) |>
  plyranges::as_granges()

mm_rse_filt <- annot_from_gr(
  mm_rse_filt,
  redi_gr,
  cols_to_map = c(
    'redi_ref',
    'redi_ed',
    'redi_func_gencode',
    'redi_gene_gencode'
  )
)

```

### Variants in genic regions

```{r annotate_variants, eval=FALSE}
txdb <- TxDb.Mmusculus.UCSC.mm10.ensGene
common <- intersect(seqlevels(txdb), seqlevels(rowRanges(mm_rse_filt)))
txdb_common <- keepSeqlevels(txdb, common)
          
txdb_annot_gr <- VariantAnnotation::locateVariants(
  rowRanges(mm_rse_filt),
  txdb_common,
  AllVariants()
)

mm_rse_filt <- annot_from_gr(
  mm_rse_filt,
  txdb_annot_gr,
  cols_to_map = c(
    'LOCATION', 'GENEID'
  )  
)
```

### Differential editing

```{r diff_ed}
# prepare the data
colData(mm_rse_filt)$sample <- c(
  str_c('virus-', c(1,2,3)),
  str_c('mock-', c(1,2,3))
)

colData(mm_rse_filt)$condition <- c(rep('virus', 3), rep('mock', 3))
colData(mm_rse_filt)

se_filt <- calc_edit_frequency(
  mm_rse_filt,
  edit_from = "A",
  edit_to = "G",
  drop = TRUE
)

dse <- prep_for_de(se_filt)

# run the diffferential editing. edgeR yields 0 sites, DESeq2 about 250.
# not sure what the diff is.
de_res <- perform_de(
  dse,
  condition_control = "mock",
  condition_treatment = "virus",
  type = 'DESeq2'
)

sig_de_res <- 
  rownames_to_column(de_res$sig_results, "site") |>
  as_tibble() |>
  dplyr::rename(lfc = log2FoldChange) |>
  janitor::clean_names() |>
  rename_with(~str_c("de_", .), .cols = -site) |>
  arrange(de_padj) |>
  mutate(new_site = site) |>
  separate(
    new_site,
    into = c('seqnames', 'start', 'strand'),
    sep = '_',
    convert = TRUE) |>
  mutate(end = start)

sig_de_gr <- as_granges(sig_de_res)

mm_rse_filt <- annot_from_gr(
  mm_rse_filt,
  sig_de_gr,
  cols_to_map = c(
    'de_lfc', 'de_padj'
  )
)
```

### Save annotated data
```{r, eval=FALSE}
qs::qsave(mm_rse_filt, fs::path(ext_path, "/rds/mm_rse_filt.rds"))
```

## Data tidying

```{r tidy_data, eval=FALSE}

rse_assays <- c('nRef', 'nVar', 'Var', 'depth', 'edit_freq')

pull_assay <- function(rse, assay_name) {
  assay(rse, assay_name) |>
    as.data.frame() |>
    rownames_to_column("site") |>
    as_tibble() |>
    pivot_longer(-site) |>
    dplyr::rename(!! assay_name := value)
}

tidy_assays_tbl <- 
  purrr::map(rse_assays, ~ pull_assay(mm_rse_filt, .x)) |>
  purrr::reduce(left_join, by = c('site', 'name'))

tidy_annots_tbl <-
  rowData(mm_rse_filt) |>
  as.data.frame() |>
  rownames_to_column("site") |>
  as_tibble() |>
  dplyr::select(-Ref, -Var, -snpId)
  
tidy_all_tbl <-
  left_join(tidy_assays_tbl, tidy_annots_tbl, by = c('site'))

tidy_mm_tbl <-
  janitor::clean_names(tidy_all_tbl) |>
  # filter out ambiguous var
  filter(var != '-' & !str_detect(var, '^N')) |>
  # polish columns in variables
  separate(name, into = c('treat', 'rep'), sep = 1) |>
  # use intuitive var, i.e. `AT` -> `A>T`
  separate(var, into = c('ref', 'alt'), sep = 1) |>
  unite(vnt, ref:alt, sep = '>') |>
  # use intuitive treatment names
  mutate(
    treat = case_when(
      treat == 'A' ~ 'virus',
      treat == 'M' ~ 'mock'
    )
  ) |>
  separate(site, into = c('chrom', 'pos', 'strand'), sep = '_') |>
  arrange(vnt) |>
  dplyr::select(-rbpz, -vpb) |>
  dplyr::rename(
    n_vnt = n_var,
    ens_gene_id = geneid,
    ens_location = location
    ) |>
  mutate(across(c(treat, rep, vnt, rep_family), as_factor)) |>
  mutate(across(c(de_lfc, de_padj), as.numeric)) |>
  mutate(across(c(pos), as.integer)) |>
  mutate(
    across(
      c(redi_gene_gencode, redi_func_gencode, ens_gene_id, ens_location, ),
      ~ str_split_i(.x, ',', 1)
    )
  )

qs::qsave(tidy_mm_tbl, file = fs::path(ext_path, '/rds/tidy_mm_tbl.qs'), nthreads = 8)    
```

## Exploratory analysis

### Load data 

```{r load_rds}
mm_rse_filt <- qs::qread(fs::path(ext_path, '/rds/mm_rse_filt.rds'), nthreads = 10)
tidy_mm_tbl <- qs::qread(fs::path(ext_path, '/rds/tidy_mm_tbl.qs'), nthreads = 10)    
```

### Variants identified by the pileup

```{r plot_counts}
vnt_counts <-
  tidy_mm_tbl |>
  group_by(treat, rep) |>
  dplyr::count(vnt) |>
  arrange(vnt) |>
  mutate(
    treat = fct_relevel(treat, c('mock', 'virus'))
  ) |> 
  ungroup()

ggplot(
  # add a name column that highlights `A>G` in bold, red text
  mutate(
    vnt_counts,
    plt_name = case_when(
      vnt == 'A>G' ~ "<b style='color:red'>A>G</b>"
   ),
   plt_name = coalesce(plt_name, vnt),
   plt_name = fct_relevel(
    as_factor(plt_name),
    "<b style='color:red'>A>G</b>",
    after = 1
    )
  ),
  aes(x = plt_name, y = n, fill = rep)
  ) +
  geom_bar(stat = 'identity', position = 'dodge') +
  facet_grid(~ treat) +
  scale_fill_manual(values = c('grey70', 'grey50', 'grey30')) +
  labs(
    title = 'Variants from raer pileup of Mm / CHIKV scRNA-seq',
    x = '',
    y = '',
    fill = 'replicate'
  ) + 
  theme_cowplot() +
  theme(
    axis.text.x = element_markdown(angle = 45, hjust = 1),
    legend.position='top', 
    legend.justification='left',
    legend.direction='horizontal',
    strip.background = element_blank(),
    strip.text = element_textbox(
      size = 14,
      color = "black", fill = "grey90", box.color = "grey90",
      halign = 0.5, linetype = 1, r = unit(5, "pt"), width = unit(1, "npc"),
      padding = margin(2, 0, 1, 0), margin = margin(3, 3, 3, 3)
    )
  )
```

The plot of counts above shows us a few things:

* There are ~4,000 `A>G` sites per replicate.
    
  ```{r}
  filter(vnt_counts, vnt == 'A>G') 
  ```
    
* There's a smaller signal for `T>C`, which comes from second-strand
  synthesis issues for abundant transcripts. The `X>A` variants are a mis-priming
  artifact that is largely mitigated by the earlier `vbpz` filtering step.

### Inspect editing frequency and depth

```{r plot_depth_edit_freq}
depth_edit_tbl <-
  tidy_mm_tbl |>
    mutate(log_depth = log10(depth)) |>
    dplyr::select(
      treat, rep, vnt, log_depth, edit_freq,
      
      ) |>
    pivot_longer(edit_freq:log_depth) |>
    arrange(vnt) |>
    mutate(
      treat = fct_relevel(treat, c('mock', 'virus'))
    )
  
ggplot(
  # add a name column that highlights `A>G` in bold, red text
  mutate(
    depth_edit_tbl,
    plt_name = case_when(
      vnt == 'A>G' ~ "<b style='color:red'>A>G</b>"
   ),
   plt_name = coalesce(plt_name, vnt),
   plt_name = fct_relevel(
    as_factor(plt_name),
    "<b style='color:red'>A>G</b>",
    after = 1
    )
  ),
  aes(x = plt_name, value, fill = rep)
  ) +
  geom_boxplot(outlier.shape = NA) +
  facet_grid(name ~ treat, scales = 'free_y') +
  theme_cowplot() +
  scale_fill_manual(values = c('grey70', 'grey50', 'grey30')) +
  theme(
    axis.text.x = element_markdown(angle = 45, hjust = 1)
  ) + 
  labs(
    x = ''
  )
  
```

- `A>G` has one of the *lowest* editing frequency.
  
- All variants have have apprxoimately the same `log_depth` .

### Differential editing

#### Heatmap of edit frequencies

```{r de_site_heatmap}
heatmap_tbl <- 
  filter(tidy_mm_tbl, vnt == 'A>G', !is.na(de_padj)) |>
  dplyr::select(chrom, pos, treat, rep, edit_freq) |>
  arrange(chrom, pos) |>
  unite(smpl, c(treat, rep)) |>
  pivot_wider(names_from = smpl, values_from = edit_freq, values_fill = 0) |>
  unite(rowname, c(chrom, pos)) 

heatmap_mtx <- as.matrix(dplyr::select(heatmap_tbl, -rowname))
rownames(heatmap_mtx) <- heatmap_tbl$rowname
# colnames(heatmap_mtx) <- str_split_i(colnames(heatmap_mtx), '_', 2)
heatmap_mtx <- t(heatmap_mtx)

nc <- ncol(heatmap_mtx)

hm <- ComplexHeatmap::Heatmap(
  heatmap_mtx,
  column_title = glue('Differentially edited A>G sites (n = {nc}; FDR = 0.01)'),
  row_labels = rownames(heatmap_mtx),
  col = viridis::viridis(10),
  cluster_rows = FALSE,
  cluster_columns = TRUE,
  show_column_dend = FALSE,
  show_column_names = FALSE,
  show_row_names = TRUE,
  show_row_dend = FALSE,
  row_names_side = "left",
  # row_split = 2,
  row_title = NULL,
  # use_raster = TRUE,
  heatmap_legend_param = list(
    title = "editing frequency (nVar / nRef)",
    direction = "horizontal",
    title_position = "lefttop"
  )
)

ComplexHeatmap::draw(hm, heatmap_legend_side = "bottom")
```

#### Features associated with DE sites

```{r de_site_genes}
de_genes_tbl <-
  filter(tidy_mm_tbl, vnt == 'A>G', !is.na(de_padj)) |>
  dplyr::select(
    chrom, pos, starts_with('ens_'),
    redi_gene_gencode, rep_family
    ) |>
  unique()

DT::datatable(de_genes_tbl)
```

### Cross-referencing with REDIportal

```{r rediportal_xref, message=FALSE}

# tidy code for analysis
tidy_redi_tbl <- 
  rowData(mm_rse_filt) |>
    as.data.frame() |>
    rownames_to_column('site') |>
    as_tibble() |>
    janitor::clean_names() |>
    pivot_longer(-c(site, ref, var, rbpz, vpb)) |>
    filter(!is.na(value)) |> 
    separate(site, into = c('chrom', 'pos', 'strand'), sep = '_') |>
    dplyr::select(-ref, -strand, -rbpz, -vpb) |>
    separate(var, into = c('ref','ed'), sep = 1) |>
    unite(var, ref:ed, sep = '>') |>
    pivot_wider() |>
    filter(!is.na(redi_ref)) |>
    dplyr::select(-redi_ref, -redi_ed) |>
    rowwise() |>
    mutate(
      redi_func_gencode = str_split_1(redi_func_gencode, ',')[[1]],
      redi_gene_gencode = str_split_1(redi_gene_gencode, ',')[[1]]
    ) |>
    ungroup() |>
    dplyr::mutate(
      rep_family_collapse = case_when(
        rep_family == 'Alu' ~ 'Alu',
        rep_family == 'L1' ~ 'L1',
        rep_family == 'B2' ~ 'B2',
        rep_family == 'B4' ~ 'B4',
        str_detect(rep_family, '^ERV') ~ 'ERV family',
        is.na(rep_family) ~ 'Not in repeat',
        TRUE ~ 'Other repeat'
      )
    )
```

```{r redi_plots}
total_sites <- nrow(tidy_redi_tbl)
n_rep_family <- length(unique(tidy_redi_tbl$rep_family_collapse))

# functional annotations
func_counts <- tidy_redi_tbl |>
  dplyr::count(redi_func_gencode) |>
  dplyr::rename(func_n = n)

tidy_redi_tbl |>
  dplyr::count(
    redi_func_gencode,
    rep_family_collapse,
    sort = TRUE) |>
  left_join(func_counts) |>
  ggplot(
    aes(
      x = fct_reorder(redi_func_gencode, -func_n),
      y = n,
      fill = fct_relevel(rep_family_collapse, 'Not in repeat', after = Inf)
    )
  ) +
  geom_col() +
  theme_cowplot() +
  labs(
    x = '',
    y = '',
    title = 'REDIportal annotation of mouse A>G sites',
    subtitle = glue('n = {scales::comma(total_sites)} total sites'),
    fill = 'Repeat class'
  ) +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1),
    legend.position='top', 
    legend.justification='left',
    legend.direction='horizontal'
  ) +
  scale_fill_manual(
    values = c(
      scales::brewer_pal(palette = "Dark2")(n_rep_family - 1),
      'grey40'
    )
  )
```


## Single-cell analysis

```{r read-so, eval=FALSE}
all_so <- read_rds(fs::path(ext_path, 'so/so.rds'))
all_sce <- Seurat::as.SingleCellExperiment(all_so)
# qs::qsave(all_sce, fs::path(ext_path, '/sce/all_sce.rds'), nthreads = 10)
```

```{r single-cell, eval = FALSE}
plotUMAP(all_sce, colour_by = "cell_type")

query_sites <- filter(rowRanges(mm_rse_filt), Var == 'AG')
query_sites$ref <- 'A'
query_sites$alt <- 'G'

fp <- FilterParam(
  min_base_quality     = 30,
  library_type         = "fr-second-strand",
  trim_5p              = 5,
  trim_3p              = 5,
  min_variant_reads    = 1
)

bpp <- MulticoreParam(
  workers = 24,
  progressbar = TRUE
)

sce_samples <- c('M1', 'M2', 'M3', 'A1', 'A2', 'A3')
outs <- list()

for (i in seq_along(sce_samples)) {
  
  cli::cli_alert_info("processing sample: {sce_samples[[i]]}")
  
  sce <- all_sce[, grepl(paste0('_', i, '$'), colnames(all_sce))]
  colnames(sce) <- str_split_i(colnames(sce), paste0('_', i), i = 1)
  cbs <- colnames(sce)
  
  e_sce <- pileup_cells(
    bamfile = fs::path(ext_path, 'bam', paste0(sce_samples[[i]], '.bam')), 
    sites = query_sites,
    cell_barcodes = cbs,
    output_directory = fs::file_temp(pattern = 'sc-edits', ext = i),
    cb_tag = "CB",
    umi_tag = "UB",
    BPPARAM = bpp
  )
 
  colnames(e_sce) <- paste0(colnames(e_sce), '_', i) 
  #names(assays(e_sce)) <- str_c(names(assays(e_sce)), paste0('_', i))
  
  outs[[i]] <- e_sce
}

# qs::qsave(outs, file = fs::path(ext_path, '/rds/sce_outs.qs'))
sce_outs <- qs::qread(file = fs::path(ext_path, '/rds/sce_outs.qs'))
```



## TODO

- [ ] Count number of A>G in our data vs rediportal (common and unique to either)
