---
title: Identifying RNA editing sites using RNA and DNA sequencing data
author: 
  - name: Kent Riemondy
    affiliation: University of Colorado School of Medicine
date: '`r Sys.Date()`'
output:
  BiocStyle::html_document
package: raer 
vignette: >
  %\VignetteIndexEntry{Find-Novel-Sites}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}    
  %\VignetteDepends{TxDb.Hsapiens.UCSC.hg38.knownGene}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


```{r}
library(raer)
library(GenomicFeatures)
library(SummarizedExperiment)
library(rtracklayer)
library(TxDb.Hsapiens.UCSC.hg38.knownGene)
```

# Novel RNA editing site detection tutorial

This vignette will demonstrate how to identify novel RNA editing sites using the `raer` package.

In this vignette a public RNA-seq and Whole-Genome sequencing dataset will be analyzed. High coverage whole-genome sequencing was conducted [ERR262997](https://www.ebi.ac.uk/ena/browser/view/ERR262997?show=reads) along with paired-end RNA-seq [SRR1258218](https://www.ebi.ac.uk/ena/browser/view/SRR1258218?show=reads). 

Aligned BAM files and a genome fasta file have been preprocessed for this vignette. These files occupy ~100 Mb of space and contain alignments from the first 1MB of chromosome 4. 

```{r}
data_dir <- tempdir()
fns <- download_NA12878(data_dir)
fns
```

Additionally we will use the following additional annotation resources:  
  
  - A database of known SNPs, for example the `SNPlocs.Hsapiens.dbSNP155.GRCh38` package. Due to space and memory constraints , in this vignette we will use a BED file containing SNPs from the first 1Mb region of chr4.  
  - `TxDb.Hsapiens.UCSC.hg38.knownGene`, a database of transcript models. Alternatively these can be generated from a `.gtf` file using `makeTxDbFromGRanges()` from the `GenomicFeatures` package.  
  - RepeatMasker annotations, which can be obtained from the `AnnotationHub()` for hg38. For space reasons these annotations for chr4 have been included as a text file. 

```{r}
txdb <- TxDb.Hsapiens.UCSC.hg38.knownGene

chr4snps <- import(fns$snps)

rmskhg38 <- read.table(fns$rmsk)
rmskhg38 <- makeGRangesFromDataFrame(rmskhg38,
                                     keep.extra.columns = TRUE, 
                                     starts.in.df.are.0based = TRUE)
```


The `get_pileup()` function can accept multiple bam files, here we supply one from RNA-seq, and one from whole genome sequencing.

```{r}
wgs_bam <- fns$bams[1]
rna_bam <- fns$bams[2]
fafn <- fns$fasta
```

Filtering parameters for the `get_pileup()` function can accept multiple arguments matched to the input bam files.

```{r}
fp <- FilterParam(min_nucleotide_depth = 1,
                  min_base_quality = 30,
                  min_mapq = c(255, 30),
                  library_type = c("fr-first-strand", "genomic-unstranded"),
                  trim_5p = 5,
                  trim_3p = 5,
                  indel_dist = 4,
                  homopolymer_len = 6,
                  max_mismatch_type = c(3, 3),
                  only_keep_variants = c(TRUE, FALSE))

res <- get_pileup(c(rna_bam, wgs_bam),
                  fafile = fafn,
                  chrom = "chr4",
                  filterParam = fp)
names(res) <- c("rna", "dna")
res
```



```{r}
se <- create_se(res)
se
```

Next we filter to keep those sites with a variant in the RNA-seq, but no variant
in the DNA-seq, and a minimum of 5 reads in the DNA-seq. The DNA-seq data will 
be reported on the "+" strand, where as the RNA-seq data will be reported on either 
strand. We therefore use `subsetByOverlaps(..., ignore.strand = TRUE)` to retain sites
passing these DNA-seq based filters independent of strand. 

```{r}
to_keep <- (assay(se, "nRef")[, "dna"] >= 5 & assay(se, "Var")[, "dna"] == "-")
se <- subsetByOverlaps(se, se[to_keep,], ignore.strand = TRUE)
nrow(se)
```

Next we filter to remove any multi-allelic sites. These sites are stored as comma-separated
strings in the `Var` assay (e.g. `AG,AC`). Non-variant sites are stored as `-`. 
`remove_multiallelic()` will remove any sites that have multiple variants in the samples present in the `summarizedExperiment` object. It will add a new column to the `rowData()` to indicate the variant for each site, and will calculate an `allele_freq` assay with variant allele frequencies for each sample. 

```{r}
se <- remove_multiallelic(se)
se <- calc_edit_frequency(se)
rowData(se)
```


Next we'll remove sites in simple repeat regions. We will add repeat information to the `rowData()` using the `annot_from_gr()` function.

```{r}
se <- annot_from_gr(se, rmskhg38, cols_to_map = c(c("repName", "repClass", "repFamily")))

rowData(se)[c("repName", "repFamily")]
```


```{r}
se <- se[!rowData(se)$repFamily  %in% c("Simple_repeat", "Low Complexity")]
```

Next we'll remove sites adjacent to other sites with different variant types. For example if an A->G variant is located proximal to a C->T variant then the variants will be removed.  

```{r}
se <- remove_clustered_variants(se, txdb, variant_dist = 100)
se
```

Next, sites with 4bp of known splicing events will be excluded. These variants are low-confidence due to the possibility of mis-alignments. 

```{r}
se <- remove_splice_variants(se, txdb)
se
```
Next, we'll annotate if the site is a known SNP and remove any known SNPs. If using a SNPlocs package you can use the `annot_snp()` function. However we will use the `annot_from_gr()` function to annotate using the prebuilt `chr4snps` object.

```{r}
se <- annot_from_gr(se, chr4snps, "name")
rowData(se)[c("name")]

se <- se[is.na(rowData(se)$name), ]
se
```
Lastly, we'll further filter the edit sites to require that the editing frequency is > 0.05 and that at least 2 reads support the editing site. 

```{r}
to_keep <- assay(se, "edit_freq")[, 1] > 0.05
se <- se[to_keep, ]

se <- se[assay(se, "nVar")[, 1] >= 2]
```


```{r}
rowRanges(se)
```


```{r}
unlink(data_dir)
```


<details style="margin-bottom:10px;">
<summary>
    Session info
</summary>

```{r}
sessionInfo()
```

</details>
