---
title: Introducing the raer package
author: 
  - name: Kent Riemondy
    affiliation: University of Colorado School of Medicine
  - name: Kristen Wells-Wrasman
    affiliation: University of Colorado School of Medicine
date: '`r Sys.Date()`'
output:
  BiocStyle::html_document
package: raer 
vignette: >
  %\VignetteIndexEntry{Introduction}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}    
---

```{r, echo=FALSE, results="hide"}
knitr::opts_chunk$set(message = FALSE, warning = FALSE)
```

# Differential Editing tutorial

This vignette will demonstrate how to identify RNA editing sites with condition-specific editing frequencies using the `raer` package.

## Set up

```{r}
library(raer)
library(SummarizedExperiment)
library(DESeq2)
library(ComplexHeatmap)
library(viridis)
library(stringr)
```


In this vignette a public RNA-seq dataset will be analyzed,  [GSE99249](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE99249), which consists of ADAR1 mutants and control human cell lines, conditionally treated with Interferon-Beta. For simplicity we will examine the ADAR1 WT and KO samples both treated with IFN-B, with triplicate samples. 

Aligned BAM files and other necessary files have been preprocessed for this vignette. These files occupy ~250 Mb of space and by default they are stored in a temporary directory. Change the `data_dir` variable to a valid file path to keep the files for reuse.

```{r}
data_dir <- tempdir()
fns <- download_GSE99249(data_dir)
```

Next we'll collect a vector containing the paths to each BAM file These BAM files are a subset of the full BAM files, containing only alignments from chromosome 18.

```{r}
bam_files <- fns$bams
bam_files[1:2]
```

Next we'll need a FASTA file to quantify the editing sites. We'll use a FASTA file only containing chromosome 18 for this demo. The fasta can be compressed. 

```{r}
fafn <- fns$fasta
fafn
```

Lastly, we'll use a BED file containing the coordinates of known human editing sites from the REDIPortal database. This file can also be optionally compressed.

```{r}
bedfn <- fns$bed
bedfn
```

## Generate editing site read counts 

The `get_pileup()` function will calculate base counts across multiple BAM files. The base counts will be returned as a list of `GRanges` objects for each bam file. The `FilterParam()` function provides numerous filters to exclude reads and bases based on commonly used filters for detecting RNA-editing events. Specific regions can be queried using the `region` argument, or specific positions can be queried by providing `bedfile` a path to a BED-file. 

Internally, `get_pileup()` will generate pileups in memory. However, these files can also be stored in a tabix-indexed format, for later retrieval, or for querying specific regions.

```{r}
fp <- FilterParam(
  only_keep_variants = TRUE,
  trim_5p = 5,
  trim_3p = 5,
  min_base_quality = 30L,
  min_mapq = 255L,
  library_type = "fr-first-strand",
  min_splice_overhang = 10,
  indel_dist = 4, 
  max_mismatch_type = c(3, 3), 
)

plps <- get_pileup(bam_files,
  fafile = fafn,
  bedfile = bedfn,
  region = "chr18",
  filterParam = fp
)

plps[[1]]
```

The base counts are returned in a stand specific fashion depending on the `library-type` parameter. The `Ref` and `Var` bases are in reference to the strand. 

Pileups from multiple BAM files can be stored in a `summarizedExperiment` object, which permits comparisons across each sample. The `rowData()` and `rowRanges()` slots can be populated with information related to each editing site, and similarly the `colData()` slot can be used to store sample metadata. The summarizedExperiment objectcan also interface with other derivative classes, such as `singleCellExperiment()`. 

```{r}
names(plps) <- names(bam_files)
se <- merge_pileups(plps)
```

Each column from the pileup can be stored as an assay, which can be specified by the `assay_cols` parameter

```{r}
se
```

Next we will add sample metadata to the `summarizedExperiment`, which will be used to conduct differential editing analysis. 


```{r}
colData(se)$genotype_treatment <- rep(
  c(
    "ADAR1KO Interferon beta",
    "Wildtype Interferon beta"
  ),
  each = 3
)

colData(se)$genotype <- rep(
  c(
    "ADAR1KO",
    "Wildtype"
  ),
  each = 3
)
colData(se)
```

## Prepare for differential editing 

We next use the `calc_edit_frequency` function to identify the percent of edits for each position and sample. With the `drop = TRUE` argument we will also exclude sites without an adenosine. The editing frequencies will not be used for differential editing analysis, which will be conducted using the raw counts, however these are useful for filtering and visualization. 


```{r}
se_filtered <- calc_edit_frequency(se,
  edit_from = "A",
  edit_to = "G",
  drop = TRUE
)
```

We can next subset the `summarizedExperiment` object to exclude low frequency editing events. For this analysis we will require that an editing site shows editing in at least 1 sample, and have at least 5 counts in each sample. 

```{r}
has_editing <- rowSums(assay(se_filtered, "edit_freq") > 0) >= 1
has_depth <- rowSums(((assay(se_filtered, "nRef") +
  assay(se_filtered, "nVar")) >= 5)) == ncol(se_filtered)

se_filtered <- se_filtered[has_editing & has_depth, ]
se_filtered
```

Once the object has been filtered, you can prepare it for DE. This means making a new object that contains an assay with read counts if both the alt and ref alleles in a single matrix.

```{r}
deobj <- prep_for_de(se_filtered,
  min_prop = 0.1,
  max_prop = 0.9,
  min_samples = 3
)

assay(deobj, "counts")[1:3, ]
```

## Run differential editing (DESeq2)

At this stage, you can use the object to perform DE yourself or you can continue with our pre built functions

For differential editing, we use the design `design <- ~0 + condition:sample + condition:count`.

For the samples, you can leave as is or combine so the same sample name shows up in both the treatment and control. These results are not identical but they are close. In my hands, the same genes come out, but the p values and log fold change values are slightly different.

It is probably best to update the levels of your object, but if you don't, this will still work.

To run using `DESeq2`, set `type = DESeq2` in the `de_results` function. This function requires you to specify what your control and treatment are from your condition column of your `deobj`. `edgeR` is also supported. 

```{r}
de_results <- perform_de(deobj,
  type = "DESeq2",
  sample_col = "sample",
  condition_col = "genotype",
  condition_control = "Wildtype",
  condition_treatment = "ADAR1KO"
)
```

This returns a list containing the dds object, the full results, the significant results, and the model matrix. 

```{r}
de_results$sig_results[1:5, ]
```

```{r, fig.height=7, fig.width=5}
top_sites <- rownames(de_results$sig_results)[1:20]

Heatmap(assay(se_filtered, "edit_freq")[top_sites, ],
  col = viridis(100),
  column_labels = se_filtered$genotype_treatment
)
```

```{r}
unlink(data_dir)
```

<details style="margin-bottom:10px;">
<summary>
    Session info
</summary>

```{r}
sessionInfo()
```

</details>


