---
title: "Identifying RNA editing sites in droplet single cell datasets"
author: 
  - name: Kent Riemondy
    affiliation: University of Colorado School of Medicine
date: '`r Sys.Date()`'
output:
  BiocStyle::html_document
package: raer 
vignette: >
  %\VignetteIndexEntry{Single-cell}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}    
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


```{r message=FALSE}
library(raer)
library(scater)
library(SingleCellExperiment)
library(Rsamtools)
library(rtracklayer)
```

# Characterizing RNA editing sites in single cell data

This vignette will demonstrate how to use the `raer` package to examine RNA editing in droplet-based single cell RNA-seq data. 

## Preprocessing

For this example analysis we will use a single cell dataset containing human PBMC cells from [10x Genomics](https://www.10xgenomics.com/resources/datasets/10k-human-pbmcs-3-v3-1-chromium-x-with-intronic-reads-3-1-high). The single cell data was processed using the cellranger pipeline. The BAM file contains a tag (`CB`) which indicates the cell-barcode associated with each alignment, as well as a tag containing the inferred UMI sequence (`UB`).  

## Single cell editing analysis

A subset of a human PBMC scRNA-seq dataset from 10x Genomics, along with other
needed files can be downloaded using `download_human_pbmc()`.   

```{r}
fns <- download_human_pbmc()
fns
```

```{r}
bam_fn <- fns$bam
bed_fn <- fns$edit_sites
```

Next we'll load in a `SingleCellExperiment` with cell-type annotations.

```{r}
sce <- readRDS(fns$sce)
sce
```

```{r}
plotUMAP(sce, colour_by = "celltype")
```

Next we'll select editing sites to query. For this analysis we will use sites from the Rediportal database.

If the editing sites of interest are not known, one option is to perform a two pass operation. First, identify editing sites by treating the data as a bulk-RNA-seq experiment, using for example `pileup_sites()`. Then filter these sites to establish a set of high confidence sites to query in single cell mode. 

`raer` provides a function, `pileup_cells()`, which will quantify edited and non-edited UMI counts per cell barcode, then collect the site counts into a SingleCellExperiment. 

The sites to quantified are specified using a custom formatted GRanges object with 1 base intervals, a strand (+ or -), and supplemented with metadata columns named `ref` and `alt` containing the reference and alternate base to query. In this case we are
only interested in A->I editing, so we set the ref and alt to `A` and `G`. Note that the `ref` and `alt` bases are in reference to strand. For a `-` strand interval the bases should be the complement of the `+` strand bases.  

```{r}
gr <- import(bed_fn)
gr$name <- NULL
gr$score <- NULL
gr$REF <- "A"
gr$ALT <- "G"
gr
```

`pileup_cells()` accepts a `FilterParam()` class to specifying how to performing read and site filtering. Note that `pileup_cells()` is strand sensitive by default, so it is important to ensure that the strand of the input sites is correctly annotated, and that the library type is correct for the type of data of interest. For 10x Genomics data, the library type is set to `fr-second-strand`, indicating that the strand of the alignments matches the strand of the RNA. 

Note that `bam_flags` is set to **include** duplicate reads by default. If the bamfile has a tag with a UMI sequence, this can be supplied to the `umi_tag` argument to only count 1 read for each CB-UMI pair at each position. This strategy allows for reads with the same UMI to be counted at multiple independent sites enabling recovery of more sequence variants than counting only 1 read per UMI.  

Processing time can be reduced by operating in parallel across chromosomes, by supplying a BiocParallel backend to the `BPPARAM` argument (e.g. `MultiCoreParam()`).

```{r pileup_cells}
outdir <- file.path(tempdir(), "sc_edits")
cbs <- colnames(sce)
e_sce <- pileup_cells(
    bamfile = bam_fn,
    sites = gr,
    cell_barcodes = cbs,
    output_directory = outdir,
    cb_tag = "CB",
    umi_tag = "UB",
    param = FilterParam(
        min_base_quality = 30L,
        library_type = "fr-second-strand",
        min_variant_reads = 1L,
        trim_5p = 5L,
        trim_3p = 5L
    ),
    verbose = FALSE
)
e_sce
```

```{r}
dir(outdir)
```


Next we'll filter the pileups to find sites with at least 5 cells with an editing event, and add the editing information to the SingleCellExperiment as an `altExp()`. 

```{r}
e_sce <- e_sce[rowSums(assays(e_sce)$nAlt > 0) >= 5, ]
e_sce <- calc_edit_frequency(e_sce, edit_from = "Ref", edit_to = "Alt", replace_na = FALSE)
altExp(sce) <- e_sce[, colnames(sce)]
```

With the editing sites added to the gene expression SingleCellExperiment we can use plotting and other methods previously developed for single cell analysis. Here we'll visualize editing sites with the highest edited read counts.

```{r}
to_plot <- rownames(altExp(sce))[order(rowSums(assay(altExp(sce), "nAlt")), decreasing = TRUE)]

lapply(to_plot[1:5], function(x) {
    plotUMAP(sce, colour_by = x, by_exprs_values = "nAlt")
})
```

Alternatively we can view these top edited sites as a Heatmap, showing the average
number of edited reads per site in each cell type. 

```{r}
altExp(sce)$cell_type <- sce$celltype

plotGroupedHeatmap(altExp(sce),
    features = to_plot[1:25],
    group = "cell_type",
    exprs_values = "nAlt"
)
```

Various tools in Bioconductor can be used to provide additional annotation information about each RNA editing site.  The `VariantAnnotation` package provides powerful functionality to annotate the location of each editing site and provide information about alterations in codons or splice sites. 

<details style="margin-bottom:10px;">
<summary>
    Session info
</summary>

```{r}
sessionInfo()
```

</details>
