---
title: "Identifying RNA editing sites in droplet single cell datasets"
author: 
  - name: Kent Riemondy
    affiliation: University of Colorado School of Medicine
date: '`r Sys.Date()`'
output:
  BiocStyle::html_document
package: raer 
vignette: >
  %\VignetteIndexEntry{Single-cell}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}    
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


```{r}
library(raer)
library(scater)
library(SingleCellExperiment)
library(GenomicFeatures)
library(SummarizedExperiment)
library(rtracklayer)
```

# Characterizing RNA editing sites in single cell data

This vignette will demonstrate how to use the `raer` package to examine RNA editing in droplet-based single cell RNA-seq data. 

## Preprocessing

For this example analysis we will use a single cell dataset containing mouse brain cells generated by [10x Genomics](link_to_dataset). The single cell data was processed using the cellranger pipeline. The BAM file contains a tag (`CB`) which indicates the cell-barcode associated with each alignment, as well as a tag to encoding information about which alignments are representative of a UMI (`xf`).  

To process the bam using `raer` we will conduct a few preprocessing steps using command line tools (`samtools`). First we will remove PCR duplicates, only keeping reads that are representative of UMIs. As suggested by 10x genomics, we will filter to keep reads that have a bitwise alignment flag set to `xf:i:25`. 

**If using bamfiles generated by cellranger prior to versions v7.0.0, then you may want to reprocess the data with `--include-introns`. Intron aligned reads were not considered as potential UMIs, so would be excluded by filtering for `xf:i:25`. The default behavior for version 7.0.0 will include introns. ** 

After filtering for UMIs the bam will be sorted by the `CB` tag. This sorting strategy will enable rapid extraction of alignments from individual (or sets of) cell-barcodes. These steps can be combined as follows:

```bash
samtools view -b -d "xf:25"  possorted_genome_bam.bam| \
  samtools sort -t CB \
  > cbsorted_genome_bam.bam
```

`raer` contains functionality to build a simple index that stores the file position for each cell-barcode, enabling rapid extraction of alignments for defined cell-barcodes. The method is based on an approach from `Jared Thompson` in the `bri` commandline tool. The index file will have the `.bri` index file name. 

```r
idx_filename <- build_tag_index("cbsorted_genome_bam.bam", tag = "CB")
```

## Single cell editing analysis

For this vignette, we will use data from a human PBMC scRNA-seq dataset from 10x Genomics. 

```{r}
data_dir <- tempdir()
fns <- download_human_pbmc(data_dir)
fns
```

```{r}
bam_fn <- fns$bams
fa_fn <-  fns$fasta
bed_fn <- fns$edit_sites
```

We'll first index the bam file using `build_tag_index`.

```{r}
build_tag_index(bam_fn, tag = "CB", overwrite = FALSE)
```

We can query the index to see the cell-barcodes are present in the bam file using `show_tag_index()`, which will return a data.frame containing each `tag` value and the number of alignments associated with each tag.

For this vignette we will select cell-barcodes with at least x reads, 
```{r}
tag_df <- show_tag_index(bam_fn)
head(tag_df)
```

Next we'll load in a `SingleCellExperiment` with cell-type annotations.

```{r}
sce <- readRDS(fns$sce)

plotUMAP(sce, colour_by = "celltype")
```

Next we'll select editing sites to query. For this analysis we will use sites from the Rediportal database. To minimize processing time we will first identify sites with read coverage. `filter_by_coverage()` will examine read coverage across all alignments to avoid examining and storing information about sites with no or minimal coverage. 

When analyzing single cell data we have the option to quantify editing sites for each cell or we can quantify sites by pooling alignments across similar cells, such as cells in the same cluster. 

`raer` provides a wrapper function, `sc_editing()`, which will extract alignments per cell (or cluster) using `get_tag_bam()`, generate editing site counts (`get_pileup()`), then collect the site counts into a SummarizedExperiment (`create_se()`). 

The `cell_barcodes` argument specifies which cell-barcodes will be queried. If a character vector of cell-barcodes is supplied, editing sites will be quantified per cell-barcode. Alternatively a list containing cell-barcodes per cluster or group can be supplied, in which case editing sites will be quantified by pooling alignments per group.  

Arguments for `get_pileup()` can be passed to `sc_editing()`. Additionally, the editing site counts can be stored as `sparseMatrices` to reduce memory usage by setting `sparse = TRUE`. Lastly, processing time can be reduced by operating in parallel across groups of cells (or clusters), by supplying a BiocParallel backend to the `BPPARAM` argument (e.g. `MultiCoreParam()`).

```{r}
cbs <- colnames(sce)
se <- sc_editing(bamfile = bam_fn,
                 fafile = fa_fn,
                 bedfile = bed_fn,
                 cell_barcodes = cbs,
                 assay_cols = c("nA", "nG"),
                 filterParam = FilterParam(min_base_quality = 30L,
                                           library_type = "fr-second-strand",
                                           min_nucleotide_depth = 1L,
                                           trim_5p = 5L,
                                           trim_3p = 5L),
                 sparse = TRUE)
se
```
Next we'll filter the pileups to find sites with at least 5 cells with an editing event, and add the editing information to the SingleCellExperiment as an `altExp()`. 

```{r}
se <- se[Matrix::rowSums(assay(se, "nA") + assay(se, "nG")) != 0, ]

se <- calc_edit_frequency(se, edit_from = "A", edit_to = "G", replace_na = FALSE)
se <- se[, colnames(sce)]

small_se <- se[rowSums(assay(se, "nG") > 0) >= 5, ]
altExp(sce, "edits") <- small_se
```

With the editing sites added to the SingleCellExperiment we can use plotting and other methods previously developed for single cell analysis. Here we'll visualize editing sites with the highest edited read counts.

```{r}
to_plot <- rownames(altExp(sce))[order(rowSums(assay(altExp(sce), "nG")) , decreasing = T)]

lapply(to_plot[1:10], function(x){
  plotUMAP(sce, colour_by = x, by_exprs_values = "nG")
})
```

```{r}
altExp(sce)$cell_type <- sce$celltype

plotGroupedHeatmap(altExp(sce), 
            features = to_plot[1:25], 
            group = "cell_type",
            exprs_values = "nG")
```

Various tools in Bioconductor can be used to provide additional annotation information about each RNA editing site.  The `VariantAnnotation` package provides powerful functionality to annotate the location of each editing site, and provide information about alterations in codons or splice site.


<details style="margin-bottom:10px;">
<summary>
    Session info
</summary>

```{r}
sessionInfo()
```

</details>
