---
title: "Identifying RNA editing sites in droplet single cell datasets"
author: 
  - name: Kent Riemondy
    affiliation: University of Colorado School of Medicine
date: '`r Sys.Date()`'
output:
  BiocStyle::html_document
package: raer 
vignette: >
  %\VignetteIndexEntry{Single-cell}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}    
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


```{r message=FALSE}
library(raer)
library(scater)
library(SingleCellExperiment)
library(GenomicFeatures)
library(SummarizedExperiment)
library(rtracklayer)
```

# Characterizing RNA editing sites in single cell data

This vignette will demonstrate how to use the `raer` package to examine RNA editing in droplet-based single cell RNA-seq data. 

## Preprocessing

For this example analysis we will use a single cell dataset containing human PBMC cells from [10x Genomics](https://www.10xgenomics.com/resources/datasets/10k-human-pbmcs-3-v3-1-chromium-x-with-intronic-reads-3-1-high). The single cell data was processed using the cellranger pipeline. The BAM file contains a tag (`CB`) which indicates the cell-barcode associated with each alignment, as well as a tag containing the inferred UMI sequence (`UB`).  

To begin `raer` the bam will be sorted by the `CB` tag. This sorting strategy will enable rapid extraction of alignments from individual (or sets of) cell-barcodes. This sorting can be completed using samtools on the command-line, or in R using Rsamtools.

```bash
samtools sort -t CB possorted_genome_bam.bam \
  > cbsorted_genome_bam.bam
```

```R
# requires version >= 2.15.1 
Rsamtools::sortBam("possorted_genome_bam.bam", "cbsorted_genome_bam", byTag = "CB")
```

`raer` contains functionality to build a simple index that stores the file position for each cell-barcode, enabling rapid extraction of alignments for defined cell-barcodes. The method is based on an approach from `Jared Thompson` in the `bri` command line tool. The index file will have the `.bri` index file name. 

```r
idx_filename <- build_tag_index("cbsorted_genome_bam.bam", tag = "CB")
```

## Single cell editing analysis

A subset of the human PBMC scRNA-seq dataset from 10x Genomics, along with other
needed files can be downloaded using `download_human_pbmc()`.   

```{r}
data_dir <- tempdir()
fns <- download_human_pbmc(data_dir)
fns
```

```{r}
bam_fn <- fns$bams
fa_fn <- fns$fasta
bed_fn <- fns$edit_sites
```

We'll first index the bam file using `build_tag_index`.

```{r}
build_tag_index(bam_fn, tag = "CB", overwrite = FALSE)
```

We can query the index to see the cell-barcodes are present in the bam file using `show_tag_index()`, which will return a data.frame containing each `tag` value and the number of alignments associated with each tag. 

```{r}
tag_df <- show_tag_index(bam_fn)
head(tag_df)
```

Next we'll load in a `SingleCellExperiment` with cell-type annotations.

```{r}
sce <- readRDS(fns$sce)

plotUMAP(sce, colour_by = "celltype")
```

Next we'll select editing sites to query. For this analysis we will use sites from the Rediportal database. To minimize processing time it can be useful to first identify sites with read coverage. `filter_by_coverage()` will exclude sites that do not have sufficient
coverage. 

If the editing sites of interest are not known, one option is to perform a two pass operation. First, identify editing sites by treating the data as a bulk-RNA-seq experiment, using for example `get_pileup()`. Then filter these sites to establish 
a set of high confidence sites to query in single cell mode. 

When analyzing single cell data we have the option to quantify editing sites for each cell or we can quantify sites by pooling alignments across similar cells, such as cells in the same cluster. 

`raer` provides a wrapper function, `sc_editing()`, which will extract alignments per cell (or cluster) using `get_tag_bam()`, generate editing site counts (`get_pileup()`), then collect the site counts into a `RangedSummarizedExperiment` (`merge_pileups()`). 

The `cell_barcodes` argument specifies which cell-barcodes will be queried. If a character vector of cell-barcodes is supplied, editing sites will be quantified per cell-barcode. Alternatively a list containing cell-barcodes per cluster or group can be supplied, in which case editing sites will be quantified by pooling alignments per group.  
Additional arguments for `get_pileup()` can be passed to `sc_editing()`. Note that `bam_flags` is set to **include** duplicate reads by default. If the bamfile has a tag with a UMI sequence, this can be supplied to the `umi_tag` argument to only count 1 read for each UMI at each position. This strategy allows for reads with the same UMI to be counted at multiple independent sites enabling recovery of more sequence variants than counting only 1 read per UMI.  

Processing time can be reduced by operating in parallel across groups of cells (or clusters), by supplying a BiocParallel backend to the `BPPARAM` argument (e.g. `MultiCoreParam()`).

```{r}
cbs <- colnames(sce)
se <- sc_editing(
  bamfile = bam_fn,
  fafile = fa_fn,
  bedfile = bed_fn,
  cell_barcodes = cbs,
  assay_cols = c("nA", "nG"),
  filterParam = FilterParam(
    min_base_quality = 30L,
    library_type = "fr-second-strand",
    min_nucleotide_depth = 1L,
    trim_5p = 5L,
    trim_3p = 5L
  ),
  umi_tag = "UB"
)
se
```

Next we'll filter the pileups to find sites with at least 5 cells with an editing event, and add the editing information to the SingleCellExperiment as an `altExp()`. 

```{r}
se <- se[Matrix::rowSums(assay(se, "nA") + assay(se, "nG")) != 0, ]

se <- calc_edit_frequency(se, edit_from = "A", edit_to = "G", replace_na = FALSE)
se <- se[, colnames(sce)]

small_se <- se[rowSums(assay(se, "nG") > 0) >= 5, ]
altExp(sce, "edits") <- small_se
```

With the editing sites added to the SingleCellExperiment we can use plotting and other methods previously developed for single cell analysis. Here we'll visualize editing sites with the highest edited read counts.

```{r}
to_plot <- rownames(altExp(sce))[order(rowSums(assay(altExp(sce), "nG")), decreasing = TRUE)]

lapply(to_plot[1:10], function(x) {
  plotUMAP(sce, colour_by = x, by_exprs_values = "nG")
})
```

```{r}
altExp(sce)$cell_type <- sce$celltype

plotGroupedHeatmap(altExp(sce),
  features = to_plot[1:25],
  group = "cell_type",
  exprs_values = "nG"
)
```

Various tools in Bioconductor can be used to provide additional annotation information about each RNA editing site.  The `VariantAnnotation` package provides powerful functionality to annotate the location of each editing site, and provide information about alterations in codons or splice site. 


<details style="margin-bottom:10px;">
<summary>
    Session info
</summary>

```{r}
sessionInfo()
```

</details>
