<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="The Adenosine Editing Index describes the magnitude of A-to-I
editing in a sample. The index is a weighted average of editing events (G
bases) observed at A positions. The vast majority A-to-I editing occurs in
ALU elements in the human genome, and these regions have a high A-to-I
editing signal compared to other regions such as coding exons. This
function will perform pileup at specified repeat regions and return a
summary AEI metric."><title>Calculate the Adenosine Editing Index (AEI) — calc_AEI • raer</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/Source_Sans_Pro-0.4.8/font.css" rel="stylesheet"><link href="../deps/Source_Code_Pro-0.4.8/font.css" rel="stylesheet"><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Calculate the Adenosine Editing Index (AEI) — calc_AEI"><meta property="og:description" content="The Adenosine Editing Index describes the magnitude of A-to-I
editing in a sample. The index is a weighted average of editing events (G
bases) observed at A positions. The vast majority A-to-I editing occurs in
ALU elements in the human genome, and these regions have a high A-to-I
editing signal compared to other regions such as coding exons. This
function will perform pileup at specified repeat regions and return a
summary AEI metric."><meta property="og:image" content="https://rnabioco.github.io/raer/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><script defer data-domain="rnabioco.github.io" src="https://plausible.io/js/plausible.js"></script></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-dark"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">raer</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.1.3</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/raer.html">Get started</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/rnabioco/raer/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Calculate the Adenosine Editing Index (AEI)</h1>
      <small class="dont-index">Source: <a href="https://github.com/rnabioco/raer/blob/HEAD/R/calc_AEI.R" class="external-link"><code>R/calc_AEI.R</code></a></small>
      <div class="d-none name"><code>calc_AEI.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>The Adenosine Editing Index describes the magnitude of A-to-I
editing in a sample. The index is a weighted average of editing events (G
bases) observed at A positions. The vast majority A-to-I editing occurs in
ALU elements in the human genome, and these regions have a high A-to-I
editing signal compared to other regions such as coding exons. This
function will perform pileup at specified repeat regions and return a
summary AEI metric.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">calc_AEI</span><span class="op">(</span></span>
<span>  <span class="va">bamfiles</span>,</span>
<span>  <span class="va">fasta</span>,</span>
<span>  alu_ranges <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  txdb <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  snp_db <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  param <span class="op">=</span> <span class="fu"><a href="pileup_sites.html">FilterParam</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  BPPARAM <span class="op">=</span> <span class="fu">SerialParam</span><span class="op">(</span><span class="op">)</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>bamfiles</dt>
<dd><p>character vector of paths to indexed bam files. If a named
character vector is supplied the names will be used in the output.</p></dd>


<dt>fasta</dt>
<dd><p>fasta filename</p></dd>


<dt>alu_ranges</dt>
<dd><p><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a> with regions to query for
calculating the AEI, typically ALU repeats.</p></dd>


<dt>txdb</dt>
<dd><p>A TxDb object, if supplied, will be used to subset the
alu_ranges to those found overlapping genes. Alternatively a <a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a>
object with gene coordinates.  If the <code>library_type</code>, specified by
<code>FilterParam</code>, is <code>unstranded</code> then the TxDb will
be used to correct the strandness relative to the reference and is a
required
parameter.</p></dd>


<dt>snp_db</dt>
<dd><p>either a <a href="https://rdrr.io/pkg/BSgenome/man/SNPlocs-class.html" class="external-link">SNPlocs</a>, <a href="https://rdrr.io/pkg/GenomicRanges/man/GPos-class.html" class="external-link">GPos</a>, or <a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a> object. If supplied,
will be used to exclude polymorphic positions prior to calculating the AEI.
If <code>calc_AEI()</code> will be used many times, one will save time by first
identifying SNPs that overlap the supplied <code>alu_ranges</code>, and passing these
as a <a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a> to <code>snp_db</code> rather than supplying all known SNPs (see
<code><a href="get_overlapping_snps.html">get_overlapping_snps()</a></code>).</p></dd>


<dt>param</dt>
<dd><p>object of class <code><a href="pileup_sites.html">FilterParam()</a></code> which specify various
filters to apply to reads and sites during pileup.</p></dd>


<dt>BPPARAM</dt>
<dd><p>A BiocParallelParam object for specifying parallel options
for operating over chromosomes.</p></dd>


<dt>verbose</dt>
<dd><p>report progress on each chromosome?</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>A named list containing:</p><ul><li><p><code>AEI</code>: a matrix of AEI index values computed for all allelic
combinations, one row for each supplied bam file.</p></li>
<li><p><code>AEI_per_chrom</code>: a data.frame containing values computed for each
chromosome</p></li>
</ul></div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    <p>Roth, S.H., Levanon, E.Y. &amp; Eisenberg, E. Genome-wide
quantification of ADAR adenosine-to-inosine RNA editing activity. Nat Methods
16, 1131–1138 (2019). https://doi.org/10.1038/s41592-019-0610-9</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/Rsamtools" class="external-link">Rsamtools</a></span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">bamfn</span> <span class="op">&lt;-</span> <span class="fu"><a href="raer_example.html">raer_example</a></span><span class="op">(</span><span class="st">"SRR5564269_Aligned.sortedByCoord.out.md.bam"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">bam2fn</span> <span class="op">&lt;-</span> <span class="fu"><a href="raer_example.html">raer_example</a></span><span class="op">(</span><span class="st">"SRR5564277_Aligned.sortedByCoord.out.md.bam"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">bams</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">bamfn</span>, <span class="va">bam2fn</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">bams</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ADAR1KO"</span>, <span class="st">"WT"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">fafn</span> <span class="op">&lt;-</span> <span class="fu"><a href="raer_example.html">raer_example</a></span><span class="op">(</span><span class="st">"human.fasta"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">mock_alu_ranges</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Rsamtools/man/scanFa.html" class="external-link">scanFaIndex</a></span><span class="op">(</span><span class="va">fafn</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu">calc_AEI</span><span class="op">(</span><span class="va">bams</span>, <span class="va">fafn</span>, <span class="va">mock_alu_ranges</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">res</span><span class="op">$</span><span class="va">AEI</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                A_C       A_G       A_T       C_A        C_G       C_T</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ADAR1KO 0.00927902 0.1019746 0.0278319 0.1237964 0.00000000 0.1100564</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> WT      0.00000000 2.2985864 0.1292293 0.2142033 0.01650982 0.2635046</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>               G_A        G_C        G_T       T_A        T_C        T_G</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ADAR1KO 0.2364360 0.01247194 0.04986909 0.0108272 0.03247456 0.03247456</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> WT      0.3170769 0.00000000 0.00000000 0.0000000 0.01161980 0.01161980</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>An R package developed by the <a href="https://medschool.cuanschutz.edu/rbi" class="external-link">RNA Bioscience Initiative</a></p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

