<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="The Adenosine Editing Index describes the magnitude of A-to-I editing
in a sample. The index is a weighted average of editing events (G bases) observed
at A positions. The vast majority A-to-I editing occurs in ALU elements in the human
genome, and these regions have a high A-to-I editing signal compared to other regions
such as coding exons. This function will perform pileup at specified repeat regions and
return a summary AEI metric."><title>Calculate the Adenosine Editing Index (AEI) — calc_AEI • raer</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><link href="../deps/_Source%20Sans%20Pro-0.4.1/font.css" rel="stylesheet"><link href="../deps/_Source%20Code%20Pro-0.4.1/font.css" rel="stylesheet"><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Calculate the Adenosine Editing Index (AEI) — calc_AEI"><meta property="og:description" content="The Adenosine Editing Index describes the magnitude of A-to-I editing
in a sample. The index is a weighted average of editing events (G bases) observed
at A positions. The vast majority A-to-I editing occurs in ALU elements in the human
genome, and these regions have a high A-to-I editing signal compared to other regions
such as coding exons. This function will perform pileup at specified repeat regions and
return a summary AEI metric."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><script defer data-domain="rnabioco.github.io" src="https://plausible.io/js/plausible.js"></script></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-dark"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">raer</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/raer.html">Get started</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/novel-sites.html">Identifying RNA editing sites using RNA and DNA sequencing data</a>
  </div>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/rnabioco/raer/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Calculate the Adenosine Editing Index (AEI)</h1>
      <small class="dont-index">Source: <a href="https://github.com/rnabioco/raer/blob/HEAD/R/calc_AEI.R" class="external-link"><code>R/calc_AEI.R</code></a></small>
      <div class="d-none name"><code>calc_AEI.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>The Adenosine Editing Index describes the magnitude of A-to-I editing
in a sample. The index is a weighted average of editing events (G bases) observed
at A positions. The vast majority A-to-I editing occurs in ALU elements in the human
genome, and these regions have a high A-to-I editing signal compared to other regions
such as coding exons. This function will perform pileup at specified repeat regions and
return a summary AEI metric.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">calc_AEI</span><span class="op">(</span></span>
<span>  <span class="va">bam_fn</span>,</span>
<span>  <span class="va">fasta_fn</span>,</span>
<span>  alu_ranges <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  txdb <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  snp_db <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  filterParam <span class="op">=</span> <span class="fu"><a href="get_pileup.html">FilterParam</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  BPPARAM <span class="op">=</span> <span class="fu">SerialParam</span><span class="op">(</span><span class="op">)</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>bam_fn</dt>
<dd><p>bam file</p></dd>


<dt>fasta_fn</dt>
<dd><p>fasta</p></dd>


<dt>alu_ranges</dt>
<dd><p>GRanges or the name of a BEDfile with regions to query for calculating the AEI,
typically ALU repeats. If a BED file is supplied it will not be filtered by the txdb option.</p></dd>


<dt>txdb</dt>
<dd><p>A txdb object, if supplied, will be used to subset the alu_ranges to
those found overlapping genes. Alternatively a GRanges object with gene coordinates.</p></dd>


<dt>snp_db</dt>
<dd><p>either a SNPlocs package, GPos, or GRanges object. If supplied,
will be used to exclude polymorphic positions prior to calculating the AEI. If
<code>calc_AEI()</code> will be used many times, one could save some time by first identifying
SNPs that overlap the supplied alu_ranges, and passing these as a GRanges to snp_db
rather than supplying all known SNPs. Combined with using a bedfile for alu_ranges can
also will save time.</p></dd>


<dt>filterParam</dt>
<dd><p>object of class <code><a href="get_pileup.html">FilterParam()</a></code> which specify various
filters to apply to reads and sites during pileup.</p></dd>


<dt>BPPARAM</dt>
<dd><p>A BiocParallelParam object for specifying parallel options for
operating over chromosomes.</p></dd>


<dt>verbose</dt>
<dd><p>report progress on each chromosome?</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>A named list with the AEI index computed for all allelic combinations.
If correctly computed the signal from the A_G index should be higher than other
alleles (T_C), which are most likely derived from noise or polymorphisms.</p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    <p>Roth, S.H., Levanon, E.Y. &amp; Eisenberg, E. Genome-wide quantification of ADAR adenosine-to-inosine RNA editing activity. Nat Methods 16, 1131–1138 (2019). https://doi.org/10.1038/s41592-019-0610-9</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/Rsamtools" class="external-link">Rsamtools</a></span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">bamfn</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"SRR5564277_Aligned.sortedByCoord.out.md.bam"</span>, package <span class="op">=</span> <span class="st">"raer"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">fafn</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"human.fasta"</span>, package <span class="op">=</span> <span class="st">"raer"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">dummy_alu_ranges</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Rsamtools/man/scanFa.html" class="external-link">scanFaIndex</a></span><span class="op">(</span><span class="va">fafn</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">calc_AEI</span><span class="op">(</span><span class="va">bamfn</span>, <span class="va">fafn</span>, <span class="va">dummy_alu_ranges</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $A_C</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $A_G</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 1.879456</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $A_T</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.1319261</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $C_A</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.2426203</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $C_G</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $C_T</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.2224469</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $G_A</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.3120411</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $G_C</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $G_T</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $T_A</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $T_C</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $T_G</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>An R package developed by the <a href="https://medschool.cuanschutz.edu/rbi" class="external-link">RNA Bioscience Initiative</a></p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.5.</p>
</div>

    </footer></div>

  

  

  </body></html>

