<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="The Adenosine Editing Index describes the magnitude of A-to-I
editing in a sample. The index is a weighted average of editing events (G
bases) observed at A positions. The vast majority A-to-I editing occurs in
ALU elements in the human genome, and these regions have a high A-to-I
editing signal compared to other regions such as coding exons. This
function will examine enumerate edited and non-edited base counts at the supplied
sites and return summary AEI metric per cell. Potential editing sites within
repeat regions can be generated using get_scAEI_sites()."><title>Calculate the Adenosine Editing Index (AEI) in single cells — calc_scAEI • raer</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><link href="../deps/Source_Sans_Pro-0.4.7/font.css" rel="stylesheet"><link href="../deps/Source_Code_Pro-0.4.7/font.css" rel="stylesheet"><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Calculate the Adenosine Editing Index (AEI) in single cells — calc_scAEI"><meta property="og:description" content="The Adenosine Editing Index describes the magnitude of A-to-I
editing in a sample. The index is a weighted average of editing events (G
bases) observed at A positions. The vast majority A-to-I editing occurs in
ALU elements in the human genome, and these regions have a high A-to-I
editing signal compared to other regions such as coding exons. This
function will examine enumerate edited and non-edited base counts at the supplied
sites and return summary AEI metric per cell. Potential editing sites within
repeat regions can be generated using get_scAEI_sites()."><meta property="og:image" content="https://rnabioco.github.io/raer/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><script defer data-domain="rnabioco.github.io" src="https://plausible.io/js/plausible.js"></script></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-dark"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">raer</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.14</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/raer.html">Get started</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/rnabioco/raer/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Calculate the Adenosine Editing Index (AEI) in single cells</h1>
      <small class="dont-index">Source: <a href="https://github.com/rnabioco/raer/blob/HEAD/R/calc_scAEI.R" class="external-link"><code>R/calc_scAEI.R</code></a></small>
      <div class="d-none name"><code>calc_scAEI.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>The Adenosine Editing Index describes the magnitude of A-to-I
editing in a sample. The index is a weighted average of editing events (G
bases) observed at A positions. The vast majority A-to-I editing occurs in
ALU elements in the human genome, and these regions have a high A-to-I
editing signal compared to other regions such as coding exons. This
function will examine enumerate edited and non-edited base counts at the supplied
sites and return summary AEI metric per cell. Potential editing sites within
repeat regions can be generated using <code>get_scAEI_sites()</code>.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">calc_scAEI</span><span class="op">(</span></span>
<span>  <span class="va">bamfiles</span>,</span>
<span>  <span class="va">sites</span>,</span>
<span>  <span class="va">cell_barcodes</span>,</span>
<span>  param <span class="op">=</span> <span class="fu"><a href="pileup_sites.html">FilterParam</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  edit_from <span class="op">=</span> <span class="st">"A"</span>,</span>
<span>  edit_to <span class="op">=</span> <span class="st">"G"</span>,</span>
<span>  output_dir <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  return_sce <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">get_scAEI_sites</span><span class="op">(</span><span class="va">fasta</span>, <span class="va">genes</span>, <span class="va">alus</span>, edit_from <span class="op">=</span> <span class="st">"A"</span>, edit_to <span class="op">=</span> <span class="st">"G"</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>bamfiles</dt>
<dd><p>a path to a BAM file (for 10x libraries), or a vector of paths
to BAM files (smart-seq2). Can be supplied as a character vector, <code>BamFile</code>, or
<code>BamFileList</code>.</p></dd>


<dt>sites</dt>
<dd><p>a GRanges object produced by <code>get_scAEI_sites()</code> containing sites to process.</p></dd>


<dt>cell_barcodes</dt>
<dd><p>A character vector of single cell barcodes to process. If
processing multiple BAM files (e.g. smart-seq-2), provide a character vector
of unique identifiers for each input BAM, to name each BAM file in the output files.</p></dd>


<dt>param</dt>
<dd><p>object of class <code><a href="pileup_sites.html">FilterParam()</a></code> which specify various filters to
apply to reads and sites during pileup.</p></dd>


<dt>edit_from</dt>
<dd><p>This should correspond to the base
(<code>A</code>, <code>C</code>, <code>G</code>, <code>T</code>) you expect in the reference. Ex. for A to I
editing events, this would be <code>A</code>.</p></dd>


<dt>edit_to</dt>
<dd><p>This should correspond to the base
(<code>A</code>, <code>C</code>, <code>G</code>, <code>T</code>) you expect in an edited site. Ex. for A
to I editing events, this would be <code>G</code>.</p></dd>


<dt>output_dir</dt>
<dd><p>Output directory for <code>nRef</code> and <code>nAlt</code> sparseMatrix files.
If NULL, a temporary directory will be used.</p></dd>


<dt>return_sce</dt>
<dd><p>if <code>TRUE</code>, data is returned as a SingleCellExperiment, if
<code>FALSE</code> a <code>DataFrame</code> containing computed AEI values will be returned.</p></dd>


<dt>...</dt>
<dd><p>additional arguments to <code><a href="pileup_cells.html">pileup_cells()</a></code></p></dd>


<dt>fasta</dt>
<dd><p>Path to a genome fasta file</p></dd>


<dt>genes</dt>
<dd><p>A <code>GRanges</code> object with gene coordinates.Alternatively a <code>TxDb</code> object,
which if supplied, will be used extract gene coordinates.</p></dd>


<dt>alus</dt>
<dd><p><code>GRanges</code> with repeat regions to query for calculating the AEI,
typically ALU repeats. The strand of the supplied intervals will be ignored for
defining repeat regions.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>A <code>DataFrame</code> containing computed <code>AEI</code> values,
count of editing events (<code>n_alt</code>), and count of reference events (<code>n_ref</code>) per cell.
If <code>return_sce</code> is <code>TRUE</code>, then a <code>SingleCellExperiment</code> is returned with the
AEI values stored in the <code>colData</code>.</p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    <p>Roth, S.H., Levanon, E.Y. &amp; Eisenberg, E. Genome-wide
quantification of ADAR adenosine-to-inosine RNA editing activity. Nat Methods
16, 1131–1138 (2019). https://doi.org/10.1038/s41592-019-0610-9</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/Rsamtools" class="external-link">Rsamtools</a></span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/GenomicRanges" class="external-link">GenomicRanges</a></span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">bam_fn</span> <span class="op">&lt;-</span> <span class="fu"><a href="raer_example.html">raer_example</a></span><span class="op">(</span><span class="st">"5k_neuron_mouse_possort.bam"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">bai</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Rsamtools/man/scanBam.html" class="external-link">indexBam</a></span><span class="op">(</span><span class="va">bam_fn</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># cell barcodes to query</span></span></span>
<span class="r-in"><span><span class="va">cbs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"TGTTTGTTCCATCCGT-1"</span>, <span class="st">"CAACCAACATAATCGC-1"</span>, <span class="st">"TGGAACTCAAGCTGTT-1"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># genes used to infer transcribed strand</span></span></span>
<span class="r-in"><span><span class="va">genes_gr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    <span class="st">"2:100-400:-"</span>,</span></span>
<span class="r-in"><span>    <span class="st">"2:500-605:-"</span>,</span></span>
<span class="r-in"><span>    <span class="st">"2:600-680:+"</span></span></span>
<span class="r-in"><span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># alu intervals</span></span></span>
<span class="r-in"><span><span class="va">alus_gr</span> <span class="op">&lt;-</span>  <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    <span class="st">"2:110-380"</span>,</span></span>
<span class="r-in"><span>    <span class="st">"2:510-600"</span>,</span></span>
<span class="r-in"><span>    <span class="st">"2:610-670"</span></span></span>
<span class="r-in"><span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># genome fasta file, used to find A bases</span></span></span>
<span class="r-in"><span><span class="va">fa_fn</span> <span class="op">&lt;-</span> <span class="fu"><a href="raer_example.html">raer_example</a></span><span class="op">(</span><span class="st">"mouse_tiny.fasta"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># get positions of potential A -&gt; G changes in alus</span></span></span>
<span class="r-in"><span><span class="va">sites</span> <span class="op">&lt;-</span> <span class="fu">get_scAEI_sites</span><span class="op">(</span><span class="va">fa_fn</span>, <span class="va">genes_gr</span>, <span class="va">alus_gr</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">fp</span> <span class="op">&lt;-</span> <span class="fu"><a href="pileup_sites.html">FilterParam</a></span><span class="op">(</span>library_type <span class="op">=</span> <span class="st">"fr-second-strand"</span>,</span></span>
<span class="r-in"><span>                  min_mapq <span class="op">=</span> <span class="fl">255</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">calc_scAEI</span><span class="op">(</span><span class="va">bam_fn</span>, <span class="va">sites</span>, <span class="va">cbs</span>, <span class="va">fp</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> DataFrame with 3 rows and 3 columns</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                          AEI     n_alt     n_ref</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                    &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> TGTTTGTTCCATCCGT-1   5.08475         3        56</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> CAACCAACATAATCGC-1  20.00000         3        12</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> TGGAACTCAAGCTGTT-1   0.00000         0         9</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>An R package developed by the <a href="https://medschool.cuanschutz.edu/rbi" class="external-link">RNA Bioscience Initiative</a></p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

