<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content='Gene annotations are used to infer the
likely strand of editing sites. This function will operate
on unstranded datasets which have been processed using "genomic-unstranded"
library type which reports variants with respect to the
strand for all sites. The strand of the editing site will be
assigned the strand of overlapping features in the genes_gr
object. Sites with no-overlap, or overlapping features
with conflicting strands (+ and -) will be removed.

'><title>Apply strand correction using gene annotations — correct_strand • raer</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><link href="../deps/_Source%20Sans%20Pro-0.4.1/font.css" rel="stylesheet"><link href="../deps/_Source%20Code%20Pro-0.4.1/font.css" rel="stylesheet"><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Apply strand correction using gene annotations — correct_strand"><meta property="og:description" content='Gene annotations are used to infer the
likely strand of editing sites. This function will operate
on unstranded datasets which have been processed using "genomic-unstranded"
library type which reports variants with respect to the
strand for all sites. The strand of the editing site will be
assigned the strand of overlapping features in the genes_gr
object. Sites with no-overlap, or overlapping features
with conflicting strands (+ and -) will be removed.

'><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><script defer data-domain="rnabioco.github.io" src="https://plausible.io/js/plausible.js"></script></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-dark"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">raer</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/raer.html">Get started</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/novel-sites.html">Identifying RNA editing sites using RNA and DNA sequencing data</a>
  </div>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/rnabioco/raer/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Apply strand correction using gene annotations</h1>
      <small class="dont-index">Source: <a href="https://github.com/rnabioco/raer/blob/HEAD/R/calc_AEI.R" class="external-link"><code>R/calc_AEI.R</code></a></small>
      <div class="d-none name"><code>correct_strand.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Gene annotations are used to infer the
likely strand of editing sites. This function will operate
on unstranded datasets which have been processed using "genomic-unstranded"
library type which reports variants with respect to the</p><ul><li><p>strand for all sites. The strand of the editing site will be
assigned the strand of overlapping features in the <code>genes_gr</code>
object. Sites with no-overlap, or overlapping features
with conflicting strands (+ and -) will be removed.</p></li>
</ul></div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">correct_strand</span><span class="op">(</span><span class="va">gr</span>, <span class="va">genes_gr</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>gr</dt>
<dd><p>GRanges object containing editing sites processed with
"genomic-unstranded" setting</p></dd>


<dt>genes_gr</dt>
<dd><p>GRanges object containing reference features to
annotate the strand of the editing sites.</p></dd>

</dl></div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://bioconductor.org/packages/GenomicRanges" class="external-link">"GenomicRanges"</a></span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">bamfn</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"SRR5564269_Aligned.sortedByCoord.out.md.bam"</span>, package <span class="op">=</span> <span class="st">"raer"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">fafn</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"human.fasta"</span>, package <span class="op">=</span> <span class="st">"raer"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">fp</span> <span class="op">&lt;-</span> <span class="fu"><a href="get_pileup.html">FilterParam</a></span><span class="op">(</span>library_type <span class="op">=</span> <span class="st">"genomic-unstranded"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">plp</span> <span class="op">&lt;-</span> <span class="fu"><a href="get_pileup.html">get_pileup</a></span><span class="op">(</span><span class="va">bamfn</span>, <span class="va">fafn</span>, filterParam <span class="op">=</span> <span class="va">fp</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"DHFR:200-400:+"</span>,</span></span>
<span class="r-in"><span>  <span class="st">"SPCS3:100-200:-"</span>,</span></span>
<span class="r-in"><span>  <span class="st">"SSR3:3-10:-"</span>,</span></span>
<span class="r-in"><span>  <span class="st">"SSR3:6-12:+"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">correct_strand</span><span class="op">(</span><span class="va">plp</span>, <span class="va">genes</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GRanges object with 307 ranges and 9 metadata columns:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         seqnames    ranges strand |         Ref         Var      nRef</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>            &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;character&gt; &lt;integer&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     [1]     SSR3         3      - |           A           -        14</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     [2]     SSR3         4      - |           A           -        15</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     [3]     SSR3         5      - |           G           -        15</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     [4]     SSR3        11      + |           A           -        15</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     [5]     SSR3        12      + |           T           -        15</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     ...      ...       ...    ... .         ...         ...       ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [303]     DHFR       396      + |           T           -        40</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [304]     DHFR       397      + |           T           -        43</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [305]     DHFR       398      + |           T           -        43</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [306]     DHFR       399      + |           C           -        43</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [307]     DHFR       400      + |           T           -        45</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>              nVar        nA        nT        nC        nG        nN</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     [1]         0        14         0         0         0         0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     [2]         0        15         0         0         0         0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     [3]         0         0         0         0        15         0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     [4]         0        15         0         0         0         0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     [5]         0         0        15         0         0         0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     ...       ...       ...       ...       ...       ...       ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [303]         0         0        40         0         0         0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [304]         0         0        43         0         0         0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [305]         0         0        43         0         0         0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [306]         0         0         0        43         0         0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [307]         0         0        45         0         0         0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   -------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   seqinfo: 3 sequences from an unspecified genome</span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>An R package developed by the <a href="https://medschool.cuanschutz.edu/rbi" class="external-link">RNA Bioscience Initiative</a></p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.5.</p>
</div>

    </footer></div>

  

  

  </body></html>

