<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="OligodT will prime at A-rich regions in an RNA. Reverse transcription
from these internal priming sites will install an oligodT sequence at the 3' end
of the cDNA. Sequence variants within these internal priming sites are enriched
for variants converting the genomic sequence to the A encoded by the oligodT primer.
Trimming poly(A) from the 3' ends of reads reduces but does not eliminate these signals
This function will identify regions that are enriched for mispriming events. Reads
that were trimmed to remove poly(A) (encoded in the pa tag by 10x genomics) are
identified. The aligned 3' positions of these reads are counted, and sites passing
thresholds (at least 2 reads) are retained as possible sites of mispriming. Be default
regions 5 bases upstream and 20 bases downstream of these putative mispriming sites
are returned."><title>Find regions with oligodT mispriming — find_mispriming_sites • raer</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><link href="../deps/Source_Sans_Pro-0.4.6/font.css" rel="stylesheet"><link href="../deps/Source_Code_Pro-0.4.6/font.css" rel="stylesheet"><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Find regions with oligodT mispriming — find_mispriming_sites"><meta property="og:description" content="OligodT will prime at A-rich regions in an RNA. Reverse transcription
from these internal priming sites will install an oligodT sequence at the 3' end
of the cDNA. Sequence variants within these internal priming sites are enriched
for variants converting the genomic sequence to the A encoded by the oligodT primer.
Trimming poly(A) from the 3' ends of reads reduces but does not eliminate these signals
This function will identify regions that are enriched for mispriming events. Reads
that were trimmed to remove poly(A) (encoded in the pa tag by 10x genomics) are
identified. The aligned 3' positions of these reads are counted, and sites passing
thresholds (at least 2 reads) are retained as possible sites of mispriming. Be default
regions 5 bases upstream and 20 bases downstream of these putative mispriming sites
are returned."><meta property="og:image" content="https://rnabioco.github.io/raer/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><script defer data-domain="rnabioco.github.io" src="https://plausible.io/js/plausible.js"></script></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-dark"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">raer</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.3</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/raer.html">Get started</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/novel-sites.html">Identifying RNA editing sites using RNA and DNA sequencing data</a>
    <a class="dropdown-item" href="../articles/single-cell.html">Identifying RNA editing sites in droplet single cell datasets</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/rnabioco/raer/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Find regions with oligodT mispriming</h1>
      <small class="dont-index">Source: <a href="https://github.com/rnabioco/raer/blob/HEAD/R/utils.R" class="external-link"><code>R/utils.R</code></a></small>
      <div class="d-none name"><code>find_mispriming_sites.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>OligodT will prime at A-rich regions in an RNA. Reverse transcription
from these internal priming sites will install an oligodT sequence at the 3' end
of the cDNA. Sequence variants within these internal priming sites are enriched
for variants converting the genomic sequence to the A encoded by the oligodT primer.
Trimming poly(A) from the 3' ends of reads reduces but does not eliminate these signals</p>
<p>This function will identify regions that are enriched for mispriming events. Reads
that were trimmed to remove poly(A) (encoded in the pa tag by 10x genomics) are
identified. The aligned 3' positions of these reads are counted, and sites passing
thresholds (at least 2 reads) are retained as possible sites of mispriming. Be default
regions 5 bases upstream and 20 bases downstream of these putative mispriming sites
are returned.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">find_mispriming_sites</span><span class="op">(</span></span>
<span>  <span class="va">bamfile</span>,</span>
<span>  <span class="va">fafile</span>,</span>
<span>  pos_5p <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  pos_3p <span class="op">=</span> <span class="fl">20</span>,</span>
<span>  min_reads <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  tag <span class="op">=</span> <span class="st">"pa"</span>,</span>
<span>  tag_values <span class="op">=</span> <span class="fl">3</span><span class="op">:</span><span class="fl">300</span>,</span>
<span>  n_reads_per_chunk <span class="op">=</span> <span class="fl">1e+06</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>bamfile</dt>
<dd><p>path to bamfile</p></dd>


<dt>fafile</dt>
<dd><p>path to fasta file</p></dd>


<dt>pos_5p</dt>
<dd><p>distance 5' of mispriming site to define mispriming region</p></dd>


<dt>pos_3p</dt>
<dd><p>distance 3' of mispriming site to define mispriming region</p></dd>


<dt>min_reads</dt>
<dd><p>minimum required number of reads at a mispriming site</p></dd>


<dt>tag</dt>
<dd><p>bam tag containing number of poly(A) bases trimmed</p></dd>


<dt>tag_values</dt>
<dd><p>range of values required for read to be considered</p></dd>


<dt>n_reads_per_chunk</dt>
<dd><p>number of reads to process in memory, see
<code><a href="https://rdrr.io/pkg/Rsamtools/man/BamFile-class.html" class="external-link">Rsamtools::BamFile()</a></code></p></dd>


<dt>verbose</dt>
<dd><p>if true report progress</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>A GenomicsRanges containing regions enriched for putative mispriming
events. The <code>n_reads</code> column specifies the number of polyA trimmed reads
overlapping the mispriming region. <code>mean_pal</code> indicates the mean length of polyA
sequence trimmed from reads overlapping the region. The <code>n_regions</code> column specifies the number
overlapping independent regions found in each chunk (dictated by <code>n_reads_per_chunk</code>).
The <code>A_freq</code> column indicates the frequency of A bases within the region.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">bam_fn</span> <span class="op">&lt;-</span> <span class="fu"><a href="raer_example.html">raer_example</a></span><span class="op">(</span><span class="st">"5k_neuron_mouse_possort.bam"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">fa_fn</span> <span class="op">&lt;-</span> <span class="fu"><a href="raer_example.html">raer_example</a></span><span class="op">(</span><span class="st">"mouse_tiny.fasta"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">find_mispriming_sites</span><span class="op">(</span><span class="va">bam_fn</span>, <span class="va">fa_fn</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> working on 2:580-633:- to 2:580-633:-</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GRanges object with 1 range and 4 metadata columns:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       seqnames    ranges strand |  mean_pal   n_reads n_regions    A_freq</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt; &lt;integer&gt; &lt;integer&gt; &lt;numeric&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [1]        2   560-585      - |   34.6667         6         1  0.615385</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   -------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   seqinfo: 4 sequences from an unspecified genome</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>An R package developed by the <a href="https://medschool.cuanschutz.edu/rbi" class="external-link">RNA Bioscience Initiative</a></p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

