<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Generate base counts using pileup"><title>Generate base counts using pileup — get_pileup • raer</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><link href="../deps/_Source%20Sans%20Pro-0.4.1/font.css" rel="stylesheet"><link href="../deps/_Source%20Code%20Pro-0.4.1/font.css" rel="stylesheet"><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Generate base counts using pileup — get_pileup"><meta property="og:description" content="Generate base counts using pileup"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><script defer data-domain="rnabioco.github.io" src="https://plausible.io/js/plausible.js"></script></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-dark"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">raer</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/rnabioco/raer/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Generate base counts using pileup</h1>
      <small class="dont-index">Source: <a href="https://github.com/rnabioco/raer/blob/HEAD/R/get_pileup.R" class="external-link"><code>R/get_pileup.R</code></a></small>
      <div class="d-none name"><code>get_pileup.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Generate base counts using pileup</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">get_pileup</span><span class="op">(</span>
  <span class="va">bamfile</span>,
  <span class="va">fafile</span>,
  bedfile <span class="op">=</span> <span class="cn">NULL</span>,
  region <span class="op">=</span> <span class="cn">NULL</span>,
  chrom <span class="op">=</span> <span class="cn">NULL</span>,
  start <span class="op">=</span> <span class="cn">NULL</span>,
  end <span class="op">=</span> <span class="cn">NULL</span>,
  min_reads <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1L</span><span class="op">)</span>,
  min_base_qual <span class="op">=</span> <span class="fl">20L</span>,
  min_mapq <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0L</span><span class="op">)</span>,
  max_depth <span class="op">=</span> <span class="fl">10000</span>,
  library_type <span class="op">=</span> <span class="st">"fr-first-strand"</span>,
  outfile <span class="op">=</span> <span class="cn">NULL</span>,
  bedidx <span class="op">=</span> <span class="cn">NULL</span>,
  bam_flags <span class="op">=</span> <span class="cn">NULL</span>,
  n_align <span class="op">=</span> <span class="fl">0</span>,
  n_align_tag <span class="op">=</span> <span class="st">"NH"</span>,
  event_filters <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span>,
  only_keep_variants <span class="op">=</span> <span class="cn">FALSE</span>,
  reads <span class="op">=</span> <span class="cn">NULL</span>,
  return_data <span class="op">=</span> <span class="cn">TRUE</span>
<span class="op">)</span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>bamfile</dt>
<dd><p>paths to 1 or 2 bam files, pass 2 files to</p></dd>
<dt>fafile</dt>
<dd><p>path to fasta file</p></dd>
<dt>bedfile</dt>
<dd><p>path to bed file with sites or regions to query</p></dd>
<dt>region</dt>
<dd><p>samtools region query string (i.e. chr1:100-1000)</p></dd>
<dt>chrom</dt>
<dd><p>chrom to process, not to be used with region</p></dd>
<dt>start</dt>
<dd><p>start position, not to be used with region</p></dd>
<dt>end</dt>
<dd><p>end position, not to be used with region</p></dd>
<dt>min_reads</dt>
<dd><p>min read depth needed to report site</p></dd>
<dt>min_base_qual</dt>
<dd><p>min base quality score to consider read for pileup</p></dd>
<dt>min_mapq</dt>
<dd><p>minimum required MAPQ score</p></dd>
<dt>max_depth</dt>
<dd><p>maximum read depth considered at each site</p></dd>
<dt>library_type</dt>
<dd><p>read orientation, one of fr-first-strand,
fr-second-strand, or unstranded.</p></dd>
<dt>outfile</dt>
<dd><p>output tabix indexed file?</p></dd>
<dt>bedidx</dt>
<dd><p>A BedFile object, if supplied, pileup will use index
generated by <code><a href="indexBed.html">indexBed()</a></code></p></dd>
<dt>bam_flags</dt>
<dd><p>bam flags to filter or keep, use <code><a href="https://rdrr.io/pkg/Rsamtools/man/ScanBamParam-class.html" class="external-link">Rsamtools::scanBamFlag()</a></code>
to set</p></dd>
<dt>n_align</dt>
<dd><p>The number of valid alignments to use for each read, based on tag value
supplied to n_align_tag. NULL disables filtering by n_align_tag
value</p></dd>
<dt>n_align_tag</dt>
<dd><p>Tag name that containing # of valid alignments</p></dd>
<dt>event_filters</dt>
<dd><p>integer vector of length 4 with values indicating
distance from pileup position to features to filter. Distance to trim from
5' ends of reads, from 3' end of reads, distance from splicing events, distance from indels, and
length of homopolymer stretch used to determine if a site is in a homopolymer.
The number of read mismatch types (e.g. A-&gt;G or G-&gt;T) allowed in a single read,
and the required # of any mismatches that activate the mismatch type filter.
default is no filters applied (e.g. c(0,0,0,0,0,0,0)).</p></dd>
<dt>only_keep_variants</dt>
<dd><p>if TRUE, then only variant sites will be reported
(FALSE by default)</p></dd>
<dt>reads</dt>
<dd><p>if supplied a fasta file will be written with reads that pass filters
and contain variants</p></dd>
<dt>return_data</dt>
<dd><p>return data as a Granges table?</p></dd>
</dl></div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>To compare RNA-seq to DNA-seq 2 bam files can be supplied to the <code>bamfile</code>
argument, with the first corresponding to RNA-seq and second DNA-seq. The <code>min_reads</code>
and <code>min_mapq</code> arguments can accept 2 values (as numeric vector) corresponding to filters
for RNA and DNA-seq respectively. The <code>min_base_qual</code>, <code>n_align</code>,
<code>n_align_tag</code>, and <code>event_filters</code> params will be applied to both files.</p>
<p>The library type argument will only apply to the 1st bam file. The second will be set to
"unstranded"</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>An R package developed by the <a href="https://medschool.cuanschutz.edu/rbi" class="external-link">RNA Bioscience Initiative</a></p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.3.</p>
</div>

    </footer></div>

  

  

  </body></html>

