<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Generate base counts using pileup"><title>Generate base counts using pileup — get_pileup • raer</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><link href="../deps/Source_Sans_Pro-0.4.5/font.css" rel="stylesheet"><link href="../deps/Source_Code_Pro-0.4.5/font.css" rel="stylesheet"><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Generate base counts using pileup — get_pileup"><meta property="og:description" content="Generate base counts using pileup"><meta property="og:image" content="https://rnabioco.github.io/raer/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><script defer data-domain="rnabioco.github.io" src="https://plausible.io/js/plausible.js"></script></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-dark"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">raer</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/raer.html">Get started</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/novel-sites.html">Identifying RNA editing sites using RNA and DNA sequencing data</a>
    <a class="dropdown-item" href="../articles/single-cell.html">Identifying RNA editing sites in droplet single cell datasets</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/rnabioco/raer/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Generate base counts using pileup</h1>
      <small class="dont-index">Source: <a href="https://github.com/rnabioco/raer/blob/HEAD/R/pileup.R" class="external-link"><code>R/pileup.R</code></a></small>
      <div class="d-none name"><code>get_pileup.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Generate base counts using pileup</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">get_pileup</span><span class="op">(</span></span>
<span>  <span class="va">bamfiles</span>,</span>
<span>  <span class="va">fafile</span>,</span>
<span>  bedfile <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  region <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  chroms <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  filterParam <span class="op">=</span> <span class="fu">FilterParam</span><span class="op">(</span><span class="op">)</span>,</span>
<span>  outfile_prefix <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  bedidx <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  bam_flags <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  reads <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  return_data <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  BPPARAM <span class="op">=</span> <span class="fu">SerialParam</span><span class="op">(</span><span class="op">)</span>,</span>
<span>  use_index <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  bad_reads <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  umi_tag <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">FilterParam</span><span class="op">(</span></span>
<span>  max_depth <span class="op">=</span> <span class="fl">10000</span>,</span>
<span>  min_base_quality <span class="op">=</span> <span class="fl">20L</span>,</span>
<span>  min_mapq <span class="op">=</span> <span class="fl">0L</span>,</span>
<span>  min_nucleotide_depth <span class="op">=</span> <span class="fl">1L</span>,</span>
<span>  library_type <span class="op">=</span> <span class="st">"fr-first-strand"</span>,</span>
<span>  only_keep_variants <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  trim_5p <span class="op">=</span> <span class="fl">0L</span>,</span>
<span>  trim_3p <span class="op">=</span> <span class="fl">0L</span>,</span>
<span>  indel_dist <span class="op">=</span> <span class="fl">0L</span>,</span>
<span>  splice_dist <span class="op">=</span> <span class="fl">0L</span>,</span>
<span>  homopolymer_len <span class="op">=</span> <span class="fl">0L</span>,</span>
<span>  max_mismatch_type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0L</span>, <span class="fl">0L</span><span class="op">)</span>,</span>
<span>  min_read_bqual <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span>,</span>
<span>  min_splice_overhang <span class="op">=</span> <span class="fl">0L</span>,</span>
<span>  min_variant_reads <span class="op">=</span> <span class="fl">0L</span>,</span>
<span>  ignore_query_Ns <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>bamfiles</dt>
<dd><p>paths to 1 or more bam files</p></dd>


<dt>fafile</dt>
<dd><p>path to fasta file</p></dd>


<dt>bedfile</dt>
<dd><p>path to bed file with sites or regions to query</p></dd>


<dt>region</dt>
<dd><p>samtools region query string (i.e. chr1:100-1000)</p></dd>


<dt>chroms</dt>
<dd><p>chromosomes to process, not to be used with region</p></dd>


<dt>filterParam</dt>
<dd><p>object of class <code>FilterParam()</code> which specify various
filters to apply to reads and sites during pileup.</p></dd>


<dt>outfile_prefix</dt>
<dd><p>Output prefix for tabix indexed files. If <code>NULL</code>, no
files will be produced.</p></dd>


<dt>bedidx</dt>
<dd><p>A BedFile object, if supplied, pileup will use index generated
by <code><a href="index_bed.html">indexBed()</a></code></p></dd>


<dt>bam_flags</dt>
<dd><p>bam flags to filter or keep, use <code><a href="https://rdrr.io/pkg/Rsamtools/man/ScanBamParam-class.html" class="external-link">Rsamtools::scanBamFlag()</a></code>
to generate</p></dd>


<dt>reads</dt>
<dd><p>if supplied a fasta file will be written with reads that pass
filters and contain variants</p></dd>


<dt>return_data</dt>
<dd><p>if <code>TRUE</code>, data is returned as a GRanges, if <code>FALSE</code> a
character vector of tabix-index files, specified by <code>outfile_prefix</code>, will
be returned.</p></dd>


<dt>BPPARAM</dt>
<dd><p>A BiocParallel class to control parallel execution. Parallel
processing occurs per chromosome, so is disabled when run on a single
region.</p></dd>


<dt>use_index</dt>
<dd><p>if TRUE regions supplied in the <code>bedfile</code> will be queried
using the bam file index. By default (FALSE) all alignments are queried via
streaming and regions in the bedfile are used to select the sites to keep.
Using the index can be much faster when querying a small number of sites in
large bamfiles. With many sites in the bedfile , or with small bamfiles,
this option may be slower than streaming.</p></dd>


<dt>bad_reads</dt>
<dd><p>a textfile containing read names to exclude from pileup.
Readnames should be formated as readid_1 or readid_2 or readid for paired
end first read paired-end second read or single end data.</p></dd>


<dt>umi_tag</dt>
<dd><p>The bam tag containing a UMI sequence. If supplied, multiple
reads with the same UMI sequence will only be counted once per position.</p></dd>


<dt>verbose</dt>
<dd><p>if TRUE, then report progress and warnings.</p></dd>


<dt>max_depth</dt>
<dd><p>maximum read depth considered at each site</p></dd>


<dt>min_base_quality</dt>
<dd><p>min base quality score to consider read for pileup</p></dd>


<dt>min_mapq</dt>
<dd><p>minimum required MAPQ score, can be a vector of values
for each bam file</p></dd>


<dt>min_nucleotide_depth</dt>
<dd><p>min read depth needed to report site</p></dd>


<dt>library_type</dt>
<dd><p>read orientation, one of fr-first-strand,
fr-second-strand, unstranded, and genomic-unstranded. Can supply as a vector to specify for each
input bam. Unstranded library type will be reported based on read alignment.
genomic-unstranded will report all variants w.r.t the + strand.</p></dd>


<dt>only_keep_variants</dt>
<dd><p>if TRUE, then only variant sites will be reported
(FALSE by default), can be a vector for each input bamfile</p></dd>


<dt>trim_5p</dt>
<dd><p>Bases to trim from 5' ends of read alignments</p></dd>


<dt>trim_3p</dt>
<dd><p>Bases to trim from 3' ends of read alignments</p></dd>


<dt>indel_dist</dt>
<dd><p>Exclude read if site occurs within given
distance from indel event in the read</p></dd>


<dt>splice_dist</dt>
<dd><p>Exclude read if site occurs within given
distance from splicing event in the read</p></dd>


<dt>homopolymer_len</dt>
<dd><p>Exclude site if occurs within homopolymer of given
length</p></dd>


<dt>max_mismatch_type</dt>
<dd><p>Exclude read if it has X different mismatch types
(e.g A-to-G, G-to-C, C-to-G, is 3 mismatch types) or Y # of mismatches,
must be supplied as a integer vector of length 2. e.g.
c(X, Y).</p></dd>


<dt>min_read_bqual</dt>
<dd><p>Exclude read if more than X percent of the bases have
base qualities less than Y. Numeric vector of length 2. e.g. c(0.25, 20)</p></dd>


<dt>min_splice_overhang</dt>
<dd><p>Exclude read if site is located adjacent to splice
site with an overhang of less than given length.</p></dd>


<dt>min_variant_reads</dt>
<dd><p>Required number of reads containing a variant for a site
to be reported. Calculated per bam file, such that if 1 bam file has &gt;= min_variant_reads,
then the site will be reported.</p></dd>


<dt>ignore_query_Ns</dt>
<dd><p>ignored for now</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>A list containing a <code>GRanges</code> object for each input bam file, or a
vector of the output tabixed file names if <code>return_data</code> is FALSE.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>Multiple bam files can be processed together, with files being
written for each bam file. In this mode the output regions will be
consistent across all files. The min_mapq, only_keep_variants, and
library_type parameters can be specified for each input files.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p>Other pileup: 
<code><a href="merge_pileups.html">merge_pileups</a>()</code>,
<code><a href="pileup_cells.html">pileup_cells</a>()</code></p></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">bamfn</span> <span class="op">&lt;-</span> <span class="fu"><a href="raer_example.html">raer_example</a></span><span class="op">(</span><span class="st">"SRR5564269_Aligned.sortedByCoord.out.md.bam"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">bam2fn</span> <span class="op">&lt;-</span> <span class="fu"><a href="raer_example.html">raer_example</a></span><span class="op">(</span><span class="st">"SRR5564277_Aligned.sortedByCoord.out.md.bam"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">fafn</span> <span class="op">&lt;-</span> <span class="fu"><a href="raer_example.html">raer_example</a></span><span class="op">(</span><span class="st">"human.fasta"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">plp</span> <span class="op">&lt;-</span> <span class="fu">get_pileup</span><span class="op">(</span><span class="va">bamfn</span>, <span class="va">fafn</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">plps</span> <span class="op">&lt;-</span> <span class="fu">get_pileup</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">bamfn</span>, <span class="va">bam2fn</span><span class="op">)</span>, <span class="va">fafn</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">fp</span> <span class="op">&lt;-</span> <span class="fu">FilterParam</span><span class="op">(</span>only_keep_variants <span class="op">=</span> <span class="cn">TRUE</span>, min_nucleotide_depth <span class="op">=</span> <span class="fl">55</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">get_pileup</span><span class="op">(</span><span class="va">bamfn</span>, <span class="va">fafn</span>, filterParam <span class="op">=</span> <span class="va">fp</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GRanges object with 7 ranges and 10 metadata columns:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       seqnames    ranges strand |         Ref         Var      nRef</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;character&gt; &lt;integer&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [1]     DHFR       207      - |           G          GT        55</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [2]     DHFR       208      - |           T          TG        55</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [3]     DHFR       223      - |           G          GT        56</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [4]     DHFR       225      - |           T          TA        55</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [5]     DHFR       291      - |           A          AG        54</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [6]     DHFR       300      - |           A          AG        56</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [7]     DHFR       332      - |           A          AG        55</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>            nVar        nA        nT        nC        nG        nN</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [1]         1         0         1         0        55         0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [2]         1         0        55         0         1         0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [3]         1         0         1         0        56         0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [4]         1         1        55         0         0         0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [5]         2        54         0         0         2         0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [6]         1        56         0         0         1         0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [7]         1        55         0         0         1         0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>              nX</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       &lt;integer&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [1]         4</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [2]         4</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [3]         6</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [4]         7</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [5]         6</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [6]         4</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [7]         2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   -------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   seqinfo: 3 sequences from an unspecified genome</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>An R package developed by the <a href="https://medschool.cuanschutz.edu/rbi" class="external-link">RNA Bioscience Initiative</a></p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

