<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Function reference â€¢ raer</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><link href="../deps/Source_Sans_Pro-0.4.7/font.css" rel="stylesheet"><link href="../deps/Source_Code_Pro-0.4.7/font.css" rel="stylesheet"><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><meta property="og:image" content="https://rnabioco.github.io/raer/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><script defer data-domain="rnabioco.github.io" src="https://plausible.io/js/plausible.js"></script></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-dark"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">raer</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.14</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/raer.html">Get started</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/rnabioco/raer/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Function reference</h1>
    </div>

    <div class="section level2">
      <h2 id="pileup">Pileup<a class="anchor" aria-label="anchor" href="#pileup"></a></h2>
      
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="pileup_cells.html">pileup_cells()</a></code> 
        </dt>
        <dd>Pileup sites per cell</dd>
      </dl><dl><dt>
          
          <code><a href="pileup_sites.html">pileup_sites()</a></code> <code><a href="pileup_sites.html">FilterParam()</a></code> 
        </dt>
        <dd>Generate base counts using pileup</dd>
      </dl></div><div class="section level2">
      <h2 id="calculations">Calculations<a class="anchor" aria-label="anchor" href="#calculations"></a></h2>
      
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="calc_AEI.html">calc_AEI()</a></code> 
        </dt>
        <dd>Calculate the Adenosine Editing Index (AEI)</dd>
      </dl><dl><dt>
          
          <code><a href="calc_confidence.html">calc_confidence()</a></code> 
        </dt>
        <dd>Calculate confidence score for observing editing</dd>
      </dl><dl><dt>
          
          <code><a href="calc_edit_frequency.html">calc_edit_frequency()</a></code> 
        </dt>
        <dd>Adds editing frequencies</dd>
      </dl><dl><dt>
          
          <code><a href="calc_scAEI.html">calc_scAEI()</a></code> <code><a href="calc_scAEI.html">get_scAEI_sites()</a></code> 
        </dt>
        <dd>Calculate the Adenosine Editing Index (AEI) in single cells</dd>
      </dl></div><div class="section level2">
      <h2 id="filtering">Filtering<a class="anchor" aria-label="anchor" href="#filtering"></a></h2>
      
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="filter_clustered_variants.html">filter_clustered_variants()</a></code> 
        </dt>
        <dd>Filter out clustered sequence variants</dd>
      </dl><dl><dt>
          
          <code><a href="filter_multiallelic.html">filter_multiallelic()</a></code> 
        </dt>
        <dd>Filter out multi-allelic sites</dd>
      </dl><dl><dt>
          
          <code><a href="filter_splice_variants.html">filter_splice_variants()</a></code> 
        </dt>
        <dd>Filter out sites near splice sites</dd>
      </dl></div><div class="section level2">
      <h2 id="annotation">Annotation<a class="anchor" aria-label="anchor" href="#annotation"></a></h2>
      
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="annot_from_gr.html">annot_from_gr()</a></code> 
        </dt>
        <dd>Annotate sites using GRanges object</dd>
      </dl><dl><dt>
          
          <code><a href="annot_snps.html">annot_snps()</a></code> 
        </dt>
        <dd>Annotate known SNP positions</dd>
      </dl><dl><dt>
          
          <code><a href="get_overlapping_snps.html">get_overlapping_snps()</a></code> 
        </dt>
        <dd>Retrieve SNPs overlapping intervals</dd>
      </dl><dl><dt>
          
          <code><a href="get_splice_sites.html">get_splice_sites()</a></code> 
        </dt>
        <dd>Extract regions surrounding splice sites</dd>
      </dl></div><div class="section level2">
      <h2 id="bulk-editing-analysis">Bulk editing analysis<a class="anchor" aria-label="anchor" href="#bulk-editing-analysis"></a></h2>
      
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="find_de_sites.html">find_de_sites()</a></code> 
        </dt>
        <dd>Perform differential editing</dd>
      </dl><dl><dt>
          
          <code><a href="make_de_object.html">make_de_object()</a></code> 
        </dt>
        <dd>Make summarized experiment object for differential editing analysis</dd>
      </dl></div><div class="section level2">
      <h2 id="single-cell-editing-analysis">Single cell editing analysis<a class="anchor" aria-label="anchor" href="#single-cell-editing-analysis"></a></h2>
      
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="find_scde_sites.html">find_scde_sites()</a></code> 
        </dt>
        <dd>Identify sites with differential editing between cells in single cell datasets</dd>
      </dl><dl><dt>
          
          <code><a href="find_mispriming_sites.html">find_mispriming_sites()</a></code> 
        </dt>
        <dd>Find regions with oligodT mispriming</dd>
      </dl></div><div class="section level2">
      <h2 id="utilities">Utilities<a class="anchor" aria-label="anchor" href="#utilities"></a></h2>
      
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="correct_strand.html">correct_strand()</a></code> 
        </dt>
        <dd>Apply strand correction using gene annotations</dd>
      </dl><dl><dt>
          
          <code><a href="calc_scAEI.html">calc_scAEI()</a></code> <code><a href="calc_scAEI.html">get_scAEI_sites()</a></code> 
        </dt>
        <dd>Calculate the Adenosine Editing Index (AEI) in single cells</dd>
      </dl><dl><dt>
          
          <code><a href="mock_rse.html">mock_rse()</a></code> 
        </dt>
        <dd>Generate a small RangedSummarizedExperiment object for tests and examples</dd>
      </dl><dl><dt>
          
          <code><a href="raer_example.html">raer_example()</a></code> 
        </dt>
        <dd>Provide working directory for raer example files.</dd>
      </dl><dl><dt>
          
          <code><a href="read_sparray.html">read_sparray()</a></code> 
        </dt>
        <dd>Read sparseMatrix produced by pileup_cells()</dd>
      </dl></div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>An R package developed by the <a href="https://medschool.cuanschutz.edu/rbi" class="external-link">RNA Bioscience Initiative</a></p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

