<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="This function performs a pileup operation at specified
sites, returning counts for Reference (e.g. unedited) or Alternate (e.g.
edited) bases. pileup_cells can process either a 10x Genomic's style
library, from a aligned BAM file containing a tag with a cell-barcode and a
tag with a UMI, or smart-seq style libraries without cell-barcodes.
The sites parameter specifies sites to pileup. This must be a GRanges
object with 1 base intervals, a strand (+ or -), and supplemented with
metadata columns named REF and ALT containing the reference and
alternate base to query. See examples for an example format.
At each site, bases from overlapping reads will be examined, and counts of
each ref and alt base enumerated for each cell-barcode present. A single
base will be counted once for each UMI sequence present in each cell."><title>Pileup sites per cell — pileup_cells • raer</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><link href="../deps/Source_Sans_Pro-0.4.6/font.css" rel="stylesheet"><link href="../deps/Source_Code_Pro-0.4.6/font.css" rel="stylesheet"><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Pileup sites per cell — pileup_cells"><meta property="og:description" content="This function performs a pileup operation at specified
sites, returning counts for Reference (e.g. unedited) or Alternate (e.g.
edited) bases. pileup_cells can process either a 10x Genomic's style
library, from a aligned BAM file containing a tag with a cell-barcode and a
tag with a UMI, or smart-seq style libraries without cell-barcodes.
The sites parameter specifies sites to pileup. This must be a GRanges
object with 1 base intervals, a strand (+ or -), and supplemented with
metadata columns named REF and ALT containing the reference and
alternate base to query. See examples for an example format.
At each site, bases from overlapping reads will be examined, and counts of
each ref and alt base enumerated for each cell-barcode present. A single
base will be counted once for each UMI sequence present in each cell."><meta property="og:image" content="https://rnabioco.github.io/raer/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><script defer data-domain="rnabioco.github.io" src="https://plausible.io/js/plausible.js"></script></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-dark"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">raer</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.3</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/raer.html">Get started</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/novel-sites.html">Identifying RNA editing sites using RNA and DNA sequencing data</a>
    <a class="dropdown-item" href="../articles/single-cell.html">Identifying RNA editing sites in droplet single cell datasets</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/rnabioco/raer/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Pileup sites per cell</h1>
      <small class="dont-index">Source: <a href="https://github.com/rnabioco/raer/blob/HEAD/R/sc-pileup.R" class="external-link"><code>R/sc-pileup.R</code></a></small>
      <div class="d-none name"><code>pileup_cells.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function performs a pileup operation at specified
sites, returning counts for Reference (e.g. unedited) or Alternate (e.g.
edited) bases. <code>pileup_cells</code> can process either a 10x Genomic's style
library, from a aligned BAM file containing a tag with a cell-barcode and a
tag with a UMI, or smart-seq style libraries without cell-barcodes.</p>
<p>The <code>sites</code> parameter specifies sites to pileup. This must be a <a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a>
object with 1 base intervals, a strand (+ or -), and supplemented with
metadata columns named <code>REF</code> and <code>ALT</code> containing the reference and
alternate base to query. See examples for an example format.</p>
<p>At each site, bases from overlapping reads will be examined, and counts of
each ref and alt base enumerated for each cell-barcode present. A single
base will be counted once for each UMI sequence present in each cell.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">pileup_cells</span><span class="op">(</span></span>
<span>  <span class="va">bamfiles</span>,</span>
<span>  <span class="va">sites</span>,</span>
<span>  <span class="va">cell_barcodes</span>,</span>
<span>  <span class="va">output_directory</span>,</span>
<span>  chroms <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  umi_tag <span class="op">=</span> <span class="st">"UB"</span>,</span>
<span>  cb_tag <span class="op">=</span> <span class="st">"CB"</span>,</span>
<span>  paired_end <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  param <span class="op">=</span> <span class="fu"><a href="pileup_sites.html">FilterParam</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  BPPARAM <span class="op">=</span> <span class="fu">SerialParam</span><span class="op">(</span><span class="op">)</span>,</span>
<span>  return_sce <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>bamfiles</dt>
<dd><p>a path to a BAM file (for 10x libraries), or a vector of paths
to BAM files (smart-seq2). Can be supplied as a character vector, <a href="https://rdrr.io/pkg/Rsamtools/man/BamFile-class.html" class="external-link">BamFile</a>, or
<a href="https://rdrr.io/pkg/Rsamtools/man/BamFile-class.html" class="external-link">BamFileList</a>.</p></dd>


<dt>sites</dt>
<dd><p>a GRanges object containing sites to process. See examples for
valid formatting.</p></dd>


<dt>cell_barcodes</dt>
<dd><p>A character vector of single cell barcodes to process. If
processing multiple BAM files (e.g. smart-seq-2), provide a character vector
of unique identifiers for each input BAM, to name each BAM file in the output files.</p></dd>


<dt>output_directory</dt>
<dd><p>Output directory for output matrix files. The directory
will be generated if it doesn't exist.</p></dd>


<dt>chroms</dt>
<dd><p>A character vector of chromosomes to process. If supplied, only
sites present in the listed chromosomes will be processed</p></dd>


<dt>umi_tag</dt>
<dd><p>tag in BAM containing the UMI sequence</p></dd>


<dt>cb_tag</dt>
<dd><p>tag in BAM containing the cell-barcode sequence</p></dd>


<dt>paired_end</dt>
<dd><p>set to TRUE if data is paired-end to prevent double counting
overlapping read pairs.</p></dd>


<dt>param</dt>
<dd><p>object of class <code><a href="pileup_sites.html">FilterParam()</a></code> which specify various filters to
apply to reads and sites during pileup. Note that the <code>min_variant_reads</code>
parameter, if set &gt; 0, specifies the number of variant reads at a site
required in order to report a site. E.g. if set to 2, then at least 2 reads
(from any cell) must have a variant in order to report the site. The
default of 0 reports all sites present in the <code>sites</code> object.</p></dd>


<dt>BPPARAM</dt>
<dd><p>BiocParallel instance. Parallel computation occurs across
chromosomes.</p></dd>


<dt>return_sce</dt>
<dd><p>if <code>TRUE</code>, data is returned as a SingleCellExperiment, if
<code>FALSE</code> a character vector of the output files, specified by
<code>outfile_prefix</code>, will be returned.</p></dd>


<dt>verbose</dt>
<dd><p>Display messages</p></dd>


<dt>...</dt>
<dd><p>For the generic, further arguments to pass to specific methods.
Unused for now.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>Returns either a SingleCellExperiment or character vector of paths
to the sparseMatrix files produced. The SingleCellExperiment object is
populated with two assays, <code>nRef</code> and <code>nAlt</code>, which represent base counts
for the reference and alternate alleles. The <code>rowRanges()</code> will contain the
genomic interval for each site, along with <code>REF</code> and <code>ALT</code> columns. The
rownames will be populated with the format
<code>site_[seqnames]_[position(1-based)]_[strand]</code>, with <code>strand</code> being encoded
as 1 = +, 2 = -, and 3 = *.</p>


<p>If <code>return_sce</code> is <code>FALSE</code> then a character vector of paths to the
sparseMatrix files (<code>barcodes.txt.gz</code>, <code>sites.txt.gz</code>, <code>counts.mtx.gz</code>),
will be returned. These files can be imported using <code><a href="read_sparray.html">read_sparray()</a></code>.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p>Other pileup: 
<code><a href="pileup_sites.html">pileup_sites</a>()</code></p></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/Rsamtools" class="external-link">Rsamtools</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/GenomicRanges" class="external-link">GenomicRanges</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">bam_fn</span> <span class="op">&lt;-</span> <span class="fu"><a href="raer_example.html">raer_example</a></span><span class="op">(</span><span class="st">"5k_neuron_mouse_possort.bam"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">gr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"2:579:-"</span>, <span class="st">"2:625:-"</span>, <span class="st">"2:645:-"</span>, <span class="st">"2:589:-"</span>, <span class="st">"2:601:-"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">gr</span><span class="op">$</span><span class="va">REF</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"A"</span>, <span class="fl">4</span><span class="op">)</span>, <span class="st">"T"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">gr</span><span class="op">$</span><span class="va">ALT</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"G"</span>, <span class="fl">4</span><span class="op">)</span>, <span class="st">"C"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">cbs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Rsamtools/man/scanBam.html" class="external-link">scanBam</a></span><span class="op">(</span><span class="va">bam_fn</span>, param <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Rsamtools/man/ScanBamParam-class.html" class="external-link">ScanBamParam</a></span><span class="op">(</span>tag <span class="op">=</span> <span class="st">"CB"</span><span class="op">)</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">tag</span><span class="op">$</span><span class="va">CB</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">cbs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/na.fail.html" class="external-link">na.omit</a></span><span class="op">(</span><span class="va">cbs</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">outdir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">bai</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Rsamtools/man/scanBam.html" class="external-link">indexBam</a></span><span class="op">(</span><span class="va">bam_fn</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">fp</span> <span class="op">&lt;-</span> <span class="fu"><a href="pileup_sites.html">FilterParam</a></span><span class="op">(</span>library_type <span class="op">=</span> <span class="st">"fr-second-strand"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">pileup_cells</span><span class="op">(</span><span class="va">bam_fn</span>, <span class="va">gr</span>, <span class="va">cbs</span>, <span class="va">outdir</span>, param <span class="op">=</span> <span class="va">fp</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">sce</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> class: SingleCellExperiment </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> dim: 5 556 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> metadata(0):</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> assays(2): nRef nAlt</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rownames(5): site_2_579_2 site_2_625_2 site_2_645_2 site_2_589_2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   site_2_601_2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rowData names(2): REF ALT</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colnames(556): TGGAACTCAAGCTGTT-1 TACTTCAGTAACCCTA-1 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   TGTACAGTCTTCGTGC-1 TGTTGAGGTGACTGAG-1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colData names(0):</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> reducedDimNames(0):</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> mainExpName: NULL</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> altExpNames(0):</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># example of processing multiple smart-seq2 style libraries</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">many_small_bams</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="va">bam_fn</span>, <span class="fl">10</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">bam_ids</span> <span class="op">&lt;-</span> <span class="va">LETTERS</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span></span></span>
<span class="r-in"><span><span class="fu">pileup_cells</span><span class="op">(</span><span class="va">many_small_bams</span>,</span></span>
<span class="r-in"><span>    sites <span class="op">=</span> <span class="va">gr</span>,</span></span>
<span class="r-in"><span>    cell_barcodes <span class="op">=</span> <span class="va">bam_ids</span>,</span></span>
<span class="r-in"><span>    cb_tag <span class="op">=</span> <span class="cn">NULL</span>,</span></span>
<span class="r-in"><span>    umi_tag <span class="op">=</span> <span class="cn">NULL</span>,</span></span>
<span class="r-in"><span>    paired_end <span class="op">=</span> <span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span>    <span class="va">outdir</span>,</span></span>
<span class="r-in"><span>    param <span class="op">=</span> <span class="va">fp</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> class: SingleCellExperiment </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> dim: 5 10 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> metadata(0):</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> assays(2): nRef nAlt</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rownames(5): site_2_579_2 site_2_625_2 site_2_645_2 site_2_589_2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   site_2_601_2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rowData names(2): REF ALT</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colnames(10): A B ... I J</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colData names(0):</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> reducedDimNames(0):</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> mainExpName: NULL</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> altExpNames(0):</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/unlink.html" class="external-link">unlink</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">outdir</span>, <span class="va">bai</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>An R package developed by the <a href="https://medschool.cuanschutz.edu/rbi" class="external-link">RNA Bioscience Initiative</a></p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

