<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="raer">
<title>Identifying RNA editing sites using RNA and DNA sequencing data • raer</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><link href="../deps/Source_Sans_Pro-0.4.6/font.css" rel="stylesheet">
<link href="../deps/Source_Code_Pro-0.4.6/font.css" rel="stylesheet">
<!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Identifying RNA editing sites using RNA and DNA sequencing data">
<meta property="og:description" content="raer">
<meta property="og:image" content="https://rnabioco.github.io/raer/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><script defer data-domain="rnabioco.github.io" src="https://plausible.io/js/plausible.js"></script>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-dark"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">raer</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.2</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/raer.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/novel-sites.html">Identifying RNA editing sites using RNA and DNA sequencing data</a>
    <a class="dropdown-item" href="../articles/single-cell.html">Identifying RNA editing sites in droplet single cell datasets</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/rnabioco/raer/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Identifying RNA editing sites using RNA and DNA sequencing data</h1>
                        <h4 data-toc-skip class="author">Kent
Riemondy</h4>
            <address class="author_afil">
      University of Colorado School of Medicine<br><h4 data-toc-skip class="date">2023-06-09</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/rnabioco/raer/blob/HEAD/vignettes/novel-sites.Rmd" class="external-link"><code>vignettes/novel-sites.Rmd</code></a></small>
      <div class="d-none name"><code>novel-sites.Rmd</code></div>
    </address>
</div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://rnabioco.github.io/raer">raer</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/GenomicFeatures" class="external-link">GenomicFeatures</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/SummarizedExperiment" class="external-link">SummarizedExperiment</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">rtracklayer</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">TxDb.Hsapiens.UCSC.hg38.knownGene</span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="novel-rna-editing-site-detection-tutorial">Novel RNA editing site detection tutorial<a class="anchor" aria-label="anchor" href="#novel-rna-editing-site-detection-tutorial"></a>
</h2>
<p>This vignette will demonstrate how to identify novel RNA editing
sites using the <code>raer</code> package.</p>
<p>In this vignette a public RNA-seq and Whole-Genome sequencing dataset
will be analyzed. High coverage whole-genome sequencing was conducted <a href="https://www.ebi.ac.uk/ena/browser/view/ERR262997?show=reads" class="external-link">ERR262997</a>
along with paired-end RNA-seq <a href="https://www.ebi.ac.uk/ena/browser/view/SRR1258218?show=reads" class="external-link">SRR1258218</a>.</p>
<p>Aligned BAM files and a genome fasta file have been preprocessed for
this vignette. These files occupy ~100 Mb of space and contain
alignments from the first 1MB of chromosome 4 and they are stored in a
<code>BiocFileCache</code>.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fns</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/download_NA12878.html">download_NA12878</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">fns</span></span></code></pre></div>
<pre><code><span><span class="co">## $bams</span></span>
<span><span class="co">## [1] "/github/home/.cache/R/raer/ERR262996_dedup_chr4_sub.bam"           </span></span>
<span><span class="co">## [2] "/github/home/.cache/R/raer/SRR1258218_Aligned.sorted.dedup_sub.bam"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $bai</span></span>
<span><span class="co">## [1] "/github/home/.cache/R/raer/ERR262996_dedup_chr4_sub.bam.bai"           </span></span>
<span><span class="co">## [2] "/github/home/.cache/R/raer/SRR1258218_Aligned.sorted.dedup_sub.bam.bai"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $fasta</span></span>
<span><span class="co">## [1] "/github/home/.cache/R/raer/5ca4a248ea7_hg38_chr4.fa.bgz"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $snps</span></span>
<span><span class="co">## [1] "/github/home/.cache/R/raer/5ca4d56135f_chr4snps.bed.gz"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $rmsk</span></span>
<span><span class="co">## [1] "/github/home/.cache/R/raer/5ca118ec74_rmsk_hg38.tsv.gz"</span></span></code></pre>
<p>Additionally we will use the following additional annotation
resources:</p>
<ul>
<li>A database of known SNPs, for example the
<code>SNPlocs.Hsapiens.dbSNP155.GRCh38</code> package. Due to space and
memory constraints , in this vignette we will use a BED file containing
SNPs from the first 1Mb region of chr4.<br>
</li>
<li>
<code>TxDb.Hsapiens.UCSC.hg38.knownGene</code>, a database of
transcript models. Alternatively these can be generated from a
<code>.gtf</code> file using <code><a href="https://rdrr.io/pkg/GenomicFeatures/man/makeTxDbFromGRanges.html" class="external-link">makeTxDbFromGRanges()</a></code> from the
<code>GenomicFeatures</code> package.<br>
</li>
<li>RepeatMasker annotations, which can be obtained from the
<code>AnnotationHub()</code> for hg38. For space reasons these
annotations for chr4 have been included as a text file.</li>
</ul>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">txdb</span> <span class="op">&lt;-</span> <span class="va">TxDb.Hsapiens.UCSC.hg38.knownGene</span></span>
<span></span>
<span><span class="va">chr4snps</span> <span class="op">&lt;-</span> <span class="fu">import</span><span class="op">(</span><span class="va">fns</span><span class="op">$</span><span class="va">snps</span><span class="op">)</span></span>
<span></span>
<span><span class="va">rmskhg38</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="va">fns</span><span class="op">$</span><span class="va">rmsk</span><span class="op">)</span></span>
<span><span class="va">rmskhg38</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/makeGRangesFromDataFrame.html" class="external-link">makeGRangesFromDataFrame</a></span><span class="op">(</span><span class="va">rmskhg38</span>,</span>
<span>    keep.extra.columns <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    starts.in.df.are.0based <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>The <code><a href="../reference/pileup_sites.html">pileup_sites()</a></code> function can accept multiple bam
files, here we supply one from RNA-seq, and one from whole genome
sequencing.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">wgs_bam</span> <span class="op">&lt;-</span> <span class="va">fns</span><span class="op">$</span><span class="va">bams</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span>
<span><span class="va">rna_bam</span> <span class="op">&lt;-</span> <span class="va">fns</span><span class="op">$</span><span class="va">bams</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span></span>
<span><span class="va">fafn</span> <span class="op">&lt;-</span> <span class="va">fns</span><span class="op">$</span><span class="va">fasta</span></span></code></pre></div>
<p>Filtering parameters for the <code><a href="../reference/pileup_sites.html">pileup_sites()</a></code> function can
accept multiple arguments matched to the input bam files.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pileup_sites.html">FilterParam</a></span><span class="op">(</span></span>
<span>    min_depth <span class="op">=</span> <span class="fl">1</span>,</span>
<span>    min_base_quality <span class="op">=</span> <span class="fl">30</span>,</span>
<span>    min_mapq <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">255</span>, <span class="fl">30</span><span class="op">)</span>,</span>
<span>    library_type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"fr-first-strand"</span>, <span class="st">"genomic-unstranded"</span><span class="op">)</span>,</span>
<span>    trim_5p <span class="op">=</span> <span class="fl">5</span>,</span>
<span>    trim_3p <span class="op">=</span> <span class="fl">5</span>,</span>
<span>    indel_dist <span class="op">=</span> <span class="fl">4</span>,</span>
<span>    homopolymer_len <span class="op">=</span> <span class="fl">6</span>,</span>
<span>    max_mismatch_type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">3</span>, <span class="fl">3</span><span class="op">)</span>,</span>
<span>    read_bqual <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.25</span>, <span class="fl">20</span><span class="op">)</span>,</span>
<span>    only_keep_variants <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="cn">TRUE</span>, <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">bams</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">rna_bam</span>, <span class="va">wgs_bam</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">bams</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"rna"</span>, <span class="st">"dna"</span><span class="op">)</span></span>
<span><span class="va">rse</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pileup_sites.html">pileup_sites</a></span><span class="op">(</span><span class="va">bams</span>,</span>
<span>    fafile <span class="op">=</span> <span class="va">fafn</span>,</span>
<span>    chrom <span class="op">=</span> <span class="st">"chr4"</span>,</span>
<span>    param <span class="op">=</span> <span class="va">fp</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">rse</span></span></code></pre></div>
<pre><code><span><span class="co">## class: RangedSummarizedExperiment </span></span>
<span><span class="co">## dim: 546 2 </span></span>
<span><span class="co">## metadata(0):</span></span>
<span><span class="co">## assays(7): ALT nRef ... nC nG</span></span>
<span><span class="co">## rownames(546): site_chr4_44338_2 site_chr4_124820_1 ...</span></span>
<span><span class="co">##   site_chr4_990981_1 site_chr4_993342_1</span></span>
<span><span class="co">## rowData names(4): REF rpbz vdb sor</span></span>
<span><span class="co">## colnames(2): rna dna</span></span>
<span><span class="co">## colData names(1): sample</span></span></code></pre>
<p>Next we filter to keep those sites with a variant in the RNA-seq, but
no variant in the DNA-seq, and a minimum of 5 reads in the DNA-seq. The
DNA-seq data will be reported on the “+” strand, where as the RNA-seq
data will be reported on either strand. We therefore use
<code>subsetByOverlaps(..., ignore.strand = TRUE)</code> to retain sites
passing these DNA-seq based filters independent of strand.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">to_keep</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">rse</span>, <span class="st">"nRef"</span><span class="op">)</span><span class="op">[</span>, <span class="st">"dna"</span><span class="op">]</span> <span class="op">&gt;=</span> <span class="fl">5</span> <span class="op">&amp;</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">rse</span>, <span class="st">"ALT"</span><span class="op">)</span><span class="op">[</span>, <span class="st">"dna"</span><span class="op">]</span> <span class="op">==</span> <span class="st">"-"</span><span class="op">)</span></span>
<span><span class="va">rse</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/findOverlaps-methods.html" class="external-link">subsetByOverlaps</a></span><span class="op">(</span><span class="va">rse</span>, <span class="va">rse</span><span class="op">[</span><span class="va">to_keep</span>, <span class="op">]</span>, ignore.strand <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">rse</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 198</span></span></code></pre>
<p>Next we filter to remove any multi-allelic sites. These sites are
stored as comma-separated strings in the <code>ALT</code> assay
(e.g. <code>G,C</code>). Non-variant sites are stored as <code>-</code>.
<code><a href="../reference/filter_multiallelic.html">filter_multiallelic()</a></code> will remove any sites that have
multiple variants in the samples present in the
<code>summarizedExperiment</code> object. It will add a new column to
the <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData()</a></code> to indicate the variant for each site, and
will calculate an <code>allele_freq</code> assay with variant allele
frequencies for each sample.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rse</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filter_multiallelic.html">filter_multiallelic</a></span><span class="op">(</span><span class="va">rse</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> `filter_multiallelic()`: removed <span style="color: #0000BB;">0</span> sites from <span style="color: #0000BB;">198</span> (<span style="color: #0000BB;">198</span> remain)</span></span></code></pre>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rse</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calc_edit_frequency.html">calc_edit_frequency</a></span><span class="op">(</span><span class="va">rse</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> 48 sites had no coverage for calculating editing</span></span></code></pre>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">rse</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 150 rows and 5 columns</span></span>
<span><span class="co">##                            REF      rpbz       vdb       sor         ALT</span></span>
<span><span class="co">##                    &lt;character&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;character&gt;</span></span>
<span><span class="co">## site_chr4_124820_1           C -0.939623       Inf  1.410919           A</span></span>
<span><span class="co">## site_chr4_125213_1           G -1.584745       Inf  0.367725           C</span></span>
<span><span class="co">## site_chr4_126885_1           A  0.826528       Inf  2.271553           G</span></span>
<span><span class="co">## site_chr4_126955_1           A  1.584745       Inf  1.506342           G</span></span>
<span><span class="co">## site_chr4_134331_1           A -1.868257       0.1  0.367725           G</span></span>
<span><span class="co">## ...                        ...       ...       ...       ...         ...</span></span>
<span><span class="co">## site_chr4_894077_1           G  1.289956       Inf  1.524254           A</span></span>
<span><span class="co">## site_chr4_952419_1           G -1.362512       Inf  0.550831           C</span></span>
<span><span class="co">## site_chr4_958447_1           C  0.694986       Inf  1.436657           A</span></span>
<span><span class="co">## site_chr4_990589_1           A -1.640049       Inf  1.570697           C</span></span>
<span><span class="co">## site_chr4_993342_1           A  1.498680       Inf  1.472727           G</span></span></code></pre>
<p>Next we’ll remove sites in simple repeat regions. We will add repeat
information to the <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData()</a></code> using the
<code><a href="../reference/annot_from_gr.html">annot_from_gr()</a></code> function.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rse</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/annot_from_gr.html">annot_from_gr</a></span><span class="op">(</span><span class="va">rse</span>, <span class="va">rmskhg38</span>, cols_to_map <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"repName"</span>, <span class="st">"repClass"</span>, <span class="st">"repFamily"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">rse</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"repName"</span>, <span class="st">"repFamily"</span><span class="op">)</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 150 rows and 2 columns</span></span>
<span><span class="co">##                        repName   repFamily</span></span>
<span><span class="co">##                    &lt;character&gt; &lt;character&gt;</span></span>
<span><span class="co">## site_chr4_124820_1          NA          NA</span></span>
<span><span class="co">## site_chr4_125213_1      MER21C        ERVL</span></span>
<span><span class="co">## site_chr4_126885_1          NA          NA</span></span>
<span><span class="co">## site_chr4_126955_1          NA          NA</span></span>
<span><span class="co">## site_chr4_134331_1          NA          NA</span></span>
<span><span class="co">## ...                        ...         ...</span></span>
<span><span class="co">## site_chr4_894077_1          NA          NA</span></span>
<span><span class="co">## site_chr4_952419_1          NA          NA</span></span>
<span><span class="co">## site_chr4_958447_1          NA          NA</span></span>
<span><span class="co">## site_chr4_990589_1          NA          NA</span></span>
<span><span class="co">## site_chr4_993342_1          NA          NA</span></span></code></pre>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rse</span> <span class="op">&lt;-</span> <span class="va">rse</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">rse</span><span class="op">)</span><span class="op">$</span><span class="va">repFamily</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Simple_repeat"</span>, <span class="st">"Low Complexity"</span><span class="op">)</span><span class="op">]</span></span></code></pre></div>
<p>Next we’ll remove sites adjacent to other sites with different
variant types. For example if an A-&gt;G variant is located proximal to
a C-&gt;T variant then the variants will be removed.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rse</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filter_clustered_variants.html">filter_clustered_variants</a></span><span class="op">(</span><span class="va">rse</span>, <span class="va">txdb</span>, variant_dist <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> `filter_clustered_variants()`: removed <span style="color: #0000BB;">22</span> sites from <span style="color: #0000BB;">149</span> (<span style="color: #0000BB;">127</span> remain)</span></span></code></pre>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rse</span></span></code></pre></div>
<pre><code><span><span class="co">## class: RangedSummarizedExperiment </span></span>
<span><span class="co">## dim: 127 2 </span></span>
<span><span class="co">## metadata(0):</span></span>
<span><span class="co">## assays(9): ALT nRef ... depth edit_freq</span></span>
<span><span class="co">## rownames(127): site_chr4_124820_1 site_chr4_125213_1 ...</span></span>
<span><span class="co">##   site_chr4_990589_1 site_chr4_993342_1</span></span>
<span><span class="co">## rowData names(8): REF rpbz ... repClass repFamily</span></span>
<span><span class="co">## colnames(2): rna dna</span></span>
<span><span class="co">## colData names(3): sample n_sites edit_idx</span></span></code></pre>
<p>Next, sites within 4bp of known splicing events will be excluded.
These sites are low-confidence due to the possibility of mis-alignments
near splicing events.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rse</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filter_splice_variants.html">filter_splice_variants</a></span><span class="op">(</span><span class="va">rse</span>, <span class="va">txdb</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> `filter_splice_variants()`: removed <span style="color: #0000BB;">2</span> sites from <span style="color: #0000BB;">127</span> (<span style="color: #0000BB;">125</span> remain)</span></span></code></pre>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rse</span></span></code></pre></div>
<pre><code><span><span class="co">## class: RangedSummarizedExperiment </span></span>
<span><span class="co">## dim: 125 2 </span></span>
<span><span class="co">## metadata(0):</span></span>
<span><span class="co">## assays(9): ALT nRef ... depth edit_freq</span></span>
<span><span class="co">## rownames(125): site_chr4_124820_1 site_chr4_125213_1 ...</span></span>
<span><span class="co">##   site_chr4_990589_1 site_chr4_993342_1</span></span>
<span><span class="co">## rowData names(8): REF rpbz ... repClass repFamily</span></span>
<span><span class="co">## colnames(2): rna dna</span></span>
<span><span class="co">## colData names(3): sample n_sites edit_idx</span></span></code></pre>
<p>Next, we’ll annotate if the site is a known SNP and remove any known
SNPs. If using a SNPlocs package you can use the
<code>annot_snp()</code> function. However we will use the
<code><a href="../reference/annot_from_gr.html">annot_from_gr()</a></code> function to annotate using the prebuilt
<code>chr4snps</code> object.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rse</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/annot_from_gr.html">annot_from_gr</a></span><span class="op">(</span><span class="va">rse</span>, <span class="va">chr4snps</span>, <span class="st">"name"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">rse</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"name"</span><span class="op">)</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 125 rows and 1 column</span></span>
<span><span class="co">##                            name</span></span>
<span><span class="co">##                     &lt;character&gt;</span></span>
<span><span class="co">## site_chr4_124820_1 rs1553807640</span></span>
<span><span class="co">## site_chr4_125213_1 rs1286128437</span></span>
<span><span class="co">## site_chr4_126885_1           NA</span></span>
<span><span class="co">## site_chr4_126955_1           NA</span></span>
<span><span class="co">## site_chr4_134331_1 rs1715464292</span></span>
<span><span class="co">## ...                         ...</span></span>
<span><span class="co">## site_chr4_894077_1           NA</span></span>
<span><span class="co">## site_chr4_952419_1 rs1577433212</span></span>
<span><span class="co">## site_chr4_958447_1 rs1052651854</span></span>
<span><span class="co">## site_chr4_990589_1           NA</span></span>
<span><span class="co">## site_chr4_993342_1           NA</span></span></code></pre>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rse</span> <span class="op">&lt;-</span> <span class="va">rse</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">rse</span><span class="op">)</span><span class="op">$</span><span class="va">name</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="va">rse</span></span></code></pre></div>
<pre><code><span><span class="co">## class: RangedSummarizedExperiment </span></span>
<span><span class="co">## dim: 78 2 </span></span>
<span><span class="co">## metadata(0):</span></span>
<span><span class="co">## assays(9): ALT nRef ... depth edit_freq</span></span>
<span><span class="co">## rownames(78): site_chr4_126885_1 site_chr4_126955_1 ...</span></span>
<span><span class="co">##   site_chr4_990589_1 site_chr4_993342_1</span></span>
<span><span class="co">## rowData names(9): REF rpbz ... repFamily name</span></span>
<span><span class="co">## colnames(2): rna dna</span></span>
<span><span class="co">## colData names(3): sample n_sites edit_idx</span></span></code></pre>
<p>Lastly, we’ll further filter the edit sites to require that the
editing frequency is &gt; 0.05 and that at least 2 reads support the
editing site.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">to_keep</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">rse</span>, <span class="st">"edit_freq"</span><span class="op">)</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span> <span class="op">&gt;</span> <span class="fl">0.05</span></span>
<span><span class="va">rse</span> <span class="op">&lt;-</span> <span class="va">rse</span><span class="op">[</span><span class="va">to_keep</span>, <span class="op">]</span></span>
<span></span>
<span><span class="va">rse</span> <span class="op">&lt;-</span> <span class="va">rse</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">rse</span>, <span class="st">"nAlt"</span><span class="op">)</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span> <span class="op">&gt;=</span> <span class="fl">2</span><span class="op">]</span></span></code></pre></div>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/stopifnot.html" class="external-link">stopifnot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/all.html" class="external-link">all</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">rse</span><span class="op">)</span><span class="op">$</span><span class="va">REF</span> <span class="op">==</span> <span class="st">"A"</span> <span class="op">&amp;</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">rse</span><span class="op">)</span><span class="op">$</span><span class="va">ALT</span> <span class="op">==</span> <span class="st">"G"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">rowRanges</span><span class="op">(</span><span class="va">rse</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 12 ranges and 9 metadata columns:</span></span>
<span><span class="co">##                      seqnames    ranges strand |         REF       rpbz</span></span>
<span><span class="co">##                         &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;  &lt;numeric&gt;</span></span>
<span><span class="co">##   site_chr4_134353_1     chr4    134353      + |           A -1.1811989</span></span>
<span><span class="co">##   site_chr4_377951_1     chr4    377951      + |           A  0.3595421</span></span>
<span><span class="co">##   site_chr4_378004_1     chr4    378004      + |           A -1.9668785</span></span>
<span><span class="co">##   site_chr4_378011_1     chr4    378011      + |           A -0.0481455</span></span>
<span><span class="co">##   site_chr4_378018_1     chr4    378018      + |           A -0.3099735</span></span>
<span><span class="co">##                  ...      ...       ...    ... .         ...        ...</span></span>
<span><span class="co">##   site_chr4_378368_1     chr4    378368      + |           A   1.096651</span></span>
<span><span class="co">##   site_chr4_378409_1     chr4    378409      + |           A   0.731101</span></span>
<span><span class="co">##   site_chr4_380392_1     chr4    380392      + |           A   0.332543</span></span>
<span><span class="co">##   site_chr4_778739_2     chr4    778739      - |           A   0.577350</span></span>
<span><span class="co">##   site_chr4_778780_2     chr4    778780      - |           A  -0.866025</span></span>
<span><span class="co">##                            vdb       sor         ALT     repName</span></span>
<span><span class="co">##                      &lt;numeric&gt; &lt;numeric&gt; &lt;character&gt; &lt;character&gt;</span></span>
<span><span class="co">##   site_chr4_134353_1  0.100000  0.527355           G        &lt;NA&gt;</span></span>
<span><span class="co">##   site_chr4_377951_1  0.580000  2.285077           G       AluSc</span></span>
<span><span class="co">##   site_chr4_378004_1  0.060000  0.464813           G       AluSc</span></span>
<span><span class="co">##   site_chr4_378011_1  0.840000  0.550831           G       AluSc</span></span>
<span><span class="co">##   site_chr4_378018_1  0.535497  1.070610           G       AluSc</span></span>
<span><span class="co">##                  ...       ...       ...         ...         ...</span></span>
<span><span class="co">##   site_chr4_378368_1  0.180000  0.268990           G        &lt;NA&gt;</span></span>
<span><span class="co">##   site_chr4_378409_1  0.580000  0.891998           G        &lt;NA&gt;</span></span>
<span><span class="co">##   site_chr4_380392_1  0.520000  0.424565           G      AluSq2</span></span>
<span><span class="co">##   site_chr4_778739_2  0.230191  0.668541           G       AluSg</span></span>
<span><span class="co">##   site_chr4_778780_2  0.303109  1.885595           G       AluSg</span></span>
<span><span class="co">##                         repClass   repFamily        name</span></span>
<span><span class="co">##                      &lt;character&gt; &lt;character&gt; &lt;character&gt;</span></span>
<span><span class="co">##   site_chr4_134353_1        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;</span></span>
<span><span class="co">##   site_chr4_377951_1        SINE         Alu        &lt;NA&gt;</span></span>
<span><span class="co">##   site_chr4_378004_1        SINE         Alu        &lt;NA&gt;</span></span>
<span><span class="co">##   site_chr4_378011_1        SINE         Alu        &lt;NA&gt;</span></span>
<span><span class="co">##   site_chr4_378018_1        SINE         Alu        &lt;NA&gt;</span></span>
<span><span class="co">##                  ...         ...         ...         ...</span></span>
<span><span class="co">##   site_chr4_378368_1        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;</span></span>
<span><span class="co">##   site_chr4_378409_1        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;</span></span>
<span><span class="co">##   site_chr4_380392_1        SINE         Alu        &lt;NA&gt;</span></span>
<span><span class="co">##   site_chr4_778739_2        SINE         Alu        &lt;NA&gt;</span></span>
<span><span class="co">##   site_chr4_778780_2        SINE         Alu        &lt;NA&gt;</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 194 sequences from an unspecified genome</span></span></code></pre>
<details style="margin-bottom:10px;"><summary>
Session info
</summary><div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.3.0 (2023-04-21)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">## Running under: Ubuntu 22.04.2 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] TxDb.Hsapiens.UCSC.hg38.knownGene_3.17.0</span></span>
<span><span class="co">##  [2] rtracklayer_1.60.0                      </span></span>
<span><span class="co">##  [3] SummarizedExperiment_1.30.2             </span></span>
<span><span class="co">##  [4] MatrixGenerics_1.12.1                   </span></span>
<span><span class="co">##  [5] matrixStats_1.0.0                       </span></span>
<span><span class="co">##  [6] GenomicFeatures_1.52.0                  </span></span>
<span><span class="co">##  [7] AnnotationDbi_1.62.1                    </span></span>
<span><span class="co">##  [8] Biobase_2.60.0                          </span></span>
<span><span class="co">##  [9] GenomicRanges_1.52.0                    </span></span>
<span><span class="co">## [10] GenomeInfoDb_1.36.0                     </span></span>
<span><span class="co">## [11] IRanges_2.34.0                          </span></span>
<span><span class="co">## [12] S4Vectors_0.38.1                        </span></span>
<span><span class="co">## [13] BiocGenerics_0.46.0                     </span></span>
<span><span class="co">## [14] raer_0.99.2                             </span></span>
<span><span class="co">## [15] BiocStyle_2.28.0                        </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##  [1] tidyselect_1.2.0            dplyr_1.1.2                </span></span>
<span><span class="co">##  [3] blob_1.2.4                  R.utils_2.12.2             </span></span>
<span><span class="co">##  [5] filelock_1.0.2              Biostrings_2.68.1          </span></span>
<span><span class="co">##  [7] bitops_1.0-7                SingleCellExperiment_1.22.0</span></span>
<span><span class="co">##  [9] fastmap_1.1.1               RCurl_1.98-1.12            </span></span>
<span><span class="co">## [11] BiocFileCache_2.8.0         GenomicAlignments_1.36.0   </span></span>
<span><span class="co">## [13] XML_3.99-0.14               digest_0.6.31              </span></span>
<span><span class="co">## [15] lifecycle_1.0.3             KEGGREST_1.40.0            </span></span>
<span><span class="co">## [17] RSQLite_2.3.1               magrittr_2.0.3             </span></span>
<span><span class="co">## [19] compiler_4.3.0              progress_1.2.2             </span></span>
<span><span class="co">## [21] rlang_1.1.1                 sass_0.4.6                 </span></span>
<span><span class="co">## [23] tools_4.3.0                 utf8_1.2.3                 </span></span>
<span><span class="co">## [25] yaml_2.3.7                  data.table_1.14.8          </span></span>
<span><span class="co">## [27] knitr_1.43                  prettyunits_1.1.1          </span></span>
<span><span class="co">## [29] S4Arrays_1.0.4              bit_4.0.5                  </span></span>
<span><span class="co">## [31] curl_5.0.1                  DelayedArray_0.26.3        </span></span>
<span><span class="co">## [33] xml2_1.3.4                  BiocParallel_1.34.2        </span></span>
<span><span class="co">## [35] withr_2.5.0                 purrr_1.0.1                </span></span>
<span><span class="co">## [37] R.oo_1.25.0                 desc_1.4.2                 </span></span>
<span><span class="co">## [39] grid_4.3.0                  fansi_1.0.4                </span></span>
<span><span class="co">## [41] biomaRt_2.56.0              cli_3.6.1                  </span></span>
<span><span class="co">## [43] rmarkdown_2.22              crayon_1.5.2               </span></span>
<span><span class="co">## [45] ragg_1.2.5                  generics_0.1.3             </span></span>
<span><span class="co">## [47] httr_1.4.6                  rjson_0.2.21               </span></span>
<span><span class="co">## [49] DBI_1.1.3                   cachem_1.0.8               </span></span>
<span><span class="co">## [51] stringr_1.5.0               zlibbioc_1.46.0            </span></span>
<span><span class="co">## [53] parallel_4.3.0              BiocManager_1.30.20        </span></span>
<span><span class="co">## [55] XVector_0.40.0              restfulr_0.0.15            </span></span>
<span><span class="co">## [57] vctrs_0.6.2                 Matrix_1.5-4.1             </span></span>
<span><span class="co">## [59] jsonlite_1.8.5              bookdown_0.34              </span></span>
<span><span class="co">## [61] hms_1.1.3                   bit64_4.0.5                </span></span>
<span><span class="co">## [63] systemfonts_1.0.4           jquerylib_0.1.4            </span></span>
<span><span class="co">## [65] glue_1.6.2                  pkgdown_2.0.7              </span></span>
<span><span class="co">## [67] codetools_0.2-19            stringi_1.7.12             </span></span>
<span><span class="co">## [69] BiocIO_1.10.0               tibble_3.2.1               </span></span>
<span><span class="co">## [71] pillar_1.9.0                rappdirs_0.3.3             </span></span>
<span><span class="co">## [73] htmltools_0.5.5             GenomeInfoDbData_1.2.10    </span></span>
<span><span class="co">## [75] BSgenome_1.68.0             R6_2.5.1                   </span></span>
<span><span class="co">## [77] dbplyr_2.3.2                textshaping_0.3.6          </span></span>
<span><span class="co">## [79] rprojroot_2.0.3             evaluate_0.21              </span></span>
<span><span class="co">## [81] lattice_0.21-8              R.methodsS3_1.8.2          </span></span>
<span><span class="co">## [83] png_0.1-8                   Rsamtools_2.16.0           </span></span>
<span><span class="co">## [85] memoise_2.0.1               bslib_0.5.0                </span></span>
<span><span class="co">## [87] xfun_0.39                   fs_1.6.2                   </span></span>
<span><span class="co">## [89] pkgconfig_2.0.3</span></span></code></pre>
</details>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>An R package developed by the <a href="https://medschool.cuanschutz.edu/rbi" class="external-link">RNA Bioscience Initiative</a></p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
