<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="raer">
<title>Identifying RNA editing sites using RNA and DNA sequencing data • raer</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><link href="../deps/Source_Sans_Pro-0.4.4/font.css" rel="stylesheet">
<link href="../deps/Source_Code_Pro-0.4.4/font.css" rel="stylesheet">
<!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Identifying RNA editing sites using RNA and DNA sequencing data">
<meta property="og:description" content="raer">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><script defer data-domain="rnabioco.github.io" src="https://plausible.io/js/plausible.js"></script>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-dark"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">raer</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/raer.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/novel-sites.html">Identifying RNA editing sites using RNA and DNA sequencing data</a>
    <a class="dropdown-item" href="../articles/single-cell.html">Identifying RNA editing sites in droplet single cell datasets</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/rnabioco/raer/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Identifying RNA editing sites using RNA and DNA sequencing data</h1>
                        <h4 data-toc-skip class="author">Kent
Riemondy</h4>
            <address class="author_afil">
      University of Colorado School of Medicine<br><h4 data-toc-skip class="date">2023-01-21</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/rnabioco/raer/blob/HEAD/vignettes/novel-sites.Rmd" class="external-link"><code>vignettes/novel-sites.Rmd</code></a></small>
      <div class="d-none name"><code>novel-sites.Rmd</code></div>
    </address>
</div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://rnabioco.github.io/raer">raer</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/GenomicFeatures" class="external-link">GenomicFeatures</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/SummarizedExperiment" class="external-link">SummarizedExperiment</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">rtracklayer</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">TxDb.Hsapiens.UCSC.hg38.knownGene</span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="novel-rna-editing-site-detection-tutorial">Novel RNA editing site detection tutorial<a class="anchor" aria-label="anchor" href="#novel-rna-editing-site-detection-tutorial"></a>
</h2>
<p>This vignette will demonstrate how to identify novel RNA editing
sites using the <code>raer</code> package.</p>
<p>In this vignette a public RNA-seq and Whole-Genome sequencing dataset
will be analyzed. High coverage whole-genome sequencing was conducted <a href="https://www.ebi.ac.uk/ena/browser/view/ERR262997?show=reads" class="external-link">ERR262997</a>
along with paired-end RNA-seq <a href="https://www.ebi.ac.uk/ena/browser/view/SRR1258218?show=reads" class="external-link">SRR1258218</a>.</p>
<p>Aligned BAM files and a genome fasta file have been preprocessed for
this vignette. These files occupy ~100 Mb of space and contain
alignments from the first 1MB of chromosome 4. By default these files
are stored in a temporary directory. Change the <code>data_dir</code>
variable to a valid file path to keep the files for reuse.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">fns</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/download_data.html">download_NA12878</a></span><span class="op">(</span><span class="va">data_dir</span><span class="op">)</span></span>
<span><span class="va">fns</span></span></code></pre></div>
<pre><code><span><span class="co">## $bams</span></span>
<span><span class="co">##                              ERR262996_dedup_chr4_sub.bam </span></span>
<span><span class="co">##            "/tmp/RtmpYoW8YX/ERR262996_dedup_chr4_sub.bam" </span></span>
<span><span class="co">##                   SRR1258218_Aligned.sorted.dedup_sub.bam </span></span>
<span><span class="co">## "/tmp/RtmpYoW8YX/SRR1258218_Aligned.sorted.dedup_sub.bam" </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $bai</span></span>
<span><span class="co">##                              ERR262996_dedup_chr4_sub.bam.bai </span></span>
<span><span class="co">##            "/tmp/RtmpYoW8YX/ERR262996_dedup_chr4_sub.bam.bai" </span></span>
<span><span class="co">##                   SRR1258218_Aligned.sorted.dedup_sub.bam.bai </span></span>
<span><span class="co">## "/tmp/RtmpYoW8YX/SRR1258218_Aligned.sorted.dedup_sub.bam.bai" </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $fasta</span></span>
<span><span class="co">##                   hg38_chr4.fa.bgz </span></span>
<span><span class="co">## "/tmp/RtmpYoW8YX/hg38_chr4.fa.bgz" </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $snps</span></span>
<span><span class="co">##                   chr4snps.bed.gz </span></span>
<span><span class="co">## "/tmp/RtmpYoW8YX/chr4snps.bed.gz" </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $rmsk</span></span>
<span><span class="co">##                   rmsk_hg38.tsv.gz </span></span>
<span><span class="co">## "/tmp/RtmpYoW8YX/rmsk_hg38.tsv.gz"</span></span></code></pre>
<p>Additionally we will use the following additional annotation
resources:</p>
<ul>
<li>A database of known SNPs, for example the
<code>SNPlocs.Hsapiens.dbSNP155.GRCh38</code> package. Due to space and
memory constraints , in this vignette we will use a BED file containing
SNPs from the first 1Mb region of chr4.<br>
</li>
<li>
<code>TxDb.Hsapiens.UCSC.hg38.knownGene</code>, a database of
transcript models. Alternatively these can be generated from a
<code>.gtf</code> file using <code><a href="https://rdrr.io/pkg/GenomicFeatures/man/makeTxDbFromGRanges.html" class="external-link">makeTxDbFromGRanges()</a></code> from the
<code>GenomicFeatures</code> package.<br>
</li>
<li>RepeatMasker annotations, which can be obtained from the
<code>AnnotationHub()</code> for hg38. For space reasons these
annotations for chr4 have been included as a text file.</li>
</ul>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">txdb</span> <span class="op">&lt;-</span> <span class="va">TxDb.Hsapiens.UCSC.hg38.knownGene</span></span>
<span></span>
<span><span class="va">chr4snps</span> <span class="op">&lt;-</span> <span class="fu">import</span><span class="op">(</span><span class="va">fns</span><span class="op">$</span><span class="va">snps</span><span class="op">)</span></span>
<span></span>
<span><span class="va">rmskhg38</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="va">fns</span><span class="op">$</span><span class="va">rmsk</span><span class="op">)</span></span>
<span><span class="va">rmskhg38</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/makeGRangesFromDataFrame.html" class="external-link">makeGRangesFromDataFrame</a></span><span class="op">(</span><span class="va">rmskhg38</span>,</span>
<span>  keep.extra.columns <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  starts.in.df.are.0based <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>The <code><a href="../reference/get_pileup.html">get_pileup()</a></code> function can accept multiple bam files,
here we supply one from RNA-seq, and one from whole genome
sequencing.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">wgs_bam</span> <span class="op">&lt;-</span> <span class="va">fns</span><span class="op">$</span><span class="va">bams</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span>
<span><span class="va">rna_bam</span> <span class="op">&lt;-</span> <span class="va">fns</span><span class="op">$</span><span class="va">bams</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span></span>
<span><span class="va">fafn</span> <span class="op">&lt;-</span> <span class="va">fns</span><span class="op">$</span><span class="va">fasta</span></span></code></pre></div>
<p>Filtering parameters for the <code><a href="../reference/get_pileup.html">get_pileup()</a></code> function can
accept multiple arguments matched to the input bam files.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_pileup.html">FilterParam</a></span><span class="op">(</span></span>
<span>  min_nucleotide_depth <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  min_base_quality <span class="op">=</span> <span class="fl">30</span>,</span>
<span>  min_mapq <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">255</span>, <span class="fl">30</span><span class="op">)</span>,</span>
<span>  library_type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"fr-first-strand"</span>, <span class="st">"genomic-unstranded"</span><span class="op">)</span>,</span>
<span>  trim_5p <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  trim_3p <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  indel_dist <span class="op">=</span> <span class="fl">4</span>,</span>
<span>  homopolymer_len <span class="op">=</span> <span class="fl">6</span>,</span>
<span>  max_mismatch_type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">3</span>, <span class="fl">3</span><span class="op">)</span>,</span>
<span>  min_read_bqual <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.25</span>, <span class="fl">20</span><span class="op">)</span>,</span>
<span>  only_keep_variants <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="cn">TRUE</span>, <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">plps</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_pileup.html">get_pileup</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">rna_bam</span>, <span class="va">wgs_bam</span><span class="op">)</span>,</span>
<span>  fafile <span class="op">=</span> <span class="va">fafn</span>,</span>
<span>  chrom <span class="op">=</span> <span class="st">"chr4"</span>,</span>
<span>  filterParam <span class="op">=</span> <span class="va">fp</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">plps</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"rna"</span>, <span class="st">"dna"</span><span class="op">)</span></span>
<span><span class="va">plps</span></span></code></pre></div>
<pre><code><span><span class="co">## $rna</span></span>
<span><span class="co">## GRanges object with 547 ranges and 10 metadata columns:</span></span>
<span><span class="co">##         seqnames    ranges strand |         Ref         Var      nRef</span></span>
<span><span class="co">##            &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;character&gt; &lt;integer&gt;</span></span>
<span><span class="co">##     [1]     chr4     44338      - |           T          TA         0</span></span>
<span><span class="co">##     [2]     chr4    124820      + |           C          CA         3</span></span>
<span><span class="co">##     [3]     chr4    125213      + |           G          GC         3</span></span>
<span><span class="co">##     [4]     chr4    126106      + |           C          CT         0</span></span>
<span><span class="co">##     [5]     chr4    126885      + |           A          AG         9</span></span>
<span><span class="co">##     ...      ...       ...    ... .         ...         ...       ...</span></span>
<span><span class="co">##   [543]     chr4    986676      - |           T          TC         0</span></span>
<span><span class="co">##   [544]     chr4    989824      + |           C          CT         1</span></span>
<span><span class="co">##   [545]     chr4    990589      + |           A          AC         1</span></span>
<span><span class="co">##   [546]     chr4    990981      + |           C          CT         0</span></span>
<span><span class="co">##   [547]     chr4    993342      + |           A          AG         0</span></span>
<span><span class="co">##              nVar        nA        nT        nC        nG        nN</span></span>
<span><span class="co">##         &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;</span></span>
<span><span class="co">##     [1]         1         1         0         0         0         0</span></span>
<span><span class="co">##     [2]         1         1         0         3         0         0</span></span>
<span><span class="co">##     [3]         1         0         0         1         3         0</span></span>
<span><span class="co">##     [4]         1         0         1         0         0         0</span></span>
<span><span class="co">##     [5]         1         9         0         0         1         0</span></span>
<span><span class="co">##     ...       ...       ...       ...       ...       ...       ...</span></span>
<span><span class="co">##   [543]         1         0         0         1         0         0</span></span>
<span><span class="co">##   [544]         1         0         1         1         0         0</span></span>
<span><span class="co">##   [545]         1         1         0         1         0         0</span></span>
<span><span class="co">##   [546]         1         0         1         0         0         0</span></span>
<span><span class="co">##   [547]         1         0         0         0         1         0</span></span>
<span><span class="co">##                nX</span></span>
<span><span class="co">##         &lt;integer&gt;</span></span>
<span><span class="co">##     [1]         0</span></span>
<span><span class="co">##     [2]         0</span></span>
<span><span class="co">##     [3]         0</span></span>
<span><span class="co">##     [4]         0</span></span>
<span><span class="co">##     [5]         1</span></span>
<span><span class="co">##     ...       ...</span></span>
<span><span class="co">##   [543]         0</span></span>
<span><span class="co">##   [544]         0</span></span>
<span><span class="co">##   [545]         0</span></span>
<span><span class="co">##   [546]         0</span></span>
<span><span class="co">##   [547]         0</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 194 sequences from an unspecified genome</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $dna</span></span>
<span><span class="co">## GRanges object with 547 ranges and 10 metadata columns:</span></span>
<span><span class="co">##         seqnames    ranges strand |         Ref         Var      nRef</span></span>
<span><span class="co">##            &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;character&gt; &lt;integer&gt;</span></span>
<span><span class="co">##     [1]     chr4     44338      - |           T           -         0</span></span>
<span><span class="co">##     [2]     chr4    124820      + |           C           -        20</span></span>
<span><span class="co">##     [3]     chr4    125213      + |           G           -         7</span></span>
<span><span class="co">##     [4]     chr4    126106      + |           C          CT        11</span></span>
<span><span class="co">##     [5]     chr4    126885      + |           A           -        13</span></span>
<span><span class="co">##     ...      ...       ...    ... .         ...         ...       ...</span></span>
<span><span class="co">##   [543]     chr4    986676      - |           T           -         0</span></span>
<span><span class="co">##   [544]     chr4    989824      + |           C           -        10</span></span>
<span><span class="co">##   [545]     chr4    990589      + |           A           -        16</span></span>
<span><span class="co">##   [546]     chr4    990981      + |           C           -        15</span></span>
<span><span class="co">##   [547]     chr4    993342      + |           A           -        21</span></span>
<span><span class="co">##              nVar        nA        nT        nC        nG        nN</span></span>
<span><span class="co">##         &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;</span></span>
<span><span class="co">##     [1]         0         0         0         0         0         0</span></span>
<span><span class="co">##     [2]         0         0         0        20         0         0</span></span>
<span><span class="co">##     [3]         0         0         0         0         7         0</span></span>
<span><span class="co">##     [4]         5         0         5        11         0         0</span></span>
<span><span class="co">##     [5]         0        13         0         0         0         0</span></span>
<span><span class="co">##     ...       ...       ...       ...       ...       ...       ...</span></span>
<span><span class="co">##   [543]         0         0         0         0         0         0</span></span>
<span><span class="co">##   [544]         0         0         0        10         0         0</span></span>
<span><span class="co">##   [545]         0        16         0         0         0         0</span></span>
<span><span class="co">##   [546]         0         0         0        15         0         0</span></span>
<span><span class="co">##   [547]         0        21         0         0         0         0</span></span>
<span><span class="co">##                nX</span></span>
<span><span class="co">##         &lt;integer&gt;</span></span>
<span><span class="co">##     [1]         0</span></span>
<span><span class="co">##     [2]         2</span></span>
<span><span class="co">##     [3]         1</span></span>
<span><span class="co">##     [4]         2</span></span>
<span><span class="co">##     [5]         0</span></span>
<span><span class="co">##     ...       ...</span></span>
<span><span class="co">##   [543]         0</span></span>
<span><span class="co">##   [544]         2</span></span>
<span><span class="co">##   [545]         6</span></span>
<span><span class="co">##   [546]         4</span></span>
<span><span class="co">##   [547]         4</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 194 sequences from an unspecified genome</span></span></code></pre>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/create_se.html">create_se</a></span><span class="op">(</span><span class="va">plps</span><span class="op">)</span></span>
<span><span class="va">se</span></span></code></pre></div>
<pre><code><span><span class="co">## class: RangedSummarizedExperiment </span></span>
<span><span class="co">## dim: 547 2 </span></span>
<span><span class="co">## metadata(0):</span></span>
<span><span class="co">## assays(7): Var nRef ... nC nG</span></span>
<span><span class="co">## rownames(547): chr4_44338_- chr4_124820_+ ... chr4_990981_+</span></span>
<span><span class="co">##   chr4_993342_+</span></span>
<span><span class="co">## rowData names(1): Ref</span></span>
<span><span class="co">## colnames(2): rna dna</span></span>
<span><span class="co">## colData names(1): sample</span></span></code></pre>
<p>Next we filter to keep those sites with a variant in the RNA-seq, but
no variant in the DNA-seq, and a minimum of 5 reads in the DNA-seq. The
DNA-seq data will be reported on the “+” strand, where as the RNA-seq
data will be reported on either strand. We therefore use
<code>subsetByOverlaps(..., ignore.strand = TRUE)</code> to retain sites
passing these DNA-seq based filters independent of strand.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">to_keep</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">se</span>, <span class="st">"nRef"</span><span class="op">)</span><span class="op">[</span>, <span class="st">"dna"</span><span class="op">]</span> <span class="op">&gt;=</span> <span class="fl">5</span> <span class="op">&amp;</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">se</span>, <span class="st">"Var"</span><span class="op">)</span><span class="op">[</span>, <span class="st">"dna"</span><span class="op">]</span> <span class="op">==</span> <span class="st">"-"</span><span class="op">)</span></span>
<span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/findOverlaps-methods.html" class="external-link">subsetByOverlaps</a></span><span class="op">(</span><span class="va">se</span>, <span class="va">se</span><span class="op">[</span><span class="va">to_keep</span>, <span class="op">]</span>, ignore.strand <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 197</span></span></code></pre>
<p>Next we filter to remove any multi-allelic sites. These sites are
stored as comma-separated strings in the <code>Var</code> assay
(e.g. <code>AG,AC</code>). Non-variant sites are stored as
<code>-</code>. <code><a href="../reference/filter_se.html">remove_multiallelic()</a></code> will remove any sites
that have multiple variants in the samples present in the
<code>summarizedExperiment</code> object. It will add a new column to
the <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData()</a></code> to indicate the variant for each site, and
will calculate an <code>allele_freq</code> assay with variant allele
frequencies for each sample.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filter_se.html">remove_multiallelic</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span></span>
<span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calc_edit_frequency.html">calc_edit_frequency</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 197 rows and 2 columns</span></span>
<span><span class="co">##                       Ref         Var</span></span>
<span><span class="co">##               &lt;character&gt; &lt;character&gt;</span></span>
<span><span class="co">## chr4_124820_+           C          CA</span></span>
<span><span class="co">## chr4_125213_+           G          GC</span></span>
<span><span class="co">## chr4_126885_+           A          AG</span></span>
<span><span class="co">## chr4_126885_-           T          TC</span></span>
<span><span class="co">## chr4_126955_+           A          AG</span></span>
<span><span class="co">## ...                   ...         ...</span></span>
<span><span class="co">## chr4_958447_+           C          CA</span></span>
<span><span class="co">## chr4_989824_+           C          CT</span></span>
<span><span class="co">## chr4_990589_+           A          AC</span></span>
<span><span class="co">## chr4_990981_+           C          CT</span></span>
<span><span class="co">## chr4_993342_+           A          AG</span></span></code></pre>
<p>Next we’ll remove sites in simple repeat regions. We will add repeat
information to the <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData()</a></code> using the
<code><a href="../reference/annot_from_gr.html">annot_from_gr()</a></code> function.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/annot_from_gr.html">annot_from_gr</a></span><span class="op">(</span><span class="va">se</span>, <span class="va">rmskhg38</span>, cols_to_map <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"repName"</span>, <span class="st">"repClass"</span>, <span class="st">"repFamily"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"repName"</span>, <span class="st">"repFamily"</span><span class="op">)</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 197 rows and 2 columns</span></span>
<span><span class="co">##                   repName   repFamily</span></span>
<span><span class="co">##               &lt;character&gt; &lt;character&gt;</span></span>
<span><span class="co">## chr4_124820_+          NA          NA</span></span>
<span><span class="co">## chr4_125213_+      MER21C        ERVL</span></span>
<span><span class="co">## chr4_126885_+          NA          NA</span></span>
<span><span class="co">## chr4_126885_-      AluSc5         Alu</span></span>
<span><span class="co">## chr4_126955_+          NA          NA</span></span>
<span><span class="co">## ...                   ...         ...</span></span>
<span><span class="co">## chr4_958447_+          NA          NA</span></span>
<span><span class="co">## chr4_989824_+          NA          NA</span></span>
<span><span class="co">## chr4_990589_+          NA          NA</span></span>
<span><span class="co">## chr4_990981_+          NA          NA</span></span>
<span><span class="co">## chr4_993342_+          NA          NA</span></span></code></pre>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se</span> <span class="op">&lt;-</span> <span class="va">se</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">$</span><span class="va">repFamily</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Simple_repeat"</span>, <span class="st">"Low Complexity"</span><span class="op">)</span><span class="op">]</span></span></code></pre></div>
<p>Next we’ll remove sites adjacent to other sites with different
variant types. For example if an A-&gt;G variant is located proximal to
a C-&gt;T variant then the variants will be removed.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filter_se.html">remove_clustered_variants</a></span><span class="op">(</span><span class="va">se</span>, <span class="va">txdb</span>, variant_dist <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning in .merge_two_Seqinfo_objects(x, y): Each of the 2 combined objects has sequence levels not in the other:</span></span>
<span><span class="co">##   - in 'x': GL000008.2, GL000009.2, GL000194.1, GL000195.1, GL000205.2, GL000208.1, GL000213.1, GL000214.1, GL000216.2, GL000218.1, GL000219.1, GL000220.1, GL000221.1, GL000224.1, GL000225.1, GL000226.1, KI270302.1, KI270303.1, KI270304.1, KI270305.1, KI270310.1, KI270311.1, KI270312.1, KI270315.1, KI270316.1, KI270317.1, KI270320.1, KI270322.1, KI270329.1, KI270330.1, KI270333.1, KI270334.1, KI270335.1, KI270336.1, KI270337.1, KI270338.1, KI270340.1, KI270362.1, KI270363.1, KI270364.1, KI270366.1, KI270371.1, KI270372.1, KI270373.1, KI270374.1, KI270375.1, KI270376.1, KI270378.1, KI270379.1, KI270381.1, KI270382.1, KI270383.1, KI270384.1, KI270385.1, KI270386.1, KI270387.1, KI270388.1, KI270389.1, KI270390.1, KI270391.1, KI270392.1, KI270393.1, KI270394.1, KI270395.1, KI270396.1, KI270411.1, KI270412.1, KI270414.1, KI270417.1, KI270418.1, KI270419.1, KI270420.1, KI270422.1, KI270423.1, KI270424.1, KI270425.1, KI270429.1, KI270435.1, KI270438.1, KI270442.1, KI270448.1, KI270465.1, KI270466.1, KI270467.1, KI270468.1, KI270507.1, KI270508.1, KI270509.1, KI270510.1, KI270511.1, KI270512.1, KI270515.1, KI270516.1, KI270517.1, KI270518.1, KI270519.1, KI270521.1, KI270522.1, KI270528.1, KI270529.1, KI270530.1, KI270538.1, KI270539.1, KI270544.1, KI270548.1, KI270579.1, KI270580.1, KI270581.1, KI270582.1, KI270583.1, KI270584.1, KI270587.1, KI270588.1, KI270589.1, KI270590.1, KI270591.1, KI270593.1, KI270706.1, KI270707.1, KI270708.1, KI270709.1, KI270710.1, KI270711.1, KI270712.1, KI270713.1, KI270714.1, KI270715.1, KI270716.1, KI270717.1, KI270718.1, KI270719.1, KI270720.1, KI270721.1, KI270722.1, KI270723.1, KI270724.1, KI270725.1, KI270726.1, KI270727.1, KI270728.1, KI270729.1, KI270730.1, KI270731.1, KI270732.1, KI270733.1, KI270734.1, KI270735.1, KI270736.1, KI270737.1, KI270738.1, KI270739.1, KI270740.1, KI270741.1, KI270742.1, KI270743.1, KI270744.1, KI270745.1, KI270746.1, KI270747.1, KI270748.1, KI270749.1, KI270750.1, KI270751.1, KI270752.1, KI270753.1, KI270754.1, KI270755.1, KI270756.1, KI270757.1</span></span>
<span><span class="co">##   - in 'y': chr1_GL383518v1_alt, chr1_GL383519v1_alt, chr1_GL383520v2_alt, chr1_KI270706v1_random, chr1_KI270707v1_random, chr1_KI270708v1_random, chr1_KI270709v1_random, chr1_KI270710v1_random, chr1_KI270711v1_random, chr1_KI270712v1_random, chr1_KI270713v1_random, chr1_KI270714v1_random, chr1_KI270759v1_alt, chr1_KI270760v1_alt, chr1_KI270761v1_alt, chr1_KI270762v1_alt, chr1_KI270763v1_alt, chr1_KI270764v1_alt, chr1_KI270765v1_alt, chr1_KI270766v1_alt, chr1_KI270892v1_alt, chr1_KN196472v1_fix, chr1_KN196473v1_fix, chr1_KN196474v1_fix, chr1_KN538360v1_fix, chr1_KN538361v1_fix, chr1_KQ031383v1_fix, chr1_KQ458382v1_alt, chr1_KQ458383v1_alt, chr1_KQ458384v1_alt, chr1_KQ983255v1_alt, chr1_KV880763v1_alt, chr1_KZ208904v1_alt, chr1_KZ208905v1_alt, chr1_KZ208906v1_fix, chr1_KZ559100v1_fix, chr2_GL383521v1_alt, chr2_GL383522v1_alt, chr2_GL582966v2_alt, chr2_KI270715v1_random, chr2_KI270716v1_random, chr2_KI270767v1_alt, chr2_KI270768v1_alt, chr2_KI270769v1_alt, chr2_KI270770v1_alt, chr2_KI270771v1_alt, chr2_KI270772v1_alt, chr2_KI270773v1_alt, chr2_KI270774v1_alt, chr2_KI270775v1_alt, chr2_KI270776v1_alt, chr2_KI270893v1_alt, chr2_KI270894v1_alt, chr2_KN538362v1_fix, chr2_KN538363v1_fix, chr2_KQ031384v1_fix, chr2_KQ983256v1_alt, chr2_KZ208907v1_alt, chr2_KZ208908v1_alt, chr2_ML143341v1_fix, chr2_ML143342v1_fix, chr3_GL000221v1_random, chr3_GL383526v1_alt, chr3_JH636055v2_alt, chr3_KI270777v1_alt, chr3_KI270778v1_alt, chr3_KI270779v1_alt, chr3_KI270780v1_alt, chr3_KI270781v1_alt, chr3_KI270782v1_alt, chr3_KI270783v1_alt, chr3_KI270784v1_alt, chr3_KI270895v1_alt, chr3_KI270924v1_alt, chr3_KI270934v1_alt, chr3_KI270935v1_alt, chr3_KI270936v1_alt, chr3_KI270937v1_alt, chr3_KN196475v1_fix, chr3_KN196476v1_fix, chr3_KN538364v1_fix, chr3_KQ031385v1_fix, chr3_KQ031386v1_fix, chr3_KV766192v1_fix, chr3_KZ208909v1_alt, chr3_KZ559101v1_alt, chr3_KZ559102v1_alt, chr3_KZ559103v1_alt, chr3_KZ559104v1_fix, chr3_KZ559105v1_alt, chr3_ML143343v1_alt, chr4_GL000008v2_random, chr4_GL000257v2_alt, chr4_GL383527v1_alt, chr4_GL383528v1_alt, chr4_KI270785v1_alt, chr4_KI270786v1_alt, chr4_KI270787v1_alt, chr4_KI270788v1_alt, chr4_KI270789v1_alt, chr4_KI270790v1_alt, chr4_KI270896v1_alt, chr4_KI270925v1_alt, chr4_KQ090013v1_alt, chr4_KQ090014v1_alt, chr4_KQ090015v1_alt, chr4_KQ983257v1_fix, chr4_KQ983258v1_alt, chr4_KV766193v1_alt, chr4_ML143344v1_fix, chr4_ML143345v1_fix, chr4_ML143346v1_fix, chr4_ML143347v1_fix, chr4_ML143348v1_fix, chr4_ML143349v1_fix, chr5_GL000208v1_random, chr5_GL339449v2_alt, chr5_GL383530v1_alt, chr5_GL383531v1_alt, chr5_GL383532v1_alt, chr5_GL949742v1_alt, chr5_KI270791v1_alt, chr5_KI270792v1_alt, chr5_KI270793v1_alt, chr5_KI270794v1_alt, chr5_KI270795v1_alt, chr5_KI270796v1_alt, chr5_KI270897v1_alt, chr5_KI270898v1_alt, chr5_KN196477v1_alt, chr5_KV575243v1_alt, chr5_KV575244v1_fix, chr5_KZ208910v1_alt, chr5_ML143350v1_fix, chr6_GL000250v2_alt, chr6_GL000251v2_alt, chr6_GL000252v2_alt, chr6_GL000253v2_alt, chr6_GL000254v2_alt, chr6_GL000255v2_alt, chr6_GL000256v2_alt, chr6_GL383533v1_alt, chr6_KB021644v2_alt, chr6_KI270758v1_alt, chr6_KI270797v1_alt, chr6_KI270798v1_alt, chr6_KI270799v1_alt, chr6_KI270800v1_alt, chr6_KI270801v1_alt, chr6_KI270802v1_alt, chr6_KN196478v1_fix, chr6_KQ031387v1_fix, chr6_KQ090016v1_fix, chr6_KQ090017v1_alt, chr6_KV766194v1_fix, chr6_KZ208911v1_fix, chr6_ML143351v1_fix, chr7_GL383534v2_alt, chr7_KI270803v1_alt, chr7_KI270804v1_alt, chr7_KI270805v1_alt, chr7_KI270806v1_alt, chr7_KI270807v1_alt, chr7_KI270808v1_alt, chr7_KI270809v1_alt, chr7_KI270899v1_alt, chr7_KQ031388v1_fix, chr7_KV880764v1_fix, chr7_KV880765v1_fix, chr7_KZ208912v1_fix, chr7_KZ208913v1_alt, chr7_KZ559106v1_alt, chr7_ML143352v1_fix, chr8_KI270810v1_alt, chr8_KI270811v1_alt, chr8_KI270812v1_alt, chr8_KI270813v1_alt, chr8_KI270814v1_alt, chr8_KI270815v1_alt, chr8_KI270816v1_alt, chr8_KI270817v1_alt, chr8_KI270818v1_alt, chr8_KI270819v1_alt, chr8_KI270820v1_alt, chr8_KI270821v1_alt, chr8_KI270822v1_alt, chr8_KI270900v1_alt, chr8_KI270901v1_alt, chr8_KI270926v1_alt, chr8_KV880766v1_fix, chr8_KV880767v1_fix, chr8_KZ208914v1_fix, chr8_KZ208915v1_fix, chr8_KZ559107v1_alt, chr9_GL383539v1_alt, chr9_GL383540v1_alt, chr9_GL383541v1_alt, chr9_GL383542v1_alt, chr9_KI270717v1_random, chr9_KI270718v1_random, chr9_KI270719v1_random, chr9_KI270720v1_random, chr9_KI270823v1_alt, chr9_KN196479v1_fix, chr9_KQ090018v1_alt, chr9_KQ090019v1_alt, chr9_ML143353v1_fix, chr10_GL383545v1_alt, chr10_GL383546v1_alt, chr10_KI270824v1_alt, chr10_KI270825v1_alt, chr10_KN196480v1_fix, chr10_KN538365v1_fix, chr10_KN538366v1_fix, chr10_KN538367v1_fix, chr10_KQ090020v1_alt, chr10_KQ090021v1_fix, chr10_ML143354v1_fix, chr10_ML143355v1_fix, chr11_GL383547v1_alt, chr11_JH159136v1_alt, chr11_JH159137v1_alt, chr11_KI270721v1_random, chr11_KI270826v1_alt, chr11_KI270827v1_alt, chr11_KI270829v1_alt, chr11_KI270830v1_alt, chr11_KI270831v1_alt, chr11_KI270832v1_alt, chr11_KI270902v1_alt, chr11_KI270903v1_alt, chr11_KI270927v1_alt, chr11_KN196481v1_fix, chr11_KN538368v1_alt, chr11_KQ090022v1_fix, chr11_KQ759759v1_fix, chr11_KV766195v1_fix, chr11_KZ559108v1_fix, chr11_KZ559109v1_fix, chr11_KZ559110v1_alt, chr11_KZ559111v1_alt, chr11_ML143356v1_fix, chr11_ML143357v1_fix, chr11_ML143358v1_fix, chr11_ML143359v1_fix, chr11_ML143360v1_fix, chr12_GL383549v1_alt, chr12_GL383550v2_alt, chr12_GL383551v1_alt, chr12_GL383552v1_alt, chr12_GL383553v2_alt, chr12_GL877875v1_alt, chr12_GL877876v1_alt, chr12_KI270833v1_alt, chr12_KI270834v1_alt, chr12_KI270835v1_alt, chr12_KI270836v1_alt, chr12_KI270837v1_alt, chr12_KI270904v1_alt, chr12_KN196482v1_fix, chr12_KN538369v1_fix, chr12_KN538370v1_fix, chr12_KQ090023v1_alt, chr12_KQ759760v1_fix, chr12_KZ208916v1_fix, chr12_KZ208917v1_fix, chr12_KZ208918v1_alt, chr12_KZ559112v1_alt, chr12_ML143361v1_fix, chr12_ML143362v1_fix, chr13_KI270838v1_alt, chr13_KI270839v1_alt, chr13_KI270840v1_alt, chr13_KI270841v1_alt, chr13_KI270842v1_alt, chr13_KI270843v1_alt, chr13_KN196483v1_fix, chr13_KN538371v1_fix, chr13_KN538372v1_fix, chr13_KN538373v1_fix, chr13_KQ090024v1_alt, chr13_KQ090025v1_alt, chr13_ML143363v1</span></span></code></pre>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se</span></span></code></pre></div>
<pre><code><span><span class="co">## class: RangedSummarizedExperiment </span></span>
<span><span class="co">## dim: 153 2 </span></span>
<span><span class="co">## metadata(0):</span></span>
<span><span class="co">## assays(9): Var nRef ... depth edit_freq</span></span>
<span><span class="co">## rownames(153): chr4_124820_+ chr4_125213_+ ... chr4_990981_+</span></span>
<span><span class="co">##   chr4_993342_+</span></span>
<span><span class="co">## rowData names(5): Ref Var repName repClass repFamily</span></span>
<span><span class="co">## colnames(2): rna dna</span></span>
<span><span class="co">## colData names(3): sample n_sites edit_idx</span></span></code></pre>
<p>Next, sites with 4bp of known splicing events will be excluded. These
variants are low-confidence due to the possibility of
mis-alignments.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filter_se.html">remove_splice_variants</a></span><span class="op">(</span><span class="va">se</span>, <span class="va">txdb</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning in .merge_two_Seqinfo_objects(x, y): Each of the 2 combined objects has sequence levels not in the other:</span></span>
<span><span class="co">##   - in 'x': GL000008.2, GL000009.2, GL000194.1, GL000195.1, GL000205.2, GL000208.1, GL000213.1, GL000214.1, GL000216.2, GL000218.1, GL000219.1, GL000220.1, GL000221.1, GL000224.1, GL000225.1, GL000226.1, KI270302.1, KI270303.1, KI270304.1, KI270305.1, KI270310.1, KI270311.1, KI270312.1, KI270315.1, KI270316.1, KI270317.1, KI270320.1, KI270322.1, KI270329.1, KI270330.1, KI270333.1, KI270334.1, KI270335.1, KI270336.1, KI270337.1, KI270338.1, KI270340.1, KI270362.1, KI270363.1, KI270364.1, KI270366.1, KI270371.1, KI270372.1, KI270373.1, KI270374.1, KI270375.1, KI270376.1, KI270378.1, KI270379.1, KI270381.1, KI270382.1, KI270383.1, KI270384.1, KI270385.1, KI270386.1, KI270387.1, KI270388.1, KI270389.1, KI270390.1, KI270391.1, KI270392.1, KI270393.1, KI270394.1, KI270395.1, KI270396.1, KI270411.1, KI270412.1, KI270414.1, KI270417.1, KI270418.1, KI270419.1, KI270420.1, KI270422.1, KI270423.1, KI270424.1, KI270425.1, KI270429.1, KI270435.1, KI270438.1, KI270442.1, KI270448.1, KI270465.1, KI270466.1, KI270467.1, KI270468.1, KI270507.1, KI270508.1, KI270509.1, KI270510.1, KI270511.1, KI270512.1, KI270515.1, KI270516.1, KI270517.1, KI270518.1, KI270519.1, KI270521.1, KI270522.1, KI270528.1, KI270529.1, KI270530.1, KI270538.1, KI270539.1, KI270544.1, KI270548.1, KI270579.1, KI270580.1, KI270581.1, KI270582.1, KI270583.1, KI270584.1, KI270587.1, KI270588.1, KI270589.1, KI270590.1, KI270591.1, KI270593.1, KI270706.1, KI270707.1, KI270708.1, KI270709.1, KI270710.1, KI270711.1, KI270712.1, KI270713.1, KI270714.1, KI270715.1, KI270716.1, KI270717.1, KI270718.1, KI270719.1, KI270720.1, KI270721.1, KI270722.1, KI270723.1, KI270724.1, KI270725.1, KI270726.1, KI270727.1, KI270728.1, KI270729.1, KI270730.1, KI270731.1, KI270732.1, KI270733.1, KI270734.1, KI270735.1, KI270736.1, KI270737.1, KI270738.1, KI270739.1, KI270740.1, KI270741.1, KI270742.1, KI270743.1, KI270744.1, KI270745.1, KI270746.1, KI270747.1, KI270748.1, KI270749.1, KI270750.1, KI270751.1, KI270752.1, KI270753.1, KI270754.1, KI270755.1, KI270756.1, KI270757.1</span></span>
<span><span class="co">##   - in 'y': chr1_GL383518v1_alt, chr1_GL383519v1_alt, chr1_GL383520v2_alt, chr1_KI270706v1_random, chr1_KI270707v1_random, chr1_KI270708v1_random, chr1_KI270709v1_random, chr1_KI270710v1_random, chr1_KI270711v1_random, chr1_KI270712v1_random, chr1_KI270713v1_random, chr1_KI270714v1_random, chr1_KI270759v1_alt, chr1_KI270760v1_alt, chr1_KI270761v1_alt, chr1_KI270762v1_alt, chr1_KI270763v1_alt, chr1_KI270764v1_alt, chr1_KI270765v1_alt, chr1_KI270766v1_alt, chr1_KI270892v1_alt, chr1_KN196472v1_fix, chr1_KN196473v1_fix, chr1_KN196474v1_fix, chr1_KN538360v1_fix, chr1_KN538361v1_fix, chr1_KQ031383v1_fix, chr1_KQ458382v1_alt, chr1_KQ458383v1_alt, chr1_KQ458384v1_alt, chr1_KQ983255v1_alt, chr1_KV880763v1_alt, chr1_KZ208904v1_alt, chr1_KZ208905v1_alt, chr1_KZ208906v1_fix, chr1_KZ559100v1_fix, chr2_GL383521v1_alt, chr2_GL383522v1_alt, chr2_GL582966v2_alt, chr2_KI270715v1_random, chr2_KI270716v1_random, chr2_KI270767v1_alt, chr2_KI270768v1_alt, chr2_KI270769v1_alt, chr2_KI270770v1_alt, chr2_KI270771v1_alt, chr2_KI270772v1_alt, chr2_KI270773v1_alt, chr2_KI270774v1_alt, chr2_KI270775v1_alt, chr2_KI270776v1_alt, chr2_KI270893v1_alt, chr2_KI270894v1_alt, chr2_KN538362v1_fix, chr2_KN538363v1_fix, chr2_KQ031384v1_fix, chr2_KQ983256v1_alt, chr2_KZ208907v1_alt, chr2_KZ208908v1_alt, chr2_ML143341v1_fix, chr2_ML143342v1_fix, chr3_GL000221v1_random, chr3_GL383526v1_alt, chr3_JH636055v2_alt, chr3_KI270777v1_alt, chr3_KI270778v1_alt, chr3_KI270779v1_alt, chr3_KI270780v1_alt, chr3_KI270781v1_alt, chr3_KI270782v1_alt, chr3_KI270783v1_alt, chr3_KI270784v1_alt, chr3_KI270895v1_alt, chr3_KI270924v1_alt, chr3_KI270934v1_alt, chr3_KI270935v1_alt, chr3_KI270936v1_alt, chr3_KI270937v1_alt, chr3_KN196475v1_fix, chr3_KN196476v1_fix, chr3_KN538364v1_fix, chr3_KQ031385v1_fix, chr3_KQ031386v1_fix, chr3_KV766192v1_fix, chr3_KZ208909v1_alt, chr3_KZ559101v1_alt, chr3_KZ559102v1_alt, chr3_KZ559103v1_alt, chr3_KZ559104v1_fix, chr3_KZ559105v1_alt, chr3_ML143343v1_alt, chr4_GL000008v2_random, chr4_GL000257v2_alt, chr4_GL383527v1_alt, chr4_GL383528v1_alt, chr4_KI270785v1_alt, chr4_KI270786v1_alt, chr4_KI270787v1_alt, chr4_KI270788v1_alt, chr4_KI270789v1_alt, chr4_KI270790v1_alt, chr4_KI270896v1_alt, chr4_KI270925v1_alt, chr4_KQ090013v1_alt, chr4_KQ090014v1_alt, chr4_KQ090015v1_alt, chr4_KQ983257v1_fix, chr4_KQ983258v1_alt, chr4_KV766193v1_alt, chr4_ML143344v1_fix, chr4_ML143345v1_fix, chr4_ML143346v1_fix, chr4_ML143347v1_fix, chr4_ML143348v1_fix, chr4_ML143349v1_fix, chr5_GL000208v1_random, chr5_GL339449v2_alt, chr5_GL383530v1_alt, chr5_GL383531v1_alt, chr5_GL383532v1_alt, chr5_GL949742v1_alt, chr5_KI270791v1_alt, chr5_KI270792v1_alt, chr5_KI270793v1_alt, chr5_KI270794v1_alt, chr5_KI270795v1_alt, chr5_KI270796v1_alt, chr5_KI270897v1_alt, chr5_KI270898v1_alt, chr5_KN196477v1_alt, chr5_KV575243v1_alt, chr5_KV575244v1_fix, chr5_KZ208910v1_alt, chr5_ML143350v1_fix, chr6_GL000250v2_alt, chr6_GL000251v2_alt, chr6_GL000252v2_alt, chr6_GL000253v2_alt, chr6_GL000254v2_alt, chr6_GL000255v2_alt, chr6_GL000256v2_alt, chr6_GL383533v1_alt, chr6_KB021644v2_alt, chr6_KI270758v1_alt, chr6_KI270797v1_alt, chr6_KI270798v1_alt, chr6_KI270799v1_alt, chr6_KI270800v1_alt, chr6_KI270801v1_alt, chr6_KI270802v1_alt, chr6_KN196478v1_fix, chr6_KQ031387v1_fix, chr6_KQ090016v1_fix, chr6_KQ090017v1_alt, chr6_KV766194v1_fix, chr6_KZ208911v1_fix, chr6_ML143351v1_fix, chr7_GL383534v2_alt, chr7_KI270803v1_alt, chr7_KI270804v1_alt, chr7_KI270805v1_alt, chr7_KI270806v1_alt, chr7_KI270807v1_alt, chr7_KI270808v1_alt, chr7_KI270809v1_alt, chr7_KI270899v1_alt, chr7_KQ031388v1_fix, chr7_KV880764v1_fix, chr7_KV880765v1_fix, chr7_KZ208912v1_fix, chr7_KZ208913v1_alt, chr7_KZ559106v1_alt, chr7_ML143352v1_fix, chr8_KI270810v1_alt, chr8_KI270811v1_alt, chr8_KI270812v1_alt, chr8_KI270813v1_alt, chr8_KI270814v1_alt, chr8_KI270815v1_alt, chr8_KI270816v1_alt, chr8_KI270817v1_alt, chr8_KI270818v1_alt, chr8_KI270819v1_alt, chr8_KI270820v1_alt, chr8_KI270821v1_alt, chr8_KI270822v1_alt, chr8_KI270900v1_alt, chr8_KI270901v1_alt, chr8_KI270926v1_alt, chr8_KV880766v1_fix, chr8_KV880767v1_fix, chr8_KZ208914v1_fix, chr8_KZ208915v1_fix, chr8_KZ559107v1_alt, chr9_GL383539v1_alt, chr9_GL383540v1_alt, chr9_GL383541v1_alt, chr9_GL383542v1_alt, chr9_KI270717v1_random, chr9_KI270718v1_random, chr9_KI270719v1_random, chr9_KI270720v1_random, chr9_KI270823v1_alt, chr9_KN196479v1_fix, chr9_KQ090018v1_alt, chr9_KQ090019v1_alt, chr9_ML143353v1_fix, chr10_GL383545v1_alt, chr10_GL383546v1_alt, chr10_KI270824v1_alt, chr10_KI270825v1_alt, chr10_KN196480v1_fix, chr10_KN538365v1_fix, chr10_KN538366v1_fix, chr10_KN538367v1_fix, chr10_KQ090020v1_alt, chr10_KQ090021v1_fix, chr10_ML143354v1_fix, chr10_ML143355v1_fix, chr11_GL383547v1_alt, chr11_JH159136v1_alt, chr11_JH159137v1_alt, chr11_KI270721v1_random, chr11_KI270826v1_alt, chr11_KI270827v1_alt, chr11_KI270829v1_alt, chr11_KI270830v1_alt, chr11_KI270831v1_alt, chr11_KI270832v1_alt, chr11_KI270902v1_alt, chr11_KI270903v1_alt, chr11_KI270927v1_alt, chr11_KN196481v1_fix, chr11_KN538368v1_alt, chr11_KQ090022v1_fix, chr11_KQ759759v1_fix, chr11_KV766195v1_fix, chr11_KZ559108v1_fix, chr11_KZ559109v1_fix, chr11_KZ559110v1_alt, chr11_KZ559111v1_alt, chr11_ML143356v1_fix, chr11_ML143357v1_fix, chr11_ML143358v1_fix, chr11_ML143359v1_fix, chr11_ML143360v1_fix, chr12_GL383549v1_alt, chr12_GL383550v2_alt, chr12_GL383551v1_alt, chr12_GL383552v1_alt, chr12_GL383553v2_alt, chr12_GL877875v1_alt, chr12_GL877876v1_alt, chr12_KI270833v1_alt, chr12_KI270834v1_alt, chr12_KI270835v1_alt, chr12_KI270836v1_alt, chr12_KI270837v1_alt, chr12_KI270904v1_alt, chr12_KN196482v1_fix, chr12_KN538369v1_fix, chr12_KN538370v1_fix, chr12_KQ090023v1_alt, chr12_KQ759760v1_fix, chr12_KZ208916v1_fix, chr12_KZ208917v1_fix, chr12_KZ208918v1_alt, chr12_KZ559112v1_alt, chr12_ML143361v1_fix, chr12_ML143362v1_fix, chr13_KI270838v1_alt, chr13_KI270839v1_alt, chr13_KI270840v1_alt, chr13_KI270841v1_alt, chr13_KI270842v1_alt, chr13_KI270843v1_alt, chr13_KN196483v1_fix, chr13_KN538371v1_fix, chr13_KN538372v1_fix, chr13_KN538373v1_fix, chr13_KQ090024v1_alt, chr13_KQ090025v1_alt, chr13_ML143363v1</span></span></code></pre>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se</span></span></code></pre></div>
<pre><code><span><span class="co">## class: RangedSummarizedExperiment </span></span>
<span><span class="co">## dim: 146 2 </span></span>
<span><span class="co">## metadata(0):</span></span>
<span><span class="co">## assays(9): Var nRef ... depth edit_freq</span></span>
<span><span class="co">## rownames(146): chr4_124820_+ chr4_125213_+ ... chr4_990981_+</span></span>
<span><span class="co">##   chr4_993342_+</span></span>
<span><span class="co">## rowData names(5): Ref Var repName repClass repFamily</span></span>
<span><span class="co">## colnames(2): rna dna</span></span>
<span><span class="co">## colData names(3): sample n_sites edit_idx</span></span></code></pre>
<p>Next, we’ll annotate if the site is a known SNP and remove any known
SNPs. If using a SNPlocs package you can use the
<code>annot_snp()</code> function. However we will use the
<code><a href="../reference/annot_from_gr.html">annot_from_gr()</a></code> function to annotate using the prebuilt
<code>chr4snps</code> object.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/annot_from_gr.html">annot_from_gr</a></span><span class="op">(</span><span class="va">se</span>, <span class="va">chr4snps</span>, <span class="st">"name"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"name"</span><span class="op">)</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 146 rows and 1 column</span></span>
<span><span class="co">##                       name</span></span>
<span><span class="co">##                &lt;character&gt;</span></span>
<span><span class="co">## chr4_124820_+ rs1553807640</span></span>
<span><span class="co">## chr4_125213_+ rs1286128437</span></span>
<span><span class="co">## chr4_126885_+           NA</span></span>
<span><span class="co">## chr4_126885_-           NA</span></span>
<span><span class="co">## chr4_126955_+           NA</span></span>
<span><span class="co">## ...                    ...</span></span>
<span><span class="co">## chr4_958447_+ rs1052651854</span></span>
<span><span class="co">## chr4_989824_+   rs73219719</span></span>
<span><span class="co">## chr4_990589_+           NA</span></span>
<span><span class="co">## chr4_990981_+  rs116514520</span></span>
<span><span class="co">## chr4_993342_+           NA</span></span></code></pre>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se</span> <span class="op">&lt;-</span> <span class="va">se</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">$</span><span class="va">name</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="va">se</span></span></code></pre></div>
<pre><code><span><span class="co">## class: RangedSummarizedExperiment </span></span>
<span><span class="co">## dim: 86 2 </span></span>
<span><span class="co">## metadata(0):</span></span>
<span><span class="co">## assays(9): Var nRef ... depth edit_freq</span></span>
<span><span class="co">## rownames(86): chr4_126885_+ chr4_126885_- ... chr4_990589_+</span></span>
<span><span class="co">##   chr4_993342_+</span></span>
<span><span class="co">## rowData names(6): Ref Var ... repFamily name</span></span>
<span><span class="co">## colnames(2): rna dna</span></span>
<span><span class="co">## colData names(3): sample n_sites edit_idx</span></span></code></pre>
<p>Lastly, we’ll further filter the edit sites to require that the
editing frequency is &gt; 0.05 and that at least 2 reads support the
editing site.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">to_keep</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">se</span>, <span class="st">"edit_freq"</span><span class="op">)</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span> <span class="op">&gt;</span> <span class="fl">0.05</span></span>
<span><span class="va">se</span> <span class="op">&lt;-</span> <span class="va">se</span><span class="op">[</span><span class="va">to_keep</span>, <span class="op">]</span></span>
<span></span>
<span><span class="va">se</span> <span class="op">&lt;-</span> <span class="va">se</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">se</span>, <span class="st">"nVar"</span><span class="op">)</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span> <span class="op">&gt;=</span> <span class="fl">2</span><span class="op">]</span></span></code></pre></div>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/stopifnot.html" class="external-link">stopifnot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/all.html" class="external-link">all</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">$</span><span class="va">Var</span> <span class="op">==</span> <span class="st">"AG"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/MatrixGenerics/man/rowRanges.html" class="external-link">rowRanges</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 10 ranges and 6 metadata columns:</span></span>
<span><span class="co">##                 seqnames    ranges strand |         Ref         Var</span></span>
<span><span class="co">##                    &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;character&gt;</span></span>
<span><span class="co">##   chr4_134353_+     chr4    134353      + |           A          AG</span></span>
<span><span class="co">##   chr4_378011_+     chr4    378011      + |           A          AG</span></span>
<span><span class="co">##   chr4_378018_+     chr4    378018      + |           A          AG</span></span>
<span><span class="co">##   chr4_378090_+     chr4    378090      + |           A          AG</span></span>
<span><span class="co">##   chr4_378362_+     chr4    378362      + |           A          AG</span></span>
<span><span class="co">##   chr4_378368_+     chr4    378368      + |           A          AG</span></span>
<span><span class="co">##   chr4_378409_+     chr4    378409      + |           A          AG</span></span>
<span><span class="co">##   chr4_380392_+     chr4    380392      + |           A          AG</span></span>
<span><span class="co">##   chr4_778739_-     chr4    778739      - |           A          AG</span></span>
<span><span class="co">##   chr4_778780_-     chr4    778780      - |           A          AG</span></span>
<span><span class="co">##                     repName    repClass   repFamily        name</span></span>
<span><span class="co">##                 &lt;character&gt; &lt;character&gt; &lt;character&gt; &lt;character&gt;</span></span>
<span><span class="co">##   chr4_134353_+        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;</span></span>
<span><span class="co">##   chr4_378011_+       AluSc        SINE         Alu        &lt;NA&gt;</span></span>
<span><span class="co">##   chr4_378018_+       AluSc        SINE         Alu        &lt;NA&gt;</span></span>
<span><span class="co">##   chr4_378090_+       AluSc        SINE         Alu        &lt;NA&gt;</span></span>
<span><span class="co">##   chr4_378362_+        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;</span></span>
<span><span class="co">##   chr4_378368_+        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;</span></span>
<span><span class="co">##   chr4_378409_+        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;</span></span>
<span><span class="co">##   chr4_380392_+      AluSq2        SINE         Alu        &lt;NA&gt;</span></span>
<span><span class="co">##   chr4_778739_-       AluSg        SINE         Alu        &lt;NA&gt;</span></span>
<span><span class="co">##   chr4_778780_-       AluSg        SINE         Alu        &lt;NA&gt;</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 194 sequences from an unspecified genome</span></span></code></pre>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/unlink.html" class="external-link">unlink</a></span><span class="op">(</span><span class="va">data_dir</span><span class="op">)</span></span></code></pre></div>
<details style="margin-bottom:10px;"><summary>
Session info
</summary><div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R Under development (unstable) (2023-01-18 r83635)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">## Running under: Ubuntu 22.04.1 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] TxDb.Hsapiens.UCSC.hg38.knownGene_3.16.0</span></span>
<span><span class="co">##  [2] rtracklayer_1.59.1                      </span></span>
<span><span class="co">##  [3] SummarizedExperiment_1.29.1             </span></span>
<span><span class="co">##  [4] MatrixGenerics_1.11.0                   </span></span>
<span><span class="co">##  [5] matrixStats_0.63.0                      </span></span>
<span><span class="co">##  [6] GenomicFeatures_1.51.2                  </span></span>
<span><span class="co">##  [7] AnnotationDbi_1.61.0                    </span></span>
<span><span class="co">##  [8] Biobase_2.59.0                          </span></span>
<span><span class="co">##  [9] GenomicRanges_1.51.4                    </span></span>
<span><span class="co">## [10] GenomeInfoDb_1.35.13                    </span></span>
<span><span class="co">## [11] IRanges_2.33.0                          </span></span>
<span><span class="co">## [12] S4Vectors_0.37.3                        </span></span>
<span><span class="co">## [13] BiocGenerics_0.45.0                     </span></span>
<span><span class="co">## [14] raer_0.99.0                             </span></span>
<span><span class="co">## [15] BiocStyle_2.27.1                        </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##  [1] tidyselect_1.2.0         dplyr_1.0.10            </span></span>
<span><span class="co">##  [3] blob_1.2.3               R.utils_2.12.2          </span></span>
<span><span class="co">##  [5] filelock_1.0.2           Biostrings_2.67.0       </span></span>
<span><span class="co">##  [7] bitops_1.0-7             fastmap_1.1.0           </span></span>
<span><span class="co">##  [9] RCurl_1.98-1.9           BiocFileCache_2.7.1     </span></span>
<span><span class="co">## [11] GenomicAlignments_1.35.0 XML_3.99-0.13           </span></span>
<span><span class="co">## [13] digest_0.6.31            lifecycle_1.0.3         </span></span>
<span><span class="co">## [15] ellipsis_0.3.2           KEGGREST_1.39.0         </span></span>
<span><span class="co">## [17] RSQLite_2.2.20           magrittr_2.0.3          </span></span>
<span><span class="co">## [19] compiler_4.3.0           rlang_1.0.6             </span></span>
<span><span class="co">## [21] sass_0.4.4               progress_1.2.2          </span></span>
<span><span class="co">## [23] tools_4.3.0              utf8_1.2.2              </span></span>
<span><span class="co">## [25] yaml_2.3.6               data.table_1.14.6       </span></span>
<span><span class="co">## [27] knitr_1.41               prettyunits_1.1.1       </span></span>
<span><span class="co">## [29] curl_5.0.0               bit_4.0.5               </span></span>
<span><span class="co">## [31] DelayedArray_0.25.0      xml2_1.3.3              </span></span>
<span><span class="co">## [33] BiocParallel_1.33.9      purrr_1.0.1             </span></span>
<span><span class="co">## [35] R.oo_1.25.0              desc_1.4.2              </span></span>
<span><span class="co">## [37] grid_4.3.0               fansi_1.0.3             </span></span>
<span><span class="co">## [39] biomaRt_2.55.0           cli_3.6.0               </span></span>
<span><span class="co">## [41] rmarkdown_2.20           crayon_1.5.2            </span></span>
<span><span class="co">## [43] ragg_1.2.5               generics_0.1.3          </span></span>
<span><span class="co">## [45] httr_1.4.4               rjson_0.2.21            </span></span>
<span><span class="co">## [47] DBI_1.1.3                cachem_1.0.6            </span></span>
<span><span class="co">## [49] stringr_1.5.0            zlibbioc_1.45.0         </span></span>
<span><span class="co">## [51] assertthat_0.2.1         parallel_4.3.0          </span></span>
<span><span class="co">## [53] BiocManager_1.30.19      XVector_0.39.0          </span></span>
<span><span class="co">## [55] restfulr_0.0.15          vctrs_0.5.1             </span></span>
<span><span class="co">## [57] Matrix_1.5-3             jsonlite_1.8.4          </span></span>
<span><span class="co">## [59] bookdown_0.32            hms_1.1.2               </span></span>
<span><span class="co">## [61] bit64_4.0.5              systemfonts_1.0.4       </span></span>
<span><span class="co">## [63] jquerylib_0.1.4          glue_1.6.2              </span></span>
<span><span class="co">## [65] pkgdown_2.0.7            codetools_0.2-18        </span></span>
<span><span class="co">## [67] stringi_1.7.12           BiocIO_1.9.2            </span></span>
<span><span class="co">## [69] tibble_3.1.8             pillar_1.8.1            </span></span>
<span><span class="co">## [71] rappdirs_0.3.3           htmltools_0.5.4         </span></span>
<span><span class="co">## [73] GenomeInfoDbData_1.2.9   BSgenome_1.67.3         </span></span>
<span><span class="co">## [75] R6_2.5.1                 dbplyr_2.3.0            </span></span>
<span><span class="co">## [77] textshaping_0.3.6        rprojroot_2.0.3         </span></span>
<span><span class="co">## [79] evaluate_0.20            lattice_0.20-45         </span></span>
<span><span class="co">## [81] R.methodsS3_1.8.2        png_0.1-8               </span></span>
<span><span class="co">## [83] Rsamtools_2.15.1         memoise_2.0.1           </span></span>
<span><span class="co">## [85] bslib_0.4.2              Rcpp_1.0.9              </span></span>
<span><span class="co">## [87] xfun_0.36                fs_1.5.2                </span></span>
<span><span class="co">## [89] pkgconfig_2.0.3</span></span></code></pre>
</details>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>An R package developed by the <a href="https://medschool.cuanschutz.edu/rbi" class="external-link">RNA Bioscience Initiative</a></p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
