<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="raer">
<title>Introducing the raer package • raer</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><link href="../deps/Source_Sans_Pro-0.4.5/font.css" rel="stylesheet">
<link href="../deps/Source_Code_Pro-0.4.5/font.css" rel="stylesheet">
<!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Introducing the raer package">
<meta property="og:description" content="raer">
<meta property="og:image" content="https://rnabioco.github.io/raer/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><script defer data-domain="rnabioco.github.io" src="https://plausible.io/js/plausible.js"></script>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-dark"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">raer</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item">
  <a class="nav-link" href="../articles/raer.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/novel-sites.html">Identifying RNA editing sites using RNA and DNA sequencing data</a>
    <a class="dropdown-item" href="../articles/single-cell.html">Identifying RNA editing sites in droplet single cell datasets</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/rnabioco/raer/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Introducing the raer package</h1>
                        <h4 data-toc-skip class="author">Kent
Riemondy</h4>
            <address class="author_afil">
      University of Colorado School of
Medicine<br><h4 data-toc-skip class="author">Kristen
Wells-Wrasman</h4>
            <address class="author_afil">
      University of Colorado School of Medicine<br><h4 data-toc-skip class="date">2023-03-10</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/rnabioco/raer/blob/HEAD/vignettes/raer.Rmd" class="external-link"><code>vignettes/raer.Rmd</code></a></small>
      <div class="d-none name"><code>raer.Rmd</code></div>
    </address>
</address>
</div>

    
    
<div class="section level2">
<h2 id="differential-editing-tutorial">Differential Editing tutorial<a class="anchor" aria-label="anchor" href="#differential-editing-tutorial"></a>
</h2>
<p>This vignette will demonstrate how to identify RNA editing sites with
condition-specific editing frequencies using the <code>raer</code>
package.</p>
<div class="section level3">
<h3 id="set-up">Set up<a class="anchor" aria-label="anchor" href="#set-up"></a>
</h3>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://rnabioco.github.io/raer">raer</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/SummarizedExperiment" class="external-link">SummarizedExperiment</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/mikelove/DESeq2" class="external-link">DESeq2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jokergoo/ComplexHeatmap" class="external-link">ComplexHeatmap</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://sjmgarnier.github.io/viridis/" class="external-link">viridis</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">rtracklayer</span><span class="op">)</span></span></code></pre></div>
<p>In this vignette a public RNA-seq dataset will be analyzed, <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE99249" class="external-link">GSE99249</a>,
which consists of ADAR1 mutants and control human cell lines,
conditionally treated with Interferon-Beta. For simplicity we will
examine the ADAR1 WT and KO samples both treated with IFN-B, with
triplicate samples.</p>
<p>Aligned BAM files and other necessary files have been preprocessed
for this vignette. These files occupy ~250 Mb of space and by default
they are stored in a temporary directory. Change the
<code>data_dir</code> variable to a valid file path to keep the files
for reuse.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">fns</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/download_GSE99249.html">download_GSE99249</a></span><span class="op">(</span><span class="va">data_dir</span><span class="op">)</span></span></code></pre></div>
<p>Next we’ll collect a vector containing the paths to each BAM file
These BAM files are a subset of the full BAM files, containing only
alignments from chromosome 18.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bam_files</span> <span class="op">&lt;-</span> <span class="va">fns</span><span class="op">$</span><span class="va">bams</span></span>
<span><span class="va">bam_files</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##                   SRR5564260_dedup_sorted.bam </span></span>
<span><span class="co">## "/tmp/RtmpBSlfIB/SRR5564260_dedup_sorted.bam" </span></span>
<span><span class="co">##                   SRR5564261_dedup_sorted.bam </span></span>
<span><span class="co">## "/tmp/RtmpBSlfIB/SRR5564261_dedup_sorted.bam"</span></span></code></pre>
<p>Next we’ll need a FASTA file to quantify the editing sites. We’ll use
a FASTA file only containing chromosome 18 for this demo. The fasta can
be compressed.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fafn</span> <span class="op">&lt;-</span> <span class="va">fns</span><span class="op">$</span><span class="va">fasta</span></span>
<span><span class="va">fafn</span></span></code></pre></div>
<pre><code><span><span class="co">##                   chr18.fasta.bgz </span></span>
<span><span class="co">## "/tmp/RtmpBSlfIB/chr18.fasta.bgz"</span></span></code></pre>
<p>Lastly, we’ll use a BED file containing the coordinates of known
human editing sites from the REDIPortal database. This file can also be
optionally compressed.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sites</span> <span class="op">&lt;-</span> <span class="fu">import</span><span class="op">(</span><span class="va">fns</span><span class="op">$</span><span class="va">bed</span><span class="op">)</span></span>
<span><span class="va">sites</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 250442 ranges and 2 metadata columns:</span></span>
<span><span class="co">##            seqnames    ranges strand |             name     score</span></span>
<span><span class="co">##               &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; |      &lt;character&gt; &lt;numeric&gt;</span></span>
<span><span class="co">##        [1]    chr18     22898      + |    chr18:22898:+         0</span></span>
<span><span class="co">##        [2]    chr18     22909      + |    chr18:22909:+         0</span></span>
<span><span class="co">##        [3]    chr18     22932      + |    chr18:22932:+         0</span></span>
<span><span class="co">##        [4]    chr18     22942      + |    chr18:22942:+         0</span></span>
<span><span class="co">##        [5]    chr18     22948      + |    chr18:22948:+         0</span></span>
<span><span class="co">##        ...      ...       ...    ... .              ...       ...</span></span>
<span><span class="co">##   [250438]    chr18  80251604      + | chr18:80251604:+         0</span></span>
<span><span class="co">##   [250439]    chr18  80251607      + | chr18:80251607:+         0</span></span>
<span><span class="co">##   [250440]    chr18  80251609      + | chr18:80251609:+         0</span></span>
<span><span class="co">##   [250441]    chr18  80251634      + | chr18:80251634:+         0</span></span>
<span><span class="co">##   [250442]    chr18  80251647      + | chr18:80251647:+         0</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="generate-editing-site-read-counts">Generate editing site read counts<a class="anchor" aria-label="anchor" href="#generate-editing-site-read-counts"></a>
</h3>
<p>The <code><a href="../reference/pileup_sites.html">pileup_sites()</a></code> function will calculate base counts
across multiple BAM files. The base counts will be returned as a list of
<code>GRanges</code> objects for each bam file. The
<code><a href="../reference/pileup_sites.html">FilterParam()</a></code> function provides numerous filters to exclude
reads and bases based on commonly used filters for detecting RNA-editing
events. Specific regions can be queried using the <code>region</code>
argument, or specific positions can be queried by providing
<code>bedfile</code> a path to a BED-file.</p>
<p>Internally, <code><a href="../reference/pileup_sites.html">pileup_sites()</a></code> will generate pileups in
memory. However, these files can also be stored in a tabix-indexed
format, for later retrieval, or for querying specific regions.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pileup_sites.html">FilterParam</a></span><span class="op">(</span></span>
<span>    only_keep_variants <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    trim_5p <span class="op">=</span> <span class="fl">5</span>,</span>
<span>    trim_3p <span class="op">=</span> <span class="fl">5</span>,</span>
<span>    min_base_quality <span class="op">=</span> <span class="fl">30L</span>,</span>
<span>    min_mapq <span class="op">=</span> <span class="fl">255L</span>,</span>
<span>    library_type <span class="op">=</span> <span class="st">"fr-first-strand"</span>,</span>
<span>    min_splice_overhang <span class="op">=</span> <span class="fl">10</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">rse</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pileup_sites.html">pileup_sites</a></span><span class="op">(</span><span class="va">bam_files</span>,</span>
<span>    fafile <span class="op">=</span> <span class="va">fafn</span>,</span>
<span>    sites <span class="op">=</span> <span class="va">sites</span>,</span>
<span>    region <span class="op">=</span> <span class="st">"chr18"</span>,</span>
<span>    param <span class="op">=</span> <span class="va">fp</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Pileups data is stored in a <code>summarizedExperiment</code> object,
which permits comparisons across each sample. The <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData()</a></code>
and <code><a href="https://rdrr.io/pkg/MatrixGenerics/man/rowRanges.html" class="external-link">rowRanges()</a></code> slots can be populated with information
related to each editing site, and similarly the <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData()</a></code>
slot can be used to store sample metadata. The summarizedExperiment
object can also interface with other derivative classes, such as
<code>singleCellExperiment()</code>.</p>
<p>The base counts are counted in a stand specific fashion depending on
the <code>library-type</code> parameter. The <code>REF</code> and
<code>ALT</code> bases are in reference to the strand.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rse</span></span></code></pre></div>
<pre><code><span><span class="co">## class: RangedSummarizedExperiment </span></span>
<span><span class="co">## dim: 6287 6 </span></span>
<span><span class="co">## metadata(0):</span></span>
<span><span class="co">## assays(7): ALT nRef ... nC nG</span></span>
<span><span class="co">## rownames(6287): chr18_168574_+ chr18_168577_+ ...</span></span>
<span><span class="co">##   chr18_80172518_- chr18_80174441_-</span></span>
<span><span class="co">## rowData names(4): REF rbpz vpb sor</span></span>
<span><span class="co">## colnames(6): SRR5564260_dedup_sorted.bam</span></span>
<span><span class="co">##   SRR5564261_dedup_sorted.bam ... SRR5564271_dedup_sorted.bam</span></span>
<span><span class="co">##   SRR5564277_dedup_sorted.bam</span></span>
<span><span class="co">## colData names(1): sample</span></span></code></pre>
<p>Next we will add sample metadata to the
<code>summarizedExperiment</code>, which will be used to conduct
differential editing analysis.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">rse</span><span class="op">)</span><span class="op">$</span><span class="va">genotype_treatment</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>        <span class="st">"ADAR1KO Interferon beta"</span>,</span>
<span>        <span class="st">"Wildtype Interferon beta"</span></span>
<span>    <span class="op">)</span>,</span>
<span>    each <span class="op">=</span> <span class="fl">3</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">rse</span><span class="op">)</span><span class="op">$</span><span class="va">genotype</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>        <span class="st">"ADAR1KO"</span>,</span>
<span>        <span class="st">"Wildtype"</span></span>
<span>    <span class="op">)</span>,</span>
<span>    each <span class="op">=</span> <span class="fl">3</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">rse</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 6 rows and 3 columns</span></span>
<span><span class="co">##                                             sample     genotype_treatment</span></span>
<span><span class="co">##                                        &lt;character&gt;            &lt;character&gt;</span></span>
<span><span class="co">## SRR5564260_dedup_sorted.bam SRR5564260_dedup_sor.. ADAR1KO Interferon b..</span></span>
<span><span class="co">## SRR5564261_dedup_sorted.bam SRR5564261_dedup_sor.. ADAR1KO Interferon b..</span></span>
<span><span class="co">## SRR5564269_dedup_sorted.bam SRR5564269_dedup_sor.. ADAR1KO Interferon b..</span></span>
<span><span class="co">## SRR5564270_dedup_sorted.bam SRR5564270_dedup_sor.. Wildtype Interferon ..</span></span>
<span><span class="co">## SRR5564271_dedup_sorted.bam SRR5564271_dedup_sor.. Wildtype Interferon ..</span></span>
<span><span class="co">## SRR5564277_dedup_sorted.bam SRR5564277_dedup_sor.. Wildtype Interferon ..</span></span>
<span><span class="co">##                                genotype</span></span>
<span><span class="co">##                             &lt;character&gt;</span></span>
<span><span class="co">## SRR5564260_dedup_sorted.bam     ADAR1KO</span></span>
<span><span class="co">## SRR5564261_dedup_sorted.bam     ADAR1KO</span></span>
<span><span class="co">## SRR5564269_dedup_sorted.bam     ADAR1KO</span></span>
<span><span class="co">## SRR5564270_dedup_sorted.bam    Wildtype</span></span>
<span><span class="co">## SRR5564271_dedup_sorted.bam    Wildtype</span></span>
<span><span class="co">## SRR5564277_dedup_sorted.bam    Wildtype</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="prepare-for-differential-editing">Prepare for differential editing<a class="anchor" aria-label="anchor" href="#prepare-for-differential-editing"></a>
</h3>
<p>We next use the <code>calc_edit_frequency</code> function to identify
the percent of edits for each position and sample. With the
<code>drop = TRUE</code> argument we will also exclude sites without an
adenosine. The editing frequencies will not be used for differential
editing analysis, which will be conducted using the raw counts, however
these are useful for filtering and visualization.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se_filtered</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calc_edit_frequency.html">calc_edit_frequency</a></span><span class="op">(</span><span class="va">rse</span>,</span>
<span>    edit_from <span class="op">=</span> <span class="st">"A"</span>,</span>
<span>    edit_to <span class="op">=</span> <span class="st">"G"</span>,</span>
<span>    drop <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>We can next subset the <code>summarizedExperiment</code> object to
exclude low frequency editing events. For this analysis we will require
that an editing site shows editing in at least 1 sample, and have at
least 5 counts in each sample.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">has_editing</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">se_filtered</span>, <span class="st">"edit_freq"</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="fl">1</span></span>
<span><span class="va">has_depth</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="op">(</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">se_filtered</span>, <span class="st">"nRef"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">se_filtered</span>, <span class="st">"nAlt"</span><span class="op">)</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="fl">5</span><span class="op">)</span><span class="op">)</span> <span class="op">==</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">se_filtered</span><span class="op">)</span></span>
<span></span>
<span><span class="va">se_filtered</span> <span class="op">&lt;-</span> <span class="va">se_filtered</span><span class="op">[</span><span class="va">has_editing</span> <span class="op">&amp;</span> <span class="va">has_depth</span>, <span class="op">]</span></span>
<span><span class="va">se_filtered</span></span></code></pre></div>
<pre><code><span><span class="co">## class: RangedSummarizedExperiment </span></span>
<span><span class="co">## dim: 608 6 </span></span>
<span><span class="co">## metadata(0):</span></span>
<span><span class="co">## assays(9): ALT nRef ... depth edit_freq</span></span>
<span><span class="co">## rownames(608): chr18_204625_+ chr18_204626_+ ... chr18_80158602_-</span></span>
<span><span class="co">##   chr18_80158603_-</span></span>
<span><span class="co">## rowData names(4): REF rbpz vpb sor</span></span>
<span><span class="co">## colnames(6): SRR5564260_dedup_sorted.bam</span></span>
<span><span class="co">##   SRR5564261_dedup_sorted.bam ... SRR5564271_dedup_sorted.bam</span></span>
<span><span class="co">##   SRR5564277_dedup_sorted.bam</span></span>
<span><span class="co">## colData names(5): sample genotype_treatment genotype n_sites</span></span>
<span><span class="co">##   edit_idx</span></span></code></pre>
<p>Once the object has been filtered, you can prepare it for DE. This
means making a new object that contains an assay with read counts if
both the alt and ref alleles in a single matrix.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">deobj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/prep_for_de.html">prep_for_de</a></span><span class="op">(</span><span class="va">se_filtered</span>,</span>
<span>    min_prop <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>    max_prop <span class="op">=</span> <span class="fl">0.9</span>,</span>
<span>    min_samples <span class="op">=</span> <span class="fl">3</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">deobj</span>, <span class="st">"counts"</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##                  SRR5564260_dedup_sorted.bam_ref</span></span>
<span><span class="co">## chr18_692390_-                                 6</span></span>
<span><span class="co">## chr18_692402_-                                 5</span></span>
<span><span class="co">## chr18_21631169_+                               8</span></span>
<span><span class="co">##                  SRR5564261_dedup_sorted.bam_ref</span></span>
<span><span class="co">## chr18_692390_-                                12</span></span>
<span><span class="co">## chr18_692402_-                                10</span></span>
<span><span class="co">## chr18_21631169_+                              10</span></span>
<span><span class="co">##                  SRR5564269_dedup_sorted.bam_ref</span></span>
<span><span class="co">## chr18_692390_-                                13</span></span>
<span><span class="co">## chr18_692402_-                                13</span></span>
<span><span class="co">## chr18_21631169_+                              15</span></span>
<span><span class="co">##                  SRR5564270_dedup_sorted.bam_ref</span></span>
<span><span class="co">## chr18_692390_-                                18</span></span>
<span><span class="co">## chr18_692402_-                                14</span></span>
<span><span class="co">## chr18_21631169_+                              10</span></span>
<span><span class="co">##                  SRR5564271_dedup_sorted.bam_ref</span></span>
<span><span class="co">## chr18_692390_-                                13</span></span>
<span><span class="co">## chr18_692402_-                                10</span></span>
<span><span class="co">## chr18_21631169_+                               3</span></span>
<span><span class="co">##                  SRR5564277_dedup_sorted.bam_ref</span></span>
<span><span class="co">## chr18_692390_-                                 8</span></span>
<span><span class="co">## chr18_692402_-                                 3</span></span>
<span><span class="co">## chr18_21631169_+                               4</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="run-differential-editing-deseq2">Run differential editing (DESeq2)<a class="anchor" aria-label="anchor" href="#run-differential-editing-deseq2"></a>
</h3>
<p>At this stage, you can use the object to perform DE yourself or you
can continue with our pre built functions</p>
<p>For differential editing, we use the design
<code>design &lt;- ~0 + condition:sample + condition:count</code>.</p>
<p>For the samples, you can leave as is or combine so the same sample
name shows up in both the treatment and control. These results are not
identical but they are close. In my hands, the same genes come out, but
the p values and log fold change values are slightly different.</p>
<p>It is probably best to update the levels of your object, but if you
don’t, this will still work.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">de_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/perform_de.html">perform_de</a></span><span class="op">(</span><span class="va">deobj</span>,</span>
<span>    sample_col <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>    condition_col <span class="op">=</span> <span class="st">"genotype"</span>,</span>
<span>    condition_control <span class="op">=</span> <span class="st">"Wildtype"</span>,</span>
<span>    condition_treatment <span class="op">=</span> <span class="st">"ADAR1KO"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>This returns a list containing the dds object, the full results, the
significant results, and the model matrix.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">de_results</span><span class="op">$</span><span class="va">sig_results</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##                      logFC   logCPM        LR       PValue      FDR</span></span>
<span><span class="co">## chr18_50393235_+ -8.033976 21.73892 11.360332 0.0007502942 0.020463</span></span>
<span><span class="co">## chr18_35248579_+ -8.129393 22.69356 10.096657 0.0014853852 0.020463</span></span>
<span><span class="co">## chr18_23297504_- -7.692014 21.72512 10.028647 0.0015412421 0.020463</span></span>
<span><span class="co">## chr18_21631272_+ -7.091754 20.75260  9.796561 0.0017483853 0.020463</span></span>
<span><span class="co">## chr18_50392804_+ -7.446234 21.07996  9.785101 0.0017593156 0.020463</span></span></code></pre>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">top_sites</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">de_results</span><span class="op">$</span><span class="va">sig_results</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">20</span><span class="op">]</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html" class="external-link">Heatmap</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">se_filtered</span>, <span class="st">"edit_freq"</span><span class="op">)</span><span class="op">[</span><span class="va">top_sites</span>, <span class="op">]</span>,</span>
<span>    col <span class="op">=</span> <span class="fu"><a href="https://sjmgarnier.github.io/viridisLite/reference/viridis.html" class="external-link">viridis</a></span><span class="op">(</span><span class="fl">100</span><span class="op">)</span>,</span>
<span>    column_labels <span class="op">=</span> <span class="va">se_filtered</span><span class="op">$</span><span class="va">genotype_treatment</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="raer_files/figure-html/unnamed-chunk-15-1.png" width="480"></p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/unlink.html" class="external-link">unlink</a></span><span class="op">(</span><span class="va">data_dir</span><span class="op">)</span></span></code></pre></div>
<details style="margin-bottom:10px;"><summary>
Session info
</summary><div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R Under development (unstable) (2023-03-02 r83926)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">## Running under: Ubuntu 22.04.2 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] grid      stats4    stats     graphics  grDevices utils     datasets </span></span>
<span><span class="co">## [8] methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] rtracklayer_1.59.1          viridis_0.6.2              </span></span>
<span><span class="co">##  [3] viridisLite_0.4.1           ComplexHeatmap_2.15.0      </span></span>
<span><span class="co">##  [5] DESeq2_1.39.8               SummarizedExperiment_1.29.1</span></span>
<span><span class="co">##  [7] Biobase_2.59.0              GenomicRanges_1.51.4       </span></span>
<span><span class="co">##  [9] GenomeInfoDb_1.35.15        IRanges_2.33.0             </span></span>
<span><span class="co">## [11] S4Vectors_0.37.4            BiocGenerics_0.45.0        </span></span>
<span><span class="co">## [13] MatrixGenerics_1.11.0       matrixStats_0.63.0         </span></span>
<span><span class="co">## [15] raer_0.99.0                 BiocStyle_2.27.1           </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] RColorBrewer_1.1-3       jsonlite_1.8.4          </span></span>
<span><span class="co">##   [3] shape_1.4.6              magrittr_2.0.3          </span></span>
<span><span class="co">##   [5] GenomicFeatures_1.51.4   rmarkdown_2.20          </span></span>
<span><span class="co">##   [7] GlobalOptions_0.1.2      fs_1.6.1                </span></span>
<span><span class="co">##   [9] BiocIO_1.9.2             zlibbioc_1.45.0         </span></span>
<span><span class="co">##  [11] ragg_1.2.5               vctrs_0.5.2             </span></span>
<span><span class="co">##  [13] memoise_2.0.1            Cairo_1.6-0             </span></span>
<span><span class="co">##  [15] Rsamtools_2.15.1         RCurl_1.98-1.10         </span></span>
<span><span class="co">##  [17] htmltools_0.5.4          progress_1.2.2          </span></span>
<span><span class="co">##  [19] curl_5.0.0               sass_0.4.5              </span></span>
<span><span class="co">##  [21] bslib_0.4.2              desc_1.4.2              </span></span>
<span><span class="co">##  [23] cachem_1.0.7             GenomicAlignments_1.35.1</span></span>
<span><span class="co">##  [25] lifecycle_1.0.3          iterators_1.0.14        </span></span>
<span><span class="co">##  [27] pkgconfig_2.0.3          Matrix_1.5-3            </span></span>
<span><span class="co">##  [29] R6_2.5.1                 fastmap_1.1.1           </span></span>
<span><span class="co">##  [31] GenomeInfoDbData_1.2.9   clue_0.3-64             </span></span>
<span><span class="co">##  [33] digest_0.6.31            colorspace_2.1-0        </span></span>
<span><span class="co">##  [35] AnnotationDbi_1.61.1     rprojroot_2.0.3         </span></span>
<span><span class="co">##  [37] textshaping_0.3.6        RSQLite_2.3.0           </span></span>
<span><span class="co">##  [39] filelock_1.0.2           fansi_1.0.4             </span></span>
<span><span class="co">##  [41] httr_1.4.5               compiler_4.3.0          </span></span>
<span><span class="co">##  [43] bit64_4.0.5              doParallel_1.0.17       </span></span>
<span><span class="co">##  [45] BiocParallel_1.33.9      DBI_1.1.3               </span></span>
<span><span class="co">##  [47] highr_0.10               R.utils_2.12.2          </span></span>
<span><span class="co">##  [49] biomaRt_2.55.0           rappdirs_0.3.3          </span></span>
<span><span class="co">##  [51] DelayedArray_0.25.0      rjson_0.2.21            </span></span>
<span><span class="co">##  [53] tools_4.3.0              R.oo_1.25.0             </span></span>
<span><span class="co">##  [55] glue_1.6.2               restfulr_0.0.15         </span></span>
<span><span class="co">##  [57] cluster_2.1.4            generics_0.1.3          </span></span>
<span><span class="co">##  [59] gtable_0.3.1             BSgenome_1.67.4         </span></span>
<span><span class="co">##  [61] R.methodsS3_1.8.2        data.table_1.14.8       </span></span>
<span><span class="co">##  [63] hms_1.1.2                xml2_1.3.3              </span></span>
<span><span class="co">##  [65] utf8_1.2.3               XVector_0.39.0          </span></span>
<span><span class="co">##  [67] foreach_1.5.2            pillar_1.8.1            </span></span>
<span><span class="co">##  [69] stringr_1.5.0            limma_3.55.5            </span></span>
<span><span class="co">##  [71] circlize_0.4.15          splines_4.3.0           </span></span>
<span><span class="co">##  [73] dplyr_1.1.0              BiocFileCache_2.7.2     </span></span>
<span><span class="co">##  [75] lattice_0.20-45          bit_4.0.5               </span></span>
<span><span class="co">##  [77] tidyselect_1.2.0         locfit_1.5-9.7          </span></span>
<span><span class="co">##  [79] Biostrings_2.67.0        knitr_1.42              </span></span>
<span><span class="co">##  [81] gridExtra_2.3            bookdown_0.33           </span></span>
<span><span class="co">##  [83] edgeR_3.41.6             xfun_0.37               </span></span>
<span><span class="co">##  [85] stringi_1.7.12           yaml_2.3.7              </span></span>
<span><span class="co">##  [87] evaluate_0.20            codetools_0.2-19        </span></span>
<span><span class="co">##  [89] tibble_3.2.0             BiocManager_1.30.20     </span></span>
<span><span class="co">##  [91] cli_3.6.0                systemfonts_1.0.4       </span></span>
<span><span class="co">##  [93] munsell_0.5.0            jquerylib_0.1.4         </span></span>
<span><span class="co">##  [95] Rcpp_1.0.10              dbplyr_2.3.1            </span></span>
<span><span class="co">##  [97] png_0.1-8                XML_3.99-0.13           </span></span>
<span><span class="co">##  [99] parallel_4.3.0           ellipsis_0.3.2          </span></span>
<span><span class="co">## [101] pkgdown_2.0.7            ggplot2_3.4.1           </span></span>
<span><span class="co">## [103] blob_1.2.3               prettyunits_1.1.1       </span></span>
<span><span class="co">## [105] bitops_1.0-7             scales_1.2.1            </span></span>
<span><span class="co">## [107] purrr_1.0.1              crayon_1.5.2            </span></span>
<span><span class="co">## [109] GetoptLong_1.0.5         rlang_1.0.6             </span></span>
<span><span class="co">## [111] KEGGREST_1.39.0</span></span></code></pre>
</details>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>An R package developed by the <a href="https://medschool.cuanschutz.edu/rbi" class="external-link">RNA Bioscience Initiative</a></p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
