<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="raer">
<title>Introducing the raer package • raer</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><link href="../deps/Source_Sans_Pro-0.4.6/font.css" rel="stylesheet">
<link href="../deps/Source_Code_Pro-0.4.6/font.css" rel="stylesheet">
<!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Introducing the raer package">
<meta property="og:description" content="raer">
<meta property="og:image" content="https://rnabioco.github.io/raer/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><script defer data-domain="rnabioco.github.io" src="https://plausible.io/js/plausible.js"></script>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-dark"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">raer</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.2</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item">
  <a class="nav-link" href="../articles/raer.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/novel-sites.html">Identifying RNA editing sites using RNA and DNA sequencing data</a>
    <a class="dropdown-item" href="../articles/single-cell.html">Identifying RNA editing sites in droplet single cell datasets</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/rnabioco/raer/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Introducing the raer package</h1>
                        <h4 data-toc-skip class="author">Kent
Riemondy</h4>
            <address class="author_afil">
      University of Colorado School of
Medicine<br><h4 data-toc-skip class="author">Kristen
Wells-Wrasman</h4>
            <address class="author_afil">
      University of Colorado School of Medicine<br><h4 data-toc-skip class="date">2023-06-13</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/rnabioco/raer/blob/HEAD/vignettes/raer.Rmd" class="external-link"><code>vignettes/raer.Rmd</code></a></small>
      <div class="d-none name"><code>raer.Rmd</code></div>
    </address>
</address>
</div>

    
    
<div class="section level2">
<h2 id="differential-editing-tutorial">Differential Editing tutorial<a class="anchor" aria-label="anchor" href="#differential-editing-tutorial"></a>
</h2>
<p>This vignette will demonstrate how to identify RNA editing sites with
condition-specific editing frequencies using the <code>raer</code>
package.</p>
<div class="section level3">
<h3 id="set-up">Set up<a class="anchor" aria-label="anchor" href="#set-up"></a>
</h3>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://rnabioco.github.io/raer">raer</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/SummarizedExperiment" class="external-link">SummarizedExperiment</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/mikelove/DESeq2" class="external-link">DESeq2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jokergoo/ComplexHeatmap" class="external-link">ComplexHeatmap</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://sjmgarnier.github.io/viridis/" class="external-link">viridis</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">rtracklayer</span><span class="op">)</span></span></code></pre></div>
<p>In this vignette a public RNA-seq dataset will be analyzed, <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE99249" class="external-link">GSE99249</a>,
which consists of ADAR1 mutants and control human cell lines,
conditionally treated with Interferon-Beta. For simplicity we will
examine the ADAR1 WT and KO samples both treated with IFN-B, with
triplicate samples.</p>
<p>Aligned BAM files and other necessary files have been preprocessed
for this vignette. These files occupy ~250 Mb of space and they are
stored in a <code>BiocFileCache</code>.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fns</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/download_GSE99249.html">download_GSE99249</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>Next we’ll collect a vector containing the paths to each BAM file
These BAM files are a subset of the full BAM files, containing only
alignments from chromosome 18.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bam_files</span> <span class="op">&lt;-</span> <span class="va">fns</span><span class="op">$</span><span class="va">bams</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">bam_files</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">sub</a></span><span class="op">(</span><span class="st">"_dedup_sorted.bam"</span>, <span class="st">""</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/path.html" class="external-link">basename</a></span><span class="op">(</span><span class="va">bam_files</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">bam_files</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##                                               SRR5564260 </span></span>
<span><span class="co">## "/github/home/.cache/R/raer/SRR5564260_dedup_sorted.bam" </span></span>
<span><span class="co">##                                               SRR5564261 </span></span>
<span><span class="co">## "/github/home/.cache/R/raer/SRR5564261_dedup_sorted.bam"</span></span></code></pre>
<p>Next we’ll need a FASTA file to quantify the editing sites. We’ll use
a FASTA file only containing chromosome 18 for this demo. The fasta can
be compressed.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fafn</span> <span class="op">&lt;-</span> <span class="va">fns</span><span class="op">$</span><span class="va">fasta</span></span>
<span><span class="va">fafn</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "/github/home/.cache/R/raer/6ca40f6f166_chr18.fasta.bgz"</span></span></code></pre>
<p>Lastly, we’ll use a BED file containing the coordinates of known
human editing sites from the REDIPortal database. This file can also be
optionally compressed.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sites</span> <span class="op">&lt;-</span> <span class="fu">import</span><span class="op">(</span><span class="va">fns</span><span class="op">$</span><span class="va">bed</span><span class="op">)</span></span>
<span><span class="va">sites</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 250442 ranges and 2 metadata columns:</span></span>
<span><span class="co">##            seqnames    ranges strand |             name     score</span></span>
<span><span class="co">##               &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; |      &lt;character&gt; &lt;numeric&gt;</span></span>
<span><span class="co">##        [1]    chr18     22898      + |    chr18:22898:+         0</span></span>
<span><span class="co">##        [2]    chr18     22909      + |    chr18:22909:+         0</span></span>
<span><span class="co">##        [3]    chr18     22932      + |    chr18:22932:+         0</span></span>
<span><span class="co">##        [4]    chr18     22942      + |    chr18:22942:+         0</span></span>
<span><span class="co">##        [5]    chr18     22948      + |    chr18:22948:+         0</span></span>
<span><span class="co">##        ...      ...       ...    ... .              ...       ...</span></span>
<span><span class="co">##   [250438]    chr18  80251604      + | chr18:80251604:+         0</span></span>
<span><span class="co">##   [250439]    chr18  80251607      + | chr18:80251607:+         0</span></span>
<span><span class="co">##   [250440]    chr18  80251609      + | chr18:80251609:+         0</span></span>
<span><span class="co">##   [250441]    chr18  80251634      + | chr18:80251634:+         0</span></span>
<span><span class="co">##   [250442]    chr18  80251647      + | chr18:80251647:+         0</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="generate-editing-site-read-counts">Generate editing site read counts<a class="anchor" aria-label="anchor" href="#generate-editing-site-read-counts"></a>
</h3>
<p>The <code><a href="../reference/pileup_sites.html">pileup_sites()</a></code> function will calculate base counts
across multiple BAM files. The base counts will be returned as a list of
<code>GRanges</code> objects for each bam file. The
<code><a href="../reference/pileup_sites.html">FilterParam()</a></code> function provides numerous filters to exclude
reads and bases based on commonly used filters for detecting RNA-editing
events. Specific regions can be queried using the <code>region</code>
argument, or specific positions can be queried by providing
<code>bedfile</code> a path to a BED-file.</p>
<p>Internally, <code><a href="../reference/pileup_sites.html">pileup_sites()</a></code> will generate pileups in
memory. However, these files can also be stored in a tabix-indexed
format, for later retrieval, or for querying specific regions.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pileup_sites.html">FilterParam</a></span><span class="op">(</span></span>
<span>    only_keep_variants <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    trim_5p <span class="op">=</span> <span class="fl">5</span>,</span>
<span>    trim_3p <span class="op">=</span> <span class="fl">5</span>,</span>
<span>    min_base_quality <span class="op">=</span> <span class="fl">30L</span>,</span>
<span>    min_mapq <span class="op">=</span> <span class="fl">255L</span>,</span>
<span>    library_type <span class="op">=</span> <span class="st">"fr-first-strand"</span>,</span>
<span>    min_splice_overhang <span class="op">=</span> <span class="fl">10</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">rse</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pileup_sites.html">pileup_sites</a></span><span class="op">(</span><span class="va">bam_files</span>,</span>
<span>    fafile <span class="op">=</span> <span class="va">fafn</span>,</span>
<span>    sites <span class="op">=</span> <span class="va">sites</span>,</span>
<span>    region <span class="op">=</span> <span class="st">"chr18"</span>,</span>
<span>    param <span class="op">=</span> <span class="va">fp</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Pileups data is stored in a <code>summarizedExperiment</code> object,
which permits comparisons across each sample. The <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData()</a></code>
and <code>rowRanges()</code> slots can be populated with information
related to each editing site, and similarly the <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData()</a></code>
slot can be used to store sample metadata. The summarizedExperiment
object can also interface with other derivative classes, such as
<code>singleCellExperiment()</code>.</p>
<p>The base counts are counted in a stand specific fashion depending on
the <code>library-type</code> parameter. The <code>REF</code> and
<code>ALT</code> bases are in reference to the strand.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rse</span></span></code></pre></div>
<pre><code><span><span class="co">## class: RangedSummarizedExperiment </span></span>
<span><span class="co">## dim: 6287 6 </span></span>
<span><span class="co">## metadata(0):</span></span>
<span><span class="co">## assays(7): ALT nRef ... nC nG</span></span>
<span><span class="co">## rownames(6287): site_chr18_168574_1 site_chr18_168577_1 ...</span></span>
<span><span class="co">##   site_chr18_80172518_2 site_chr18_80174441_2</span></span>
<span><span class="co">## rowData names(4): REF rpbz vdb sor</span></span>
<span><span class="co">## colnames(6): SRR5564260 SRR5564261 ... SRR5564271 SRR5564277</span></span>
<span><span class="co">## colData names(1): sample</span></span></code></pre>
<p>Next we will add sample metadata to the
<code>summarizedExperiment</code>, which will be used to conduct
differential editing analysis.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">rse</span><span class="op">)</span><span class="op">$</span><span class="va">genotype_treatment</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>        <span class="st">"ADAR1KO Interferon beta"</span>,</span>
<span>        <span class="st">"Wildtype Interferon beta"</span></span>
<span>    <span class="op">)</span>,</span>
<span>    each <span class="op">=</span> <span class="fl">3</span></span>
<span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">rse</span><span class="op">)</span><span class="op">$</span><span class="va">genotype</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>        <span class="st">"ADAR1KO"</span>,</span>
<span>        <span class="st">"Wildtype"</span></span>
<span>    <span class="op">)</span>,</span>
<span>    each <span class="op">=</span> <span class="fl">3</span></span>
<span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">rse</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 6 rows and 3 columns</span></span>
<span><span class="co">##                 sample       genotype_treatment genotype</span></span>
<span><span class="co">##            &lt;character&gt;                 &lt;factor&gt; &lt;factor&gt;</span></span>
<span><span class="co">## SRR5564260  SRR5564260 ADAR1KO Interferon beta  ADAR1KO </span></span>
<span><span class="co">## SRR5564261  SRR5564261 ADAR1KO Interferon beta  ADAR1KO </span></span>
<span><span class="co">## SRR5564269  SRR5564269 ADAR1KO Interferon beta  ADAR1KO </span></span>
<span><span class="co">## SRR5564270  SRR5564270 Wildtype Interferon beta Wildtype</span></span>
<span><span class="co">## SRR5564271  SRR5564271 Wildtype Interferon beta Wildtype</span></span>
<span><span class="co">## SRR5564277  SRR5564277 Wildtype Interferon beta Wildtype</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="prepare-for-differential-editing">Prepare for differential editing<a class="anchor" aria-label="anchor" href="#prepare-for-differential-editing"></a>
</h3>
<p>We next use the <code>calc_edit_frequency</code> function to identify
the percent of edits for each position and sample. With the
<code>drop = TRUE</code> argument we will also exclude sites without an
adenosine. The editing frequencies will not be used for differential
editing analysis, which will be conducted using the raw counts, however
these are useful for filtering and visualization.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se_filtered</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calc_edit_frequency.html">calc_edit_frequency</a></span><span class="op">(</span><span class="va">rse</span>,</span>
<span>    edit_from <span class="op">=</span> <span class="st">"A"</span>,</span>
<span>    edit_to <span class="op">=</span> <span class="st">"G"</span>,</span>
<span>    drop <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>We can next subset the <code>summarizedExperiment</code> object to
exclude low frequency editing events. For this analysis we will require
that an editing site shows editing in at least 1 sample, and have at
least 5 counts in each sample.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">has_editing</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MatrixGenerics/man/rowSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">se_filtered</span>, <span class="st">"edit_freq"</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="fl">1</span></span>
<span><span class="va">has_depth</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MatrixGenerics/man/rowSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="op">(</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">se_filtered</span>, <span class="st">"nRef"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">se_filtered</span>, <span class="st">"nAlt"</span><span class="op">)</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="fl">5</span><span class="op">)</span><span class="op">)</span> <span class="op">==</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">se_filtered</span><span class="op">)</span></span>
<span></span>
<span><span class="va">se_filtered</span> <span class="op">&lt;-</span> <span class="va">se_filtered</span><span class="op">[</span><span class="va">has_editing</span> <span class="op">&amp;</span> <span class="va">has_depth</span>, <span class="op">]</span></span>
<span><span class="va">se_filtered</span></span></code></pre></div>
<pre><code><span><span class="co">## class: RangedSummarizedExperiment </span></span>
<span><span class="co">## dim: 608 6 </span></span>
<span><span class="co">## metadata(0):</span></span>
<span><span class="co">## assays(9): ALT nRef ... depth edit_freq</span></span>
<span><span class="co">## rownames(608): site_chr18_204625_1 site_chr18_204626_1 ...</span></span>
<span><span class="co">##   site_chr18_80158602_2 site_chr18_80158603_2</span></span>
<span><span class="co">## rowData names(4): REF rpbz vdb sor</span></span>
<span><span class="co">## colnames(6): SRR5564260 SRR5564261 ... SRR5564271 SRR5564277</span></span>
<span><span class="co">## colData names(5): sample genotype_treatment genotype n_sites</span></span>
<span><span class="co">##   edit_idx</span></span></code></pre>
<p>Once the object has been filtered, you can prepare it for DE. This
means making a new object that contains an assay with read counts if
both the alt and ref alleles in a single matrix.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">deobj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/make_de_object.html">make_de_object</a></span><span class="op">(</span><span class="va">se_filtered</span>, min_prop <span class="op">=</span> <span class="fl">0.05</span>, min_samples <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">deobj</span>, <span class="st">"counts"</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##                       SRR5564260_ref SRR5564261_ref SRR5564269_ref</span></span>
<span><span class="co">## site_chr18_692390_2                6             12             13</span></span>
<span><span class="co">## site_chr18_692402_2                5             10             13</span></span>
<span><span class="co">## site_chr18_21631169_1              8             10             15</span></span>
<span><span class="co">##                       SRR5564270_ref SRR5564271_ref SRR5564277_ref</span></span>
<span><span class="co">## site_chr18_692390_2               18             13              8</span></span>
<span><span class="co">## site_chr18_692402_2               14             10              3</span></span>
<span><span class="co">## site_chr18_21631169_1             10              3              4</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="run-differential-editing-deseq2">Run differential editing (DESeq2)<a class="anchor" aria-label="anchor" href="#run-differential-editing-deseq2"></a>
</h3>
<p>At this stage, you can use the object to perform DE yourself or use
<code><a href="../reference/find_de_sites.html">find_de_sites()</a></code> to use <code>edgeR</code> or
<code>DESeq2</code> to identify condition specific editing events. For
differential editing, we use the design
<code>design &lt;- ~0 + condition:sample + condition:count</code> and
perform testing to compare the edited read counts against unedited read
counts.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">deobj</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 12 rows and 6 columns</span></span>
<span><span class="co">##                     sample       genotype_treatment genotype   n_sites</span></span>
<span><span class="co">##                &lt;character&gt;                 &lt;factor&gt; &lt;factor&gt; &lt;numeric&gt;</span></span>
<span><span class="co">## SRR5564260_ref  SRR5564260 ADAR1KO Interferon beta  ADAR1KO        248</span></span>
<span><span class="co">## SRR5564261_ref  SRR5564261 ADAR1KO Interferon beta  ADAR1KO        137</span></span>
<span><span class="co">## SRR5564269_ref  SRR5564269 ADAR1KO Interferon beta  ADAR1KO        294</span></span>
<span><span class="co">## SRR5564270_ref  SRR5564270 Wildtype Interferon beta Wildtype      2366</span></span>
<span><span class="co">## SRR5564271_ref  SRR5564271 Wildtype Interferon beta Wildtype      1965</span></span>
<span><span class="co">## ...                    ...                      ...      ...       ...</span></span>
<span><span class="co">## SRR5564261_alt  SRR5564261 ADAR1KO Interferon beta  ADAR1KO        137</span></span>
<span><span class="co">## SRR5564269_alt  SRR5564269 ADAR1KO Interferon beta  ADAR1KO        294</span></span>
<span><span class="co">## SRR5564270_alt  SRR5564270 Wildtype Interferon beta Wildtype      2366</span></span>
<span><span class="co">## SRR5564271_alt  SRR5564271 Wildtype Interferon beta Wildtype      1965</span></span>
<span><span class="co">## SRR5564277_alt  SRR5564277 Wildtype Interferon beta Wildtype      2378</span></span>
<span><span class="co">##                  edit_idx    count</span></span>
<span><span class="co">##                 &lt;numeric&gt; &lt;factor&gt;</span></span>
<span><span class="co">## SRR5564260_ref 0.00869646      ref</span></span>
<span><span class="co">## SRR5564261_ref 0.00790857      ref</span></span>
<span><span class="co">## SRR5564269_ref 0.00792891      ref</span></span>
<span><span class="co">## SRR5564270_ref 0.05709817      ref</span></span>
<span><span class="co">## SRR5564271_ref 0.06075750      ref</span></span>
<span><span class="co">## ...                   ...      ...</span></span>
<span><span class="co">## SRR5564261_alt 0.00790857      alt</span></span>
<span><span class="co">## SRR5564269_alt 0.00792891      alt</span></span>
<span><span class="co">## SRR5564270_alt 0.05709817      alt</span></span>
<span><span class="co">## SRR5564271_alt 0.06075750      alt</span></span>
<span><span class="co">## SRR5564277_alt 0.05874736      alt</span></span></code></pre>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">deobj</span><span class="op">$</span><span class="va">sample</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">deobj</span><span class="op">$</span><span class="va">sample</span><span class="op">)</span></span>
<span><span class="va">de_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/find_de_sites.html">find_de_sites</a></span><span class="op">(</span><span class="va">deobj</span>, </span>
<span>    test <span class="op">=</span> <span class="st">"DESeq2"</span>,</span>
<span>    sample_col <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>    condition_col <span class="op">=</span> <span class="st">"genotype"</span>,</span>
<span>    condition_control <span class="op">=</span> <span class="st">"Wildtype"</span>,</span>
<span>    condition_treatment <span class="op">=</span> <span class="st">"ADAR1KO"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>This returns a list containing the dds object, the full results, the
significant results, and the model matrix.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">de_results</span><span class="op">$</span><span class="va">sig_results</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##                        baseMean log2FoldChange    lfcSE      stat</span></span>
<span><span class="co">## site_chr18_32070144_1  6.750000      -2.494722 1.086252 -2.296633</span></span>
<span><span class="co">## site_chr18_35247800_1  8.333333      -2.856115 1.128355 -2.531220</span></span>
<span><span class="co">## site_chr18_23296417_2 14.583333      -3.058616 0.875257 -3.494535</span></span>
<span><span class="co">## site_chr18_32120079_1  7.916667      -3.291919 1.454411 -2.263403</span></span>
<span><span class="co">## site_chr18_21632002_1  9.250000      -3.318768 1.549462 -2.141883</span></span>
<span><span class="co">##                             pvalue        padj</span></span>
<span><span class="co">## site_chr18_32070144_1 0.0216397271 0.031738266</span></span>
<span><span class="co">## site_chr18_35247800_1 0.0113666451 0.021282229</span></span>
<span><span class="co">## site_chr18_23296417_2 0.0004748879 0.002088297</span></span>
<span><span class="co">## site_chr18_32120079_1 0.0236108504 0.034061555</span></span>
<span><span class="co">## site_chr18_21632002_1 0.0322028647 0.044981779</span></span></code></pre>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">top_sites</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">de_results</span><span class="op">$</span><span class="va">sig_results</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">20</span><span class="op">]</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html" class="external-link">Heatmap</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">se_filtered</span>, <span class="st">"edit_freq"</span><span class="op">)</span><span class="op">[</span><span class="va">top_sites</span>, <span class="op">]</span>, </span>
<span>        name <span class="op">=</span> <span class="st">"editing frequency"</span>,</span>
<span>        col <span class="op">=</span> <span class="fu"><a href="https://sjmgarnier.github.io/viridisLite/reference/viridis.html" class="external-link">viridis</a></span><span class="op">(</span><span class="fl">100</span><span class="op">)</span>,</span>
<span>        column_labels <span class="op">=</span> <span class="va">se_filtered</span><span class="op">$</span><span class="va">genotype_treatment</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="raer_files/figure-html/unnamed-chunk-16-1.png" width="480"></p>
</div>
<div class="section level3">
<h3 id="examine-overall-editing-activites-using-the-alu-editing-index">Examine overall editing activites using the Alu Editing Index<a class="anchor" aria-label="anchor" href="#examine-overall-editing-activites-using-the-alu-editing-index"></a>
</h3>
<p>For some studies it may be informative to assess the overall ADAR
editing activity rather than examining individual editing sites. The
<strong>Alu Editing Index</strong> (AEI), developed by <span class="citation">Roth, Levanon, and Eisenberg (2019)</span>, is a metric
that summarizes that amount of editing occurring at ALU elements which
account for the vast majority of A-to-I editing (&gt; 99%) in
humans.</p>
<p><code>raer</code> provides <code><a href="../reference/calc_AEI.html">calc_AEI()</a></code> to calculate the
AEI metric. Many of the same parameters used for
<code><a href="../reference/pileup_sites.html">pileup_sites()</a></code> are available in
<code><a href="../reference/calc_AEI.html">calc_AEI()</a></code>.</p>
<p>First we will use the <code>AnnotationHub</code> to obtain
coordinates for ALU elements. For this example we will only examine a
subset of ALUs on <code>chr18</code>. Then we will use a
<code>SNPlocs</code> package to identify SNPs overlapping the ALU
elements from the dbSNP database, which will be excluded from the AEI
calculation. The SNPlocs coordinates are <code>NCBI</code> based,
whereas the <code>ALU</code> elements are based on <code>hg38</code>, we
will therefore convert between the two as needed to obtain SNP and ALU
element coordinates based on <code>hg38</code>.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">AnnotationHub</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SNPlocs.Hsapiens.dbSNP144.GRCh38</span><span class="op">)</span></span>
<span></span>
<span><span class="va">ah</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationHub/man/AnnotationHub-class.html" class="external-link">AnnotationHub</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">rmsk_hg38</span> <span class="op">&lt;-</span> <span class="va">ah</span><span class="op">[[</span><span class="st">"AH99003"</span><span class="op">]</span><span class="op">]</span></span>
<span></span>
<span><span class="va">alus</span> <span class="op">&lt;-</span> <span class="va">rmsk_hg38</span><span class="op">[</span><span class="va">rmsk_hg38</span><span class="op">$</span><span class="va">repFamily</span> <span class="op">==</span> <span class="st">"Alu"</span>, <span class="op">]</span></span>
<span><span class="va">alus</span> <span class="op">&lt;-</span> <span class="va">alus</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqnames</a></span><span class="op">(</span><span class="va">alus</span><span class="op">)</span> <span class="op">==</span> <span class="st">"chr18"</span>, <span class="op">]</span></span>
<span><span class="va">alus</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqlevels-wrappers.html" class="external-link">keepStandardChromosomes</a></span><span class="op">(</span><span class="va">alus</span><span class="op">)</span> </span>
<span><span class="va">alus</span> <span class="op">&lt;-</span> <span class="va">alus</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">1000</span>, <span class="op">]</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqlevelsStyle.html" class="external-link">seqlevelsStyle</a></span><span class="op">(</span><span class="va">alus</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"NCBI"</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">genome</a></span><span class="op">(</span><span class="va">alus</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"GRCh38.p2"</span></span>
<span></span>
<span><span class="va">alu_snps</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_overlapping_snps.html">get_overlapping_snps</a></span><span class="op">(</span><span class="va">alus</span>, <span class="va">SNPlocs.Hsapiens.dbSNP144.GRCh38</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqlevelsStyle.html" class="external-link">seqlevelsStyle</a></span><span class="op">(</span><span class="va">alu_snps</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"UCSC"</span></span>
<span><span class="va">alu_snps</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, <span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## UnstitchedGPos object with 3 positions and 0 metadata columns:</span></span>
<span><span class="co">##       seqnames       pos strand</span></span>
<span><span class="co">##          &lt;Rle&gt; &lt;integer&gt;  &lt;Rle&gt;</span></span>
<span><span class="co">##   [1]    chr18     21651      *</span></span>
<span><span class="co">##   [2]    chr18     21654      *</span></span>
<span><span class="co">##   [3]    chr18     21667      *</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 25 sequences (1 circular) from hg38 genome</span></span></code></pre>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqlevelsStyle.html" class="external-link">seqlevelsStyle</a></span><span class="op">(</span><span class="va">alus</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"UCSC"</span></span>
<span><span class="va">alus</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, <span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 3 ranges and 11 metadata columns:</span></span>
<span><span class="co">##       seqnames      ranges strand |   swScore  milliDiv  milliDel</span></span>
<span><span class="co">##          &lt;Rle&gt;   &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">##   [1]    chr18 21645-21819      + |      1319       114         0</span></span>
<span><span class="co">##   [2]    chr18 26052-26327      + |      1539       199         7</span></span>
<span><span class="co">##   [3]    chr18 31708-32021      + |      2192       140         0</span></span>
<span><span class="co">##        milliIns  genoLeft     repName    repClass   repFamily  repStart</span></span>
<span><span class="co">##       &lt;numeric&gt; &lt;integer&gt; &lt;character&gt; &lt;character&gt; &lt;character&gt; &lt;integer&gt;</span></span>
<span><span class="co">##   [1]         0 -80351466      AluSq2        SINE         Alu       136</span></span>
<span><span class="co">##   [2]         0 -80346958       AluJr        SINE         Alu         1</span></span>
<span><span class="co">##   [3]         3 -80341264       AluSp        SINE         Alu         1</span></span>
<span><span class="co">##          repEnd   repLeft</span></span>
<span><span class="co">##       &lt;integer&gt; &lt;integer&gt;</span></span>
<span><span class="co">##   [1]       310        -3</span></span>
<span><span class="co">##   [2]       278       -34</span></span>
<span><span class="co">##   [3]       313         0</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 25 sequences (1 circular) from hg38 genome</span></span></code></pre>
<p><code><a href="../reference/calc_AEI.html">calc_AEI()</a></code> will return a matrix containing the AEI
calculated for all allelic combinations and a more detailed table
containing values for each chromosome.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">alu_index</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calc_AEI.html">calc_AEI</a></span><span class="op">(</span>bam_fn <span class="op">=</span> <span class="va">bam_files</span>, </span>
<span>                      fasta_fn <span class="op">=</span> <span class="va">fafn</span>, </span>
<span>                      snp_db <span class="op">=</span> <span class="va">alu_snps</span>,</span>
<span>                      alu_ranges <span class="op">=</span> <span class="va">alus</span>, </span>
<span>                      param <span class="op">=</span> <span class="va">fp</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">alu_index</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "AEI"           "AEI_per_chrom"</span></span></code></pre>
<p>The AEI in wild-type samples is highest for <code>A-to-G</code>, as
expected, and sharply reduced in the <code>ADAR1KO</code> samples.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html" class="external-link">Heatmap</a></span><span class="op">(</span><span class="va">alu_index</span><span class="op">$</span><span class="va">AEI</span>, </span>
<span>        name <span class="op">=</span> <span class="st">"AEI"</span>,</span>
<span>        col <span class="op">=</span> <span class="fu"><a href="https://sjmgarnier.github.io/viridisLite/reference/viridis.html" class="external-link">viridis</a></span><span class="op">(</span><span class="fl">100</span><span class="op">)</span>, </span>
<span>        row_labels <span class="op">=</span> <span class="va">rse</span><span class="op">$</span><span class="va">genotype_treatment</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">match</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">alu_index</span><span class="op">$</span><span class="va">AEI</span><span class="op">)</span>,</span>
<span>                                                  <span class="va">rse</span><span class="op">$</span><span class="va">sample</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="raer_files/figure-html/unnamed-chunk-19-1.png" width="700"></p>
<details style="margin-bottom:10px;"><summary>
Session info
</summary><div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.3.0 (2023-04-21)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">## Running under: Ubuntu 22.04.2 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] grid      stats4    stats     graphics  grDevices utils     datasets </span></span>
<span><span class="co">## [8] methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] SNPlocs.Hsapiens.dbSNP144.GRCh38_0.99.20</span></span>
<span><span class="co">##  [2] BSgenome_1.68.0                         </span></span>
<span><span class="co">##  [3] Biostrings_2.68.1                       </span></span>
<span><span class="co">##  [4] XVector_0.40.0                          </span></span>
<span><span class="co">##  [5] AnnotationHub_3.8.0                     </span></span>
<span><span class="co">##  [6] BiocFileCache_2.8.0                     </span></span>
<span><span class="co">##  [7] dbplyr_2.3.2                            </span></span>
<span><span class="co">##  [8] rtracklayer_1.60.0                      </span></span>
<span><span class="co">##  [9] viridis_0.6.3                           </span></span>
<span><span class="co">## [10] viridisLite_0.4.2                       </span></span>
<span><span class="co">## [11] ComplexHeatmap_2.16.0                   </span></span>
<span><span class="co">## [12] DESeq2_1.40.1                           </span></span>
<span><span class="co">## [13] SummarizedExperiment_1.30.2             </span></span>
<span><span class="co">## [14] Biobase_2.60.0                          </span></span>
<span><span class="co">## [15] GenomicRanges_1.52.0                    </span></span>
<span><span class="co">## [16] GenomeInfoDb_1.36.0                     </span></span>
<span><span class="co">## [17] IRanges_2.34.0                          </span></span>
<span><span class="co">## [18] S4Vectors_0.38.1                        </span></span>
<span><span class="co">## [19] BiocGenerics_0.46.0                     </span></span>
<span><span class="co">## [20] MatrixGenerics_1.12.2                   </span></span>
<span><span class="co">## [21] matrixStats_1.0.0                       </span></span>
<span><span class="co">## [22] raer_0.99.2                             </span></span>
<span><span class="co">## [23] BiocStyle_2.28.0                        </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] RColorBrewer_1.1-3            jsonlite_1.8.5               </span></span>
<span><span class="co">##   [3] shape_1.4.6                   magrittr_2.0.3               </span></span>
<span><span class="co">##   [5] GenomicFeatures_1.52.0        rmarkdown_2.22               </span></span>
<span><span class="co">##   [7] GlobalOptions_0.1.2           fs_1.6.2                     </span></span>
<span><span class="co">##   [9] BiocIO_1.10.0                 zlibbioc_1.46.0              </span></span>
<span><span class="co">##  [11] ragg_1.2.5                    vctrs_0.6.2                  </span></span>
<span><span class="co">##  [13] memoise_2.0.1                 Cairo_1.6-0                  </span></span>
<span><span class="co">##  [15] Rsamtools_2.16.0              RCurl_1.98-1.12              </span></span>
<span><span class="co">##  [17] htmltools_0.5.5               S4Arrays_1.0.4               </span></span>
<span><span class="co">##  [19] progress_1.2.2                curl_5.0.1                   </span></span>
<span><span class="co">##  [21] sass_0.4.6                    bslib_0.5.0                  </span></span>
<span><span class="co">##  [23] desc_1.4.2                    cachem_1.0.8                 </span></span>
<span><span class="co">##  [25] GenomicAlignments_1.36.0      mime_0.12                    </span></span>
<span><span class="co">##  [27] lifecycle_1.0.3               iterators_1.0.14             </span></span>
<span><span class="co">##  [29] pkgconfig_2.0.3               Matrix_1.5-4.1               </span></span>
<span><span class="co">##  [31] R6_2.5.1                      fastmap_1.1.1                </span></span>
<span><span class="co">##  [33] shiny_1.7.4                   GenomeInfoDbData_1.2.10      </span></span>
<span><span class="co">##  [35] clue_0.3-64                   digest_0.6.31                </span></span>
<span><span class="co">##  [37] colorspace_2.1-0              AnnotationDbi_1.62.1         </span></span>
<span><span class="co">##  [39] rprojroot_2.0.3               textshaping_0.3.6            </span></span>
<span><span class="co">##  [41] RSQLite_2.3.1                 filelock_1.0.2               </span></span>
<span><span class="co">##  [43] fansi_1.0.4                   httr_1.4.6                   </span></span>
<span><span class="co">##  [45] compiler_4.3.0                bit64_4.0.5                  </span></span>
<span><span class="co">##  [47] withr_2.5.0                   doParallel_1.0.17            </span></span>
<span><span class="co">##  [49] BiocParallel_1.34.2           DBI_1.1.3                    </span></span>
<span><span class="co">##  [51] highr_0.10                    R.utils_2.12.2               </span></span>
<span><span class="co">##  [53] biomaRt_2.56.1                rappdirs_0.3.3               </span></span>
<span><span class="co">##  [55] DelayedArray_0.26.3           rjson_0.2.21                 </span></span>
<span><span class="co">##  [57] tools_4.3.0                   interactiveDisplayBase_1.38.0</span></span>
<span><span class="co">##  [59] httpuv_1.6.11                 R.oo_1.25.0                  </span></span>
<span><span class="co">##  [61] glue_1.6.2                    restfulr_0.0.15              </span></span>
<span><span class="co">##  [63] promises_1.2.0.1              cluster_2.1.4                </span></span>
<span><span class="co">##  [65] generics_0.1.3                gtable_0.3.3                 </span></span>
<span><span class="co">##  [67] R.methodsS3_1.8.2             data.table_1.14.8            </span></span>
<span><span class="co">##  [69] hms_1.1.3                     xml2_1.3.4                   </span></span>
<span><span class="co">##  [71] utf8_1.2.3                    BiocVersion_3.17.1           </span></span>
<span><span class="co">##  [73] foreach_1.5.2                 pillar_1.9.0                 </span></span>
<span><span class="co">##  [75] stringr_1.5.0                 later_1.3.1                  </span></span>
<span><span class="co">##  [77] circlize_0.4.15               dplyr_1.1.2                  </span></span>
<span><span class="co">##  [79] lattice_0.21-8                bit_4.0.5                    </span></span>
<span><span class="co">##  [81] tidyselect_1.2.0              SingleCellExperiment_1.22.0  </span></span>
<span><span class="co">##  [83] locfit_1.5-9.8                knitr_1.43                   </span></span>
<span><span class="co">##  [85] gridExtra_2.3                 bookdown_0.34                </span></span>
<span><span class="co">##  [87] xfun_0.39                     stringi_1.7.12               </span></span>
<span><span class="co">##  [89] yaml_2.3.7                    evaluate_0.21                </span></span>
<span><span class="co">##  [91] codetools_0.2-19              tibble_3.2.1                 </span></span>
<span><span class="co">##  [93] BiocManager_1.30.21           cli_3.6.1                    </span></span>
<span><span class="co">##  [95] xtable_1.8-4                  systemfonts_1.0.4            </span></span>
<span><span class="co">##  [97] munsell_0.5.0                 jquerylib_0.1.4              </span></span>
<span><span class="co">##  [99] Rcpp_1.0.10                   png_0.1-8                    </span></span>
<span><span class="co">## [101] XML_3.99-0.14                 parallel_4.3.0               </span></span>
<span><span class="co">## [103] ellipsis_0.3.2                pkgdown_2.0.7                </span></span>
<span><span class="co">## [105] ggplot2_3.4.2                 blob_1.2.4                   </span></span>
<span><span class="co">## [107] prettyunits_1.1.1             bitops_1.0-7                 </span></span>
<span><span class="co">## [109] scales_1.2.1                  purrr_1.0.1                  </span></span>
<span><span class="co">## [111] crayon_1.5.2                  GetoptLong_1.0.5             </span></span>
<span><span class="co">## [113] rlang_1.1.1                   KEGGREST_1.40.0</span></span></code></pre>
</details><div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Roth2019-yu" class="csl-entry">
Roth, Shalom Hillel, Erez Y Levanon, and Eli Eisenberg. 2019.
<span>“Genome-Wide Quantification of <span>ADAR</span>
Adenosine-to-Inosine <span>RNA</span> Editing Activity.”</span> <em>Nat.
Methods</em> 16 (11): 1131–38. <a href="https://doi.org/10.1038/s41592-019-0610-9" class="external-link">https://doi.org/10.1038/s41592-019-0610-9</a>.
</div>
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>An R package developed by the <a href="https://medschool.cuanschutz.edu/rbi" class="external-link">RNA Bioscience Initiative</a></p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
