<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="raer">
<title>Introducing the raer package • raer</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><link href="../deps/_Source%20Sans%20Pro-0.4.1/font.css" rel="stylesheet">
<link href="../deps/_Source%20Code%20Pro-0.4.1/font.css" rel="stylesheet">
<!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Introducing the raer package">
<meta property="og:description" content="raer">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><script defer data-domain="rnabioco.github.io" src="https://plausible.io/js/plausible.js"></script>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-dark"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">raer</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item">
  <a class="nav-link" href="../articles/raer.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/rnabioco/raer/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Introducing the raer package</h1>
                        <h4 data-toc-skip class="author">Kent Riemondy</h4>
            <address class="author_afil">
      University of Colorado School of Medicine<br><h4 data-toc-skip class="date">Revised: 2 May 2022</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/rnabioco/raer/blob/HEAD/vignettes/raer.Rmd" class="external-link"><code>vignettes/raer.Rmd</code></a></small>
      <div class="d-none name"><code>raer.Rmd</code></div>
    </address>
</div>

    
    
<div class="section level2">
<h2 id="de-tutorial">DE tutorial<a class="anchor" aria-label="anchor" href="#de-tutorial"></a>
</h2>
<p>Steps to perform differential editing using <code>Raer</code></p>
<div class="section level3">
<h3 id="set-up">Set up<a class="anchor" aria-label="anchor" href="#set-up"></a>
</h3>
<p>First load in required packages</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://github.com/rnabioco/raer" class="external-link">raer</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://here.r-lib.org/" class="external-link">here</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/SummarizedExperiment" class="external-link">SummarizedExperiment</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioinf.wehi.edu.au/edgeR" class="external-link">edgeR</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/mikelove/DESeq2" class="external-link">DESeq2</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">ComplexHeatmap</span><span class="op">)</span></code></pre></div>
<p>Set directories and read in the data (eventually, we should make this downloadable)</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">res_dir</span> <span class="op">&lt;-</span> <span class="fu">here</span><span class="op">(</span><span class="st">"../working_dir/results/GSE99249/"</span><span class="op">)</span>

<span class="co"># Read in data</span>
<span class="va">se</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">res_dir</span>, <span class="st">"merged_se_with_meta.rds"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="preprocess">Preprocess<a class="anchor" aria-label="anchor" href="#preprocess"></a>
</h3>
<p>We start by using the <code>add_editing_frequencies</code> function to identify the percent of edits for each position and sample. Several pre-built options are available: A to I (“AI”), U to C (“UC”), C to U (“CU”), and G to A (“GA”). You can specify A to I editing by setting <code>type = AI</code> (this is also the default).</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">se_filtered</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/add_editing_frequencies.html">add_editing_frequencies</a></span><span class="op">(</span><span class="va">se</span>, type <span class="op">=</span> <span class="st">"AI"</span>,
                                       edit_frequency <span class="op">=</span> <span class="fl">0.01</span>,
                                       min_count <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></code></pre></div>
<p>If you are interested in some other kind of editing, you can use a custom event. To do this, set <code>type = "none"</code> and <code>edit_from = nucleotide_1</code> and <code>edit_to = nucleotide_2</code>. Here, the <code>edit_from</code> should be the nucleotide in the reference (for A to I this would be <code>A</code>) and <code>edit_to</code> should be the nucleotide after editing (for A to I this would be <code>G</code>). Both <code>edit_from</code> and <code>edit_to</code> must be a valid nucleotide (A, T, C, G).</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">se_filtered</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/add_editing_frequencies.html">add_editing_frequencies</a></span><span class="op">(</span><span class="va">se</span>, type <span class="op">=</span> <span class="st">"none"</span>,
                                       edit_frequency <span class="op">=</span> <span class="fl">0.01</span>,
                                       min_count <span class="op">=</span> <span class="fl">10</span>,
                                       edit_from <span class="op">=</span> <span class="st">"C"</span>, edit_to <span class="op">=</span> <span class="st">"T"</span><span class="op">)</span></code></pre></div>
<p>You can also make plots that show the number of edited sites and the frequency of editing per sample. You can make these plots using <code>make_plots = TRUE</code>. If <code>make_plots = TRUE</code> and no save dir is specified, the plots and the filtered object will be returned as a list</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">se_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/add_editing_frequencies.html">add_editing_frequencies</a></span><span class="op">(</span><span class="va">se</span>, type <span class="op">=</span> <span class="st">"AI"</span>,
                                      edit_frequency <span class="op">=</span> <span class="fl">0.01</span>,
                                      min_count <span class="op">=</span> <span class="fl">10</span>,
                                      make_plots <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
                                      
<span class="va">se_filtered</span> <span class="op">&lt;-</span> <span class="va">se_results</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>
<span class="va">site_plot</span> <span class="op">&lt;-</span> <span class="va">se_results</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>
<span class="va">frequency_plot</span> <span class="op">&lt;-</span> <span class="va">se_results</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span></code></pre></div>
<p>If you would rather directly save these plots, you can provide a save directory. The plots will be saved as “number_of_sites.pdf” and “editing_index.pdf”. If the plots are saved, only the filtered object will be returned.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">se_filtered</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/add_editing_frequencies.html">add_editing_frequencies</a></span><span class="op">(</span><span class="va">se</span>, type <span class="op">=</span> <span class="st">"AI"</span>,
                                       edit_frequency <span class="op">=</span> <span class="fl">0.01</span>,
                                       min_count <span class="op">=</span> <span class="fl">10</span>,
                                       make_plots <span class="op">=</span> <span class="cn">TRUE</span>,
                                       save_dir <span class="op">=</span> <span class="st">"results"</span><span class="op">)</span></code></pre></div>
<p>Once your object has been filtered, you can prepare it for DE. This means making an assay of counts that contains both your alt and ref allele.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">deobj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/prep_for_de.html">prep_for_de</a></span><span class="op">(</span><span class="va">se_filtered</span>,
                     min_prop <span class="op">=</span> <span class="fl">0.1</span>,
                     max_prop <span class="op">=</span> <span class="fl">0.9</span>,
                     min_samples <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="run-differential-editing-deseq2">Run differential editing (DESeq2)<a class="anchor" aria-label="anchor" href="#run-differential-editing-deseq2"></a>
</h3>
<p>At this stage, you can use the object to perform DE yourself or you can continue with our pre built functions</p>
<p>For differential editing, we use the design <code>design &lt;- ~0 + condition:sample + condition:count</code>.</p>
<p>For the samples, you can leave as is or combine so the same sample name shows up in both the treatment and control. These results are not identical but they are close. In my hands, the same genes come out, but the p values and log fold change values are slightly different.</p>
<p>It is probably best to update the levels of your object, but if you don’t, this will still work.</p>
<p>To run using <code>DESEq2</code>, set <code>type = DESEq2</code> in the <code>de_results</code> function. This function requires you to specify what your control and treatment are from your condition column of your <code>deobj</code>.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">de_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/perform_de.html">perform_de</a></span><span class="op">(</span><span class="va">deobj</span>, type <span class="op">=</span> <span class="st">"DESeq2"</span>, sample_col <span class="op">=</span> <span class="st">"sample"</span>,
                         condition_col <span class="op">=</span> <span class="st">"genotype"</span>,
                         condition_control <span class="op">=</span> <span class="st">"Wildtype"</span>,
                         condition_treatment <span class="op">=</span> <span class="st">"ADAR1KO"</span><span class="op">)</span></code></pre></div>
<p>This returns a list of the dds object, the full results, the significant results, and the model matrix.</p>
<p>If you’d rather use the sample information as described in the <code>DESEq2</code> vignette, you can make your sample names go across your treatment and control</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sample_mapping</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"SRR5564260"</span> <span class="op">=</span> <span class="st">"s1"</span>,
                   <span class="st">"SRR5564270"</span> <span class="op">=</span> <span class="st">"s1"</span>,
                   <span class="st">"SRR5564261"</span> <span class="op">=</span> <span class="st">"s2"</span>,
                   <span class="st">"SRR5564271"</span> <span class="op">=</span> <span class="st">"s2"</span>,
                   <span class="st">"SRR5564268"</span> <span class="op">=</span> <span class="st">"s3"</span>,
                   <span class="st">"SRR5564274"</span> <span class="op">=</span> <span class="st">"s3"</span>,
                   <span class="st">"SRR5564269"</span> <span class="op">=</span> <span class="st">"s4"</span>,
                   <span class="st">"SRR5564275"</span> <span class="op">=</span> <span class="st">"s4"</span>,
                   <span class="st">"SRR5564272"</span> <span class="op">=</span> <span class="st">"s5"</span>,
                   <span class="st">"SRR5564276"</span> <span class="op">=</span> <span class="st">"s5"</span>,
                   <span class="st">"SRR5564273"</span> <span class="op">=</span> <span class="st">"s6"</span>,
                   <span class="st">"SRR5564277"</span> <span class="op">=</span> <span class="st">"s6"</span><span class="op">)</span>

<span class="va">deobj</span><span class="op">$</span><span class="va">sample_new</span> <span class="op">&lt;-</span> <span class="va">sample_mapping</span><span class="op">[</span><span class="va">deobj</span><span class="op">$</span><span class="va">sample</span><span class="op">]</span>

<span class="va">deobj</span><span class="op">$</span><span class="va">sample_new</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">deobj</span><span class="op">$</span><span class="va">sample_new</span><span class="op">)</span>

<span class="va">de_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/perform_de.html">perform_de</a></span><span class="op">(</span><span class="va">deobj</span>, type <span class="op">=</span> <span class="st">"DESeq2"</span>, sample_col <span class="op">=</span> <span class="st">"sample_new"</span>,
                         condition_col <span class="op">=</span> <span class="st">"genotype"</span>,
                         condition_control <span class="op">=</span> <span class="st">"Wildtype"</span>,
                         condition_treatment <span class="op">=</span> <span class="st">"ADAR1KO"</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="run-differential-editing-edger">Run differential editing (edgeR)<a class="anchor" aria-label="anchor" href="#run-differential-editing-edger"></a>
</h3>
<p>At this stage, you can use the object to perform DE yourself or you can continue with our pre built functions</p>
<p>For differential editing, we use the design <code>design &lt;- ~0 + condition:sample + condition:count</code>.</p>
<p>For the samples, you can leave as is or combine so the same sample name shows up in both the treatment and control. These results are not identical but they are close. In my hands, the same genes come out, but the p values and log fold change values are slightly different.</p>
<p>It is probably best to update the levels of your object, but if you don’t, this will still work.</p>
<p>To run using <code>edgeR</code>, set <code>type = edgeR</code> in the <code>de_results</code> function. This function requires you to specify what your control and treatment are from your condition column of your <code>deobj</code>.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">de_edger_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/perform_de.html">perform_de</a></span><span class="op">(</span><span class="va">deobj</span>, type <span class="op">=</span> <span class="st">"edgeR"</span>,
                               sample_col <span class="op">=</span> <span class="st">"sample"</span>,
                               condition_col <span class="op">=</span> <span class="st">"genotype"</span>,
                               condition_control <span class="op">=</span> <span class="st">"Wildtype"</span>,
                               condition_treatment <span class="op">=</span> <span class="st">"ADAR1KO"</span><span class="op">)</span></code></pre></div>
<p>This returns a list of the dds object, the full results, the significant results, and the model matrix.</p>
<p>If you’d rather use the sample information as described in the <code>DESEq2</code> vignette, you can make your sample names go across your treatment and control</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sample_mapping</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"SRR5564260"</span> <span class="op">=</span> <span class="st">"s1"</span>,
                   <span class="st">"SRR5564270"</span> <span class="op">=</span> <span class="st">"s1"</span>,
                   <span class="st">"SRR5564261"</span> <span class="op">=</span> <span class="st">"s2"</span>,
                   <span class="st">"SRR5564271"</span> <span class="op">=</span> <span class="st">"s2"</span>,
                   <span class="st">"SRR5564268"</span> <span class="op">=</span> <span class="st">"s3"</span>,
                   <span class="st">"SRR5564274"</span> <span class="op">=</span> <span class="st">"s3"</span>,
                   <span class="st">"SRR5564269"</span> <span class="op">=</span> <span class="st">"s4"</span>,
                   <span class="st">"SRR5564275"</span> <span class="op">=</span> <span class="st">"s4"</span>,
                   <span class="st">"SRR5564272"</span> <span class="op">=</span> <span class="st">"s5"</span>,
                   <span class="st">"SRR5564276"</span> <span class="op">=</span> <span class="st">"s5"</span>,
                   <span class="st">"SRR5564273"</span> <span class="op">=</span> <span class="st">"s6"</span>,
                   <span class="st">"SRR5564277"</span> <span class="op">=</span> <span class="st">"s6"</span><span class="op">)</span>

<span class="va">deobj</span><span class="op">$</span><span class="va">sample_new</span> <span class="op">&lt;-</span> <span class="va">sample_mapping</span><span class="op">[</span><span class="va">deobj</span><span class="op">$</span><span class="va">sample</span><span class="op">]</span>

<span class="va">deobj</span><span class="op">$</span><span class="va">sample_new</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">deobj</span><span class="op">$</span><span class="va">sample_new</span><span class="op">)</span>

<span class="va">de_edger_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/perform_de.html">perform_de</a></span><span class="op">(</span><span class="va">deobj</span>, type <span class="op">=</span> <span class="st">"edgeR"</span>,
                               sample_col <span class="op">=</span> <span class="st">"sample_new"</span>,
                               condition_col <span class="op">=</span> <span class="st">"genotype"</span>,
                               condition_control <span class="op">=</span> <span class="st">"Wildtype"</span>,
                               condition_treatment <span class="op">=</span> <span class="st">"ADAR1KO"</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="perform-your-own-comparisons-deseq2">Perform your own comparisons (DESeq2)<a class="anchor" aria-label="anchor" href="#perform-your-own-comparisons-deseq2"></a>
</h3>
<p>I have found it more straight forward to pull out comparisions of interest following <a href="https://github.com/tavareshugo/tutorial_DESeq2_contrasts/blob/main/DESeq2_contrasts.md" class="external-link">this tutorial</a>. To ask for other comparisons with this dataset, you first need to identify the specific matrix design of the comparison.</p>
<p>The model matrix used for your experiment will be returned from the <code>perform_de</code> function. To get these results, first pull them out of the return list</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">mod_mat</span> <span class="op">&lt;-</span> <span class="va">de_results</span><span class="op">$</span><span class="va">model_matrix</span></code></pre></div>
<p>We will also need to get the DESeq object from this list</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">dds</span> <span class="op">&lt;-</span> <span class="va">de_results</span><span class="op">$</span><span class="va">deseq_obj</span></code></pre></div>
<p>We now can subset this matrix to the conditions we are interested in</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">alt_treatment</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colMeans</a></span><span class="op">(</span><span class="va">mod_mat</span><span class="op">[</span><span class="va">dds</span><span class="op">$</span><span class="va">condition</span> <span class="op">==</span> <span class="va">condition_treatment</span> <span class="op">&amp;</span>
                                  <span class="va">dds</span><span class="op">$</span><span class="va">count</span> <span class="op">==</span> <span class="st">"alt"</span>,<span class="op">]</span><span class="op">)</span>
<span class="va">ref_treatment</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colMeans</a></span><span class="op">(</span><span class="va">mod_mat</span><span class="op">[</span><span class="va">dds</span><span class="op">$</span><span class="va">condition</span> <span class="op">==</span> <span class="va">condition_treatment</span> <span class="op">&amp;</span>
                                  <span class="va">dds</span><span class="op">$</span><span class="va">count</span> <span class="op">==</span> <span class="st">"ref"</span>,<span class="op">]</span><span class="op">)</span>

<span class="va">alt_control</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colMeans</a></span><span class="op">(</span><span class="va">mod_mat</span><span class="op">[</span><span class="va">dds</span><span class="op">$</span><span class="va">condition</span> <span class="op">==</span> <span class="va">condition_control</span> <span class="op">&amp;</span>
                                <span class="va">dds</span><span class="op">$</span><span class="va">count</span> <span class="op">==</span> <span class="st">"alt"</span>,<span class="op">]</span><span class="op">)</span>
<span class="va">ref_control</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colMeans</a></span><span class="op">(</span><span class="va">mod_mat</span><span class="op">[</span><span class="va">dds</span><span class="op">$</span><span class="va">condition</span> <span class="op">==</span> <span class="va">condition_control</span> <span class="op">&amp;</span>
                                <span class="va">dds</span><span class="op">$</span><span class="va">count</span> <span class="op">==</span> <span class="st">"ref"</span>,<span class="op">]</span><span class="op">)</span></code></pre></div>
<p>We can now use these to pull out specific comparisions using the <code>results</code> function.</p>
<p>We can look at the differences between treatment and control for just the alt allele:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">alt_treat_control</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/results.html" class="external-link">results</a></span><span class="op">(</span><span class="va">dds</span>, contrast <span class="op">=</span> <span class="va">alt_treatment</span> <span class="op">-</span> <span class="va">alt_control</span><span class="op">)</span></code></pre></div>
<p>Or just the ref allele:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ref_treat_control</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/results.html" class="external-link">results</a></span><span class="op">(</span><span class="va">dds</span>, contrast <span class="op">=</span> <span class="va">ref_treatment</span> <span class="op">-</span> <span class="va">ref_control</span><span class="op">)</span></code></pre></div>
<p>Or just find the alt vs ref allele for treatment:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">alt_ref_treat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/results.html" class="external-link">results</a></span><span class="op">(</span><span class="va">dds</span>, contrast <span class="op">=</span> <span class="va">alt_treatment</span> <span class="op">-</span> <span class="va">ref_treatment</span><span class="op">)</span></code></pre></div>
<p>To find the alt vs ref specific to the treatment (what we returned as the results from <code>perform_de</code>):</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R">  <span class="va">treatment_vs_control</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/results.html" class="external-link">results</a></span><span class="op">(</span><span class="va">dds</span>,
                                  contrast <span class="op">=</span> <span class="op">(</span><span class="va">alt_treatment</span> <span class="op">-</span> <span class="va">ref_treatment</span><span class="op">)</span> <span class="op">-</span> 
                                    <span class="op">(</span><span class="va">alt_control</span> <span class="op">-</span> <span class="va">ref_control</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="perform-your-own-comparisons-edger">Perform your own comparisons (edgeR)<a class="anchor" aria-label="anchor" href="#perform-your-own-comparisons-edger"></a>
</h3>
<p>I have found it more straight forward to pull out comparisions of interest following <a href="https://github.com/tavareshugo/tutorial_DESeq2_contrasts/blob/main/DESeq2_contrasts.md" class="external-link">this tutorial</a>. To ask for other comparisons with this dataset, you first need to identify the specific matrix design of the comparison.</p>
<p>The model matrix used for your experiment will be returned from the <code>perform_de</code> function. To get these results, first pull them out of the return list</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">mod_mat</span> <span class="op">&lt;-</span> <span class="va">de_results</span><span class="op">$</span><span class="va">model_matrix</span></code></pre></div>
<p>We will also need to get the DESeq object from this list</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">deobj</span> <span class="op">&lt;-</span> <span class="va">de_results</span><span class="op">$</span><span class="va">deseq_obj</span></code></pre></div>
<p>We now can subset this matrix to the conditions we are interested in</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">alt_treatment</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colMeans</a></span><span class="op">(</span><span class="va">mod_mat</span><span class="op">[</span><span class="va">deobj</span><span class="op">$</span><span class="va">condition</span> <span class="op">==</span> <span class="va">condition_treatment</span> <span class="op">&amp;</span>
                                 <span class="va">deobj</span><span class="op">$</span><span class="va">count</span> <span class="op">==</span> <span class="st">"alt"</span>,<span class="op">]</span><span class="op">)</span>
<span class="va">ref_treatment</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colMeans</a></span><span class="op">(</span><span class="va">mod_mat</span><span class="op">[</span><span class="va">deobj</span><span class="op">$</span><span class="va">condition</span> <span class="op">==</span> <span class="va">condition_treatment</span> <span class="op">&amp;</span>
                                 <span class="va">deobj</span><span class="op">$</span><span class="va">count</span> <span class="op">==</span> <span class="st">"ref"</span>,<span class="op">]</span><span class="op">)</span>

<span class="va">alt_control</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colMeans</a></span><span class="op">(</span><span class="va">mod_mat</span><span class="op">[</span><span class="va">deobj</span><span class="op">$</span><span class="va">condition</span> <span class="op">==</span> <span class="va">condition_control</span> <span class="op">&amp;</span>
                               <span class="va">deobj</span><span class="op">$</span><span class="va">count</span> <span class="op">==</span> <span class="st">"alt"</span>,<span class="op">]</span><span class="op">)</span>
<span class="va">ref_control</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colMeans</a></span><span class="op">(</span><span class="va">mod_mat</span><span class="op">[</span><span class="va">deobj</span><span class="op">$</span><span class="va">condition</span> <span class="op">==</span> <span class="va">condition_control</span> <span class="op">&amp;</span>
                               <span class="va">deobj</span><span class="op">$</span><span class="va">count</span> <span class="op">==</span> <span class="st">"ref"</span>,<span class="op">]</span><span class="op">)</span></code></pre></div>
<p>We can now use these to pull out specific comparisions using the <code>results</code> function.</p>
<p>We can look at the differences between treatment and control for just the alt allele:</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">alt_treat_control</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/glmfit.html" class="external-link">glmLRT</a></span><span class="op">(</span><span class="va">deobj</span>, contrast <span class="op">=</span> <span class="va">alt_treatment</span> <span class="op">-</span> <span class="va">alt_control</span><span class="op">)</span>

<span class="va">alt_treat_control</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/topTags.html" class="external-link">topTags</a></span><span class="op">(</span><span class="va">alt_treat_control</span>, n <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">alt_treat_control</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>Or just the ref allele:</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ref_treat_control</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/glmfit.html" class="external-link">glmLRT</a></span><span class="op">(</span><span class="va">deobj</span>, contrast <span class="op">=</span> <span class="va">ref_treatment</span> <span class="op">-</span> <span class="va">ref_control</span><span class="op">)</span>

<span class="va">ref_treat_control</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/topTags.html" class="external-link">topTags</a></span><span class="op">(</span><span class="va">ref_treat_control</span>, n <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">ref_treat_control</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>Or just find the alt vs ref allele for treatment:</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">alt_ref_treat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/glmfit.html" class="external-link">glmLRT</a></span><span class="op">(</span><span class="va">deobj</span>, contrast <span class="op">=</span> <span class="va">alt_treatment</span> <span class="op">-</span> <span class="va">ref_treatment</span><span class="op">)</span>

<span class="va">alt_ref_treat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/topTags.html" class="external-link">topTags</a></span><span class="op">(</span><span class="va">alt_ref_treat</span>, n <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">alt_ref_treat</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>To find the alt vs ref specific to the treatment (what we returned as the results from <code>perform_de</code>):</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">treatment_vs_control</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/glmfit.html" class="external-link">glmLRT</a></span><span class="op">(</span><span class="va">deobj</span>, contrast <span class="op">=</span> <span class="op">(</span><span class="va">alt_treatment</span> <span class="op">-</span> <span class="va">ref_treatment</span><span class="op">)</span> <span class="op">-</span> 
                  <span class="op">(</span><span class="va">alt_control</span> <span class="op">-</span> <span class="va">ref_control</span><span class="op">)</span><span class="op">)</span>

<span class="va">treatment_vs_control</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/topTags.html" class="external-link">topTags</a></span><span class="op">(</span><span class="va">treatment_vs_control</span>,
                              n <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">treatment_vs_control</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>An R package developed by the <a href="https://medschool.cuanschutz.edu/rbi" class="external-link">RNA Bioscience Initiative</a></p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.3.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
