<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="raer">
<title>Introducing the raer package • raer</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><link href="../deps/Source_Sans_Pro-0.4.7/font.css" rel="stylesheet">
<link href="../deps/Source_Code_Pro-0.4.7/font.css" rel="stylesheet">
<!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Introducing the raer package">
<meta property="og:description" content="raer">
<meta property="og:image" content="https://rnabioco.github.io/raer/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><script defer data-domain="rnabioco.github.io" src="https://plausible.io/js/plausible.js"></script>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-dark"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">raer</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.13</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item">
  <a class="nav-link" href="../articles/raer.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/rnabioco/raer/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Introducing the raer package</h1>
                        <h4 data-toc-skip class="author">Kent
Riemondy</h4>
            <address class="author_afil">
      University of Colorado School of
Medicine<br><h4 data-toc-skip class="author">Kristen
Wells-Wrasman</h4>
            <address class="author_afil">
      University of Colorado School of
Medicine<br><h4 data-toc-skip class="author">Jay
Hesselberth</h4>
            <address class="author_afil">
      University of Colorado School of Medicine<br><h4 data-toc-skip class="date">2023-09-22</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/rnabioco/raer/blob/HEAD/vignettes/raer.Rmd" class="external-link"><code>vignettes/raer.Rmd</code></a></small>
      <div class="d-none name"><code>raer.Rmd</code></div>
    </address>
</address>
</address>
</div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>The <em>raer</em> (RNA Adenosine editing in R) package provides tools
to characterize A-to-I editing in single cell and bulk RNA-sequencing
datasets. Both novel and known editing sites can be detected and
quantified beginning with BAM alignment files. At it’s core the
<em>raer</em> package uses the pileup routines from the HTSlib C library
(<span class="citation">Bonfield et al. (2021)</span>) to identify
candidate RNA editing sites, and leverages the annotation resources in
the Bioconductor ecosystem to further characterize and identify
high-confidence RNA editing sites.</p>
<p>Here we demonstrate how to use the <em>raer</em> package to a)
quantify known RNA editing in droplet scRNA-seq dataset, b) identify
editing sites with condition specific editing in bulk RNA-seq data, and
c) predict novel editing sites from bulk RNA-seq.</p>
</div>
<div class="section level2">
<h2 id="characterizing-rna-editing-sites-in-scrna-seq-data">Characterizing RNA editing sites in scRNA-seq data<a class="anchor" aria-label="anchor" href="#characterizing-rna-editing-sites-in-scrna-seq-data"></a>
</h2>
<p>Here we will use the <em>raer</em> package to examine RNA editing in
droplet-based single cell RNA-seq data. For this example we will examine
a scRNA-seq dataset from human PBMC cells provided by <a href="https://www.10xgenomics.com/resources/datasets/10k-human-pbmcs-3-v3-1-chromium-x-with-intronic-reads-3-1-high" class="external-link">10x
Genomics</a>. The single cell data was aligned and processed using the
10x Genomics cellranger pipeline.</p>
<p>The PBMC scRNA-seq dataset from 10x Genomics, along with other needed
files will downloaded and cached using <code><a href="https://rdrr.io/pkg/raerdata/man/pbmc_10x.html" class="external-link">pbmc_10x()</a></code>from the
<em>raerdata</em> ExperimentHub package. For this vignette, the BAM file
was subset to retain 2 million alignments that overlap human RNA editing
sites on chromosome 16.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://rnabioco.github.io/raer">raer</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/rnabioco/raerdata" class="external-link">raerdata</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">TxDb.Hsapiens.UCSC.hg38.knownGene</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">AnnotationHub</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SNPlocs.Hsapiens.dbSNP144.GRCh38</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jokergoo/ComplexHeatmap" class="external-link">ComplexHeatmap</a></span><span class="op">)</span></span></code></pre></div>
<p><code><a href="https://rdrr.io/pkg/raerdata/man/pbmc_10x.html" class="external-link">pbmc_10x()</a></code> returns a list containing a
<code>BamFile</code> object, a <code>GRanges</code> object with known
RNA editing sites from the <code>REDIportal</code> database, and a
<code>SingleCellExperiment</code> populated with the gene expression
data and cell type annotations.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/raerdata/man/pbmc_10x.html" class="external-link">pbmc_10x</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">pbmc_bam</span> <span class="op">&lt;-</span> <span class="va">pbmc</span><span class="op">$</span><span class="va">bam</span></span>
<span><span class="va">editing_sites</span> <span class="op">&lt;-</span> <span class="va">pbmc</span><span class="op">$</span><span class="va">sites</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">pbmc</span><span class="op">$</span><span class="va">sce</span></span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotUMAP</a></span><span class="op">(</span><span class="va">sce</span>, colour_by <span class="op">=</span> <span class="st">"celltype"</span><span class="op">)</span></span></code></pre></div>
<p><img src="raer_files/figure-html/unnamed-chunk-4-1.png" width="700"></p>
<div class="section level3">
<h3 id="specifying-sites-to-quantify">Specifying sites to quantify<a class="anchor" aria-label="anchor" href="#specifying-sites-to-quantify"></a>
</h3>
<p>Next we’ll select editing sites to query. For this analysis we will
use RNA editing sites cataloged in the REDIportal database <span class="citation">Mansi et al. (2021)</span>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">editing_sites</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 15638648 ranges and 0 metadata columns:</span></span>
<span><span class="co">##              seqnames    ranges strand</span></span>
<span><span class="co">##                 &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;</span></span>
<span><span class="co">##          [1]     chr1     87158      -</span></span>
<span><span class="co">##          [2]     chr1     87168      -</span></span>
<span><span class="co">##          [3]     chr1     87171      -</span></span>
<span><span class="co">##          [4]     chr1     87189      -</span></span>
<span><span class="co">##          [5]     chr1     87218      -</span></span>
<span><span class="co">##          ...      ...       ...    ...</span></span>
<span><span class="co">##   [15638644]     chrY  56885715      +</span></span>
<span><span class="co">##   [15638645]     chrY  56885716      +</span></span>
<span><span class="co">##   [15638646]     chrY  56885728      +</span></span>
<span><span class="co">##   [15638647]     chrY  56885841      +</span></span>
<span><span class="co">##   [15638648]     chrY  56885850      +</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 44 sequences from hg38 genome; no seqlengths</span></span></code></pre>
<p>The sites to quantify are specified using a custom formatted GRanges
object with 1 base intervals, a strand (+ or -), and supplemented with
metadata columns named <code>REF</code> and <code>ALT</code> containing
the reference and alternate base to query. In this case we are only
interested in A-&gt;I editing, so we set the ref and alt to
<code>A</code> and <code>G</code>. Note that the <code>REF</code> and
<code>ALT</code> bases are in reference to strand. For a <code>-</code>
strand interval the bases should be the complement of the <code>+</code>
strand bases. Also note that these bases can be stored as traditional
character vectors or as <code><a href="https://rdrr.io/pkg/S4Vectors/man/Rle-class.html" class="external-link">Rle()</a></code> objects to save memory.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">editing_sites</span><span class="op">$</span><span class="va">REF</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Rle-class.html" class="external-link">Rle</a></span><span class="op">(</span><span class="st">"A"</span><span class="op">)</span></span>
<span><span class="va">editing_sites</span><span class="op">$</span><span class="va">ALT</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Rle-class.html" class="external-link">Rle</a></span><span class="op">(</span><span class="st">"G"</span><span class="op">)</span></span>
<span><span class="va">editing_sites</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 15638648 ranges and 2 metadata columns:</span></span>
<span><span class="co">##              seqnames    ranges strand |   REF   ALT</span></span>
<span><span class="co">##                 &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;Rle&gt; &lt;Rle&gt;</span></span>
<span><span class="co">##          [1]     chr1     87158      - |     A     G</span></span>
<span><span class="co">##          [2]     chr1     87168      - |     A     G</span></span>
<span><span class="co">##          [3]     chr1     87171      - |     A     G</span></span>
<span><span class="co">##          [4]     chr1     87189      - |     A     G</span></span>
<span><span class="co">##          [5]     chr1     87218      - |     A     G</span></span>
<span><span class="co">##          ...      ...       ...    ... .   ...   ...</span></span>
<span><span class="co">##   [15638644]     chrY  56885715      + |     A     G</span></span>
<span><span class="co">##   [15638645]     chrY  56885716      + |     A     G</span></span>
<span><span class="co">##   [15638646]     chrY  56885728      + |     A     G</span></span>
<span><span class="co">##   [15638647]     chrY  56885841      + |     A     G</span></span>
<span><span class="co">##   [15638648]     chrY  56885850      + |     A     G</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 44 sequences from hg38 genome; no seqlengths</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="quantifying-sites-in-single-cells-using-pileup_cells">Quantifying sites in single cells using <em>pileup_cells</em><a class="anchor" aria-label="anchor" href="#quantifying-sites-in-single-cells-using-pileup_cells"></a>
</h3>
<p><code><a href="../reference/pileup_cells.html">pileup_cells()</a></code> quantifies edited and non-edited UMI
counts per cell barcode, then organizes the site counts into a
<code>SingleCellExperiment</code> object. <code><a href="../reference/pileup_cells.html">pileup_cells()</a></code>
accepts a <code><a href="../reference/pileup_sites.html">FilterParam()</a></code> object that specifies parameters
for multiple read-level and site-level filtering and processing options.
Note that <code><a href="../reference/pileup_cells.html">pileup_cells()</a></code> is strand sensitive by default, so
it is important to ensure that the strand of the input sites is
correctly annotated, and that the <code>library-type</code> is set
correctly for the strandedness of the sequencing library. For 10x
Genomics data, the library type is set to <code>fr-second-strand</code>,
indicating that the strand of the BAM alignments is the same strand as
the RNA. See <a href="#ss2">quantifying smart-seq scRNA-seq
libraries</a> for an example of using <em>pileup_cells()</em> to handle
unstranded data and data from libraries that produce 1 BAM file for each
cell.</p>
<p>To exclude duplicate reads derived from PCR,
<code><a href="../reference/pileup_cells.html">pileup_cells()</a></code> can use a UMI sequence, supplied via the
<code>umi_tag</code> argument, to only count 1 read for each CB-UMI pair
at each editing site position. Note however that by default the
<code>bam_flags</code> argument for the <code>FilterParam</code> class
is set to <strong>include</strong> duplicate reads when using
<code><a href="../reference/pileup_cells.html">pileup_cells()</a></code>. Droplet single cell libraries produce
multiple cDNA fragments from a single reverse transcription event. The
cDNA fragments have different alignment positions due to fragmentation
despite being derived from a single RNA molecule. scRNA-seq data
processed by cellranger from 10x Genomics will set the “Not primary
alignment” bam flag for every read except one read for each UMI. If
duplicates are removed based on this bam flag, then only 1
representative fragment for a single UMI will be examined, which will
exclude many valid regions.</p>
<p>To reduce processing time many functions in the <em>raer</em> package
operate in parallel across multiple chromosomes. To enable parallel
processing, a <code>BiocParallel</code> backend can be supplied via the
<code>BPPARAM</code> argument (e.g. <code>MultiCoreParam()</code>).</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">outdir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"sc_edits"</span><span class="op">)</span></span>
<span><span class="va">cbs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span></span>
<span><span class="va">params</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pileup_sites.html">FilterParam</a></span><span class="op">(</span>min_mapq <span class="op">=</span> <span class="fl">255</span>, <span class="co"># required alignment MAPQ score</span></span>
<span>                      library_type <span class="op">=</span> <span class="st">"fr-second-strand"</span>, <span class="co">#library type</span></span>
<span>                      min_variant_reads <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="va">e_sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pileup_cells.html">pileup_cells</a></span><span class="op">(</span></span>
<span>    bamfile <span class="op">=</span> <span class="va">pbmc_bam</span>,</span>
<span>    sites <span class="op">=</span> <span class="va">editing_sites</span>,</span>
<span>    cell_barcodes <span class="op">=</span> <span class="va">cbs</span>,</span>
<span>    output_directory <span class="op">=</span> <span class="va">outdir</span>,</span>
<span>    cb_tag <span class="op">=</span> <span class="st">"CB"</span>,</span>
<span>    umi_tag <span class="op">=</span> <span class="st">"UB"</span>, </span>
<span>    param <span class="op">=</span> <span class="va">params</span></span>
<span><span class="op">)</span></span>
<span><span class="va">e_sce</span></span></code></pre></div>
<pre><code><span><span class="co">## class: SingleCellExperiment </span></span>
<span><span class="co">## dim: 3849 500 </span></span>
<span><span class="co">## metadata(0):</span></span>
<span><span class="co">## assays(2): nRef nAlt</span></span>
<span><span class="co">## rownames(3849): site_chr16_83540_1_AG site_chr16_83621_1_AG ...</span></span>
<span><span class="co">##   site_chr16_31453268_2_AG site_chr16_31454303_2_AG</span></span>
<span><span class="co">## rowData names(2): REF ALT</span></span>
<span><span class="co">## colnames(500): TGTTTGTCAGTTAGGG-1 ATCTCTACAAGCTACT-1 ...</span></span>
<span><span class="co">##   GGGCGTTTCAGGACGA-1 CTATAGGAGATTGTGA-1</span></span>
<span><span class="co">## colData names(0):</span></span>
<span><span class="co">## reducedDimNames(0):</span></span>
<span><span class="co">## mainExpName: NULL</span></span>
<span><span class="co">## altExpNames(0):</span></span></code></pre>
<p>The outputs from <code><a href="../reference/pileup_cells.html">pileup_cells()</a></code> are a
<code>SingleCellExperiment</code> object populated with
<code>nRef</code> and <code>nAlt</code> assays containing the base
counts for the reference (unedited) and alternate (edited) alleles at
each position.</p>
<p>The sparseMatrices are also written to files, at a directory
specified by <code>output_directory</code>, which can be loaded into R
using the <code><a href="../reference/read_sparray.html">read_sparray()</a></code> function.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">dir</a></span><span class="op">(</span><span class="va">outdir</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "barcodes.txt.gz" "counts.mtx.gz"   "sites.txt.gz"</span></span></code></pre>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/read_sparray.html">read_sparray</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outdir</span>, <span class="st">"counts.mtx.gz"</span><span class="op">)</span>,</span>
<span>             <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outdir</span>, <span class="st">"sites.txt.gz"</span><span class="op">)</span>,</span>
<span>             <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outdir</span>, <span class="st">"barcodes.txt.gz"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## class: SingleCellExperiment </span></span>
<span><span class="co">## dim: 3849 500 </span></span>
<span><span class="co">## metadata(0):</span></span>
<span><span class="co">## assays(2): nRef nAlt</span></span>
<span><span class="co">## rownames(3849): site_chr16_83540_1_AG site_chr16_83621_1_AG ...</span></span>
<span><span class="co">##   site_chr16_31453268_2_AG site_chr16_31454303_2_AG</span></span>
<span><span class="co">## rowData names(2): REF ALT</span></span>
<span><span class="co">## colnames(500): TGTTTGTCAGTTAGGG-1 ATCTCTACAAGCTACT-1 ...</span></span>
<span><span class="co">##   GGGCGTTTCAGGACGA-1 CTATAGGAGATTGTGA-1</span></span>
<span><span class="co">## colData names(0):</span></span>
<span><span class="co">## reducedDimNames(0):</span></span>
<span><span class="co">## mainExpName: NULL</span></span>
<span><span class="co">## altExpNames(0):</span></span></code></pre>
<p>Next we’ll filter the single cell editing dataset to find sites with
an editing event in at least 5 cells and add the editing counts to the
gene expression SingleCellExperiment as an <code><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html" class="external-link">altExp()</a></code>.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">e_sce</span> <span class="op">&lt;-</span> <span class="va">e_sce</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/MatrixGenerics/man/rowSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assays</a></span><span class="op">(</span><span class="va">e_sce</span><span class="op">)</span><span class="op">$</span><span class="va">nAlt</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="fl">5</span>, <span class="op">]</span></span>
<span><span class="va">e_sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calc_edit_frequency.html">calc_edit_frequency</a></span><span class="op">(</span><span class="va">e_sce</span>, edit_from <span class="op">=</span> <span class="st">"Ref"</span>, edit_to <span class="op">=</span> <span class="st">"Alt"</span>, replace_na <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html" class="external-link">altExp</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">e_sce</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">]</span></span></code></pre></div>
<p>With the editing sites added to the gene expression
SingleCellExperiment we can use plotting and other methods previously
developed for single cell analysis. Here we’ll visualize editing sites
with the highest edited read counts.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">to_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html" class="external-link">altExp</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/order.html" class="external-link">order</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/MatrixGenerics/man/rowSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html" class="external-link">altExp</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>, <span class="st">"nAlt"</span><span class="op">)</span><span class="op">)</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">to_plot</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotUMAP</a></span><span class="op">(</span><span class="va">sce</span>, colour_by <span class="op">=</span> <span class="va">x</span>, by_exprs_values <span class="op">=</span> <span class="st">"nAlt"</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [[1]]</span></span></code></pre>
<p><img src="raer_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## [[2]]</span></span></code></pre>
<p><img src="raer_files/figure-html/unnamed-chunk-9-2.png" width="700"></p>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## [[3]]</span></span></code></pre>
<p><img src="raer_files/figure-html/unnamed-chunk-9-3.png" width="700"></p>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## [[4]]</span></span></code></pre>
<p><img src="raer_files/figure-html/unnamed-chunk-9-4.png" width="700"></p>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## [[5]]</span></span></code></pre>
<p><img src="raer_files/figure-html/unnamed-chunk-9-5.png" width="700"></p>
<p>Alternatively we can view these top edited sites as a Heatmap,
showing the average number of edited reads per site in each cell
type.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html" class="external-link">altExp</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">$</span><span class="va">celltype</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">$</span><span class="va">celltype</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotGroupedHeatmap.html" class="external-link">plotGroupedHeatmap</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html" class="external-link">altExp</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>,</span>
<span>    features <span class="op">=</span> <span class="va">to_plot</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">25</span><span class="op">]</span>,</span>
<span>    group <span class="op">=</span> <span class="st">"celltype"</span>,</span>
<span>    exprs_values <span class="op">=</span> <span class="st">"nAlt"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="raer_files/figure-html/unnamed-chunk-10-1.png" width="700"></p>
<p><em>raer</em> provides additional tools to examine cell type specific
editing.</p>
<ul>
<li><p><code><a href="../reference/find_scde_sites.html">find_scde_sites()</a></code> will perform statistical testing
to identify sites with different editing frequencies between
clusters/cell types.</p></li>
<li><p><code><a href="../reference/calc_scAEI.html">calc_scAEI()</a></code> will calculate the Alu Editing Index
(<code>AEI</code>) metric in single cells.</p></li>
</ul>
<p><strong>If the editing sites of interest are not known, we recommend
the following approach. First, treat the single cell data as a bulk
RNA-seq experiment, and follow the <a href="#bulk">bulk RNA-seq
workflow</a> to identify putative editing sites. Then query these sites
in single cell mode using <code><a href="../reference/pileup_cells.html">pileup_cells()</a></code></strong></p>
</div>
<div class="section level3">
<h3 id="ss2">Quantifying sites in Smart-seq libaries<a class="anchor" aria-label="anchor" href="#ss2"></a>
</h3>
<p><code><a href="../reference/pileup_cells.html">pileup_cells()</a></code> can also process Smart-seq style single
cell libraries. These datasets typically store data from each cell in
separate BAM files and the library type for these alignments are
generally unstranded. To process these datasets the
<code>library-type</code> should be set to <code>unstranded</code>, and
the reference editing sites need to be reported all on the
<code>+</code> strand.</p>
<p>For example, the editing sites on the minus strand will need to be
complemented (set as T -&gt; C rather than A -&gt; G). Additionally the
<code>umi_tag</code> and <code>cb_tag</code> arguments should be set as
follows to disable UMI and cell barcode detection.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Note these steps are not run </span></span>
<span><span class="va">is_minus</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/strand.html" class="external-link">strand</a></span><span class="op">(</span><span class="va">editing_sites</span><span class="op">)</span> <span class="op">==</span> <span class="st">"-"</span></span>
<span><span class="va">editing_sites</span><span class="op">[</span><span class="va">is_minus</span><span class="op">]</span><span class="op">$</span><span class="va">REF</span> <span class="op">&lt;-</span> <span class="st">"T"</span> </span>
<span><span class="va">editing_sites</span><span class="op">[</span><span class="va">is_minus</span><span class="op">]</span><span class="op">$</span><span class="va">ALT</span> <span class="op">&lt;-</span> <span class="st">"C"</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/strand.html" class="external-link">strand</a></span><span class="op">(</span><span class="va">editing_sites</span><span class="op">[</span><span class="va">is_minus</span><span class="op">]</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"+"</span></span>
<span></span>
<span><span class="va">vector_of_bam_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cell1.bam"</span>, <span class="st">"cell2.bam"</span>, <span class="st">"cell3.bam"</span><span class="op">)</span></span>
<span><span class="va">vector_of_cell_ids</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cell1"</span>, <span class="st">"cell2"</span>, <span class="st">"cell3"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/pileup_cells.html">pileup_cells</a></span><span class="op">(</span>bamfiles <span class="op">=</span> <span class="va">vector_of_bam_files</span>,</span>
<span>             cell_barcodes <span class="op">=</span> <span class="va">vector_of_cell_ids</span>, </span>
<span>             sites <span class="op">=</span> <span class="va">editing_sites</span>, </span>
<span>             umi_tag <span class="op">=</span> <span class="cn">NULL</span>,  <span class="co"># no UMI tag in most smart-seq libraries</span></span>
<span>             cb_tag <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span>   <span class="co"># no cell barcode in most smart-seq libraries</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="bulk">Quantifying RNA editing sites in bulk RNA-Seq<a class="anchor" aria-label="anchor" href="#bulk"></a>
</h2>
<p>Next we will perform a reanalysis of a published bulk RNA-seq
dataset, <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE99249" class="external-link">GSE99249</a>,
which consists of ADAR1 mutants and control human cell lines,
conditionally treated with Interferon-Beta. We will examine data from
two genotypes, ADAR1 WT and KO, both treated with Ifn-B, with triplicate
samples.</p>
<p>Aligned BAM files and other necessary files have been preprocessed
for this vignette and are available using <code><a href="https://rdrr.io/pkg/raerdata/man/GSE99249.html" class="external-link">GSE99249()</a></code> from
the <code>raerdata</code> package. Calling <code><a href="https://rdrr.io/pkg/raerdata/man/GSE99249.html" class="external-link">GSE99249()</a></code> will
downloaded and cache the necessary files and return a list containing
the data.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ifnb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/raerdata/man/GSE99249.html" class="external-link">GSE99249</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">ifnb</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "bams"  "fasta" "sites"</span></span></code></pre>
<p><code>bams</code> contains a vector of <code>BamFile</code> objects
with the paths to each BAM file. These BAM files are a subset of the
full BAM files, containing alignments from chromosome 18.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bam_files</span> <span class="op">&lt;-</span> <span class="va">ifnb</span><span class="op">$</span><span class="va">bams</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">bam_files</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "SRR5564260" "SRR5564261" "SRR5564269" "SRR5564270" "SRR5564271"</span></span>
<span><span class="co">## [6] "SRR5564277"</span></span></code></pre>
<p>To quantify editing sites we will need a FASTA file to compare read
alignments to the reference sequence. For space reasons we’ll use a
FASTA file containing only chromosome 18 for this demo.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fafn</span> <span class="op">&lt;-</span> <span class="va">ifnb</span><span class="op">$</span><span class="va">fasta</span></span></code></pre></div>
<p>We will again use the database of known human editing sites from
REDIPortal, only processing those from <code>chr18</code>.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">editing_sites</span> <span class="op">&lt;-</span> <span class="va">ifnb</span><span class="op">$</span><span class="va">sites</span></span>
<span><span class="va">chr_18_editing_sites</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqlevels-wrappers.html" class="external-link">keepSeqlevels</a></span><span class="op">(</span><span class="va">editing_sites</span>, <span class="st">"chr18"</span>, pruning.mode<span class="op">=</span><span class="st">"coarse"</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="generate-editing-site-read-counts-using-pileup_sites">Generate editing site read counts using <em>pileup_sites</em><a class="anchor" aria-label="anchor" href="#generate-editing-site-read-counts-using-pileup_sites"></a>
</h3>
<p>The <code><a href="../reference/pileup_sites.html">pileup_sites()</a></code> function will process BAM files and
calculate base counts at each supplied position. The
<code><a href="../reference/pileup_sites.html">FilterParam()</a></code> will again be used to specify parameters to
exclude reads and bases based on commonly used filters for detecting
RNA-editing events. Specific regions can also be queried using the
<code>region</code> argument which accepts a samtools style region
specification string (e.g. <code>chr</code> or
<code>chr:start-end</code>).</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pileup_sites.html">FilterParam</a></span><span class="op">(</span></span>
<span>    only_keep_variants <span class="op">=</span> <span class="cn">TRUE</span>,        <span class="co"># only report sites with variants</span></span>
<span>    trim_5p <span class="op">=</span> <span class="fl">5</span>,                      <span class="co"># bases to remove from 5' or 3' end </span></span>
<span>    trim_3p <span class="op">=</span> <span class="fl">5</span>,</span>
<span>    min_base_quality <span class="op">=</span> <span class="fl">30</span>,           <span class="co"># minimum base quality score</span></span>
<span>    min_mapq <span class="op">=</span> <span class="fl">255</span>,                  <span class="co"># minimum MAPQ read score</span></span>
<span>    library_type <span class="op">=</span> <span class="st">"fr-first-strand"</span>, <span class="co"># library type</span></span>
<span>    min_splice_overhang <span class="op">=</span> <span class="fl">10</span>          <span class="co"># minimum required splice site overhang</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">rse</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pileup_sites.html">pileup_sites</a></span><span class="op">(</span><span class="va">bam_files</span>,</span>
<span>    fasta <span class="op">=</span> <span class="va">fafn</span>,</span>
<span>    sites <span class="op">=</span> <span class="va">chr_18_editing_sites</span>,</span>
<span>    chroms <span class="op">=</span> <span class="st">"chr18"</span>,</span>
<span>    param <span class="op">=</span> <span class="va">fp</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">rse</span></span></code></pre></div>
<pre><code><span><span class="co">## class: RangedSummarizedExperiment </span></span>
<span><span class="co">## dim: 6192 6 </span></span>
<span><span class="co">## metadata(0):</span></span>
<span><span class="co">## assays(7): ALT nRef ... nC nG</span></span>
<span><span class="co">## rownames(6192): site_chr18_178100_1 site_chr18_184553_1 ...</span></span>
<span><span class="co">##   site_chr18_80172518_2 site_chr18_80174441_2</span></span>
<span><span class="co">## rowData names(4): REF rpbz vdb sor</span></span>
<span><span class="co">## colnames(6): SRR5564260 SRR5564261 ... SRR5564271 SRR5564277</span></span>
<span><span class="co">## colData names(1): sample</span></span></code></pre>
<p>Pileup data is stored in a <code>RangedSummarizedExperiment</code>
object which facilitates comparisons across samples and conveniently
stores genomic coordinate information. The <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData()</a></code> and
<code>rowRanges()</code> slots are populated with the reference base
(<code>REF</code>) and information related to each editing site, and
similarly the <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData()</a></code> slot can be used to store sample
metadata.</p>
<p>The base counts and other information are stored in different assays
within the object. <code>REF</code> and <code>ALT</code> bases and base
count data are all provided in a stand specific fashion depending on the
supplied <code>library-type</code> parameter. The <code>REF</code> and
<code>ALT</code> bases are in reference to the strand.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assays</a></span><span class="op">(</span><span class="va">rse</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## List of length 7</span></span>
<span><span class="co">## names(7): ALT nRef nAlt nA nT nC nG</span></span></code></pre>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">rse</span>, <span class="st">"nA"</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##                     SRR5564260 SRR5564261 SRR5564269 SRR5564270</span></span>
<span><span class="co">## site_chr18_178100_1          2          0          0          1</span></span>
<span><span class="co">## site_chr18_184553_1          0          1          1          2</span></span>
<span><span class="co">## site_chr18_184659_1          2          0          1          1</span></span>
<span><span class="co">## site_chr18_184747_1          1          4          2          4</span></span>
<span><span class="co">## site_chr18_185203_1          1          0          1          1</span></span>
<span><span class="co">##                     SRR5564271 SRR5564277</span></span>
<span><span class="co">## site_chr18_178100_1          1          2</span></span>
<span><span class="co">## site_chr18_184553_1          3          3</span></span>
<span><span class="co">## site_chr18_184659_1          2          1</span></span>
<span><span class="co">## site_chr18_184747_1          3          2</span></span>
<span><span class="co">## site_chr18_185203_1          1          0</span></span></code></pre>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">rse</span>, <span class="st">"nG"</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##                     SRR5564260 SRR5564261 SRR5564269 SRR5564270</span></span>
<span><span class="co">## site_chr18_178100_1          0          0          0          0</span></span>
<span><span class="co">## site_chr18_184553_1          0          0          0          0</span></span>
<span><span class="co">## site_chr18_184659_1          0          0          0          0</span></span>
<span><span class="co">## site_chr18_184747_1          0          0          0          0</span></span>
<span><span class="co">## site_chr18_185203_1          0          0          0          0</span></span>
<span><span class="co">##                     SRR5564271 SRR5564277</span></span>
<span><span class="co">## site_chr18_178100_1          1          0</span></span>
<span><span class="co">## site_chr18_184553_1          1          0</span></span>
<span><span class="co">## site_chr18_184659_1          1          0</span></span>
<span><span class="co">## site_chr18_184747_1          0          1</span></span>
<span><span class="co">## site_chr18_185203_1          0          0</span></span></code></pre>
<p>Next we’ll add sample information which will be needed for identify
sites with differential editing frequencies across genotypes.</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">rse</span><span class="op">)</span><span class="op">$</span><span class="va">treatment</span> <span class="op">&lt;-</span> <span class="st">"Interferon beta"</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">rse</span><span class="op">)</span><span class="op">$</span><span class="va">genotype</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ADAR1KO"</span>,<span class="st">"Wildtype"</span><span class="op">)</span>, each <span class="op">=</span> <span class="fl">3</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">rse</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 6 rows and 3 columns</span></span>
<span><span class="co">##                 sample       treatment genotype</span></span>
<span><span class="co">##            &lt;character&gt;     &lt;character&gt; &lt;factor&gt;</span></span>
<span><span class="co">## SRR5564260  SRR5564260 Interferon beta ADAR1KO </span></span>
<span><span class="co">## SRR5564261  SRR5564261 Interferon beta ADAR1KO </span></span>
<span><span class="co">## SRR5564269  SRR5564269 Interferon beta ADAR1KO </span></span>
<span><span class="co">## SRR5564270  SRR5564270 Interferon beta Wildtype</span></span>
<span><span class="co">## SRR5564271  SRR5564271 Interferon beta Wildtype</span></span>
<span><span class="co">## SRR5564277  SRR5564277 Interferon beta Wildtype</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="prepare-for-differential-editing">Prepare for differential editing<a class="anchor" aria-label="anchor" href="#prepare-for-differential-editing"></a>
</h3>
<p><em>raer</em> provides the <code>calc_edit_frequency</code> function
to calculate the editing percentage and read depth at each position.
With the <code>drop = TRUE</code> argument we will also exclude
non-adenosine sites. The editing frequencies will not be used for
differential editing analysis, which will be conducted using the raw
counts, however these are useful for filtering and visualization.
<code>calc_edit_frequency</code> will add two additional assays to the
object, the editing frequency (<code>edit_freq</code>) and read
<code>depth</code>, both computed based on the <code>edit_to</code> and
<code>edit_from</code> counts.</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rse</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calc_edit_frequency.html">calc_edit_frequency</a></span><span class="op">(</span><span class="va">rse</span>,</span>
<span>    edit_from <span class="op">=</span> <span class="st">"A"</span>,</span>
<span>    edit_to <span class="op">=</span> <span class="st">"G"</span>,</span>
<span>    drop <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>We’ll next filter to exclude low frequency editing events. For this
analysis we require that an editing site shows editing in at least 1
sample and has at least 5 counts in each sample.</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">has_editing</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MatrixGenerics/man/rowSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">rse</span>, <span class="st">"edit_freq"</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="fl">1</span></span>
<span><span class="va">has_depth</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MatrixGenerics/man/rowSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">rse</span>, <span class="st">"depth"</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="fl">5</span><span class="op">)</span> <span class="op">==</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">rse</span><span class="op">)</span></span>
<span></span>
<span><span class="va">rse</span> <span class="op">&lt;-</span> <span class="va">rse</span><span class="op">[</span><span class="va">has_editing</span> <span class="op">&amp;</span> <span class="va">has_depth</span>, <span class="op">]</span></span>
<span><span class="va">rse</span></span></code></pre></div>
<pre><code><span><span class="co">## class: RangedSummarizedExperiment </span></span>
<span><span class="co">## dim: 612 6 </span></span>
<span><span class="co">## metadata(0):</span></span>
<span><span class="co">## assays(9): ALT nRef ... depth edit_freq</span></span>
<span><span class="co">## rownames(612): site_chr18_204626_1 site_chr18_212426_1 ...</span></span>
<span><span class="co">##   site_chr18_79984359_2 site_chr18_79984760_2</span></span>
<span><span class="co">## rowData names(4): REF rpbz vdb sor</span></span>
<span><span class="co">## colnames(6): SRR5564260 SRR5564261 ... SRR5564271 SRR5564277</span></span>
<span><span class="co">## colData names(5): sample treatment genotype n_sites edit_idx</span></span></code></pre>
<p>Once the object has been filtered, we will transform it into an
alternative data structure for differential editing analysis that
contains an assay with read counts of both the <code>ALT</code> and
<code>REF</code> alleles in a single matrix.</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">deobj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/make_de_object.html">make_de_object</a></span><span class="op">(</span><span class="va">rse</span>, min_prop <span class="op">=</span> <span class="fl">0.05</span>, min_samples <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">deobj</span>, <span class="st">"counts"</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">7</span>,<span class="fl">2</span>,<span class="fl">8</span><span class="op">)</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##                     SRR5564260_ref SRR5564260_alt SRR5564261_ref</span></span>
<span><span class="co">## site_chr18_691546_2              8              0              6</span></span>
<span><span class="co">## site_chr18_691578_2              8              0              7</span></span>
<span><span class="co">## site_chr18_692372_2              6              0              9</span></span>
<span><span class="co">##                     SRR5564261_alt</span></span>
<span><span class="co">## site_chr18_691546_2              0</span></span>
<span><span class="co">## site_chr18_691578_2              0</span></span>
<span><span class="co">## site_chr18_692372_2              0</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="run-differential-editing">Run differential editing<a class="anchor" aria-label="anchor" href="#run-differential-editing"></a>
</h3>
<p>At this stage, you can use the object to perform differential
yourself or use <code><a href="../reference/find_de_sites.html">find_de_sites()</a></code> to use <code>edgeR</code>
or <code>DESeq2</code> to identify condition specific editing events.
For differential editing, we use the design
<code>design &lt;- ~0 + condition:sample + condition:count</code> and
perform testing to compare the edited read counts against unedited read
counts.</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">deobj</span><span class="op">$</span><span class="va">sample</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">deobj</span><span class="op">$</span><span class="va">sample</span><span class="op">)</span></span>
<span><span class="va">de_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/find_de_sites.html">find_de_sites</a></span><span class="op">(</span><span class="va">deobj</span>, </span>
<span>    test <span class="op">=</span> <span class="st">"DESeq2"</span>,</span>
<span>    sample_col <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>    condition_col <span class="op">=</span> <span class="st">"genotype"</span>,</span>
<span>    condition_control <span class="op">=</span> <span class="st">"Wildtype"</span>,</span>
<span>    condition_treatment <span class="op">=</span> <span class="st">"ADAR1KO"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>This returns a list containing the dds object, the full results, the
significant results, and the model matrix.</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">de_results</span><span class="op">$</span><span class="va">sig_results</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##                        baseMean log2FoldChange     lfcSE      stat</span></span>
<span><span class="co">## site_chr18_23296417_2 15.500000      -2.450652 0.8459822 -2.896813</span></span>
<span><span class="co">## site_chr18_32070144_1  6.666667      -2.631946 1.0760283 -2.445983</span></span>
<span><span class="co">## site_chr18_21632030_1 10.000000      -3.377617 1.4959882 -2.257783</span></span>
<span><span class="co">## site_chr18_21631237_1  6.666667      -3.381984 1.5541279 -2.176130</span></span>
<span><span class="co">## site_chr18_35263230_2  9.333333      -3.403508 1.4894626 -2.285057</span></span>
<span><span class="co">##                            pvalue        padj</span></span>
<span><span class="co">## site_chr18_23296417_2 0.003769742 0.008216103</span></span>
<span><span class="co">## site_chr18_32070144_1 0.014445798 0.024557856</span></span>
<span><span class="co">## site_chr18_21632030_1 0.023959185 0.036474262</span></span>
<span><span class="co">## site_chr18_21631237_1 0.029545546 0.044059148</span></span>
<span><span class="co">## site_chr18_35263230_2 0.022309459 0.035633191</span></span></code></pre>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">top_sites</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">de_results</span><span class="op">$</span><span class="va">sig_results</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">20</span><span class="op">]</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html" class="external-link">Heatmap</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">rse</span>, <span class="st">"edit_freq"</span><span class="op">)</span><span class="op">[</span><span class="va">top_sites</span>, <span class="op">]</span>, </span>
<span>        name <span class="op">=</span> <span class="st">"editing frequency"</span>,</span>
<span>        column_labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">rse</span><span class="op">$</span><span class="va">genotype</span>, <span class="st">"-"</span>, <span class="va">rse</span><span class="op">$</span><span class="va">treatment</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="raer_files/figure-html/unnamed-chunk-24-1.png" width="480"></p>
<p>As anticipated the top identified sites are those with greatly
reduced editing in the ADAR1KO samples.</p>
</div>
<div class="section level3">
<h3 id="examine-overall-editing-activites-using-the-alu-editing-index">Examine overall editing activites using the Alu Editing Index<a class="anchor" aria-label="anchor" href="#examine-overall-editing-activites-using-the-alu-editing-index"></a>
</h3>
<p>For some studies it is informative to assess the overall ADAR editing
activity in addition to examining individual editing sites. The
<strong>Alu Editing Index</strong> (AEI), developed by <span class="citation">Roth, Levanon, and Eisenberg (2019)</span>, is a metric
that summarizes that amount of editing occurring at ALU elements which
account for the vast majority of A-to-I editing (&gt; 99%) in
humans.</p>
<p><em>raer</em> provides <code><a href="../reference/calc_AEI.html">calc_AEI()</a></code>, based on this
approach, to calculate the AEI metric. Many of the same parameters used
for <code><a href="../reference/pileup_sites.html">pileup_sites()</a></code> are available in
<code><a href="../reference/calc_AEI.html">calc_AEI()</a></code>.</p>
<p>First we will use the <code>AnnotationHub</code> package to obtain
coordinates for ALU elements in the human genome. For this example we
will only examine a subset of ALUs on <code>chr18</code>. We will also
use a <code>SNPlocs</code> package, based on the dbSNP database, to
exclude any SNPs overlapping the ALU elements from the AEI calculation.
The SNP coordinates are <code>NCBI</code> based, whereas the
<code>ALU</code> elements are based on <code>hg38</code>, we will
therefore convert between the two as needed to obtain SNP and ALU
element coordinates based on <code>hg38</code>.</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ah</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationHub/man/AnnotationHub-class.html" class="external-link">AnnotationHub</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">rmsk_hg38</span> <span class="op">&lt;-</span> <span class="va">ah</span><span class="op">[[</span><span class="st">"AH99003"</span><span class="op">]</span><span class="op">]</span></span>
<span></span>
<span><span class="va">alus</span> <span class="op">&lt;-</span> <span class="va">rmsk_hg38</span><span class="op">[</span><span class="va">rmsk_hg38</span><span class="op">$</span><span class="va">repFamily</span> <span class="op">==</span> <span class="st">"Alu"</span>, <span class="op">]</span></span>
<span><span class="va">alus</span> <span class="op">&lt;-</span> <span class="va">alus</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqnames</a></span><span class="op">(</span><span class="va">alus</span><span class="op">)</span> <span class="op">==</span> <span class="st">"chr18"</span>, <span class="op">]</span></span>
<span><span class="va">alus</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqlevels-wrappers.html" class="external-link">keepStandardChromosomes</a></span><span class="op">(</span><span class="va">alus</span><span class="op">)</span> </span>
<span><span class="va">alus</span> <span class="op">&lt;-</span> <span class="va">alus</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">1000</span>, <span class="op">]</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqlevelsStyle.html" class="external-link">seqlevelsStyle</a></span><span class="op">(</span><span class="va">alus</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"NCBI"</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">genome</a></span><span class="op">(</span><span class="va">alus</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"GRCh38.p2"</span></span>
<span></span>
<span><span class="va">alu_snps</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_overlapping_snps.html">get_overlapping_snps</a></span><span class="op">(</span><span class="va">alus</span>, <span class="va">SNPlocs.Hsapiens.dbSNP144.GRCh38</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqlevelsStyle.html" class="external-link">seqlevelsStyle</a></span><span class="op">(</span><span class="va">alu_snps</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"UCSC"</span></span>
<span><span class="va">alu_snps</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, <span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## UnstitchedGPos object with 3 positions and 0 metadata columns:</span></span>
<span><span class="co">##       seqnames       pos strand</span></span>
<span><span class="co">##          &lt;Rle&gt; &lt;integer&gt;  &lt;Rle&gt;</span></span>
<span><span class="co">##   [1]    chr18     21651      *</span></span>
<span><span class="co">##   [2]    chr18     21654      *</span></span>
<span><span class="co">##   [3]    chr18     21667      *</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 25 sequences (1 circular) from hg38 genome</span></span></code></pre>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqlevelsStyle.html" class="external-link">seqlevelsStyle</a></span><span class="op">(</span><span class="va">alus</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"UCSC"</span></span>
<span><span class="va">alus</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, <span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 3 ranges and 11 metadata columns:</span></span>
<span><span class="co">##       seqnames      ranges strand |   swScore  milliDiv  milliDel</span></span>
<span><span class="co">##          &lt;Rle&gt;   &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">##   [1]    chr18 21645-21819      + |      1319       114         0</span></span>
<span><span class="co">##   [2]    chr18 26052-26327      + |      1539       199         7</span></span>
<span><span class="co">##   [3]    chr18 31708-32021      + |      2192       140         0</span></span>
<span><span class="co">##        milliIns  genoLeft     repName    repClass   repFamily  repStart</span></span>
<span><span class="co">##       &lt;numeric&gt; &lt;integer&gt; &lt;character&gt; &lt;character&gt; &lt;character&gt; &lt;integer&gt;</span></span>
<span><span class="co">##   [1]         0 -80351466      AluSq2        SINE         Alu       136</span></span>
<span><span class="co">##   [2]         0 -80346958       AluJr        SINE         Alu         1</span></span>
<span><span class="co">##   [3]         3 -80341264       AluSp        SINE         Alu         1</span></span>
<span><span class="co">##          repEnd   repLeft</span></span>
<span><span class="co">##       &lt;integer&gt; &lt;integer&gt;</span></span>
<span><span class="co">##   [1]       310        -3</span></span>
<span><span class="co">##   [2]       278       -34</span></span>
<span><span class="co">##   [3]       313         0</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 25 sequences (1 circular) from hg38 genome</span></span></code></pre>
<p><code><a href="../reference/calc_AEI.html">calc_AEI()</a></code> returns a matrix containing the AEI
calculated for all allelic combinations and a more detailed table
containing values for each chromosome.</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">alu_index</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calc_AEI.html">calc_AEI</a></span><span class="op">(</span><span class="va">bam_files</span>, </span>
<span>                      fasta <span class="op">=</span> <span class="va">fafn</span>, </span>
<span>                      snp_db <span class="op">=</span> <span class="va">alu_snps</span>,</span>
<span>                      alu_ranges <span class="op">=</span> <span class="va">alus</span>, </span>
<span>                      param <span class="op">=</span> <span class="va">fp</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">alu_index</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "AEI"           "AEI_per_chrom"</span></span></code></pre>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html" class="external-link">Heatmap</a></span><span class="op">(</span><span class="va">alu_index</span><span class="op">$</span><span class="va">AEI</span>, </span>
<span>        name <span class="op">=</span> <span class="st">"AEI"</span>,</span>
<span>        row_labels <span class="op">=</span> <span class="va">rse</span><span class="op">$</span><span class="va">genotype</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">match</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">alu_index</span><span class="op">$</span><span class="va">AEI</span><span class="op">)</span>, <span class="va">rse</span><span class="op">$</span><span class="va">sample</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="raer_files/figure-html/unnamed-chunk-27-1.png" width="700"></p>
<p>The AEI in the <code>Wildtype</code> samples is highest for
<code>A-to-G</code>, and sharply reduced in the <code>ADAR1KO</code>
samples as expected.</p>
</div>
</div>
<div class="section level2">
<h2 id="novel-rna-editing-site-detection">Novel RNA editing site detection<a class="anchor" aria-label="anchor" href="#novel-rna-editing-site-detection"></a>
</h2>
<p>Next we will demonstrate how to identify novel RNA editing sites
using the <em>raer</em> package. In this analysis a published RNA-seq
and whole genome sequencing dataset will be analyzed. High coverage
whole-genome sequencing was conducted <a href="https://www.ebi.ac.uk/ena/browser/view/ERR262997?show=reads" class="external-link">ERR262997</a>
along with paired-end RNA-seq <a href="https://www.ebi.ac.uk/ena/browser/view/SRR1258218?show=reads" class="external-link">SRR1258218</a>
in a human cell line (<code>NA12878</code>).</p>
<p>Aligned BAM files, a genome fasta file, and a GRanges object
containing SNPs corresponding to the first 1Mb region of chr4 have been
prepared for this vignette and can be downloaded and cached using
<code><a href="https://rdrr.io/pkg/raerdata/man/NA12878.html" class="external-link">NA12878()</a></code>.</p>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna_wgs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/raerdata/man/NA12878.html" class="external-link">NA12878</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">rna_wgs</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "bams"  "fasta" "snps"</span></span></code></pre>
<p>Additionally we will use the following additional annotation
resources:</p>
<ul>
<li><p>A database of known SNPs, for example the
<code>SNPlocs.Hsapiens.dbSNP155.GRCh38</code> package. Due to space and
memory constraints in this vignette we will only examine SNPs from the
first 1Mb region of chr4.</p></li>
<li><p><code>TxDb.Hsapiens.UCSC.hg38.knownGene</code>, a database of
transcript models. Alternatively these can be generated from a
<code>.gtf</code> file using <code><a href="https://rdrr.io/pkg/GenomicFeatures/man/makeTxDbFromGRanges.html" class="external-link">makeTxDbFromGRanges()</a></code> from the
<code>GenomicFeatures</code> package.</p></li>
<li><p>RepeatMasker annotations, which can be obtained from the
<code><a href="https://rdrr.io/pkg/AnnotationHub/man/AnnotationHub-class.html" class="external-link">AnnotationHub()</a></code> for hg38, as shown in the <a href="#bulk">bulk RNA-seq</a> tutorial.</p></li>
</ul>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">txdb</span> <span class="op">&lt;-</span> <span class="va">TxDb.Hsapiens.UCSC.hg38.knownGene</span></span>
<span><span class="va">chr4snps</span> <span class="op">&lt;-</span> <span class="va">rna_wgs</span><span class="op">$</span><span class="va">snps</span></span></code></pre></div>
<p>The <code><a href="../reference/pileup_sites.html">pileup_sites()</a></code> function accept multiple bam files,
here we supply one from RNA-seq, and one from whole genome sequencing. A
subset of the filtering parameters (<code><a href="../reference/pileup_sites.html">FilterParam()</a></code>) can
accept multiple arguments matched to each of the input bam files. This
allows us to have distinct settings for the WGS and RNA-seq BAM
files.</p>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bams</span> <span class="op">&lt;-</span> <span class="va">rna_wgs</span><span class="op">$</span><span class="va">bams</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">bams</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"rna"</span>, <span class="st">"dna"</span><span class="op">)</span></span>
<span><span class="va">fp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pileup_sites.html">FilterParam</a></span><span class="op">(</span></span>
<span>    min_depth <span class="op">=</span> <span class="fl">1</span>,               <span class="co"># minimum read depth across all samples</span></span>
<span>    min_base_quality <span class="op">=</span> <span class="fl">30</span>,       <span class="co"># minimum base quality</span></span>
<span>    min_mapq <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">255</span>, <span class="fl">30</span><span class="op">)</span>,       <span class="co"># minimum MAPQ for each bam file</span></span>
<span>    library_type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"fr-first-strand"</span>, <span class="st">"unstranded"</span><span class="op">)</span>, <span class="co"># library-type for each bam file</span></span>
<span>    trim_5p <span class="op">=</span> <span class="fl">5</span>,                 <span class="co"># bases to trim from 5' end of alignment</span></span>
<span>    trim_3p <span class="op">=</span> <span class="fl">5</span>,                 <span class="co"># bases to trim from 3' end of alignment</span></span>
<span>    indel_dist <span class="op">=</span> <span class="fl">4</span>,              <span class="co"># ignore read if contains an indel within distance from site</span></span>
<span>    min_splice_overhang <span class="op">=</span> <span class="fl">10</span>,    <span class="co"># required alignment overhang in order to count read with splice </span></span>
<span>    read_bqual <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.25</span>, <span class="fl">20</span><span class="op">)</span>,    <span class="co"># minimum fraction of the read (0.25) that must have base quality of (20)</span></span>
<span>    only_keep_variants <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="cn">TRUE</span>, <span class="cn">FALSE</span><span class="op">)</span>, <span class="co"># report site if rnaseq bam has variant</span></span>
<span>    report_multiallelic <span class="op">=</span> <span class="cn">FALSE</span>, <span class="co"># do not report sites with multiple variant alleles</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">rse</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pileup_sites.html">pileup_sites</a></span><span class="op">(</span><span class="va">bams</span>,</span>
<span>                    fasta <span class="op">=</span> <span class="va">rna_wgs</span><span class="op">$</span><span class="va">fasta</span>,</span>
<span>                    chroms <span class="op">=</span> <span class="st">"chr4"</span>,</span>
<span>                    param <span class="op">=</span> <span class="va">fp</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">rse</span></span></code></pre></div>
<pre><code><span><span class="co">## class: RangedSummarizedExperiment </span></span>
<span><span class="co">## dim: 1035 2 </span></span>
<span><span class="co">## metadata(0):</span></span>
<span><span class="co">## assays(7): ALT nRef ... nC nG</span></span>
<span><span class="co">## rownames(1035): site_chr4_40244_2 site_chr4_44338_2 ...</span></span>
<span><span class="co">##   site_chr4_995145_1 site_chr4_998975_1</span></span>
<span><span class="co">## rowData names(4): REF rpbz vdb sor</span></span>
<span><span class="co">## colnames(2): rna dna</span></span>
<span><span class="co">## colData names(1): sample</span></span></code></pre>
<p>Next we filter to keep those sites with a variant in the RNA-seq, but
no variant in the DNA-seq, and a minimum of 5 reads covering the site in
the DNA-seq. The DNA-seq data is unstranded, and therefore will be
reported on the “+” strand whereas the RNA-seq data will be reported on
expressing RNA strand. We therefore use
<code>subsetByOverlaps(..., ignore.strand = TRUE)</code> to retain sites
passing these DNA-seq based filters independent of strand.</p>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">to_keep</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">rse</span>, <span class="st">"nRef"</span><span class="op">)</span><span class="op">[</span>, <span class="st">"dna"</span><span class="op">]</span> <span class="op">&gt;=</span> <span class="fl">5</span> <span class="op">&amp;</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">rse</span>, <span class="st">"ALT"</span><span class="op">)</span><span class="op">[</span>, <span class="st">"dna"</span><span class="op">]</span> <span class="op">==</span> <span class="st">"-"</span><span class="op">)</span></span>
<span><span class="va">rse</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/findOverlaps-methods.html" class="external-link">subsetByOverlaps</a></span><span class="op">(</span><span class="va">rse</span>, <span class="va">rse</span><span class="op">[</span><span class="va">to_keep</span>, <span class="op">]</span>, ignore.strand <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">rse</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 339</span></span></code></pre>
<p>Next we filter to remove any multiallelic sites. These sites are
stored as comma-separated strings in the <code>ALT</code> assay
(e.g. <code>G,C</code>). Non-variant sites are stored as <code>-</code>.
<code><a href="../reference/filter_multiallelic.html">filter_multiallelic()</a></code> will remove any sites that have
multiple variants in the samples present in the
<code>summarizedExperiment</code> object. It will add a new column to
the <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData()</a></code> to indicate the variant for each site, and
will calculate an <code>edit_freq</code> assay with variant allele
frequencies for each sample.</p>
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rse</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filter_multiallelic.html">filter_multiallelic</a></span><span class="op">(</span><span class="va">rse</span><span class="op">)</span></span>
<span><span class="va">rse</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calc_edit_frequency.html">calc_edit_frequency</a></span><span class="op">(</span><span class="va">rse</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">rse</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 260 rows and 5 columns</span></span>
<span><span class="co">##                            REF      rpbz       vdb       sor         ALT</span></span>
<span><span class="co">##                    &lt;character&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;character&gt;</span></span>
<span><span class="co">## site_chr4_124551_1           G -1.138153       Inf   1.42563           A</span></span>
<span><span class="co">## site_chr4_124611_1           T -0.266758       Inf   1.60944           G</span></span>
<span><span class="co">## site_chr4_124781_1           T -1.656293       Inf   1.45499           G</span></span>
<span><span class="co">## site_chr4_124820_1           C -0.600205       Inf   1.51787           A</span></span>
<span><span class="co">## site_chr4_124940_1           C -0.754898       Inf   1.24398           G</span></span>
<span><span class="co">## ...                        ...       ...       ...       ...         ...</span></span>
<span><span class="co">## site_chr4_992535_1           G -0.661008       Inf  1.499465           T</span></span>
<span><span class="co">## site_chr4_993342_1           A  0.802887       Inf  0.260455           G</span></span>
<span><span class="co">## site_chr4_994239_1           A  1.412126       Inf  1.425634           G</span></span>
<span><span class="co">## site_chr4_995144_1           A  0.387298       Inf  1.491655           G</span></span>
<span><span class="co">## site_chr4_995145_1           A  0.522233       Inf  1.464766           G</span></span></code></pre>
<p>Next we’ll remove sites in simple repeat regions. We will add repeat
information to the <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData()</a></code> using the
<code><a href="../reference/annot_from_gr.html">annot_from_gr()</a></code> function.</p>
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># subset both to chromosome 4 to avoid warning about different seqlevels</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlevels</a></span><span class="op">(</span><span class="va">rse</span>, pruning.mode <span class="op">=</span> <span class="st">"coarse"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"chr4"</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlevels</a></span><span class="op">(</span><span class="va">rmsk_hg38</span>, pruning.mode <span class="op">=</span> <span class="st">"coarse"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"chr4"</span></span>
<span></span>
<span><span class="va">rse</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/annot_from_gr.html">annot_from_gr</a></span><span class="op">(</span><span class="va">rse</span>, <span class="va">rmsk_hg38</span>, cols_to_map <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"repName"</span>, <span class="st">"repClass"</span>, <span class="st">"repFamily"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">rse</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"repName"</span>, <span class="st">"repFamily"</span><span class="op">)</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 260 rows and 2 columns</span></span>
<span><span class="co">##                    repName repFamily</span></span>
<span><span class="co">##                      &lt;Rle&gt;     &lt;Rle&gt;</span></span>
<span><span class="co">## site_chr4_124551_1      NA        NA</span></span>
<span><span class="co">## site_chr4_124611_1      NA        NA</span></span>
<span><span class="co">## site_chr4_124781_1      NA        NA</span></span>
<span><span class="co">## site_chr4_124820_1      NA        NA</span></span>
<span><span class="co">## site_chr4_124940_1      NA        NA</span></span>
<span><span class="co">## ...                    ...       ...</span></span>
<span><span class="co">## site_chr4_992535_1      NA        NA</span></span>
<span><span class="co">## site_chr4_993342_1      NA        NA</span></span>
<span><span class="co">## site_chr4_994239_1   AluJr       Alu</span></span>
<span><span class="co">## site_chr4_995144_1      NA        NA</span></span>
<span><span class="co">## site_chr4_995145_1      NA        NA</span></span></code></pre>
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rse</span> <span class="op">&lt;-</span> <span class="va">rse</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">rse</span><span class="op">)</span><span class="op">$</span><span class="va">repFamily</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Simple_repeat"</span>, <span class="st">"Low_complexity"</span><span class="op">)</span><span class="op">]</span></span></code></pre></div>
<p>Next we’ll remove sites adjacent to other sites with different
variant types. For example if an A-&gt;G variant is located proximal to
a C-&gt;T variant then the variants will be removed.</p>
<div class="sourceCode" id="cb67"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlevels</a></span><span class="op">(</span><span class="va">txdb</span>, pruning.mode <span class="op">=</span> <span class="st">"coarse"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"chr4"</span></span>
<span><span class="va">rse</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filter_clustered_variants.html">filter_clustered_variants</a></span><span class="op">(</span><span class="va">rse</span>, <span class="va">txdb</span>, variant_dist <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span>
<span><span class="va">rse</span></span></code></pre></div>
<pre><code><span><span class="co">## class: RangedSummarizedExperiment </span></span>
<span><span class="co">## dim: 159 2 </span></span>
<span><span class="co">## metadata(0):</span></span>
<span><span class="co">## assays(9): ALT nRef ... depth edit_freq</span></span>
<span><span class="co">## rownames(159): site_chr4_124940_1 site_chr4_126885_1 ...</span></span>
<span><span class="co">##   site_chr4_995144_1 site_chr4_995145_1</span></span>
<span><span class="co">## rowData names(8): REF rpbz ... repClass repFamily</span></span>
<span><span class="co">## colnames(2): rna dna</span></span>
<span><span class="co">## colData names(3): sample n_sites edit_idx</span></span></code></pre>
<p>Next, we’ll annotate if the site is a known SNP and remove any known
SNPs. If using a SNPlocs package you can use the
<code><a href="../reference/annot_snps.html">annot_snps()</a></code> function, which also allows one to compare the
variant base to the SNP variant base. Here we will use the
<code><a href="../reference/annot_from_gr.html">annot_from_gr()</a></code> function to annotate using the
<code>chr4snps</code> object and coarsely remove any editing sites
overlapping the same position as a SNP.</p>
<div class="sourceCode" id="cb69"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rse</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/annot_from_gr.html">annot_from_gr</a></span><span class="op">(</span><span class="va">rse</span>, <span class="va">chr4snps</span>, <span class="st">"name"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">rse</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"name"</span><span class="op">)</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 159 rows and 1 column</span></span>
<span><span class="co">##                            name</span></span>
<span><span class="co">##                           &lt;Rle&gt;</span></span>
<span><span class="co">## site_chr4_124940_1 rs1361822847</span></span>
<span><span class="co">## site_chr4_126885_1           NA</span></span>
<span><span class="co">## site_chr4_126955_1           NA</span></span>
<span><span class="co">## site_chr4_133310_1           NA</span></span>
<span><span class="co">## site_chr4_133971_1  rs998760640</span></span>
<span><span class="co">## ...                         ...</span></span>
<span><span class="co">## site_chr4_992535_1 rs1714447589</span></span>
<span><span class="co">## site_chr4_993342_1           NA</span></span>
<span><span class="co">## site_chr4_994239_1           NA</span></span>
<span><span class="co">## site_chr4_995144_1 rs1217215121</span></span>
<span><span class="co">## site_chr4_995145_1           NA</span></span></code></pre>
<div class="sourceCode" id="cb71"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rse</span> <span class="op">&lt;-</span> <span class="va">rse</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">rse</span><span class="op">)</span><span class="op">$</span><span class="va">name</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="va">rse</span></span></code></pre></div>
<pre><code><span><span class="co">## class: RangedSummarizedExperiment </span></span>
<span><span class="co">## dim: 100 2 </span></span>
<span><span class="co">## metadata(0):</span></span>
<span><span class="co">## assays(9): ALT nRef ... depth edit_freq</span></span>
<span><span class="co">## rownames(100): site_chr4_126885_1 site_chr4_126955_1 ...</span></span>
<span><span class="co">##   site_chr4_994239_1 site_chr4_995145_1</span></span>
<span><span class="co">## rowData names(9): REF rpbz ... repFamily name</span></span>
<span><span class="co">## colnames(2): rna dna</span></span>
<span><span class="co">## colData names(3): sample n_sites edit_idx</span></span></code></pre>
<p>Lastly, we’ll further filter the edit sites to require that the
editing frequency is &gt; 0.05 and that at least 2 reads support the
editing site.</p>
<div class="sourceCode" id="cb73"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">to_keep</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">rse</span>, <span class="st">"edit_freq"</span><span class="op">)</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span> <span class="op">&gt;</span> <span class="fl">0.05</span></span>
<span><span class="va">rse</span> <span class="op">&lt;-</span> <span class="va">rse</span><span class="op">[</span><span class="va">to_keep</span>, <span class="op">]</span></span>
<span></span>
<span><span class="va">rse</span> <span class="op">&lt;-</span> <span class="va">rse</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">rse</span>, <span class="st">"nAlt"</span><span class="op">)</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span> <span class="op">&gt;=</span> <span class="fl">2</span><span class="op">]</span></span></code></pre></div>
<p>With the above filtering approach we obtain a set of putative editing
sites. The specificity of the filtering can be estimated by examining
the number of A-to-G changes relative to other variants. A-to-I RNA
editing is more common than other types of editing (e.g. C-&gt;U editing
by APOBEC enzymes) in human datasets so the majority of the variants
should by A-to-G. In this vignette data all of the identified sites are
A-to-G.</p>
<div class="sourceCode" id="cb74"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">rowRanges</span><span class="op">(</span><span class="va">rse</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 10 ranges and 9 metadata columns:</span></span>
<span><span class="co">##                      seqnames    ranges strand |         REF      rpbz</span></span>
<span><span class="co">##                         &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;numeric&gt;</span></span>
<span><span class="co">##   site_chr4_126885_1     chr4    126885      + |           A  0.342818</span></span>
<span><span class="co">##   site_chr4_134353_1     chr4    134353      + |           A -0.264728</span></span>
<span><span class="co">##   site_chr4_348458_1     chr4    348458      + |           A -0.252166</span></span>
<span><span class="co">##   site_chr4_378548_1     chr4    378548      + |           A  1.406115</span></span>
<span><span class="co">##   site_chr4_379540_1     chr4    379540      + |           A -1.949107</span></span>
<span><span class="co">##   site_chr4_379558_1     chr4    379558      + |           A -0.803727</span></span>
<span><span class="co">##   site_chr4_380392_1     chr4    380392      + |           A -0.207575</span></span>
<span><span class="co">##   site_chr4_380573_1     chr4    380573      + |           A -1.918805</span></span>
<span><span class="co">##   site_chr4_778739_2     chr4    778739      - |           A  1.109806</span></span>
<span><span class="co">##   site_chr4_778780_2     chr4    778780      - |           A -1.469323</span></span>
<span><span class="co">##                            vdb       sor         ALT repName repClass</span></span>
<span><span class="co">##                      &lt;numeric&gt; &lt;numeric&gt; &lt;character&gt;   &lt;Rle&gt;    &lt;Rle&gt;</span></span>
<span><span class="co">##   site_chr4_126885_1 0.7000000  1.061473           G    &lt;NA&gt;     &lt;NA&gt;</span></span>
<span><span class="co">##   site_chr4_134353_1 0.4310148  0.252306           G    &lt;NA&gt;     &lt;NA&gt;</span></span>
<span><span class="co">##   site_chr4_348458_1 0.8000000  0.168623           G  AluSx1     SINE</span></span>
<span><span class="co">##   site_chr4_378548_1 0.2800000  2.833213           G    &lt;NA&gt;     &lt;NA&gt;</span></span>
<span><span class="co">##   site_chr4_379540_1 0.2200000  1.243978           G    &lt;NA&gt;     &lt;NA&gt;</span></span>
<span><span class="co">##   site_chr4_379558_1 0.0200000  0.446287           G    &lt;NA&gt;     &lt;NA&gt;</span></span>
<span><span class="co">##   site_chr4_380392_1 0.5689852  0.910414           G  AluSq2     SINE</span></span>
<span><span class="co">##   site_chr4_380573_1 0.1000000  2.285077           G  AluSq2     SINE</span></span>
<span><span class="co">##   site_chr4_778739_2 0.1918587  0.604657           G   AluSg     SINE</span></span>
<span><span class="co">##   site_chr4_778780_2 0.0490877  2.821974           G   AluSg     SINE</span></span>
<span><span class="co">##                      repFamily  name</span></span>
<span><span class="co">##                          &lt;Rle&gt; &lt;Rle&gt;</span></span>
<span><span class="co">##   site_chr4_126885_1      &lt;NA&gt;  &lt;NA&gt;</span></span>
<span><span class="co">##   site_chr4_134353_1      &lt;NA&gt;  &lt;NA&gt;</span></span>
<span><span class="co">##   site_chr4_348458_1       Alu  &lt;NA&gt;</span></span>
<span><span class="co">##   site_chr4_378548_1      &lt;NA&gt;  &lt;NA&gt;</span></span>
<span><span class="co">##   site_chr4_379540_1      &lt;NA&gt;  &lt;NA&gt;</span></span>
<span><span class="co">##   site_chr4_379558_1      &lt;NA&gt;  &lt;NA&gt;</span></span>
<span><span class="co">##   site_chr4_380392_1       Alu  &lt;NA&gt;</span></span>
<span><span class="co">##   site_chr4_380573_1       Alu  &lt;NA&gt;</span></span>
<span><span class="co">##   site_chr4_778739_2       Alu  &lt;NA&gt;</span></span>
<span><span class="co">##   site_chr4_778780_2       Alu  &lt;NA&gt;</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 1 sequence from an unspecified genome</span></span></code></pre>
<p>Finally once a set of sites has been identified, additional packges
in the bioconductor ecosystem, such the <a href="https://bioconductor.org/packages/release/bioc/html/VariantAnnotation.html" class="external-link">VariantAnnotation</a>
package, can be used to determine the genomic context and potential
molecular consequences of the editing event.</p>
<details style="margin-bottom:10px;"><summary>
Session info
</summary><div class="sourceCode" id="cb76"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.3.1 (2023-06-16)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">## Running under: Ubuntu 22.04.3 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] grid      stats4    stats     graphics  grDevices utils     datasets </span></span>
<span><span class="co">## [8] methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] Rsamtools_2.17.0                        </span></span>
<span><span class="co">##  [2] ComplexHeatmap_2.17.0                   </span></span>
<span><span class="co">##  [3] SNPlocs.Hsapiens.dbSNP144.GRCh38_0.99.20</span></span>
<span><span class="co">##  [4] BSgenome_1.69.0                         </span></span>
<span><span class="co">##  [5] rtracklayer_1.61.1                      </span></span>
<span><span class="co">##  [6] Biostrings_2.69.2                       </span></span>
<span><span class="co">##  [7] XVector_0.41.1                          </span></span>
<span><span class="co">##  [8] AnnotationHub_3.9.2                     </span></span>
<span><span class="co">##  [9] BiocFileCache_2.9.1                     </span></span>
<span><span class="co">## [10] dbplyr_2.3.3                            </span></span>
<span><span class="co">## [11] TxDb.Hsapiens.UCSC.hg38.knownGene_3.18.0</span></span>
<span><span class="co">## [12] GenomicFeatures_1.53.2                  </span></span>
<span><span class="co">## [13] AnnotationDbi_1.63.2                    </span></span>
<span><span class="co">## [14] scater_1.29.4                           </span></span>
<span><span class="co">## [15] ggplot2_3.4.3                           </span></span>
<span><span class="co">## [16] scuttle_1.11.2                          </span></span>
<span><span class="co">## [17] SingleCellExperiment_1.23.0             </span></span>
<span><span class="co">## [18] SummarizedExperiment_1.31.1             </span></span>
<span><span class="co">## [19] Biobase_2.61.0                          </span></span>
<span><span class="co">## [20] GenomicRanges_1.53.1                    </span></span>
<span><span class="co">## [21] GenomeInfoDb_1.37.4                     </span></span>
<span><span class="co">## [22] IRanges_2.35.2                          </span></span>
<span><span class="co">## [23] S4Vectors_0.39.1                        </span></span>
<span><span class="co">## [24] BiocGenerics_0.47.0                     </span></span>
<span><span class="co">## [25] MatrixGenerics_1.13.1                   </span></span>
<span><span class="co">## [26] matrixStats_1.0.0                       </span></span>
<span><span class="co">## [27] raerdata_0.99.2                         </span></span>
<span><span class="co">## [28] raer_0.99.13                            </span></span>
<span><span class="co">## [29] BiocStyle_2.29.2                        </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] later_1.3.1                   BiocIO_1.11.0                </span></span>
<span><span class="co">##   [3] bitops_1.0-7                  filelock_1.0.2               </span></span>
<span><span class="co">##   [5] tibble_3.2.1                  R.oo_1.25.0                  </span></span>
<span><span class="co">##   [7] XML_3.99-0.14                 lifecycle_1.0.3              </span></span>
<span><span class="co">##   [9] doParallel_1.0.17             rprojroot_2.0.3              </span></span>
<span><span class="co">##  [11] lattice_0.21-8                magrittr_2.0.3               </span></span>
<span><span class="co">##  [13] sass_0.4.7                    rmarkdown_2.25               </span></span>
<span><span class="co">##  [15] jquerylib_0.1.4               yaml_2.3.7                   </span></span>
<span><span class="co">##  [17] httpuv_1.6.11                 DBI_1.1.3                    </span></span>
<span><span class="co">##  [19] RColorBrewer_1.1-3            abind_1.4-5                  </span></span>
<span><span class="co">##  [21] zlibbioc_1.47.0               purrr_1.0.2                  </span></span>
<span><span class="co">##  [23] R.utils_2.12.2                RCurl_1.98-1.12              </span></span>
<span><span class="co">##  [25] rappdirs_0.3.3                circlize_0.4.15              </span></span>
<span><span class="co">##  [27] GenomeInfoDbData_1.2.10       ggrepel_0.9.3                </span></span>
<span><span class="co">##  [29] irlba_2.3.5.1                 pheatmap_1.0.12              </span></span>
<span><span class="co">##  [31] pkgdown_2.0.7                 DelayedMatrixStats_1.23.4    </span></span>
<span><span class="co">##  [33] codetools_0.2-19              DelayedArray_0.27.10         </span></span>
<span><span class="co">##  [35] xml2_1.3.5                    tidyselect_1.2.0             </span></span>
<span><span class="co">##  [37] shape_1.4.6                   farver_2.1.1                 </span></span>
<span><span class="co">##  [39] ScaledMatrix_1.9.1            viridis_0.6.4                </span></span>
<span><span class="co">##  [41] GenomicAlignments_1.37.0      jsonlite_1.8.7               </span></span>
<span><span class="co">##  [43] GetoptLong_1.0.5              BiocNeighbors_1.19.0         </span></span>
<span><span class="co">##  [45] ellipsis_0.3.2                iterators_1.0.14             </span></span>
<span><span class="co">##  [47] systemfonts_1.0.4             foreach_1.5.2                </span></span>
<span><span class="co">##  [49] tools_4.3.1                   progress_1.2.2               </span></span>
<span><span class="co">##  [51] ragg_1.2.5                    Rcpp_1.0.11                  </span></span>
<span><span class="co">##  [53] glue_1.6.2                    gridExtra_2.3                </span></span>
<span><span class="co">##  [55] SparseArray_1.1.12            xfun_0.40                    </span></span>
<span><span class="co">##  [57] DESeq2_1.41.10                dplyr_1.1.3                  </span></span>
<span><span class="co">##  [59] withr_2.5.0                   BiocManager_1.30.22          </span></span>
<span><span class="co">##  [61] fastmap_1.1.1                 fansi_1.0.4                  </span></span>
<span><span class="co">##  [63] digest_0.6.33                 rsvd_1.0.5                   </span></span>
<span><span class="co">##  [65] R6_2.5.1                      mime_0.12                    </span></span>
<span><span class="co">##  [67] textshaping_0.3.6             colorspace_2.1-0             </span></span>
<span><span class="co">##  [69] Cairo_1.6-1                   biomaRt_2.57.1               </span></span>
<span><span class="co">##  [71] RSQLite_2.3.1                 R.methodsS3_1.8.2            </span></span>
<span><span class="co">##  [73] utf8_1.2.3                    generics_0.1.3               </span></span>
<span><span class="co">##  [75] data.table_1.14.8             prettyunits_1.1.1            </span></span>
<span><span class="co">##  [77] httr_1.4.7                    S4Arrays_1.1.6               </span></span>
<span><span class="co">##  [79] pkgconfig_2.0.3               gtable_0.3.4                 </span></span>
<span><span class="co">##  [81] blob_1.2.4                    htmltools_0.5.6              </span></span>
<span><span class="co">##  [83] bookdown_0.35                 clue_0.3-64                  </span></span>
<span><span class="co">##  [85] scales_1.2.1                  png_0.1-8                    </span></span>
<span><span class="co">##  [87] knitr_1.44                    rjson_0.2.21                 </span></span>
<span><span class="co">##  [89] curl_5.0.2                    cachem_1.0.8                 </span></span>
<span><span class="co">##  [91] GlobalOptions_0.1.2           stringr_1.5.0                </span></span>
<span><span class="co">##  [93] BiocVersion_3.18.0            parallel_4.3.1               </span></span>
<span><span class="co">##  [95] vipor_0.4.5                   restfulr_0.0.15              </span></span>
<span><span class="co">##  [97] desc_1.4.2                    pillar_1.9.0                 </span></span>
<span><span class="co">##  [99] vctrs_0.6.3                   promises_1.2.1               </span></span>
<span><span class="co">## [101] BiocSingular_1.17.1           beachmat_2.17.16             </span></span>
<span><span class="co">## [103] xtable_1.8-4                  cluster_2.1.4                </span></span>
<span><span class="co">## [105] beeswarm_0.4.0                evaluate_0.21                </span></span>
<span><span class="co">## [107] cli_3.6.1                     locfit_1.5-9.8               </span></span>
<span><span class="co">## [109] compiler_4.3.1                rlang_1.1.1                  </span></span>
<span><span class="co">## [111] crayon_1.5.2                  labeling_0.4.3               </span></span>
<span><span class="co">## [113] fs_1.6.3                      ggbeeswarm_0.7.2             </span></span>
<span><span class="co">## [115] stringi_1.7.12                viridisLite_0.4.2            </span></span>
<span><span class="co">## [117] BiocParallel_1.35.4           munsell_0.5.0                </span></span>
<span><span class="co">## [119] Matrix_1.6-1.1                ExperimentHub_2.9.1          </span></span>
<span><span class="co">## [121] hms_1.1.3                     sparseMatrixStats_1.13.4     </span></span>
<span><span class="co">## [123] bit64_4.0.5                   KEGGREST_1.41.0              </span></span>
<span><span class="co">## [125] shiny_1.7.5                   interactiveDisplayBase_1.39.0</span></span>
<span><span class="co">## [127] memoise_2.0.1                 bslib_0.5.1                  </span></span>
<span><span class="co">## [129] bit_4.0.5</span></span></code></pre>
</details><div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Bonfield2021-yo" class="csl-entry">
Bonfield, James K, John Marshall, Petr Danecek, Heng Li, Valeriu Ohan,
Andrew Whitwham, Thomas Keane, and Robert M Davies. 2021.
<span>“<span>HTSlib</span>: <span>C</span> Library for Reading/Writing
High-Throughput Sequencing Data.”</span> <em>Gigascience</em> 10 (2). <a href="https://doi.org/10.1093/gigascience/giab007" class="external-link">https://doi.org/10.1093/gigascience/giab007</a>.
</div>
<div id="ref-Mansi2021-za" class="csl-entry">
Mansi, Luigi, Marco Antonio Tangaro, Claudio Lo Giudice, Tiziano Flati,
Eli Kopel, Amos Avraham Schaffer, Tiziana Castrignanò, Giovanni
Chillemi, Graziano Pesole, and Ernesto Picardi. 2021.
<span>“<span>REDIportal</span>: Millions of Novel <span class="nocase">A-to-I</span> <span>RNA</span> Editing Events from
Thousands of <span>RNAseq</span> Experiments.”</span> <em>Nucleic Acids
Res.</em> 49 (D1): D1012–19. <a href="https://doi.org/10.1093/nar/gkaa916" class="external-link">https://doi.org/10.1093/nar/gkaa916</a>.
</div>
<div id="ref-Roth2019-yu" class="csl-entry">
Roth, Shalom Hillel, Erez Y Levanon, and Eli Eisenberg. 2019.
<span>“Genome-Wide Quantification of <span>ADAR</span>
Adenosine-to-Inosine <span>RNA</span> Editing Activity.”</span> <em>Nat.
Methods</em> 16 (11): 1131–38. <a href="https://doi.org/10.1038/s41592-019-0610-9" class="external-link">https://doi.org/10.1038/s41592-019-0610-9</a>.
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>An R package developed by the <a href="https://medschool.cuanschutz.edu/rbi" class="external-link">RNA Bioscience Initiative</a></p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
