<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="raer">
<title>Identifying RNA editing sites in droplet single cell datasets • raer</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><link href="../deps/Source_Sans_Pro-0.4.7/font.css" rel="stylesheet">
<link href="../deps/Source_Code_Pro-0.4.7/font.css" rel="stylesheet">
<!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Identifying RNA editing sites in droplet single cell datasets">
<meta property="og:description" content="raer">
<meta property="og:image" content="https://rnabioco.github.io/raer/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><script defer data-domain="rnabioco.github.io" src="https://plausible.io/js/plausible.js"></script>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-dark"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">raer</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.8</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/raer.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/novel-sites.html">Identifying RNA editing sites using RNA and DNA sequencing data</a>
    <a class="dropdown-item" href="../articles/single-cell.html">Identifying RNA editing sites in droplet single cell datasets</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/rnabioco/raer/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Identifying RNA editing sites in droplet single cell datasets</h1>
                        <h4 data-toc-skip class="author">Kent
Riemondy</h4>
            <address class="author_afil">
      University of Colorado School of Medicine<br><h4 data-toc-skip class="date">2023-08-21</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/rnabioco/raer/blob/HEAD/vignettes/single-cell.Rmd" class="external-link"><code>vignettes/single-cell.Rmd</code></a></small>
      <div class="d-none name"><code>single-cell.Rmd</code></div>
    </address>
</div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://rnabioco.github.io/raer">raer</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/Rsamtools" class="external-link">Rsamtools</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">rtracklayer</span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="characterizing-rna-editing-sites-in-single-cell-data">Characterizing RNA editing sites in single cell data<a class="anchor" aria-label="anchor" href="#characterizing-rna-editing-sites-in-single-cell-data"></a>
</h2>
<p>This vignette will demonstrate how to use the <code>raer</code>
package to examine RNA editing in droplet-based single cell RNA-seq
data.</p>
<div class="section level3">
<h3 id="preprocessing">Preprocessing<a class="anchor" aria-label="anchor" href="#preprocessing"></a>
</h3>
<p>For this example analysis we will use a single cell dataset
containing human PBMC cells from <a href="https://www.10xgenomics.com/resources/datasets/10k-human-pbmcs-3-v3-1-chromium-x-with-intronic-reads-3-1-high" class="external-link">10x
Genomics</a>. The single cell data was processed using the cellranger
pipeline. The BAM file contains a tag (<code>CB</code>) which indicates
the cell-barcode associated with each alignment, as well as a tag
containing the inferred UMI sequence (<code>UB</code>).</p>
</div>
<div class="section level3">
<h3 id="single-cell-editing-analysis">Single cell editing analysis<a class="anchor" aria-label="anchor" href="#single-cell-editing-analysis"></a>
</h3>
<p>A subset of a human PBMC scRNA-seq dataset from 10x Genomics, along
with other needed files can be downloaded using
<code><a href="../reference/download_human_pbmc.html">download_human_pbmc()</a></code>.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fns</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/download_human_pbmc.html">download_human_pbmc</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">fns</span></span></code></pre></div>
<pre><code><span><span class="co">## $bam</span></span>
<span><span class="co">## [1] "/github/home/.cache/R/raer/10k_PBMC_3p_nextgem_Chromium_X_intron_possorted_chr16_rp.bam"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $bai</span></span>
<span><span class="co">## [1] "/github/home/.cache/R/raer/10k_PBMC_3p_nextgem_Chromium_X_intron_possorted_chr16_rp.bam.bai"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $edit_sites</span></span>
<span><span class="co">## [1] "/github/home/.cache/R/raer/133d16b712fd_rediportal_chr16.bed.gz"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $sce</span></span>
<span><span class="co">## [1] "/github/home/.cache/R/raer/133d9190e9_sce.rds"</span></span></code></pre>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bam_fn</span> <span class="op">&lt;-</span> <span class="va">fns</span><span class="op">$</span><span class="va">bam</span></span>
<span><span class="va">bed_fn</span> <span class="op">&lt;-</span> <span class="va">fns</span><span class="op">$</span><span class="va">edit_sites</span></span></code></pre></div>
<p>Next we’ll load in a <code>SingleCellExperiment</code> with cell-type
annotations.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="va">fns</span><span class="op">$</span><span class="va">sce</span><span class="op">)</span></span>
<span><span class="va">sce</span></span></code></pre></div>
<pre><code><span><span class="co">## class: SingleCellExperiment </span></span>
<span><span class="co">## dim: 36601 500 </span></span>
<span><span class="co">## metadata(2): Samples mkrs</span></span>
<span><span class="co">## assays(2): counts logcounts</span></span>
<span><span class="co">## rownames(36601): MIR1302-2HG FAM138A ... AC007325.4 AC007325.2</span></span>
<span><span class="co">## rowData names(3): ID Symbol Type</span></span>
<span><span class="co">## colnames(500): TGTTTGTCAGTTAGGG-1 ATCTCTACAAGCTACT-1 ...</span></span>
<span><span class="co">##   GGGCGTTTCAGGACGA-1 CTATAGGAGATTGTGA-1</span></span>
<span><span class="co">## colData names(8): Sample Barcode ... r celltype</span></span>
<span><span class="co">## reducedDimNames(2): PCA UMAP</span></span>
<span><span class="co">## mainExpName: NULL</span></span>
<span><span class="co">## altExpNames(0):</span></span></code></pre>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotUMAP</a></span><span class="op">(</span><span class="va">sce</span>, colour_by <span class="op">=</span> <span class="st">"celltype"</span><span class="op">)</span></span></code></pre></div>
<p><img src="single-cell_files/figure-html/unnamed-chunk-5-1.png" width="700"></p>
<p>Next we’ll select editing sites to query. For this analysis we will
use sites from the Rediportal database.</p>
<p>If the editing sites of interest are not known, one option is to
perform a two pass operation. First, identify editing sites by treating
the data as a bulk-RNA-seq experiment, using for example
<code><a href="../reference/pileup_sites.html">pileup_sites()</a></code>. Then filter these sites to establish a set
of high confidence sites to query in single cell mode.</p>
<p><code>raer</code> provides a function, <code><a href="../reference/pileup_cells.html">pileup_cells()</a></code>,
which will quantify edited and non-edited UMI counts per cell barcode,
then collect the site counts into a SingleCellExperiment.</p>
<p>The sites to quantified are specified using a custom formatted
GRanges object with 1 base intervals, a strand (+ or -), and
supplemented with metadata columns named <code>REF</code> and
<code>ALT</code> containing the reference and alternate base to query.
In this case we are only interested in A-&gt;I editing, so we set the
ref and alt to <code>A</code> and <code>G</code>. Note that the
<code>REF</code> and <code>ALT</code> bases are in reference to strand.
For a <code>-</code> strand interval the bases should be the complement
of the <code>+</code> strand bases.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gr</span> <span class="op">&lt;-</span> <span class="fu">import</span><span class="op">(</span><span class="va">bed_fn</span><span class="op">)</span></span>
<span><span class="va">gr</span><span class="op">$</span><span class="va">name</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span>
<span><span class="va">gr</span><span class="op">$</span><span class="va">score</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span>
<span><span class="va">gr</span><span class="op">$</span><span class="va">REF</span> <span class="op">&lt;-</span> <span class="st">"A"</span></span>
<span><span class="va">gr</span><span class="op">$</span><span class="va">ALT</span> <span class="op">&lt;-</span> <span class="st">"G"</span></span>
<span><span class="va">gr</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 785208 ranges and 2 metadata columns:</span></span>
<span><span class="co">##            seqnames    ranges strand |         REF         ALT</span></span>
<span><span class="co">##               &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;character&gt;</span></span>
<span><span class="co">##        [1]    chr16     22083      - |           A           G</span></span>
<span><span class="co">##        [2]    chr16     22101      - |           A           G</span></span>
<span><span class="co">##        [3]    chr16     22107      - |           A           G</span></span>
<span><span class="co">##        [4]    chr16     22113      - |           A           G</span></span>
<span><span class="co">##        [5]    chr16     22114      - |           A           G</span></span>
<span><span class="co">##        ...      ...       ...    ... .         ...         ...</span></span>
<span><span class="co">##   [785204]    chr16  90219996      + |           A           G</span></span>
<span><span class="co">##   [785205]    chr16  90220035      + |           A           G</span></span>
<span><span class="co">##   [785206]    chr16  90220081      + |           A           G</span></span>
<span><span class="co">##   [785207]    chr16  90220095      + |           A           G</span></span>
<span><span class="co">##   [785208]    chr16  90220135      + |           A           G</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></span></code></pre>
<p><code><a href="../reference/pileup_cells.html">pileup_cells()</a></code> accepts a <code><a href="../reference/pileup_sites.html">FilterParam()</a></code>
class to specifying how to performing read and site filtering. Note that
<code><a href="../reference/pileup_cells.html">pileup_cells()</a></code> is strand sensitive by default, so it is
important to ensure that the strand of the input sites is correctly
annotated, and that the library type is correct for the type of data of
interest. For 10x Genomics data, the library type is set to
<code>fr-second-strand</code>, indicating that the strand of the
alignments matches the strand of the RNA.</p>
<p>Note that <code>bam_flags</code> is set to <strong>include</strong>
duplicate reads by default. If the bamfile has a tag with a UMI
sequence, this can be supplied to the <code>umi_tag</code> argument to
only count 1 read for each CB-UMI pair at each position. This strategy
allows for reads with the same UMI to be counted at multiple independent
sites enabling recovery of more sequence variants than counting only 1
read per UMI.</p>
<p>Processing time can be reduced by operating in parallel across
chromosomes, by supplying a BiocParallel backend to the
<code>BPPARAM</code> argument (e.g. <code>MultiCoreParam()</code>).</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">outdir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"sc_edits"</span><span class="op">)</span></span>
<span><span class="va">cbs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span><span class="va">e_sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pileup_cells.html">pileup_cells</a></span><span class="op">(</span></span>
<span>    bamfile <span class="op">=</span> <span class="va">bam_fn</span>,</span>
<span>    sites <span class="op">=</span> <span class="va">gr</span>,</span>
<span>    cell_barcodes <span class="op">=</span> <span class="va">cbs</span>,</span>
<span>    output_directory <span class="op">=</span> <span class="va">outdir</span>,</span>
<span>    cb_tag <span class="op">=</span> <span class="st">"CB"</span>,</span>
<span>    umi_tag <span class="op">=</span> <span class="st">"UB"</span>,</span>
<span>    param <span class="op">=</span> <span class="fu"><a href="../reference/pileup_sites.html">FilterParam</a></span><span class="op">(</span></span>
<span>        min_mapq <span class="op">=</span> <span class="fl">255L</span>,</span>
<span>        library_type <span class="op">=</span> <span class="st">"fr-second-strand"</span>,</span>
<span>        min_variant_reads <span class="op">=</span> <span class="fl">1L</span></span>
<span>    <span class="op">)</span>,</span>
<span>    verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span><span class="va">e_sce</span></span></code></pre></div>
<pre><code><span><span class="co">## class: SingleCellExperiment </span></span>
<span><span class="co">## dim: 3850 500 </span></span>
<span><span class="co">## metadata(0):</span></span>
<span><span class="co">## assays(2): nRef nAlt</span></span>
<span><span class="co">## rownames(3850): site_chr16_83540_1_AG site_chr16_83621_1_AG ...</span></span>
<span><span class="co">##   site_chr16_31453268_2_AG site_chr16_31454303_2_AG</span></span>
<span><span class="co">## rowData names(2): REF ALT</span></span>
<span><span class="co">## colnames(500): TGTTTGTCAGTTAGGG-1 ATCTCTACAAGCTACT-1 ...</span></span>
<span><span class="co">##   GGGCGTTTCAGGACGA-1 CTATAGGAGATTGTGA-1</span></span>
<span><span class="co">## colData names(0):</span></span>
<span><span class="co">## reducedDimNames(0):</span></span>
<span><span class="co">## mainExpName: NULL</span></span>
<span><span class="co">## altExpNames(0):</span></span></code></pre>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">dir</a></span><span class="op">(</span><span class="va">outdir</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "barcodes.txt.gz" "counts.mtx.gz"   "sites.txt.gz"</span></span></code></pre>
<p>Next we’ll filter the pileups to find sites with at least 5 cells
with an editing event, and add the editing information to the
SingleCellExperiment as an <code><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html" class="external-link">altExp()</a></code>.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">e_sce</span> <span class="op">&lt;-</span> <span class="va">e_sce</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/MatrixGenerics/man/rowSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assays</a></span><span class="op">(</span><span class="va">e_sce</span><span class="op">)</span><span class="op">$</span><span class="va">nAlt</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="fl">5</span>, <span class="op">]</span></span>
<span><span class="va">e_sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calc_edit_frequency.html">calc_edit_frequency</a></span><span class="op">(</span><span class="va">e_sce</span>, edit_from <span class="op">=</span> <span class="st">"Ref"</span>, edit_to <span class="op">=</span> <span class="st">"Alt"</span>, replace_na <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html" class="external-link">altExp</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">e_sce</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">]</span></span></code></pre></div>
<p>With the editing sites added to the gene expression
SingleCellExperiment we can use plotting and other methods previously
developed for single cell analysis. Here we’ll visualize editing sites
with the highest edited read counts.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">to_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html" class="external-link">altExp</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/order.html" class="external-link">order</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/MatrixGenerics/man/rowSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html" class="external-link">altExp</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>, <span class="st">"nAlt"</span><span class="op">)</span><span class="op">)</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">to_plot</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotUMAP</a></span><span class="op">(</span><span class="va">sce</span>, colour_by <span class="op">=</span> <span class="va">x</span>, by_exprs_values <span class="op">=</span> <span class="st">"nAlt"</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [[1]]</span></span></code></pre>
<p><img src="single-cell_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## [[2]]</span></span></code></pre>
<p><img src="single-cell_files/figure-html/unnamed-chunk-9-2.png" width="700"></p>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## [[3]]</span></span></code></pre>
<p><img src="single-cell_files/figure-html/unnamed-chunk-9-3.png" width="700"></p>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## [[4]]</span></span></code></pre>
<p><img src="single-cell_files/figure-html/unnamed-chunk-9-4.png" width="700"></p>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## [[5]]</span></span></code></pre>
<p><img src="single-cell_files/figure-html/unnamed-chunk-9-5.png" width="700"></p>
<p>Alternatively we can view these top edited sites as a Heatmap,
showing the average number of edited reads per site in each cell
type.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html" class="external-link">altExp</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">$</span><span class="va">celltype</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">$</span><span class="va">celltype</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotGroupedHeatmap.html" class="external-link">plotGroupedHeatmap</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html" class="external-link">altExp</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>,</span>
<span>    features <span class="op">=</span> <span class="va">to_plot</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">25</span><span class="op">]</span>,</span>
<span>    group <span class="op">=</span> <span class="st">"celltype"</span>,</span>
<span>    exprs_values <span class="op">=</span> <span class="st">"nAlt"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="single-cell_files/figure-html/unnamed-chunk-10-1.png" width="700"></p>
<p><code>raer</code> provides additional tools to examine cell type
specific editing. <code>find_celltype_de()</code> will perform
statistical testing to identify sites with different editing frequencies
between clusters/cell types. <code><a href="../reference/calc_scAEI.html">calc_scAEI()</a></code> will calculate
the <code>AEI</code> score in single cells.</p>
<details style="margin-bottom:10px;"><summary>
Session info
</summary><div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.3.0 (2023-04-21)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">## Running under: Ubuntu 22.04.2 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] rtracklayer_1.60.1          Rsamtools_2.16.0           </span></span>
<span><span class="co">##  [3] Biostrings_2.68.1           XVector_0.40.0             </span></span>
<span><span class="co">##  [5] scater_1.28.0               ggplot2_3.4.3              </span></span>
<span><span class="co">##  [7] scuttle_1.10.2              SingleCellExperiment_1.22.0</span></span>
<span><span class="co">##  [9] SummarizedExperiment_1.30.2 Biobase_2.60.0             </span></span>
<span><span class="co">## [11] GenomicRanges_1.52.0        GenomeInfoDb_1.36.1        </span></span>
<span><span class="co">## [13] IRanges_2.34.1              S4Vectors_0.38.1           </span></span>
<span><span class="co">## [15] BiocGenerics_0.46.0         MatrixGenerics_1.12.3      </span></span>
<span><span class="co">## [17] matrixStats_1.0.0           raer_0.99.8                </span></span>
<span><span class="co">## [19] BiocStyle_2.28.0           </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] RColorBrewer_1.1-3        jsonlite_1.8.7           </span></span>
<span><span class="co">##   [3] magrittr_2.0.3            ggbeeswarm_0.7.2         </span></span>
<span><span class="co">##   [5] GenomicFeatures_1.52.1    farver_2.1.1             </span></span>
<span><span class="co">##   [7] rmarkdown_2.24            fs_1.6.3                 </span></span>
<span><span class="co">##   [9] BiocIO_1.10.0             zlibbioc_1.46.0          </span></span>
<span><span class="co">##  [11] ragg_1.2.5                vctrs_0.6.3              </span></span>
<span><span class="co">##  [13] memoise_2.0.1             DelayedMatrixStats_1.22.5</span></span>
<span><span class="co">##  [15] RCurl_1.98-1.12           htmltools_0.5.6          </span></span>
<span><span class="co">##  [17] S4Arrays_1.0.5            progress_1.2.2           </span></span>
<span><span class="co">##  [19] curl_5.0.2                BiocNeighbors_1.18.0     </span></span>
<span><span class="co">##  [21] sass_0.4.7                bslib_0.5.1              </span></span>
<span><span class="co">##  [23] desc_1.4.2                cachem_1.0.8             </span></span>
<span><span class="co">##  [25] GenomicAlignments_1.36.0  lifecycle_1.0.3          </span></span>
<span><span class="co">##  [27] pkgconfig_2.0.3           rsvd_1.0.5               </span></span>
<span><span class="co">##  [29] Matrix_1.6-1              R6_2.5.1                 </span></span>
<span><span class="co">##  [31] fastmap_1.1.1             GenomeInfoDbData_1.2.10  </span></span>
<span><span class="co">##  [33] digest_0.6.33             colorspace_2.1-0         </span></span>
<span><span class="co">##  [35] AnnotationDbi_1.62.2      rprojroot_2.0.3          </span></span>
<span><span class="co">##  [37] irlba_2.3.5.1             textshaping_0.3.6        </span></span>
<span><span class="co">##  [39] RSQLite_2.3.1             beachmat_2.16.0          </span></span>
<span><span class="co">##  [41] labeling_0.4.2            filelock_1.0.2           </span></span>
<span><span class="co">##  [43] fansi_1.0.4               httr_1.4.7               </span></span>
<span><span class="co">##  [45] abind_1.4-5               compiler_4.3.0           </span></span>
<span><span class="co">##  [47] bit64_4.0.5               withr_2.5.0              </span></span>
<span><span class="co">##  [49] BiocParallel_1.34.2       viridis_0.6.4            </span></span>
<span><span class="co">##  [51] DBI_1.1.3                 highr_0.10               </span></span>
<span><span class="co">##  [53] R.utils_2.12.2            biomaRt_2.56.1           </span></span>
<span><span class="co">##  [55] rappdirs_0.3.3            DelayedArray_0.26.7      </span></span>
<span><span class="co">##  [57] rjson_0.2.21              tools_4.3.0              </span></span>
<span><span class="co">##  [59] vipor_0.4.5               beeswarm_0.4.0           </span></span>
<span><span class="co">##  [61] R.oo_1.25.0               glue_1.6.2               </span></span>
<span><span class="co">##  [63] restfulr_0.0.15           grid_4.3.0               </span></span>
<span><span class="co">##  [65] generics_0.1.3            gtable_0.3.3             </span></span>
<span><span class="co">##  [67] BSgenome_1.68.0           R.methodsS3_1.8.2        </span></span>
<span><span class="co">##  [69] data.table_1.14.8         hms_1.1.3                </span></span>
<span><span class="co">##  [71] BiocSingular_1.16.0       ScaledMatrix_1.8.1       </span></span>
<span><span class="co">##  [73] xml2_1.3.5                utf8_1.2.3               </span></span>
<span><span class="co">##  [75] ggrepel_0.9.3             pillar_1.9.0             </span></span>
<span><span class="co">##  [77] stringr_1.5.0             dplyr_1.1.2              </span></span>
<span><span class="co">##  [79] BiocFileCache_2.8.0       lattice_0.21-8           </span></span>
<span><span class="co">##  [81] bit_4.0.5                 tidyselect_1.2.0         </span></span>
<span><span class="co">##  [83] knitr_1.43                gridExtra_2.3            </span></span>
<span><span class="co">##  [85] bookdown_0.35             xfun_0.40                </span></span>
<span><span class="co">##  [87] pheatmap_1.0.12           stringi_1.7.12           </span></span>
<span><span class="co">##  [89] yaml_2.3.7                evaluate_0.21            </span></span>
<span><span class="co">##  [91] codetools_0.2-19          tibble_3.2.1             </span></span>
<span><span class="co">##  [93] BiocManager_1.30.22       cli_3.6.1                </span></span>
<span><span class="co">##  [95] systemfonts_1.0.4         munsell_0.5.0            </span></span>
<span><span class="co">##  [97] jquerylib_0.1.4           Rcpp_1.0.11              </span></span>
<span><span class="co">##  [99] dbplyr_2.3.3              png_0.1-8                </span></span>
<span><span class="co">## [101] XML_3.99-0.14             parallel_4.3.0           </span></span>
<span><span class="co">## [103] pkgdown_2.0.7             blob_1.2.4               </span></span>
<span><span class="co">## [105] prettyunits_1.1.1         sparseMatrixStats_1.12.2 </span></span>
<span><span class="co">## [107] bitops_1.0-7              viridisLite_0.4.2        </span></span>
<span><span class="co">## [109] scales_1.2.1              purrr_1.0.2              </span></span>
<span><span class="co">## [111] crayon_1.5.2              rlang_1.1.1              </span></span>
<span><span class="co">## [113] KEGGREST_1.40.0</span></span></code></pre>
</details>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>An R package developed by the <a href="https://medschool.cuanschutz.edu/rbi" class="external-link">RNA Bioscience Initiative</a></p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
